---
title: "505_WGS_pre_analysis"
author: "Ruijuan Li"
date: "8/20/2018"
output: html_document
---

### load libs
```{r} 
library(tidyverse)
library(reshape2)
```

# sample submission for WGS 
```{r}
WGS_samples <- read.csv("~/505/WGS/sample/DNA_to_Library_Submission_Form_20180508_incorporated_all.csv", header = F)  
WGS_samples <- 
t(WGS_samples) %>% as.data.frame() %>% 
  dplyr::select(V1:V12) 

WGS_samples$V2 %>% unlist() %>% as.character()

missing_sample <- c("K114", "SJ_280", "SJ_294")

geno <- read.vcfR("~/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.vcf.gz") 
geno.2 <- reform.vcf(geno) 
dim(geno.2) #  174397    135
geno.2 %>% colnames() 

RNAseq_sample <- gsub("X505.| X505_", "", colnames(geno.2))
RNAseq_sample <- gsub("_", "", RNAseq_sample)[5:135] 

sum(WGS_samples$V2 %in% RNAseq_sample)
sum(RNAseq_sample %in% WGS_samples$V2)

WGS_sample_only <- as.data.frame(t(WGS_samples[!(WGS_samples$V2 %in% c(RNAseq_sample, missing_sample)),])) 
write.csv(WGS_sample_only, file = "~/505/WGS/sample/WGS_sample_only.csv")

sum(WGS_sample_only[2,] %>% unlist() %>% as.character() %in% c(RNAseq_sample, missing_sample)) # correct 
as.data.frame(t(WGS_sample_only[c(1,2),])) %>% 
  left_join(WGS_samples[,c(1,2)]) %>% dim() 
```  

### download 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/download_array_WGS.slurm

### QC 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/QC.sh
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/trimming_array.slurm
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/QC_trimmed.sh
QC stats: 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/fastqc_extract_stats.sh 

```{r}
WGS_data <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/WGS_raw_data_info.csv")
WGS_data <- 
WGS_data %>% 
  dplyr::select(Sample, Yield..Mbases., X.....Q30, Mean.Quality)  

WGS_data <- WGS_data[2:239,] 
colnames(WGS_data) <- c("Sample", "Yield_in_Mb", "Q30_pct", "Mean_quality")
WGS_data$Yield_in_Mb <- gsub("\\,", "",as.character(WGS_data$Yield_in_Mb)) %>% as.numeric()
WGS_data$Q30_pct <- as.character(WGS_data$Q30_pct) %>% as.numeric()
WGS_data$Mean_quality <- as.character(WGS_data$Mean_quality) %>% as.numeric()

sapply(colnames(WGS_data)[2:4], function(i) {median(WGS_data[,i])})
sapply(colnames(WGS_data)[2:4], function(i) {min(WGS_data[,i])})
sapply(colnames(WGS_data)[2:4], function(i) {max(WGS_data[,i])})

# histogram based on reads number 
p.yeild <- 
WGS_data %>% 
  ggplot() + 
  geom_histogram(aes(x = Yield_in_Mb), binwidth = 100) + 
  ylab("Number of samples") 

p.yeild
ggsave(p.yeild, filename = "~/Desktop/p.yeild.png", width = 9, height = 4)
# read count generated by FastQC
read_stats <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats_238")
read_stats <- 
read_stats %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  group_by(Sample) %>% 
  summarise(Number_Reads = sum(Number_Reads)) 
dim(read_stats)

read_stats_HQ <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats_HQ_238")
read_stats_HQ <- 
read_stats_HQ %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  group_by(Sample) %>% 
  summarise(Number_Reads_HQ = sum(Number_Reads))  

# note sample with inconsistent yield report and fastqc read count 
stats_compare <- 
read_stats %>% 
  left_join(WGS_data) %>% 
  mutate(FastQC_calc_Mb = round(Number_Reads * 151 / 1000000, digits = 0)) %>% 
  dplyr::select(Sample, Yield_in_Mb, FastQC_calc_Mb, Number_Reads)

sum(stats_compare$Yield_in_Mb > stats_compare$FastQC_calc_Mb)
sum(stats_compare$Yield_in_Mb != stats_compare$FastQC_calc_Mb) 

stats_compare$note <- ifelse(stats_compare$Yield_in_Mb != stats_compare$FastQC_calc_Mb, "NO", "YES")

p.stats_compare <- 
stats_compare %>% 
  filter(note == "NO") %>% 
  mutate(From_DNA_center = Yield_in_Mb, From_FastQC = FastQC_calc_Mb) %>% 
  dplyr::select(Sample, From_DNA_center, From_FastQC) %>% 
  melt() %>% 
  ggplot() + 
  geom_col(aes(x = Sample, y = value, fill = variable)) + 
  theme(axis.text.x=element_blank()) + 
  ylab("Yield in Mb")

ggsave(p.stats_compare, filename = "~/Desktop/p.stats_compare.png", width = 9, height = 4)  
write.csv(stats_compare, file = "~/Desktop/stats_compare.csv") 

stats_compare <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/stats_compare.csv")
sample_list <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/sample_list")

Prob_samples <- 
stats_compare %>% 
  filter(note == "NO") %>% 
  dplyr::select(Sample) 

sample_list_47 <- 
sample_list %>% 
  mutate(Sample = gsub("_L003", "", V1)) %>% 
  mutate(Sample = gsub("_S.*", "", V1)) %>% 
  right_join(Prob_samples) %>% 
  dplyr::select(V1)

write.table(sample_list_47, file = "~/Desktop/sample_list_47")

# redownload 47 samples and did QC 
read_stats_47 <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats.tab")
read_stats_47 <- 
read_stats_47 %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  group_by(Sample) %>% 
  summarise(Number_Reads = sum(Number_Reads)) 
dim(read_stats_47)

read_stats_47_HQ <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats_HQ.tab")
read_stats_47_HQ <- 
read_stats_47_HQ %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  group_by(Sample) %>% 
  summarise(Number_Reads_HQ = sum(Number_Reads))  

stats_compare_47 <- 
read_stats_47 %>% 
  left_join(WGS_data) %>% 
  mutate(FastQC_calc_Mb = round(Number_Reads * 151 / 1000000, digits = 0)) %>% 
  dplyr::select(Sample, Yield_in_Mb, FastQC_calc_Mb, Number_Reads)

sum(stats_compare_47$Yield_in_Mb > stats_compare_47$FastQC_calc_Mb)
sum(stats_compare_47$Yield_in_Mb != stats_compare_47$FastQC_calc_Mb) 

read_stats_47 %>% 
  left_join(read_stats_47_HQ) %>% 
  mutate(HQ_pct = Number_Reads_HQ/Number_Reads) %>% 
  ggplot() + 
  geom_histogram(aes(x = HQ_pct))  

read_stats_summary_47 <- 
read_stats_47 %>% 
  left_join(read_stats_47_HQ) 

### for KIAT meeting 
read_stats_summary <- 
read_stats %>% 
  left_join(read_stats_HQ) 

read_stats <- 
read_stats_summary %>% 
  anti_join(read_stats_summary_47, by = "Sample") %>% 
  rbind(read_stats_summary_47) 

read_stats %>% head()  
```

### data import to SQL database 
```{r}
WGS_data <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/WGS_raw_data_info.csv")

WGS_data <- 
WGS_data %>% 
  mutate(reads = gsub(",", "", PF.Clusters)) %>%   
  dplyr::select(Sample, reads)  

read_stats <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats_238")
read_stats_SQL <- 
read_stats %>% 
  mutate(sample_ID = gsub("_S.*", "", Sample)) %>% 
  left_join(WGS_data, by = c("sample_ID" = "Sample")) 

SQL_sample_ID <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/database/sample_output.csv")

data_SQL <- 
read_stats_SQL %>% 
  left_join(SQL_sample_ID, by = c("sample_ID" = "Sample_name")) %>% 
  mutate(Sample = gsub("_fastqc", ".fastq.gz", Sample)) %>%
  dplyr::select(Sample, reads, Sample_id) 

write.csv(data_SQL, "~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/database/data_SQL.csv")
```

### mapping 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/mapping_array_WGS_cabernet.slurm 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/Prep4SNPcalling.slurm

mapping stats: 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/5055751.pts-164.barbera/mapping_stats.sh 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/bwa_extract_stats.sh

```{r}
library(reshape2) 

mapping_stats <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/BWA_Stats_new.tab")

WGS_stats <- 
mapping_stats %>% 
  mutate(Sample = gsub("_L003", "", Sample)) %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  left_join(read_stats) %>% 
  mutate(Pct_HQ = Number_Reads_HQ/Number_Reads, 
         Pct_Mapped = Number_Mapped/100, 
         Pct_Unique_Mapped = Number_Unique_Mapped/Number_Reads_HQ)

p.WGS_stats <-  
WGS_stats %>% 
  dplyr::select(starts_with("Pct")) %>% 
  melt() %>% 
  ggplot() + 
  geom_histogram(aes(x = value), binwidth = 0.01) + 
  facet_wrap(~variable, nrow = 3) + 
  ylab("Number of Samples") + xlab("")  

p.WGS_stats  

save(WGS_stats, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/WGS_stats.Rdata")
# note libs with low quailty, low mapping rate, and/or low unique mapping rate  
WGS_stats %>% 
  dplyr::select(Sample, starts_with("Pct")) %>% 
  filter(Pct_HQ < 0.85 | Pct_Mapped < 0.75 | Pct_Unique_Mapped < 0.4)   

ggsave(p.WGS_stats, filename = "~/Desktop/p.WGS_stats.png", width = 8, height = 6)  
```

### SNP calling 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/Bam_Split_By_Chrom.slurm 
Freebayes-parallel SNP calling: 
cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_parallel_Array_WGS.slurm

* 7 chroms require large memory, which mighe be due to extremely high coverage for some regions on those chrom. 

for these 7 chroms, run SNP calling on splitted regions   
```{r}
# for regions
# calculate read coverage: /share/malooflab/Ruijuan/KIAT_cabernet/505/calc_depth_subset_by_chr.slurm   

# cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500_2019/split_region.slurm    

# for chroms: 
# merge bam file from 238 samples for these 7 chroms: actually no need to merge bam files, seperate bam file works as well   
# cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/merge_index_bam_for_Freebayes.slurm

# call SNPs  
# /share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500_2019/Freebayes_parallel_Array_WGS_region.slurm (two options)
# /share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500_2019/Freebayes_parallel_Array_WGS_region_3options.slurm (two options) 
```

### vcf merge      

1) WGS: merge vcf for each chrom, use vcftool to calculate stats and filter, use tassel to convert to hapmap format 

merge and index vcf file for each chrom 
cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500/combine_vcf.sh 

reorder each vcf file and compress: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/reorder_vcf_etc.sh
concat all vcf files: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/concat.sh 
<<<<<<< HEAD
calculate stats: 
=======
calculate stats: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/calc.sh 

check stats: 
>>>>>>> 35705c820751c70c150a0946011c10afbde744e8
```{r} 
# QUAL score
QUAL <- read.delim("~/505/WGS/vcf_raw/combined/505.lqual")
hist(log10(QUAL$QUAL))    
sum(QUAL$QUAL > 40)/nrow(QUAL) # 81% 
sum(QUAL$QUAL > 30)/nrow(QUAL) # 85% 

# MAF
MAF <- read.table("~/505/WGS/vcf_raw/combined/MAF")
nrow(MAF)/nrow(QUAL) # 83% SNP out of all squence variations called from Freebayes

MAF <- 
MAF %>% 
  mutate(V7 = gsub("([[:print:]]+)(:)([[:print:]]+)", "\\3", V5), 
         V8 = gsub("([[:print:]]+)(:)([[:print:]]+)", "\\3", V6)) %>% 
  dplyr::select(V7, V8) 

MAF_minor <- as.numeric(apply(MAF, 1, min))  
sum(MAF_minor[!is.na(MAF_minor)] > 0.05) / nrow(QUAL) # 28% 
sum(MAF_minor[!is.na(MAF_minor)] > 0.01) / nrow(QUAL) # 54% # use this 
sum(MAF_minor[!is.na(MAF_minor)] > 0.001) / nrow(QUAL) # 80% # use this 

hist(as.numeric(apply(MAF, 1, min)))   

# mean read depth across individuals 
read_depth <- read.delim("~/505/WGS/vcf_raw/combined/505.ldepth.mean")
mean(read_depth$MEAN_DEPTH) # 3.56 

hist(log10(read_depth$MEAN_DEPTH), breaks = 1000)
read_depth <- 
read_depth %>% 
  filter(MEAN_DEPTH > 1 & MEAN_DEPTH < 100) 

nrow(read_depth)/nrow(QUAL) # 99% 
hist(read_depth$MEAN_DEPTH, breaks = 500)

# missing rate 
# acorss site, on individual level 
imiss <- read.delim("~/505/WGS/vcf_raw/combined/505.imiss")
imiss %>% 
  filter(F_MISS > 0.8) 

#             INDV   N_DATA N_GENOTYPES_FILTERED   N_MISS   F_MISS
# 1 K185_S290_L003 32876150                    0 32531155 0.989506
# 2 K186_S289_L003 32876150                    0 32534117 0.989596
# 3 K279_S424_L003 32876150                    0 28089705 0.854410

# across taxa, on site level 
lmiss <- read.delim("~/505/WGS/vcf_raw/combined/505.lmiss")
hist(lmiss$F_MISS) 
sum(lmiss$F_MISS < 0.5)/nrow(lmiss) # 90% left    
```

filter for each chrom seperately: biallelic SNP; QUAL score > 40, MAF >= 0.05, mean read depth across individuals greater than 1 and less than 100, missing rate across tax less than 0.5, genotype quality greater than 30 https://github.com/leejimmy93/KIAT/blob/master/505/WGS/filter.sh

vcf to hmp: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/vcf_to_hmp.sh  

2) RNAseq: compress and index vcf file for each chrom, use vcftool to filter, use tassel to convert to hapmap format 

vcf files at: cluster:/share/malooflab/Ruijuan/505/vcf_no_pop_late_silique/all
or Whitney: /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_131_sample 

filter: https://github.com/leejimmy93/KIAT/blob/master/505/filter_late_silique.sh 

vcf to hmp: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/vcf_to_hmp.sh 

3) use join function to merge every chrom's hapmap file from WGS and RNAseq 

hmp files at whitney:/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_combine_WGS_RNAseq/

edit hmp file to remove "#" after rs & assembly  
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_combine_WGS_RNAseq/WGS_filtered/hmp"
files.WGS <- list.files(path=path)  
setwd(path) 
hmp.WGS <- lapply(files.WGS, function(x) read.table(x, header = T, stringsAsFactors = F)) 
hmp.WGS %>% length()
files.WGS <- gsub("_filtered.hmp.txt", "", files.WGS)
names(hmp.WGS) <- files.WGS

path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_combine_WGS_RNAseq/RNA_filtered/hmp"
files.RNA <- list.files(path=path) 
setwd(path)
hmp.RNA <- lapply(files.RNA, function(x) read.table(x, header = T, stringsAsFactors = F)) 
hmp.RNA %>% length() 
files.RNA <- gsub("_filtered.hmp.txt", "", files.RNA) 
names(hmp.RNA) <- files.RNA

common_WGS_RNA_2 <- 
lapply(names(hmp.RNA), function(i) 
  test <- hmp.WGS[[i]] %>% semi_join(hmp.RNA[[i]], by = c("rs", "alleles")))

names(common_WGS_RNA_2) <- files.RNA

for(i in names(hmp.WGS)){  
  print(i) 
  print(dim(hmp.WGS[[i]]))
  print(dim(hmp.RNA[[i]]))
  print(dim(common_WGS_RNA_2[[i]]))
} # the overlapped SNPs between WGS and RNAseq is very few... 

WGS_sum <- 
sapply(names(hmp.WGS), function(i)
  nrow(hmp.WGS[[i]])) 

RNA_sum <- 
sapply(names(hmp.RNA), function(i)
  nrow(hmp.RNA[[i]])) 

overlap_sum <- 
sapply(names(common_WGS_RNA_2), function(i)
  nrow(common_WGS_RNA_2[[i]])) 

test <- rbind(WGS_sum, RNA_sum, overlap_sum) %>% as.data.frame()

test$type <- rownames(test)
test <- test %>% melt() 
# split to main & random chromosomes 
# main 
test.main <- test[!grepl("random", test$variable),]  
head(test.main) 

# make a plot to see how many SNPs on each chromosome 
pl.SNP.main <- ggplot(data = test.main) + 
  geom_bar(aes(x=factor(variable), y = value, fill=type), stat = "identity") +   
  facet_wrap(~type, ncol = 1) + 
  geom_text(aes(x=factor(variable), y = value, label=factor(value), size=1)) + 
  labs(list(title = "", x = "chromosome ID", y = "number of SNPs")) + 
  theme(legend.position = "none")

pl.SNP.main     

### just raw vcf file 
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/WGS/vcf_raw/reordered/hmp/" 
files.WGS <- list.files(path=path)   
setwd(path) 
hmp.WGS <- lapply(files.WGS, function(x) read.table(x, header = T)) 
hmp.WGS %>% length()
files.WGS <- gsub(".hmp.txt", "", files.WGS)
names(hmp.WGS) <- files.WGS

path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_combine_WGS_RNAseq/RNA_filtered/hmp" 
files.RNA <- list.files(path=path) 
setwd(path)
hmp.RNA <- lapply(files.RNA, function(x) read.table(x, header = T)) 
hmp.RNA %>% length() 
files.RNA <- gsub("_filtered.hmp.txt", "", files.RNA)
names(hmp.RNA) <- files.RNA

common_WGS_RNA <- 
lapply(names(hmp.RNA), function(i) 
  test <- hmp.WGS[[i]] %>% full_join(hmp.RNA[[i]], by = c("rs", "alleles")))

common_WGS_RNA[[13]] %>% dim() 
names(common_WGS_RNA) <- files.RNA

for(i in names(hmp.WGS)){
  print(i)
  print(dim(hmp.WGS[[i]]))
  print(dim(hmp.RNA[[i]]))
  print(dim(common_WGS_RNA[[i]]))
}

common_WGS_RNA_2 <- 
lapply(names(hmp.RNA), function(i) 
  test <- hmp.WGS[[i]] %>% inner_join(hmp.RNA[[i]], by = c("rs", "alleles")))

names(common_WGS_RNA_2) <- files.RNA

for(i in names(hmp.WGS)){
  print(i) 
  print(dim(hmp.WGS[[i]]))
  print(dim(hmp.RNA[[i]]))
  print(dim(common_WGS_RNA_2[[i]]))
} # the overlapped SNPs between WGS and RNAseq is very few... 

WGS_sum <- 
sapply(names(hmp.WGS), function(i)
  nrow(hmp.WGS[[i]])) 

RNA_sum <- 
sapply(names(hmp.RNA), function(i)
  nrow(hmp.RNA[[i]])) 

overlap_sum <- 
sapply(names(common_WGS_RNA_2), function(i)
  nrow(common_WGS_RNA_2[[i]])) 

test <- rbind(WGS_sum, RNA_sum, overlap_sum) %>% as.data.frame()

test$type <- rownames(test)
test <- test %>% melt() 
# split to main & random chromosomes 
# main 
test.main <- test[!grepl("random", test$variable),]  
head(test.main) 

# make a plot to see how many SNPs on each chromosome 
pl.SNP.main <- ggplot(data = test.main) + 
  geom_bar(aes(x=factor(variable), y = value, fill=type), stat = "identity") +   
  facet_wrap(~type, ncol = 1) + 
  geom_text(aes(x=factor(variable), y = value, label=factor(value), size=1)) + 
  labs(list(title = "", x = "chromosome ID", y = "number of SNPs")) + 
  theme(legend.position = "none")

pl.SNP.main   

``` 

### trouble shooting: why is there just small overlap between SNP called from RNAseq and WGS 
1) inlude population level SNP calling by excluding --report-genotype-likelihood-max but including --no-population-priors 
/share/malooflab/Ruijuan/KIAT_cabernet/505/WGS_RNA/Freebayes_parallel_Array_WGS.slurm
2) force SNP calling on all sites  
/share/malooflab/Ruijuan/KIAT_cabernet/505/WGS_RNA/Freebayes_parallel_Array_WGS_all.slurm 

tested chrA10, and the conclusion was that SNP calling from low coverage data might still has problem, because polymorphic sites might be underestimated (many polymorphic sites were not called in WGS data compared to RNAseq data)... 
select C05, A08, and C09 to see whether the result applys to different chroms 
compress the file uisng gzip --> transfer to whitney   
vcf to hmp 
format hmp 

```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/WGS/vcf_raw_pop/hmp/" 
files.WGS <- list.files(path=path)  
setwd(path)  
hmp.WGS <- lapply(files.WGS, function(x) read.table(x, header = T)) 
hmp.WGS %>% length()
files.WGS <- gsub("_population.hmp.txt", "", files.WGS)
names(hmp.WGS) <- files.WGS 

# when pop info was incorreclty used 
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/WGS/vcf_raw/reordered/hmp/" 
files.WGS <- list.files(path=path)  
files.WGS <- "chrA10.hmp.txt" 
setwd(path) 
hmp.WGS.2 <- lapply(files.WGS, function(x) read.table(x, header = T)) 
hmp.WGS.2 %>% length()
files.WGS <- gsub(".hmp.txt", "", files.WGS)
names(hmp.WGS.2) <- files.WGS

# compare 
common_WGS <- 
lapply(names(hmp.WGS.2), function(i) 
  test <- hmp.WGS[[i]] %>% anti_join(hmp.WGS.2[[i]]))

names(common_WGS) <- files.WGS

for(i in names(hmp.WGS)){
  print(i) 
  print(dim(hmp.WGS[[i]]))
  print(dim(hmp.WGS.2[[i]]))
  print(dim(common_WGS[[i]]))
} # the overlapped SNPs between WGS and RNAseq is very few... 

# when data was filtered from the wrong file (pop info incorrectly used)
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/WGS/vcf_raw/reordered/hmp"
files.WGS <- list.files(path=path)  
files.WGS <- c("chrA08.hmp.txt", "chrC05.hmp.txt", "chrC09.hmp.txt")
setwd(path) 
hmp.WGS.3 <- lapply(files.WGS, function(x) read.table(x, header = T, stringsAsFactors = F)) 
hmp.WGS.3 %>% length()
files.WGS <- gsub(".hmp.txt", "", files.WGS)
names(hmp.WGS.3) <- files.WGS 

# compare, were the over 10K SNP unique to incorrect calls left after filtering?  
common_WGS_2 <- 
lapply(names(hmp.WGS.3), function(i) 
  test <- common_WGS[[i]] %>% semi_join(hmp.WGS.3[[i]], by = c("rs", "alleles"))) # only >200 SNPs were kept, which means these over 200 calls could be genotyping error due to incorreclty not using population info 

names(common_WGS_2) <- files.WGS

for(i in names(hmp.WGS)){ 
  print(i) 
  print(dim(hmp.WGS[[i]])) 
  print(dim(hmp.WGS.3[[i]])) 
  print(dim(common_WGS_2[[i]])) 
}   

# when pop info was incorreclty used 
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/WGS_RNA/hmp/" 
files.WGS <- list.files(path=path)  
files.WGS <- c("chrA08.hmp.txt", "chrC05.hmp.txt", "chrC09.hmp.txt")
setwd(path) 
hmp.WGS.4 <- lapply(files.WGS, function(x) read.table(x, header = T, stringsAsFactors = F)) 
hmp.WGS.4 %>% length()
files.WGS <- gsub(".hmp.txt", "", files.WGS)
names(hmp.WGS.4) <- files.WGS
hmp.WGS.4[[1]] %>% dim()  

# common between WGS & RNA   
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_combine_WGS_RNAseq/RNA_filtered/hmp"
files.RNA <- list.files(path=path) 
files.RNA <- c("chrA08_filtered.hmp.txt", "chrC05_filtered.hmp.txt", "chrC09_filtered.hmp.txt")
setwd(path)
hmp.RNA <- lapply(files.RNA, function(x) read.table(x, header = T, stringsAsFactors = F)) 
hmp.RNA %>% length() 
files.RNA <- gsub("_filtered.hmp.txt", "", files.RNA)
names(hmp.RNA) <- files.RNA

common_WGS_RNA_2 <- 
lapply(names(hmp.RNA), function(i) 
  test <- hmp.WGS.4[[i]] %>% semi_join(hmp.RNA[[i]], by = c("rs"))) # called in RNAseq 

names(common_WGS_RNA_2) <- files.RNA 

common_WGS_RNA_2[[1]][1:10, 1:10] 

common_WGS_RNA_3 <- 
lapply(names(hmp.RNA), function(i) 
  test <- hmp.WGS.3[[i]] %>% semi_join(hmp.RNA[[i]], by = c("rs", "alleles"))) # called in both RNAseq and WGS 

names(common_WGS_RNA_3) <- files.WGS

for(i in names(hmp.WGS.4)){ 
  print(i)
  print(dim(hmp.WGS.3[[i]])) # raw SNP from WGS w/o filtering 
  print(dim(hmp.WGS.4[[i]])) # "SNP" from WGS including monophophic sites
  print(dim(hmp.RNA[[i]])) # SNP from RNAseq data w/ filtering 
  print(dim(common_WGS_RNA_2[[i]])) # SNPs called in RNAseq data and also have coverage in WGS data  
  print(dim(common_WGS_RNA_3[[i]])) # SNPs called in both RNAseq and WGS data 
} 

RNA_unique_WGS_absent <- 
  lapply(names(common_WGS_RNA_3), function(i) 
    common_WGS_RNA_2[[i]] %>% anti_join(common_WGS_RNA_3[[i]], by = c("rs")))
  
dim(RNA_unique_WGS_absent[[1]]); dim(RNA_unique_WGS_absent[[2]]); dim(RNA_unique_WGS_absent[[3]])

RNA_unique_WGS_absent[[1]]$chrom <- gsub("chr", "", RNA_unique_WGS_absent[[1]]$chrom)
RNA_unique_WGS_absent[[2]]$chrom <- gsub("chr", "", RNA_unique_WGS_absent[[2]]$chrom)
RNA_unique_WGS_absent[[3]]$chrom <- gsub("chr", "", RNA_unique_WGS_absent[[3]]$chrom)

RNA_unique_WGS_absent[[1]][1:10, 1:10] 
names(RNA_unique_WGS_absent) <- files.WGS

# MAF, depth, GQ, and missing rate for these sites 
# MAF 
# cat chrA08.vcf.frq | awk '{print $1 "\t" $2 "\t" $3}' > chrA08.vcf.allele
# cat chrC05.vcf.frq | awk '{print $1 "\t" $2 "\t" $3}' > chrC05.vcf.allele
# cat chrC09.vcf.frq | awk '{print $1 "\t" $2 "\t" $3}' > chrC09.vcf.allele

QUAL.tmp.tmp <- read.delim("~/505/WGS_RNA/chrA10_population_all.lqual")
MAF.tmp <- QUAL.tmp.tmp %>% left_join(MAF.tmp, c("CHROM" = "V1", "POS" = "V2")) %>% dplyr::select(CHROM, POS, MAF_minor) 
MAF.tmp$MAF_minor <- ifelse(is.na(MAF.tmp$MAF_minor), 0, MAF.tmp$MAF_minor)
MAF.tmp %>% head()

MAF.tmp <- MAF.tmp %>% semi_join(RNA_unique_WGS_absent, c("CHROM" = "chrom", "POS" = "pos")) 

hist(as.numeric(MAF.tmp$MAF_minor), breaks = 100)  

# QUAL score
QUAL.tmp <- read.delim("~/505/WGS_RNA/chrA10_population_all.lqual")
QUAL.tmp <- QUAL.tmp %>% semi_join(RNA_unique_WGS_absent, c("CHROM" = "chrom", "POS" = "pos")) 
hist(log10(QUAL.tmp$QUAL), breaks = 100) 

# mean read depth across individuals  
QUAL.tmp <- read.delim("~/505/WGS_RNA/chrA10_population_all.lqual")
QUAL.tmp <- QUAL.tmp %>% semi_join(RNA_unique_WGS_absent, c("CHROM" = "chrom", "POS" = "pos")) 

read_depth.tmp <- read.delim("~/505/WGS_RNA/chrA10_population_all.ldepth.mean")
read_depth.tmp <- read_depth.tmp %>% semi_join(RNA_unique_WGS_absent, c("CHROM" = "chrom", "POS" = "pos")) 
dim(read_depth.tmp)
mean(read_depth.tmp$MEAN_DEPTH) # 3.56 
hist(read_depth.tmp$MEAN_DEPTH, breaks = 100)

# missing rate 
# across taxa, on site level   
lmiss.tmp <- read.delim("~/505/WGS_RNA/chrA10_population_all.lmiss")
lmiss.tmp <- lmiss.tmp %>% semi_join(RNA_unique_WGS_absent, c("CHR" = "chrom", "POS" = "pos")) 
hist(lmiss.tmp$F_MISS)  
dim(lmiss.tmp)

### Tomorrow check MAF and see maybe these are 
QUAL.tmp %>% dim()
read_depth.tmp %>% dim()    
lmiss.tmp %>% dim()

all <- 
QUAL.tmp %>% left_join(read_depth.tmp) %>% left_join(MAF.tmp) %>% 
  left_join(lmiss.tmp, c("CHROM" = "CHR", "POS" = "POS")) 

all %>% 
  filter(QUAL > 40 & MEAN_DEPTH > 1 & F_MISS < 0.5) %>% 
  dim()

str(all)   

log10(all$QUAL) %>% hist(breaks = 5000) 

all %>% 
  filter(QUAL < 40) %>% 
  dplyr::select(MAF_minor) %>% 
  ggplot() +
  geom_histogram(aes(x = MAF_minor), binwidth = 0.001) 

all %>% 
  filter(QUAL < 40) %>% 
  filter(MAF_minor != 0) 

tmp <- 
all %>% 
  filter(QUAL < 40) %>% 
  dplyr::select(MAF_minor) %>% unlist() %>% as.numeric() 

MAF.tmp.RNA <- read.table("~/505/vcf_combine_WGS_RNAseq/RNA_filtered/MAF")

MAF.tmp.RNA <- 
MAF.tmp.RNA %>% 
  mutate(V7 = gsub("([[:print:]]+)(:)([[:print:]]+)", "\\3", V5), 
         V8 = gsub("([[:print:]]+)(:)([[:print:]]+)", "\\3", V6)) %>% 
  dplyr::select(V1, V2, V7, V8) 

MAF.tmp.RNA$MAF_minor <- as.numeric(apply(MAF.tmp.RNA[,c("V7", "V8")], 1, min)) 
MAF.tmp.RNA$V1 <- paste("chr", MAF.tmp.RNA$V1, sep = "")

MAF.tmp.RNA %>% semi_join(all, c("V1" = "CHROM", "V2" = "POS")) %>% 
  dplyr::select(MAF_minor) %>%  
  ggplot() +
  geom_histogram(aes(x = MAF_minor), binwidth = 0.01)  
``` 

SNP calling from low coverage data is still a problem here, which caused the missing alternate allele: most of the SNPs that were called from RNAseq data are monomorphic in WGS data.  

### trouble shooting  ######################################################################    

1) because the het ratio is too high for WGS data, we are wondering whether this is due to the mismapping problem, where reads from one sub genome is mapped to the other sub genome, and low coverage data additionaly caused the mis-genotyping. To test this, I decided to use STAR for mapping, because we used STAR for RNAseq mapping and did not observe the same problem. BWA, as reported, is the best available tool that could be used for DNA mapping. 

/share/malooflab/Ruijuan/KIAT_cabernet/505/mapping_array_WGS_STAR_2.slurm  
/share/malooflab/Ruijuan/KIAT_cabernet/505/Prep4Freebayes.slurm
/share/malooflab/Ruijuan/KIAT_cabernet/505/Bam_Split_By_Chrom.slurm  
```{r}
# star_extract_stats.sh get star mapping stats  
mapping_result <- read.table("~/505/WGS/mapping/Star_Stats_176.tab", header = T)
head(mapping_result)

# look at HQ, mapped, multimapped in detail for each sample (memory too low, cannot view histogram of total reads...)
mapping_result_505 <- mapping_result
colnames(mapping_result)
mapping_result_505_sub <- data.frame(Sample.ID = mapping_result_505$Sample, 
                                    unmapped = mapping_result$Number_Input_Reads * (as.numeric(mapping_result_505$Percent_Unmapped_Mismatches + mapping_result_505$Percent_Unmapped_Other + mapping_result_505$Percent_Unmapped_Too_Short)/100), 
                                    multi_too_many_mapped = mapping_result_505$Number_Multi_Mapped + mapping_result_505$Number_Too_Many_Multi_Mapped, 
                                    unqiuemapped = mapping_result_505$Number_Unique_Mapped
) 

mapping_result_505_sub 
dim(mapping_result_505_sub) # 85 4  

library(reshape2)
mapping_result_505_sub.long <- melt(mapping_result_505_sub) 

pl.mapping.indi.c <- ggplot(data = mapping_result_505_sub.long)
pl.mapping.indi.c <- pl.mapping.indi.c + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.c <- pl.mapping.indi.c + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_text(angle = 90, size = 8)) + ylim(0, 2e+7) 
pl.mapping.indi.c 

# histogram of unqiue mapped reads 
p.uniq.mapping <- ggplot(data=mapping_result_505)
p.uniq.mapping <- p.uniq.mapping + geom_histogram(aes(x=Percent_Unique_Mapped), binwidth = 2.5) 
p.uniq.mapping <- p.uniq.mapping + labs(list(title = "histogram of uniquely mapped reads percentage", x = "Percent of uniquely mapped reads", y = "number of samples")) 
p.uniq.mapping   

# histogram of unmapped  
p.mapping <- ggplot(data=mapping_result_505)
p.mapping <- p.un.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Unique_Mapped+mapping_result_505$Percent_Multi_Mapped+mapping_result_505$Percent_Too_Many_Multi_Mapped), binwidth = 2.5) 
p.mapping <- p.un.mapping + labs(list(title = "histogram of mapped reads percentage", x = "Percent of mapped reads", y = "number of samples")) 
p.mapping   

# compare with BWA mapping result, total mapping ratio and unique mapping ratio 
load("~/505/WGS/mapping/WGS_stats.Rdata")
mapping_result_505$Sample <- gsub("_S.*", "", mapping_result_505$Sample) 
  
WGS_stats$Sample
mapping_result_505$Sample

WGS_stats %>% dim(); mapping_result_505 %>% dim()

mapping_result_compare <- 
mapping_result_505 %>% 
  mutate(Percent_Mapped = Percent_Unique_Mapped+Percent_Multi_Mapped+Percent_Too_Many_Multi_Mapped) %>%
  left_join(WGS_stats, by = "Sample") %>% 
  dplyr::select(Sample, Percent_Mapped, Percent_Unique_Mapped, Pct_Mapped, Pct_Unique_Mapped) %>% 
  melt() 
 
grep("Pct", mapping_result_compare$variable) # BWA
mapping_result_compare$value[1:350] 
mapping_result_compare$value[351:700] <- mapping_result_compare$value[351:700] * 100 

mapping_result_compare$mapper <- c(rep("STAR", 350), rep("BWA", 350)) 

mapping_result_compare$variable <- gsub("Percent", "Pct", mapping_result_compare$variable)
mapping_result_compare$variable %>% unique()

P.mapping_result_compare <- 
mapping_result_compare %>% 
  ggplot() + 
  geom_bar(aes(x=Sample, y = value, fill=mapper), stat = "identity", position = position_dodge()) + 
  facet_wrap(~variable, ncol = 1) + 
  theme_gray()  + 
  labs(list(title = "", x = "individual sample", y = "")) + theme(axis.text.x=element_text(angle = 90, size = 8)) 

ggsave(P.mapping_result_compare, filename = "~/505/WGS/mapping/p.mapping_result_compare.png", width = 20, height = 6)

# check reads number 
# star_extract_stats.sh get star mapping stats  
mapping_result <- read.table("~/505/WGS/mapping/Star_Stats_176.tab", header = T)
head(mapping_result)

# look at HQ, mapped, multimapped in detail for each sample (memory too low, cannot view histogram of total reads...)
mapping_result_505 <- mapping_result
colnames(mapping_result)
mapping_result_505_sub <- data.frame(Sample.ID = mapping_result_505$Sample, 
                                    unmapped = mapping_result$Number_Input_Reads * (as.numeric(mapping_result_505$Percent_Unmapped_Mismatches + mapping_result_505$Percent_Unmapped_Other + mapping_result_505$Percent_Unmapped_Too_Short)/100), 
                                    multi_too_many_mapped = mapping_result_505$Number_Multi_Mapped + mapping_result_505$Number_Too_Many_Multi_Mapped, 
                                    unqiuemapped = mapping_result_505$Number_Unique_Mapped
) 

mapping_result_505_sub %>% head()
dim(mapping_result_505_sub) # 85 4  

library(reshape2)
mapping_result_505_sub.long <- melt(mapping_result_505_sub) 

# compare with BWA mapping result, total mapping ratio and unique mapping ratio 
load("~/505/WGS/mapping/WGS_stats.Rdata")
mapping_result_505$Sample <- gsub("_S.*", "", mapping_result_505$Sample) 
  
WGS_stats$Sample
mapping_result_505$Sample

WGS_stats %>% dim(); mapping_result_505 %>% dim()  

mapping_result_compare.2 <- 
mapping_result_505 %>% 
  left_join(WGS_stats, by = "Sample") %>%
  dplyr::select(Sample, Number_Unique_Mapped.x, Number_Unique_Mapped.y) %>% 
  melt() 
 
grep("x", mapping_result_compare.2$variable) # BWA
mapping_result_compare.2$value[1:175] <- mapping_result_compare.2$value[1:175] * 2

mapping_result_compare.2$mapper <- c(rep("STAR", 175), rep("BWA", 175)) 

P.mapping_result_compare.2 <- 
mapping_result_compare.2 %>% 
  ggplot() + 
  geom_bar(aes(x=Sample, y = value, fill=mapper), stat = "identity", position = position_dodge()) + 
  # facet_wrap(~variable, ncol = 1) + 
  theme_gray()  + 
  labs(list(title = "", x = "individual sample", y = "")) + theme(axis.text.x=element_text(angle = 90, size = 8)) 

ggsave(P.mapping_result_compare.2, filename = "~/505/WGS/mapping/p.mapping_result_compare.2.png", width = 20, height = 6) 
```

### SNP calling on STAR mapped reads 
/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_parallel_Array_WGS_STAR.slurm, check chrA08 
transfer to whitney --> gzip  exclude (--no-population-priors --report-genotype-likelihood-max)  
filter based on read depth, missing rate, QUAL score, etc, then calc het ratio 
convert vcf to hmp --> export to tassel to calculate het ratio for each site 
```{r}
# MAF 0.01
het_STAR <- read.delim("~/check_low_coverage/bioinformatics/STAR/het_STAR.txt")
het_STAR %>% head() 
hist(het_STAR$Proportion.Heterozygous)  

# MAF 0.05
het_STAR_A08 <- read.delim("~/check_low_coverage/bioinformatics/STAR/MAF_0.05/hmp/site_A08.txt")
hist(het_STAR_A08$Proportion.Heterozygous) 

taxa_STAR_A08 <- read.delim("~/check_low_coverage/bioinformatics/STAR/MAF_0.05/hmp/taxa_A08.txt")
hist(taxa_STAR_A08$Proportion.Heterozygous)  

# remove sites with het ratio greater than 0.2
SNP.ID.het.filterd.0.2.RNA <-  
het_STAR_A08 %>% 
  filter(Proportion.Heterozygous < 0.2) %>%
  dplyr::select(Chromosome, Physical.Position)

SNP.ID.het.filterd.0.2.RNA %>% dim()

SNP.ID.het.filterd.0.2.RNA$Chromosome <- paste("chr", SNP.ID.het.filterd.0.2.RNA$Chromosome, sep="")
SNP.ID.het.filterd.0.2.RNA$Chromosome <- gsub("NN", "nn", SNP.ID.het.filterd.0.2.RNA$Chromosome)
SNP.ID.het.filterd.0.2.RNA$Chromosome <- gsub("RANDOM", "random", SNP.ID.het.filterd.0.2.RNA$Chromosome)

write.table(SNP.ID.het.filterd.0.2.RNA, file= "~/check_low_coverage/bioinformatics/STAR/hmp/SNP.ID.het.filterd.0.2.txt")

# cat SNP.ID.het.filterd.0.2.txt | sed 's/"//g' | grep -v "Chromosome" | awk '{print $2 "\t" $3}' > SNP.ID.het.filterd.0.2  

# vcftools --positions  SNP.ID.het.filterd.0.2 --gzvcf chrA08_filtered.vcf.gz --recode --recode-INFO-all --out chrA08_filtered_het_0.2    

# gzip chrA08_filtered_het_0.2.vcf

# vcf to hmp --> format hmp 
# GWAS, no signal for the 1st run, maybe because the right samples were not included... 
```

2) The mis-usage of Freebayes might caused the problem, so re-call SNPs using the correct parameter and check result exclude (--no-population-priors --report-genotype-likelihood-max)  
/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_parallel_Array_WGS_BWA.slurm, check chrA08
transfer to whitney --> gzip  
filter based on read depth, missing rate, QUAL score, etc, then calc het ratio 
convert vcf to hmp --> export to tassel to calculate het ratio for each site  
```{r}
# MAF 0.01
het_BWA <- read.delim("~/check_low_coverage/bioinformatics/BWA/het_BWA.txt")
het_BWA %>% head()
hist(het_BWA$Proportion.Heterozygous) 

het_BWA_C03 <- read.delim("~/check_low_coverage/bioinformatics/BWA/site_C03.txt")
het_BWA_C03 %>% head()
hist(het_BWA_C03$Proportion.Heterozygous) 

taxa_BWA_C03 <- read.delim("~/check_low_coverage/bioinformatics/BWA/taxa_C03.txt")
taxa_BWA_C03 %>% head()
hist(taxa_BWA_C03$Proportion.Heterozygous)      

# MAF 0.05
het_BWA <- read.delim("~/check_low_coverage/bioinformatics/BWA/MAF_0.05/hmp/site_A08.txt")
het_BWA %>% head()
hist(het_BWA$Proportion.Heterozygous) 

het_BWA_C03 <- read.delim("~/check_low_coverage/bioinformatics/BWA/MAF_0.05/hmp/site_C03.txt")
het_BWA_C03 %>% head()
hist(het_BWA_C03$Proportion.Heterozygous) 

taxa_BWA_C03 <- read.delim("~/check_low_coverage/bioinformatics/BWA/MAF_0.05/hmp/taxa_C03.txt")
taxa_BWA_C03 %>% head()
hist(taxa_BWA_C03$Proportion.Heterozygous)      

taxa_BWA_A08 <- read.delim("~/check_low_coverage/bioinformatics/BWA/MAF_0.05/hmp/taxa_A08.txt")
taxa_BWA_A08 %>% head()
hist(taxa_BWA_A08$Proportion.Heterozygous)  

# do GWAS test on C03 becuause A08 only finished part run 
SNP.ID.het.filterd.0.2.RNA <-  
het_BWA_C03 %>% 
  filter(Proportion.Heterozygous < 0.2) %>%
  dplyr::select(Chromosome, Physical.Position)

SNP.ID.het.filterd.0.2.RNA %>% dim()

SNP.ID.het.filterd.0.2.RNA$Chromosome <- paste("chr", SNP.ID.het.filterd.0.2.RNA$Chromosome, sep="")
SNP.ID.het.filterd.0.2.RNA$Chromosome <- gsub("NN", "nn", SNP.ID.het.filterd.0.2.RNA$Chromosome)
SNP.ID.het.filterd.0.2.RNA$Chromosome <- gsub("RANDOM", "random", SNP.ID.het.filterd.0.2.RNA$Chromosome)

write.table(SNP.ID.het.filterd.0.2.RNA, file= "~/check_low_coverage/bioinformatics/BWA/MAF_0.05/hmp/SNP.ID.het.filterd.0.2.txt")

# cat SNP.ID.het.filterd.0.2.txt | sed 's/"//g' | grep -v "Chromosome" | awk '{print $2 "\t" $3}' > SNP.ID.het.filterd.0.2  

# vcftools --positions SNP.ID.het.filterd.0.2 --gzvcf chrA08_filtered.vcf.gz --recode --recode-INFO-all --out chrA08_filtered_het_0.2    

# gzip chrC03_filtered_het_0.2.vcf 
# vcf to hmp --> format hmp 
# GWAS 
```

3) We also checked the het ratio across taxa, there are some of them with low het ratio, but that's because they have a high SNP missing rate. 

We also checked the het ratio across taxa, there are some of them with low het ratio, but that's because they have a high SNP missing rate. 
```{r}
het_taxa <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/WGS/het_taxa.txt", stringsAsFactors = F)

png("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/WGS/het_taxa.png", width=6, height=6, units="in", res=300)
par(mfrow=c(3,1)) 
hist(het_taxa$Proportion.Heterozygous, breaks = 30, main = "proportion of heterozygous SNPs per line")
hist(het_taxa$Proportion.Missing, breaks = 30, main = "proportion of missing SNPs per line") 
plot(het_taxa$Proportion.Missing, het_taxa$Proportion.Heterozygous) 
dev.off()     

het_A10 <- read.delim("~/Desktop/het_chrA10_pop.txt", stringsAsFactors = F)
het_A10$Proportion.Heterozygous %>% hist()  

het_A10_nopop <- read.delim("~/Desktop/het_chrA10_nopop.txt", stringsAsFactors = F)
het_A10_nopop$Proportion.Heterozygous %>% hist() 

het_A10 <- read.delim("~/Desktop/het_chrA10_pop.txt", stringsAsFactors = F)
het_A10$Proportion.Heterozygous %>% hist()  
```

### from Biological perspective

Julin was worried that there could be pollen/seed contamination in the material, so he would like to do PCR for a couple of SNPs, see whether they are heterzygous or not. I need to generate a table for SNP to include: 

chrom, pos, quality, missing rate, allele ratio, het ratio, depth, and combine that with genotyping info for each sample 

```{r}
# function to work with vcf file 
reform.vcf <- function(temp){
  vcfbi <- is.biallelic(temp) # return with logics indicating whehter biallelic or not... 
  vcfref <- subset(getREF(temp), subset = vcfbi) # get ref allele
  vcfalt <- subset(getALT(temp), subset = vcfbi) # get alt allele
  vcfchrom <- subset(getCHROM(temp), subset = vcfbi) # get chromosome info 
  vcfpos <- subset(getPOS(temp), subset = vcfbi) # get pos info 
  vcfgts <- subset(extract.gt(temp, element = "GT", IDtoRowNames = F), subset = vcfbi) # get gt
  vcfgq <- subset(extract.gt(temp, element="GQ", IDtoRowNames=F), subset=vcfbi) # gq for each 
  vcfdp <- subset(extract.gt(temp, element="DP", IDtoRowNames=F), subset=vcfbi) # depth for each
  vcfro <- subset(extract.gt(temp, element="RO", IDtoRowNames=F), subset=vcfbi) # ref count 
  vcfao <- subset(extract.gt(temp, element="AO", IDtoRowNames=F), subset=vcfbi) # alt count 
  
  temp2 <- data.frame(cbind(vcfchrom,vcfpos,vcfref,vcfalt,vcfgts,vcfgq,vcfdp,vcfro,vcfao))
  colnames(temp2)[1:4] <- c("CHROM","POS","REF","ALT")
  
  rnames <- rownames(temp2)
  temp2 <- data.frame(sapply(temp2, function(x) sub("0/0","-1",x)))
  temp2 <- data.frame(sapply(temp2, function(x) sub("0/1","0",x)))
  temp2 <- data.frame(sapply(temp2, function(x) sub("1/1","1",x)))
  row.names(temp2) <- rnames 
  
  return(temp2) 
} 

SNP_WGS_for_Julin <- read.vcfR("~/505/WGS/vcf_raw/filtered/505_filtered.vcf")
SNP_WGS_for_Julin.2 <- reform.vcf(SNP_WGS_for_Julin) 
save(SNP_WGS_for_Julin.2, file = "~/check_low_coverage/biological/SNP_GWAS_for_Julin.2.Rdata")

dim(SNP_WGS_for_Julin.2) # 1194 

# rename col to add more info
ncol_1 <- paste(gsub("\\.1*", "", colnames(SNP_WGS_for_Julin.2[243:480])), "GQ", sep = "_")
ncol_2 <- paste(gsub("\\.2*", "", colnames(SNP_WGS_for_Julin.2[481:718])), "depth", sep = "_")
ncol_3 <- paste(gsub("\\.3*", "", colnames(SNP_WGS_for_Julin.2[719:956])), "ref_count", sep = "_")
ncol_4 <- paste(gsub("\\.4*", "", colnames(SNP_WGS_for_Julin.2[957:1194])), "alt_count", sep = "_")

colnames(SNP_WGS_for_Julin.2)[243:1194] <- c(ncol_1, ncol_2, ncol_3, ncol_4)
  
SNP_WGS_for_Julin.2[1:10, 5:242] # gt
SNP_WGS_for_Julin.2[1:10, 243:480] # GQ
SNP_WGS_for_Julin.2[1:10, 481:718] # depth
SNP_WGS_for_Julin.2[1:10, 719:956] # ref count
SNP_WGS_for_Julin.2[1:10, 957:1194] # alt count 

SNP_WGS_for_Julin.2[1:10, 1:10]

# QUAL
QUAL <- read.delim("~/505/WGS/vcf_raw/filtered/505.lqual", stringsAsFactors = F)
str(QUAL)   
str(SNP_WGS_for_Julin.2)

SNP_WGS_for_Julin.3 <- 
SNP_WGS_for_Julin.2 %>% 
  mutate(CHROM = as.character(CHROM)) %>% 
  mutate(POS = as.numeric(as.character(POS))) %>% 
  left_join(QUAL) 

# missing rate
lmiss <- read.delim("~/505/WGS/vcf_raw/filtered/505.lmiss")
lmiss %>% head()
lmiss %>% str()

SNP_WGS_for_Julin.4 <- 
lmiss %>% 
  mutate(CHROM = as.character(CHR)) %>% 
  dplyr::select(CHROM, POS, F_MISS) %>% 
  left_join(SNP_WGS_for_Julin.3) 

# depth 
read_depth <- read.delim("~/505/WGS/vcf_raw/filtered/505.ldepth.mean")
read_depth %>% head()
read_depth %>% str()

SNP_WGS_for_Julin.5 <- 
read_depth %>% 
  mutate(CHROM = as.character(CHROM)) %>% 
  dplyr::select(CHROM, POS, MEAN_DEPTH) %>% 
  left_join(SNP_WGS_for_Julin.4) 

SNP_WGS_for_Julin.5[4, 721:1196] %>% as.matrix() %>% as.numeric() %>% sum(na.rm = T) / 238
SNP_WGS_for_Julin.5[4, 483:720] %>% as.matrix() %>% as.numeric() %>% sum(na.rm = T) / 238
SNP_WGS_for_Julin.5$MEAN_DEPTH[4] 

# gt ratio
temp <- SNP_WGS_for_Julin.5[,7:244] %>% as.matrix()

SNP_WGS_for_Julin.5$homo_alt <- 
sapply(1:nrow(SNP_WGS_for_Julin.5), function(i) 
{sum(temp[i,] %>% as.numeric() == 1, na.rm = T) # homo alt
})

SNP_WGS_for_Julin.5$homo_ref <- 
sapply(1:nrow(SNP_WGS_for_Julin.5), function(i) 
{sum(temp[i,] %>% as.numeric() == -1, na.rm = T) # homo ref 
})

SNP_WGS_for_Julin.5$het <- 
sapply(1:nrow(SNP_WGS_for_Julin.5), function(i) 
{sum(temp[i,] %>% as.numeric() == 0, na.rm = T) # het
}) 

taxa_BWA <- read.delim("~/Desktop/taxa_BWA.txt", stringsAsFactors = F)
taxa_BWA$Proportion.Heterozygous %>% hist(main = "BWA") 

taxa_STAR <- read.delim("~/Desktop/taxa_STAR.txt", stringsAsFactors = F)
taxa_STAR$Proportion.Heterozygous %>% hist(main = "STAR") 

taxa_A10 <- read.delim("~/Desktop/taxa_A10.txt", stringsAsFactors = F)
taxa_A10$Proportion.Heterozygous %>% hist(main = "STAR")  

### ref/alt ratio for those with homo_ref call, homo_alt call, or het call
load("~/check_low_coverage/biological/SNP_list/SNP_WGS_for_Julin.final.Rdata")

P.gt_dist <- 
SNP_WGS_for_Julin.final %>% 
  dplyr::select(homo_alt, homo_ref, het) %>% 
  melt() %>%
  ggplot() + 
  geom_histogram(aes(x = value, fill = variable)) + 
  facet_wrap(~variable) + 
  labs(x = "number of samples", y = "number of SNPs") 

ggsave(P.gt_dist, filename = "~/check_low_coverage/biological/p.yeild.png", width = 9, height = 4)

gt <- SNP_WGS_for_Julin.final[,7:244] %>% as.matrix() 
ref <- SNP_WGS_for_Julin.final[,721:958] 
alt <- SNP_WGS_for_Julin.final[,959:1196] 

gt_alt_loc <- data.frame(sapply(SNP_WGS_for_Julin.final, function(x) x == 1))[,7:244]
gt_ref_loc <- data.frame(sapply(SNP_WGS_for_Julin.final, function(x) x == -1))[,7:244]
gt_het_loc <- data.frame(sapply(SNP_WGS_for_Julin.final, function(x) x == 0))[,7:244]

SNP_WGS_for_Julin.final[,721:958]
colnames(SNP_WGS_for_Julin.final) 
data.frame(tmp1/tmp2)
tmp1 <- ref[1:10, 1:10]
tmp2 <- alt[1:10, 1:10] 
```

### still bioinformatics 

Me: This high het ratio reminds me of the problem I came across when I first started to work on this project: unexpected high het ratio for Da-Ae and Da-Ol-1. For that problem, we finally believed it was because of the mis-assembly of genes in the reference genome, so reads could be mapped to the wrong but genes with similar sequences. Here I think this could also be the problem, and since this time non genic region is included as well, so the problem could be more serious compared to RNAseq data. To test the hypothesis, I could try to map to John's reference genome.

Julin: Also you should map Da-Ae (or John's 10X Da-Ae) reads back to the Da-Ae reference and call SNPs there.  Any SNPs found in that way can be excluded. 

1) indexed genome for BWA mapping 
/share/malooflab/John/KIAT/MASTER_DATA_ANALYSIS/BWA/Assemblies/B.napus_Final.fa.masked with repeat regions masked 

2) use BWA for mapping
/share/malooflab/Ruijuan/KIAT_cabernet/505/mapping_array_WGS_cabernet_DaAe.slurm
/share/malooflab/Ruijuan/KIAT_cabernet/505/Prep4SNPcalling.slurm 

Da-Ae's 10x trimmed reads mapped to Da-Ae  
/share/malooflab/Ruijuan/KIAT_cabernet/505/mapping_array_WGS_cabernet_DaAe_DaAe.slurm 

### 2019 resolution 
* based on the result from recall SNPs using "the correct" options works, by checking the het ratio and GWAS result. Also Julin suggest include "--hwe-priors-off" to see whether it gives better result. So I tried SNP calling using tow OR three options. 

This time, every chrom was calculated for depth, and high depth regions were excluded (> 2500)  

1) compress & index
compress_index_vcf.sh
/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500_2019/compress_index_2.slurm

2) reorder 
/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500_2019/reorder.slurm

3) combine into one big vcf file
/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500_2019/concat.slurm
output: 505.vcf.gz  

6) filter  
/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500_2019/filter.slurm
output: 505.filtered.vcf.gz

### use tassel for heterozygosity calculation --> filter based on het ratio (>0.2) --> GWAS --> impute 
vcf_to_hmp.sh # find out no vcf_to_hmp is required, now tassel can accept vcf format 

```{r}
three_options <- read.delim("~/505/WGS/vcf_filtered/three_options/505_site_summary.txt")
three_options %>% head()
hist(three_options$Proportion.Heterozygous) 

two_options <- read.delim("~/505/WGS/vcf_filtered/two_options/505_site_summary_two_options.txt")
two_options %>% head()
hist(two_options$Proportion.Heterozygous) 

dim(three_options) # 360141     37
dim(two_options) # 1703835      37 
# many more SNPs found using two options, HW matters .... 

# filter based on het ratio 
SNP.ID.het.filterd.0.2 <-  
two_options %>% 
  filter(Proportion.Heterozygous < 0.2) %>%
  dplyr::select(Chromosome, Physical.Position)

SNP.ID.het.filterd.0.2 %>% dim() # 1529974      

SNP.ID.het.filterd.0.2$Chromosome <- paste("chr", SNP.ID.het.filterd.0.2$Chromosome, sep="")
SNP.ID.het.filterd.0.2$Chromosome <- gsub("NN", "nn", SNP.ID.het.filterd.0.2$Chromosome)
SNP.ID.het.filterd.0.2$Chromosome <- gsub("RANDOM", "random", SNP.ID.het.filterd.0.2$Chromosome)

write.table(SNP.ID.het.filterd.0.2, file= "~/505/WGS/vcf_filtered/two_options/SNP.ID.het.filterd.0.2.txt")

# cat SNP.ID.het.filterd.0.2.txt | sed 's/"//g' | grep -v "Chromosome" | awk '{print $2 "\t" $3}' > SNP.ID.het.filterd.0.2  

# vcftools --positions  SNP.ID.het.filterd.0.2 --gzvcf 505_filtered.vcf.gz --recode --recode-INFO-all --out 505_filtered_het_0.2    

# gzip 505_filtered_het_0.2.recode.vcf 

# vcf to hmp --> format hmp 
# GWAS 
```






