---
title: "505_WGS_pre_analysis"
author: "Ruijuan Li"
date: "8/20/2018"
output: html_document
---

### load libs
```{r}
library(tidyverse)
library(reshape2)
```

# sample submission for WGS 
```{r}
WGS_samples <- read.csv("~/505/WGS/sample/DNA_to_Library_Submission_Form_20180508_incorporated_all.csv", header = F)
WGS_samples <- 
t(WGS_samples) %>% as.data.frame() %>% 
  dplyr::select(V1:V12) 

WGS_samples$V2 %>% unlist() %>% as.character()

missing_sample <- c("K114", "SJ_280", "SJ_294")

geno <- read.vcfR("~/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.vcf.gz") 
geno.2 <- reform.vcf(geno) 
dim(geno.2) #  174397    135
geno.2 %>% colnames() 

RNAseq_sample <- gsub("X505.| X505_", "", colnames(geno.2))
RNAseq_sample <- gsub("_", "", RNAseq_sample)[5:135] 

sum(WGS_samples$V2 %in% RNAseq_sample)
sum(RNAseq_sample %in% WGS_samples$V2)

WGS_sample_only <- as.data.frame(t(WGS_samples[!(WGS_samples$V2 %in% c(RNAseq_sample, missing_sample)),])) 
write.csv(WGS_sample_only, file = "~/505/WGS/sample/WGS_sample_only.csv")

sum(WGS_sample_only[2,] %>% unlist() %>% as.character() %in% c(RNAseq_sample, missing_sample)) # correct 
as.data.frame(t(WGS_sample_only[c(1,2),])) %>% 
  left_join(WGS_samples[,c(1,2)]) %>% dim() 
```  

### download 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/download_array_WGS.slurm

### QC 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/QC.sh
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/trimming_array.slurm
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/QC_trimmed.sh
QC stats: 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/fastqc_extract_stats.sh 

```{r}
WGS_data <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/WGS_raw_data_info.csv")
WGS_data <- 
WGS_data %>% 
  dplyr::select(Sample, Yield..Mbases., X.....Q30, Mean.Quality)  

WGS_data <- WGS_data[2:239,] 
colnames(WGS_data) <- c("Sample", "Yield_in_Mb", "Q30_pct", "Mean_quality")
WGS_data$Yield_in_Mb <- gsub("\\,", "",as.character(WGS_data$Yield_in_Mb)) %>% as.numeric()
WGS_data$Q30_pct <- as.character(WGS_data$Q30_pct) %>% as.numeric()
WGS_data$Mean_quality <- as.character(WGS_data$Mean_quality) %>% as.numeric()

sapply(colnames(WGS_data)[2:4], function(i) {median(WGS_data[,i])})
sapply(colnames(WGS_data)[2:4], function(i) {min(WGS_data[,i])})
sapply(colnames(WGS_data)[2:4], function(i) {max(WGS_data[,i])})

# histogram based on reads number 
p.yeild <- 
WGS_data %>% 
  ggplot() + 
  geom_histogram(aes(x = Yield_in_Mb), binwidth = 100) + 
  ylab("Number of samples") 

p.yeild
ggsave(p.yeild, filename = "~/Desktop/p.yeild.png", width = 9, height = 4)
# read count generated by FastQC
read_stats <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats_238")
read_stats <- 
read_stats %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  group_by(Sample) %>% 
  summarise(Number_Reads = sum(Number_Reads)) 
dim(read_stats)

read_stats_HQ <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats_HQ_238")
read_stats_HQ <- 
read_stats_HQ %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  group_by(Sample) %>% 
  summarise(Number_Reads_HQ = sum(Number_Reads))  

# note sample with inconsistent yield report and fastqc read count 
stats_compare <- 
read_stats %>% 
  left_join(WGS_data) %>% 
  mutate(FastQC_calc_Mb = round(Number_Reads * 151 / 1000000, digits = 0)) %>% 
  dplyr::select(Sample, Yield_in_Mb, FastQC_calc_Mb, Number_Reads)

sum(stats_compare$Yield_in_Mb > stats_compare$FastQC_calc_Mb)
sum(stats_compare$Yield_in_Mb != stats_compare$FastQC_calc_Mb) 

stats_compare$note <- ifelse(stats_compare$Yield_in_Mb != stats_compare$FastQC_calc_Mb, "NO", "YES")

p.stats_compare <- 
stats_compare %>% 
  filter(note == "NO") %>% 
  mutate(From_DNA_center = Yield_in_Mb, From_FastQC = FastQC_calc_Mb) %>% 
  dplyr::select(Sample, From_DNA_center, From_FastQC) %>% 
  melt() %>% 
  ggplot() + 
  geom_col(aes(x = Sample, y = value, fill = variable)) + 
  theme(axis.text.x=element_blank()) + 
  ylab("Yield in Mb")

ggsave(p.stats_compare, filename = "~/Desktop/p.stats_compare.png", width = 9, height = 4)  
write.csv(stats_compare, file = "~/Desktop/stats_compare.csv") 

stats_compare <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/stats_compare.csv")
sample_list <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/sample_list")

Prob_samples <- 
stats_compare %>% 
  filter(note == "NO") %>% 
  dplyr::select(Sample) 

sample_list_47 <- 
sample_list %>% 
  mutate(Sample = gsub("_L003", "", V1)) %>% 
  mutate(Sample = gsub("_S.*", "", V1)) %>% 
  right_join(Prob_samples) %>% 
  dplyr::select(V1)

write.table(sample_list_47, file = "~/Desktop/sample_list_47")

# redownload 47 samples and did QC 
read_stats_47 <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats.tab")
read_stats_47 <- 
read_stats_47 %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  group_by(Sample) %>% 
  summarise(Number_Reads = sum(Number_Reads)) 
dim(read_stats_47)

read_stats_47_HQ <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats_HQ.tab")
read_stats_47_HQ <- 
read_stats_47_HQ %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  group_by(Sample) %>% 
  summarise(Number_Reads_HQ = sum(Number_Reads))  

stats_compare_47 <- 
read_stats_47 %>% 
  left_join(WGS_data) %>% 
  mutate(FastQC_calc_Mb = round(Number_Reads * 151 / 1000000, digits = 0)) %>% 
  dplyr::select(Sample, Yield_in_Mb, FastQC_calc_Mb, Number_Reads)

sum(stats_compare_47$Yield_in_Mb > stats_compare_47$FastQC_calc_Mb)
sum(stats_compare_47$Yield_in_Mb != stats_compare_47$FastQC_calc_Mb) 

read_stats_47 %>% 
  left_join(read_stats_47_HQ) %>% 
  mutate(HQ_pct = Number_Reads_HQ/Number_Reads) %>% 
  ggplot() + 
  geom_histogram(aes(x = HQ_pct))  

read_stats_summary_47 <- 
read_stats_47 %>% 
  left_join(read_stats_47_HQ) 

### for KIAT meeting 
read_stats_summary <- 
read_stats %>% 
  left_join(read_stats_HQ) 

read_stats <- 
read_stats_summary %>% 
  anti_join(read_stats_summary_47, by = "Sample") %>% 
  rbind(read_stats_summary_47) 

read_stats %>% head()  
```

### data import to SQL database 
```{r}
WGS_data <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/WGS_raw_data_info.csv")

WGS_data <- 
WGS_data %>% 
  mutate(reads = gsub(",", "", PF.Clusters)) %>%   
  dplyr::select(Sample, reads)  

read_stats <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/FastQC_Stats_238")
read_stats_SQL <- 
read_stats %>% 
  mutate(sample_ID = gsub("_S.*", "", Sample)) %>% 
  left_join(WGS_data, by = c("sample_ID" = "Sample")) 

SQL_sample_ID <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/database/sample_output.csv")

data_SQL <- 
read_stats_SQL %>% 
  left_join(SQL_sample_ID, by = c("sample_ID" = "Sample_name")) %>% 
  mutate(Sample = gsub("_fastqc", ".fastq.gz", Sample)) %>%
  dplyr::select(Sample, reads, Sample_id) 

write.csv(data_SQL, "~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/database/data_SQL.csv")
```

### mapping 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/mapping_array_WGS_cabernet.slurm 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/Prep4SNPcalling.slurm

mapping stats: 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/5055751.pts-164.barbera/mapping_stats.sh 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/bwa_extract_stats.sh

```{r}
library(reshape2)

mapping_stats <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/505/WGS/data/BWA_Stats_new.tab")

WGS_stats <- 
mapping_stats %>% 
  mutate(Sample = gsub("_L003", "", Sample)) %>% 
  mutate(Sample = gsub("_S.*", "", Sample)) %>% 
  left_join(read_stats) %>% 
  mutate(Pct_HQ = Number_Reads_HQ/Number_Reads, 
         Pct_Mapped = Number_Mapped/100, 
         Pct_Unique_Mapped = Number_Unique_Mapped/Number_Reads_HQ)

p.WGS_stats <-  
WGS_stats %>% 
  dplyr::select(starts_with("Pct")) %>% 
  melt() %>% 
  ggplot() + 
  geom_histogram(aes(x = value), binwidth = 0.01) + 
  facet_wrap(~variable, nrow = 3) + 
  ylab("Number of Samples") + xlab("")  

p.WGS_stats  

# note libs with low quailty, low mapping rate, and/or low unique mapping rate  
WGS_stats %>% 
  dplyr::select(Sample, starts_with("Pct")) %>% 
  filter(Pct_HQ < 0.85 | Pct_Mapped < 0.75 | Pct_Unique_Mapped < 0.4)   

ggsave(p.WGS_stats, filename = "~/Desktop/p.WGS_stats.png", width = 8, height = 6)  
```

### SNP calling 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/Bam_Split_By_Chrom.slurm 
Freebayes-parallel SNP calling: 
cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_parallel_Array_WGS.slurm

* 7 chroms require large memory, which mighe be due to extremely high coverage for some regions on those chrom. 

for these 7 chroms, run SNP calling on splitted regions 
```{r}
# for regions
# calculate read coverage: cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/calc_depth_subset_by_chr.slurm 

# sum read coverage at certain location: cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/depth_sum_$i.sh 

# cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500/split_region.slurm 

# for chroms: 
# merge bam file from 238 samples for these 7 chroms: actually no need to merge bam files, seperate bam file works as well 
# cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/merge_index_bam_for_Freebayes.slurm

# call SNPs 
# /share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500/Freebayes_parallel_Array_WGS_merged_region_i.slurm  
```

### vcf merge 

1) WGS: merge vcf for each chrom, use vcftool to calculate stats and filter, use tassel to convert to hapmap format 

merge and index vcf file for each chrom 
cluster:/share/malooflab/Ruijuan/KIAT_cabernet/505/Freebayes_script_last/coverage_2500/combine_vcf.sh 

reorder each vcf file and compress: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/reorder_vcf_etc.sh
concat all vcf files: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/concat.sh 

```{r}

```

2) RNAseq: compress and index vcf file for each chrom, use vcftool to filter, use tassel to convert to hapmap format 

vcf files at: cluster:/share/malooflab/Ruijuan/505/vcf_no_pop_late_silique/all
or Whitney: /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_131_sample 

filter: https://github.com/leejimmy93/KIAT/blob/master/505/filter_late_silique.sh 

vcf to hmp: /usr/local/stow/tassel-5-standalone/run_pipeline.pl -SortGenotypeFilePlugin -inputFile chrA10_filtered.vcf.gz -outputFile chrA10_filtered -fileType VCF
/usr/local/stow/tassel-5-standalone/run_pipeline.pl -Xmx5g -fork1 -vcf chrA10_filtered.vcf -export -exportType Hapmap -runfork1

```{r}

```


3) use join function to merge every chrom's hapmap file from WGS and RNAseq 

hmp files at whitney:/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_combine_WGS_RNAseq/

edit hmp file to remove "#" after rs & assembly  
```{r}

```

### test 
1) 

--> GWAS & population structure + genetic diveristy  
```{r}

```





