---
title: "WGS_GWAS"
output: html_document
---

Purpose of this script is to do test GWAS on just samples with WGS data, to see whether we get the same A08 and C03 peaks 

### pheno data for all  
```{r}
# K line
fatty_acid_oil_K <- read.csv("~/505/data/phenotype/505_fatty_acid_20171128.csv", stringsAsFactors = F)
colnames(fatty_acid_oil_K)   

fatty_acid_oil_K <- 
fatty_acid_oil_K %>% 
  dplyr::select(Line.number:Erucic.acid..C22.1n9.)    

colnames(fatty_acid_oil_K) <- c("SN", "Oil_content", "Caprylic_acid", "Capric_acid", "Lauric_acid", "Myristic_acid", "Pentadecanoic_acid", "Palmitic_acid", "Palmitoliec_aicd", "Heptadecanoic_acid", "Stearic_acid", "Oleic_acid", "vaccenic_acid", "Linoleic_acid", "Arachidic_acid", "cis_11_Eicosenoic_acid", "Linolenic_acid", "Behenic_acid", "Erucic_acid")

str(fatty_acid_oil_K)
fatty_acid_oil_K[1:10, 1:10]

# SJ line 
fatty_acid_oil_SJ <- read.csv("~/505/WGS/GWAS/pheno/FA_SJ_line.csv", stringsAsFactors = F, skip = 1)
str(fatty_acid_oil_SJ)  
colnames(fatty_acid_oil_SJ)  

fatty_acid_oil_SJ <- 
fatty_acid_oil_SJ %>% 
  dplyr::select(sowing_no, crude.oil:Erucic.acid..C22.1n9.)    
colnames(fatty_acid_oil_SJ)

colnames(fatty_acid_oil_SJ) <- c("SN", "Oil_content", "Caprylic_acid", "Capric_acid", "Lauric_acid", "Myristic_acid", "Pentadecanoic_acid", "Palmitic_acid", "Palmitoliec_aicd", "Heptadecanoic_acid", "Stearic_acid", "Oleic_acid", "vaccenic_acid", "Linoleic_acid", "Arachidic_acid", "cis_11_Eicosenoic_acid", "Linolenic_acid", "Behenic_acid", "Erucic_acid")

fatty_acid_oil_SJ[1:10, 1:10] 

# TC line 
fatty_acid_oil_TC <- read.csv("~/505/WGS/GWAS/pheno/FA_TC_line.csv", stringsAsFactors = F, skip = 1)
str(fatty_acid_oil_TC)  
colnames(fatty_acid_oil_TC)  

fatty_acid_oil_TC <- 
fatty_acid_oil_TC %>% 
  dplyr::select(Line.number:Erucic.acid..C22.1n9.)    
colnames(fatty_acid_oil_TC)

colnames(fatty_acid_oil_TC) <- c("SN", "Oil_content", "Caprylic_acid", "Capric_acid", "Lauric_acid", "Myristic_acid", "Pentadecanoic_acid", "Palmitic_acid", "Palmitoliec_aicd", "Heptadecanoic_acid", "Stearic_acid", "Oleic_acid", "vaccenic_acid", "Linoleic_acid", "Arachidic_acid", "cis_11_Eicosenoic_acid", "Linolenic_acid", "Behenic_acid", "Erucic_acid")

fatty_acid_oil_TC[1:10, 1:10] 

# combine 4 pheno files 
fatty_acid_oil_K_all <- 
rbind(fatty_acid_oil_K, fatty_acid_oil_SJ, fatty_acid_oil_TC) 

# just test Erucic and Oleic acid 
fatty_acid_oil_K_all_sub <- 
fatty_acid_oil_K_all %>% 
  dplyr::select(SN, Oleic_acid, Erucic_acid) 

fatty_acid_oil_K_all_sub <- fatty_acid_oil_K_all_sub[!is.na(fatty_acid_oil_K_all_sub$Oleic_acid),] 
colnames(fatty_acid_oil_K_all_sub)[1] <- "Taxa" 

# pheno data for 238 samples 
imiss <- read.delim("~/505/WGS/vcf_raw/combined/505.imiss") 
sample_238 <- gsub("_S.*", "", imiss$INDV) 

sum(!(sample_238 %in% fatty_acid_oil_K_all_sub$Taxa)) # 24 missing sample ID from fatty acid data

fatty_acid_oil_K_all_sub <- 
fatty_acid_oil_K_all_sub[fatty_acid_oil_K_all_sub$Taxa %in% sample_238,] 

fatty_acid_oil_K_all_sub %>% dim()
sample_238 <- sample_238 %>% as.data.frame() 
colnames(sample_238) <- "Taxa"
fatty_acid_oil_K_all_sub <- sample_238 %>% left_join(fatty_acid_oil_K_all_sub) %>% unique()  
dim(fatty_acid_oil_K_all_sub) 
summary(fatty_acid_oil_K_all_sub)

save(fatty_acid_oil_K_all_sub, file="~/505/WGS/GWAS/fatty_acid_oil_K_all_sub.Rdata")  
```

### genotype data 
filter: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/filter.sh 
vcf to hmp: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/vcf_to_hmp.sh
format hmp file: https://github.com/leejimmy93/KIAT/blob/master/505/WGS/format_hmp_2.txt 
test GWAS: 
```{r}
geno_505_hmp <- read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/WGS/vcf_raw/filtered/hmp/505.hmp.txt", head=FALSE, stringsAsFactors = F)
# get chrom name to numerics
geno_505_hmp$V3 <- as.numeric(geno_505_hmp$V3)
geno_505_hmp[1,3] <- "chrom"

geno_505_hmp[1,] <- gsub("_S.*", "", unlist(geno_505_hmp[1,])) 
```

