Title: Allelic specific expression 
========================================================
### call SNPs between 414 & 415 F1 population 
```{r} 
# 1) trim & map fastq file using script trimming_mapping_01_23_2017.sh 

# 2) merge bam files from biological replicates 

# 3) extract uniquely mapped file, sort, and convert to bam file 

# 4) call SNPs for each tissue type seperately, include Da-Ae & Da-Ol 

# 5) filter to remove indel & keep biallelic SNPs only
# https://github.com/leejimmy93/KIAT/blob/master/F1/filter_SNP_biallelic_QUAL.sh
```

### load libraries 
```{r}
library(vcfR)
source("~/Desktop/Brassica_project/KIAT_RNA_seq/analysis/F1/helpler.R")  
```

### examine F1 dataset to study gene imprinting effect
### import data
```{r}
# ./process_vcf.sh input.vcf input_modified.vcf # to make the right format of vcf file 
F1.young <- read.vcfR("~/F1_SNP/vcf/with_GQ/F1_young_filtered.recode.vcf")  
F1.young.reform <- reform.vcf.F1(temp=F1.young)
dim(F1.young.reform) # 1247960      24
save(F1.young.refom, file="~/F1_SNP/output/F1.young.reform.Rdata")
```

### filter based on missing data (no missing data), GQ (>30 in each sample), DP (>10 in each sample)  
```{r} 
load("~/F1_SNP/output/F1.young.reform.Rdata") 

# 1) filter based on missing data 
F1.young.nomissing <- F1.young.reform[complete.cases(F1.young.reform),]
dim(F1.young.nomissing) # 635218     24 

# 2) transform filter columns to numeric values 
F1.young.nomissing[,(5:24)] <- lapply(F1.young.nomissing[,(5:24)], function(x) as.numeric(as.character(x))) 

# 3) filter based on GQ  
F1.young.GQ.filtered <- GQ.filter(vcf=F1.young.nomissing, n=30)
dim(F1.young.GQ.filtered) # 288085     24  

hist(as.numeric(as.character(F1.young.GQ.filtered$F1_415_GQ)))

# 4) filter based depth 
F1.young.DP.filtered <- DP.filter(vcf=F1.young.GQ.filtered, n=10)
dim(F1.young.DP.filtered) # 166691     24 

# 5) filter based on gt (homozygous in parents)
F1.young.GQ.filtered <- subset(F1.young.DP.filtered, (((Ae_GT=="-1" & Ol_GT=="1")) | ((Ae_GT=="1" & Ol_GT=="-1"))))
dim(F1.young.GQ.filtered) # 49140    24 
F1.young.GQ.filtered[1:10,]
save(F1.young.GQ.filtered, file = "~/F1_SNP/output/F1.young.GQ.filtered.Rdata") 
```

### plot to check read depth, genotye quality, and genotype of two F1s 
```{r}
library(gridExtra)
library(grid)
library(ggplot2)
library(lattice)

load("~/F1_SNP/output/F1.young.GQ.filtered.Rdata")
## check genotypes of two F1s
p.414 <- ggplot(data = F1.young.GQ.filtered) + geom_histogram(aes(x=as.factor(F1_414_GT)), stat = "count") + labs(x="gentoype of 414", y="Number of SNPs") 

p.415 <- ggplot(data = F1.young.GQ.filtered) + geom_histogram(aes(x=as.factor(F1_415_GT)), stat = "count") + labs(x="gentoype of 415", y="Number of SNPs") 

p.gt <- grid.arrange(p.414, p.415, ncol=2) 
ggsave(p.gt, filename="~/F1_SNP/output/figure/p.gt.young.png", height=5, width=7)

## check read depth of two F1s 
p.414 <- ggplot(data = F1.young.GQ.filtered) + geom_histogram(aes(x=log10(F1_414_DP))) + labs(x="log10(read_depth of 414)", y="Number of SNPs") 

p.415 <- ggplot(data = F1.young.GQ.filtered) + geom_histogram(aes(x=log10(F1_415_DP))) + labs(x="log10(read_depth of 415)", y="Number of SNPs") 

p.dp <- grid.arrange(p.414, p.415, ncol=2)
ggsave(p.dp, filename="~/F1_SNP/output/figure/p.dp.young.png", height=5, width=7)

## genotype quality 
p.414 <- ggplot(data = F1.young.GQ.filtered) + geom_histogram(aes(x=F1_414_GQ)) + labs(x="genotype quality of 414", y="Number of SNPs") 

p.415 <- ggplot(data = F1.young.GQ.filtered) + geom_histogram(aes(x=F1_415_GQ)) + labs(x="genotype quality of 415", y="Number of SNPs") 

p.dp <- grid.arrange(p.414, p.415, ncol=2)
ggsave(p.dp, filename="~/F1_SNP/output/figure/p.gq.young.png", height=5, width=7)  
```

### calculate Da-Ae allele ratio and plot it out for each genotype 
```{r}
# calculate to see whether DP == AO + RO
n1 <- sum(F1.young.GQ.filtered$F1_414_DP == F1.young.GQ.filtered$F1_414_RO + F1.young.GQ.filtered$F1_414_AO) 
n2 <- sum(F1.young.GQ.filtered$F1_414_DP < F1.young.GQ.filtered$F1_414_RO + F1.young.GQ.filtered$F1_414_AO) 
n3 <- sum(F1.young.GQ.filtered$F1_414_DP > F1.young.GQ.filtered$F1_414_RO + F1.young.GQ.filtered$F1_414_AO) 

cat("number of 414 SNPs DP equals AO + RO:", n1, "\n")
cat("number of 414 SNPs DP greater than AO + RO:", n2, "\n")
cat("number of 414 SNPs DP smaller than AO + RO:", n3, "\n") 

# plot out distribution of reference allele ratio 
p.RAR.414 <- ggplot(data = F1.young.GQ.filtered) + geom_histogram(aes(x=F1_414_RO/(F1_414_AO+F1_414_RO), group=F1_414_GT, fill=F1_414_GT)) + facet_wrap(~F1_414_GT, nrow=3) + labs(x="reference allele ratio of 414", y="Number of SNPs", title="reference allele ratio for 414F1") + theme(legend.position="none") 

p.RAR.415 <- ggplot(data = F1.young.GQ.filtered) + geom_histogram(aes(x=F1_415_RO/(F1_415_AO+F1_415_RO), group=F1_415_GT, fill=F1_415_GT)) + facet_wrap(~F1_415_GT, nrow=3) + labs(x="reference allele ratio of 415", y="Number of SNPs", title="reference allele ratio for 415F1") + theme(legend.position="none")   

ggsave(p.RAR.414, filename="~/F1_SNP/output/figure/p.RAR.414.young.png", height=8, width=5)
ggsave(p.RAR.415, filename="~/F1_SNP/output/figure/p.RAR.415.young.png", height=8, width=5)

# calculate Da-Ae allele ratio and also plot it out for 414 & 415 
F1.young.GQ.filtered_Ae_ratio <- Ae_ratio_test(F1.young.GQ.filtered)

p.Ae.allele.ratio <- 
F1.young.GQ.filtered_Ae_ratio %>% 
  dplyr::select(Ae_ratio_414, Ae_ratio_415) %>%
  gather(key="F1_type", value="Ae_ratio") %>% 
  ggplot() + 
  geom_histogram(aes(x=Ae_ratio)) + 
  facet_wrap(~F1_type, nrow=2) + 
  labs(x="Da-Ae allele ratio", y="Number of SNPs", title="Da-Ae allele ratio") 
  
ggsave(p.Ae.allele.ratio, filename="~/F1_SNP/output/figure/p.Ae.allele.ratio.png", height=8, width=5)
```

### gene annotation of SNPs 
```{r}
str(F1.young.GQ.filtered) 
F1.young.GQ.filtered$POS <- as.numeric(as.character(F1.young.GQ.filtered$POS))

F1.young.gene <- SNP.annotation(SNP_data=F1.young.GQ.filtered)
dim(F1.young.gene) # 49140    25   

### for data with Ae allele ratio
F1.young.gene.Ae_ratio <- SNP.annotation(SNP_data=F1.young.GQ.filtered_Ae_ratio)
dim(F1.young.gene.Ae_ratio) # 49140    27 
F1.young.gene.Ae_ratio %>% 
  dplyr::select(Ae_ratio_414, Ae_ratio_415, gene_ID) %>%
  head(20)

save(F1.young.gene.Ae_ratio, file="~/F1_SNP/output/F1.young.gene.Ae_ratio.Rdata")
```

### work from here for gene level ASE analysis 
```{r}
library(brms)








```

################## below is the branch for single SNP binomial test ########################
### binomial test on SNP basis
```{r}
dim(F1.young.gene)
F1.young.gene.BT <- binomial_test(F1=F1.young.gene)
save(F1.young.gene.BT, file="~/F1_SNP/output/F1.young.gene.BT.Rdata")
colnames(F1.young.gene.BT)

# stats on BT 
SNP_BT_number_414 <-
F1.young.gene.BT %>%
  filter(BT_414 < 0.05) %>%
  count() %>% 
  as.numeric()

gene_BT_number_414 <- 
F1.young.gene.BT %>%
  filter(BT_414 < 0.05) %>%
  group_by(gene_ID) %>%
  count() %>%
  nrow()

SNP_BT_number_415 <-
F1.young.gene.BT %>%
  filter(BT_415 < 0.05) %>%
  count() %>% 
  as.numeric()

gene_BT_number_415 <- 
F1.young.gene.BT %>%
  filter(BT_415 < 0.05) %>%
  group_by(gene_ID) %>%
  count() %>%
  nrow()

cat(SNP_BT_number_414, "SNPs in", gene_BT_number_414, "genes are deviated from binomial distribution in 414F1", "\n")
cat(SNP_BT_number_415, "SNPs in", gene_BT_number_415, "genes are deviated from binomial distribution in 415F1", "\n")

# plot binomial test result 
p.BT.414 <- ggplot(data = F1.young.gene.BT) + geom_histogram(aes(x=BT_414)) + facet_wrap(~F1.young.gene.BT$F1_414_GT, nrow=3) + labs(x="P-value from binomial test of 414F1", y="Number of SNPs")

p.BT.415 <- ggplot(data = F1.young.gene.BT) + geom_histogram(aes(x=BT_415)) + facet_wrap(~F1.young.gene.BT$F1_415_GT, nrow=3) + labs(x="P-value from binomial test of 415F1", y="Number of SNPs")

ggsave(p.BT.414, filename="~/F1_SNP/output/figure/p.BT.414.young.png", height=8, width=5)
ggsave(p.BT.415, filename="~/F1_SNP/output/figure/p.BT.415.young.png", height=8, width=5) 
```

### import expression data
https://github.com/leejimmy93/KIAT/blob/master/F1/KIAT_F1_data_analysis.Rmd

are SNPs inside these genes show "parent-of-origin-specific" expression pattern? check that based on binomial test result generated from the above code chunk 
```{r}
load("~/F1_SNP/output/DEgene.gt.F1.Rdata") 

# young, 414 F1 as refernece level, FDR of 0.05 
head(DEgene.young.F1)
dim(DEgene.young.F1) # 4024 

DEgene.young.F1$gene_ID <- rownames(DEgene.young.F1)

DEgene.young.F1.BT <- 
DEgene.young.F1 %>%
  left_join(F1.young.gene.BT, by="gene_ID") 

# Q: how many DEGs have allele count (ie. SNP data)? among these many (how many) loci with allele count, how many of them show allele specific expression (ie. deviated from binomial distribution, p-value from binomial test is smaller than 0.05)? 
DEgene.young.F1.BT %>%
  group_by(gene_ID) %>%
  head()

DEGs_allele_count <- 
DEgene.young.F1.BT[complete.cases(DEgene.young.F1.BT),] %>%
  group_by(gene_ID) %>%
  count() %>%
  nrow() 

SNP_number <- 
DEgene.young.F1.BT[complete.cases(DEgene.young.F1.BT),] %>%
  group_by(CHROM, POS) %>%
  nrow()

SNP_ASE_414 <- 
DEgene.young.F1.BT[complete.cases(DEgene.young.F1.BT),] %>%
  group_by(CHROM, POS) %>%
  filter(BT_414 < 0.05) %>%
  nrow() # 

SNP_ASE_415 <- 
DEgene.young.F1.BT[complete.cases(DEgene.young.F1.BT),] %>%
  group_by(CHROM, POS) %>%
  filter(BT_415 < 0.05) %>%
  nrow() # 

cat(DEGs_allele_count, "DEGs have allele count, among", SNP_number, "loci with allele count,")
cat(SNP_ASE_414, SNP_ASE_415, "show allele specific expression in 414F1 and 415F1 respecitively \n")
save(DEgene.young.F1.BT, file="~/F1_SNP/output/DEgene.young.F1.BT.Rdata")
```

what are the biological function of those genes which show ASE in DEGs? to do GO enrichment analysis 

### parent-of-origin-specific expression pattern searching
for SNP/loci that show ASE (deviated from binomial distribution), how many of them have Da-Ae-origin-specific expression? Are SNPs in the same gene show the same direction? (ie. all have Da-Ae-orginal-specific expression pattern?) 
```{r}
load("~/F1_SNP/output/F1.young.gene.BT.Rdata")
head(F1.young.gene.BT) 

F1.young.gene.ASE <- ASE_test(F1=F1.young.gene.BT)
colnames(F1.young.gene.ASE)
F1.young.gene.ASE[,c("gene_ID", "ASE_414", "ASE_415")] %>%
  head(20)

save(F1.young.gene.ASE, file="~/F1_SNP/output/F1.young.gene.ASE.Rdata") # 
```

To do: 
1) multiple test problem, use FDR correction for single SNP method 

2) binomial test on gene level, use statisical model to look for genes that have "parent-of-origin-specific" expression patterns. may need STAN, treat each SNP inside a gene as an independent observation.

############################################################################## 
use brms for ASE analysis on gene level, Below is Julin's demo on one gene  
```{r}
library(brms)
library(tidyverse)
library(stringr)
library(rethinking)
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F1/data/F1.young.gene.Ae_ratio.Rdata")
```

Calcaulte  total and Ae counts because this is what (I think) we want to work with in the binomial model
```{r}
F1.young.gene.Ae_ratio %>% head()
F1.young.gene.Ae_ratio <-  F1.young.gene.Ae_ratio %>%  
  mutate(total_F1_414.counts=F1_414_RO+F1_414_AO, # total count 
         total_F1_415.counts=F1_415_RO+F1_415_AO,
         Ae_F1_414.counts=Ae_ratio_414*total_F1_414.counts, # Ae count 
         Ae_F1_415.counts=Ae_ratio_415*total_F1_415.counts,
         SNP_ID=str_c(CHROM,POS,sep="_")) %>% # usage of str_c 
  select(SNP_ID,gene_ID,ends_with("counts"))

F1.young.gene.Ae_ratio %>% head()
```

Rearrange the data so that each line is for one SNP in one of the crosses
```{r}
Total.counts.long <- F1.young.gene.Ae_ratio %>% 
gather(key="cross",value="total.counts",total_F1_414.counts,total_F1_415.counts) %>% # the usage of gather 
  mutate(cross=str_extract(cross,"F1_41[45]")) %>% # usage of str_extract()
  select(-starts_with("Ae")) # usage of -starts_with()
Total.counts.long %>% head() 

Ae.counts.long <- F1.young.gene.Ae_ratio %>%
  gather(key="cross",value="Ae.counts",Ae_F1_414.counts,Ae_F1_415.counts) %>%
    mutate(cross=str_extract(cross,"F1_41[45]")) %>%
select(-starts_with("total"))
Ae.counts.long %>% head()

F1.counts.long <- full_join(Total.counts.long,Ae.counts.long) # full_join()
F1.counts.long %>% head()
``` 

Start by selecting a single gene to work with
```{r}
onegene <- F1.counts.long %>% filter(gene_ID=="BnaA01g00020D")
onegene
```

one gene, no random effects
```{r}
brms1 <- brm(Ae.counts | trials(total.counts) ~ cross,
             family="binomial",
             prior=set_prior("normal(0,10)", class="b"),
             data=onegene) 
```

```{r} 
brms1
plot(brms1)
```

does this make sense? 
```{r}
#Estimate for 414 probability of being Da-Ae?
logistic(fixef(brms1)[1])

#Estimate of 415 probability of being Da-Ae?
logistic(sum(fixef(brms1)))

#compare to real data...should match up...
onegene %>% group_by(cross) %>% summarize(ae.ratio=mean(Ae.counts/total.counts))
```
Yes, we understand!

We want to test whether or not the to ratios are actually different
```{r}
hyp1 <- hypothesis(brms1,"Intercept = (Intercept + crossF1_415)")
hyp1
plot(hyp1)
```
Snp as random effect?

```{r}
brms2 <- brm(Ae.counts | trials(total.counts) ~ cross + (1|SNP_ID),
             family="binomial",
             prior=set_prior("normal(0,10)", class="b"),
             data=onegene,
             control = list(adapt_delta = 0.99)) 
```

```{r}
brms2
plot(brms2)
```

### compare model 1 and model 2 
```{r}
AIC(brms1, brms2) # fixed model is better than mixed model... 
brms1
brms2
?AIC
```

```{r}
hyp2 <- hypothesis(brms2,"Intercept = (Intercept + crossF1_415)")
hyp2
plot(hyp2)
```

```{r}
gene2 <- F1.counts.long %>% filter(gene_ID=="BnaA01g00180D")
brms3 <- update(brms2,newdata=gene2)
# pairs(brms3)
```

```{r}
brms3 
``` 

### write model in rethinking format
```{r}
# mixed model 
brms2 <- brm(Ae.counts | trials(total.counts) ~ cross + (1|SNP_ID),
             family="binomial",
             prior=set_prior("normal(0,10)", class="b"),
             data=onegene,
             control = list(adapt_delta = 0.99))

brms2

str(onegene) 
onegene$Ae.counts <- as.integer(onegene$Ae.counts)
onegene$total.counts <- as.integer(onegene$total.counts)
onegene$SNP <- 1:nrow(onegene)
colnames(onegene)[4:5] <- c("total_counts", "Ae_counts")
onegene

library(rethinking)
# problem with the model below
# no equivalent brms model 
rethinking1 <- map2stan(
  alist(
    Ae_counts ~ dbinom(total_counts, p), 
    logit(p) <- a_SNP[SNP],
    a_SNP[SNP] <- dnorm(0, 5)
    ),
  data = onegene, iter = 1e4, warmup = 1000, chains = 4) 
 
precis(rethinking1) 

rethinking2 <- map2stan(
  alist(
    Ae_counts ~ dbinom(total_counts, p), 
    logit(p) <- a_SNP[SNP],
    a_SNP[SNP] <- dnorm(a, sigma),
    a ~ dnorm(0, 1),
    sigma ~ dcauchy(0, 1)
    ),
  data = onegene, iter = 1e4, warmup = 1000, chains = 4) 
 
precis(rethinking2, depth = 2)

# add cross effect 
# equivalent to mixed effect model 
onegene$cross_414 <- ifelse(onegene$cross=="F1_414", 1, 0)

rethinking3 <- map2stan(
  alist(
    Ae_counts ~ dbinom(total_counts, p), 
    logit(p) <- a_SNP[SNP] + b*cross_414, 
    a_SNP[SNP] <- dnorm(a, sigma),
    a ~ dnorm(0, 1),
    sigma ~ dcauchy(0, 1), 
    b ~ dnorm(0, 10)
    ),
  data = onegene, iter = 1e4, warmup = 1000, chains = 4) 
 
precis(rethinking3, depth = 2) 

# need to write equilavent of fixed effect model in rethinking format and compare fixed VS mixed effect model 
precis(rethinking3, depth = 2) 


compare(rethinking1, rethinking3, rethinking2) 
```

iteration among all genes,tested on my MAC  
```{r}
library(snowfall)

F1.counts.long.c <- F1.counts.long[complete.cases(F1.counts.long),]
dim(F1.counts.long)
dim(F1.counts.long.c)
length(unique(F1.counts.long.c$gene_ID)) # 12044

sfInit(parallel = TRUE, cpus = parallel::detectCores())
sfLibrary(brms)
sfExport("F1.counts.long.c","brms2")

brms_all <-  
  sapply(unique(as.character(F1.counts.long.c$gene_ID)), function(x){
    new.data <- F1.counts.long.c %>% filter(gene_ID==x)
    tryCatch(update(brms1, newdata = new.data),
    error = function(e) NA)
})    

sfStop()

brms_all[[5]]
length(brms_all) 
```

run iteration on cluster 
```{r}
# fixed effct: https://github.com/leejimmy93/KIAT_cabernet/blob/master/F1/ASE_brms_fixed.R (actually ran this on my MAC, then transfer to cabernet for analysis) 
# mixed effct: https://github.com/leejimmy93/KIAT_cabernet/blob/master/F1/ASE.slurm  
```

check result for fixed effect & mixed effect 
```{r}
# mixed effect (on cabernet)
# srun -N 1 -n 20 --mem 100000 --time 07-00:00:00 --pty /bin/bash (large memory is required to load the data)

# Questions that we are intrested: 
# 1) is there cross effect for each gene? i.e. Ae.counts is affected by the kind of cross, which might affect the expression level of those genes between the two crosses. 
# all on cabernet 
```

* gene imprinting, ie. genes inherited from one parent is methylated so that allele is not expressed or expressed at a very low level. 

After checking the result, I realized that there are many genes which show different ASE between crosses but not differentialy expressed between crosses, or genes that are differentially expressed but not show ASE.   

For genes that show DE but not different ASE, there can be several different reasons:

1) non-genetic, which is due to maternal effect, ie. the maternal environment effect
2) genetic, genes which show different ASE are transcription factors which affect the expression of other genes, although the affected genes don't show different ASE. (FLC example)

but why different ASE but no different expression? ...  
too much power to detect ASE? ... 

* as a summary of the result: 4024 genes show DE between the two corsses (414 & 415), but only around 200 genes show different ASE using brms. 

We have over 60,000 genes have detectable expression in young tissue, so can use fisher's exact test to see whether there is enrichment of ASE in differentially expressed genes. I think we are expecting a yes. 

Also, since ASE can only be detected among genes which have SNPs, so the fisher's exact test needs to be done based on genes which have SNPs. 

```{r}
ASE_DE_SNP <- data.frame(DE = c(219, 2311), 
                         no_DE = c(2797, 6336))

rownames(ASE_DE_SNP) = c("ASE", "no_ASE")

fisher.test(ASE_DE_SNP) 

2300/53700*3800 # before not filtered by genes with SNP
2311/6336*2797 # there is an enrichment in the non DE genes. wierd... 

# also tried to increase iteraions to 10,000 to see whether can get Rhat all equal to 1 and avoid "## Warning: There were 2 divergent transitions after warmup. Increasing adapt_delta above 0.8 may help.". seems helped with the error, but Rhat still greater than 1. 


```

use MBASED for ASE in F1 
```{r}
# source("https://bioconductor.org/biocLite.R")
# biocLite("MBASED")
library(MBASED) 

```





