---
title: "F1 ASE Analysis with MBASED"
author: Lynn Ly
output: html_document
---

Author comment  
File description comment, including purpose of program, inputs, and outputs  
source() and library() statements  
Function definitions  
Executed statements, if applicable (e.g., print, plot)  

Purpose: This uses the MBASED package to analyze allele specific expression.  
  Single sample analysis: Find genes that display allelic imbalance ie. not expressed 0.5/0.5. Default threshold: 0.7   
  Two sample analysis: Find genes that display different ASE ratios between two samples with the SAME HAPLOTYPE. Doesn't matter what the actual ratios are, as long as they differ by over 0.2  
Inputs: Pre-filtered VCF file, simplified gff file with just ranges and feature name  
Outputs: major allele frequencies, frequency differences, list of genes displaying ASE  

```{r setup, include=FALSE}
library(MBASED)
library(tidyverse)
```

```{r AnnotateSNPs Function}
# Modified from Ruijuan's function in helpler.R
AnnotateSNPs <- function(SNPdata, gff.mRNA){
  # Combines SNP/SNV loci with gene names
  #
  # Args:
  #   SNP.data: SNP data containing positions to be matched with genomic features
  #   gff: A gff file containing only CHROM, START, END, GeneID
  #
  # Returns:
  #   SNP.data with a new column, GeneID
  
  colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 
  
  genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),
                   ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), 
                   names = gff.mRNA$name)
  
  SNPs <- GRanges(seqnames = Rle(SNPdata$CHROM), 
                 ranges = IRanges(start = SNPdata$POS, SNPdata$POS), 
                 CHROM = SNPdata$CHROM,
                 POS = SNPdata$POS)
  
  # Overlap SNP position with gene range 
  overlappedGenes <- mergeByOverlaps(SNPs, genes)
  overlappedGenes <- overlappedGenes[, c(2, 3, 5)]
  colnames(overlappedGenes) <- c("CHROM", "POS", "GeneID")
  
  annotatedSNPdata <- SNPdata %>% 
    left_join(as.data.frame(overlappedGenes), by=c("CHROM", "POS")) 
  
  return(annotatedSNPdata)  
}
```

```{r Analyzing Results Functions}
SummarizeASEResults_1s <- function(MBASEDOutput) {
  # Output: geneOutputDF is an easier way to look at MAF and p-values at the same time
  geneOutputDF <- data.frame(
    majorAlleleFrequency = assays(MBASEDOutput)$majorAlleleFrequency[,1],
    pValueASE = assays(MBASEDOutput)$pValueASE[,1],
    pValueHeterogeneity = assays(MBASEDOutput)$pValueHeterogeneity[,1])
  
  lociOutputGR <- rowRanges(metadata(MBASEDOutput)$locusSpecificResults)
  lociOutputGR$allele1IsMajor <- assays(metadata(MBASEDOutput)$locusSpecificResults)$allele1IsMajor[,1]
  lociOutputGR$MAF <- assays(metadata(MBASEDOutput)$locusSpecificResults)$MAF[,1]
  lociOutputList <- split(lociOutputGR, factor(lociOutputGR$aseID, levels = unique(lociOutputGR$aseID)))
  return(list(geneOutput=geneOutputDF, locusOutput=lociOutputList))
}

ExtractASE <- function(MBASEDOutput) {
  # Extract only desired genes
  # Modify ASEindexes to vary the strictness of selection.
  
  results <- SummarizeASEResults_1s(MBASEDOutput)

  # Apply Benjamini-Hochberg (fdr) correction for multiple testing
  adjustedP <- p.adjust(results$geneOutput$pValueASE, method = "BH", n = length(results$geneOutput$pValueASE))
  
  ASEindexes <- adjustedP < 0.05 & 
    results$geneOutput$majorAlleleFrequency > 0.7
  
  significantResults <- list(results$geneOutput[ASEindexes, ], 
                             results$locusOutput[ASEindexes, ])
  return(significantResults)
}
```

```{r Single and Two Sample Function}
# Please change numSim to at least 1,000,000 for final analysis

SingleSample <- function(annotatedData, mySNVs, genotype, numSim = 0){
  # create RangedSummarizedExperiment object as input for runMBASED
  # then runMBASED
  
  RO <- paste(genotype, "RO", sep = "_")
  AO <- paste(genotype, "AO", sep = "_")
  
  mySample <- SummarizedExperiment(
    assays = list(lociAllele1Counts = matrix(annotatedData[, RO], ncol = 1, dimnames = list(names(mySNVs), 'mySample')),
                lociAllele2Counts = matrix(annotatedData[, AO], ncol = 1,  dimnames = list(names(mySNVs), 'mySample'))
                ),
    rowRanges=mySNVs)
  
  MBASEDOutput <- runMBASED(
    ASESummarizedExperiment = mySample,
    numSim = numSim,
    isPhased = TRUE) 

  return(MBASEDOutput)
}

TwoSample <- function(annotatedData, mySNVs, genotype1, genotype2, numSim = 0){
  RO1 <- paste(genotype1, "RO", sep = "_")
  AO1 <- paste(genotype1, "AO", sep = "_")
  RO2 <- paste(genotype2, "RO", sep = "_")
  AO2 <- paste(genotype2, "AO", sep = "_")
  
  mySample <- SummarizedExperiment(
    assays = list(lociAllele1Counts = matrix(c(annotatedData[, RO1], annotatedData[, RO2]), ncol = 2,
                                             dimnames = list(names(mySNVs), c(genotype1, genotype2))),
                                             
                  lociAllele2Counts = matrix(c(annotatedData[, AO1], annotatedData[, AO2]), ncol = 2,
                                            dimnames = list(names(mySNVs), c(genotype1, genotype2)))),
    rowRanges=mySNVs)
  
  MBASEDOutput <- runMBASED(
    ASESummarizedExperiment = mySample,
    isPhased = TRUE,
    numSim = numSim
    #BPPARAM = SerialParam() # Default: No paralellization
  )
  
  return(MBASEDOutput)
} 
```

```{r Phasing Data}
gff.mRNA <- read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/gff.mRNA")

# Data to use (basicallly, VCF data pre-filtered for quality)
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/mizukikadowaki/project/output/F1.young.GQ.filtered.Rdata")

inputVCF <- F1.young.GQ.filtered
inputVCF.Ae <- inputVCF[inputVCF$Ae_GT == 1, ]
inputVCF.Ol <- inputVCF[inputVCF$Ol_GT == 1, ]

# Switch the Ref and Alt alleles, and change the counts accordingly
colnames(inputVCF.Ol)[3:4] <- c("ALT", "REF")
colnames(inputVCF.Ol)[17:24] <- c("Ae_AO", "Ol_AO", "F1_414_AO", "F1_415_AO",
                                  "Ae_RO", "Ol_RO", "F1_414_RO", "F1_415_RO")

phasedData.unordered <- rbind(inputVCF.Ae, inputVCF.Ol)

attach(phasedData.unordered)
phasedData <- phasedData.unordered[order(CHROM, POS), ]
detach(phasedData.unordered)
```

```{r Filtering Functions}

CoverageFilter <- function(vcf) {
  # Modified from Ruijuan's function in helpler.R
  # Require at least 10 reads and at least 10% of all reads support each allele
  
  totalReads <- vcf$F1_414_DP
  alleles <- vcf[, c(19, 23)]
  vcf.filtered <- vcf[which(apply(alleles / totalReads, 1, min) > .10), ]
  
  totalReads <- vcf.filtered$F1_415_DP
  alleles <- vcf.filtered[, c(20, 24)]
  vcf.filtered <- vcf.filtered[which(apply(alleles / totalReads, 1, min) > .10), ]
  
  return(vcf.filtered) 
} 

RangeFilter <- function(vcf) {
  # Discard SNVs that are within 10 bp of another called SNV. 
  # This is necessary because if one read spans both SNVs, then our counts are not independent. 
  # There is probably a better way to do this...
  
  SNV.ranges <- GRanges(seqnames = Rle(SNPdata$CHROM), 
                    ranges = IRanges(start = SNPdata$POS - 5, end = SNPdata$POS + 5),
                    CHROM = SNPdata$CHROM,
                    POS = SNPdata$POS)
  
  overlappedGenes <- mergeByOverlaps(SNV.ranges, SNV.ranges)
  
  # Get only the 10bp overlaps that were not with itself
  actualOverlappedGenes <- overlappedGenes[which(overlappedGenes$POS != overlappedGenes$POS.1),]
  
  a <- as.data.frame(actualOverlappedGenes)[, c(8, 9, 17, 18)]
  
  # We don't want to get rid of all the SNVs though. We want to keep as many SNVs as possible that aren't overlapping. 
  # idea for tmrw morning: combine chrom and pos. 
  i <- 1
  while(i <= nrow(a)) {
    if (a$POS[i] %in% a$POS.1) {
      a <- a[-i, ]
    } else {
    i <- i + 1
    }
  }

  rangeFilteredData <- anti_join(SNPdata, as.data.frame(actualOverlappedGenes), by = c("CHROM", "POS"))
}
```


```{r Annotating and Filtering Data}
annotatedData <- AnnotateSNPs(SNPdata = phasedData, gff.mRNA = gff.mRNA)
  
# Remove SNVs with no associated genes
annotatedData <- filter(annotatedData, !is.na(GeneID))
dim(annotatedData)

# The following filters are applied to avoid spurious ASE calls.

# Require at least 5 reads and at least 10% of all reads support each allele
annotatedData <- CoverageFilter(annotatedData)
dim(annotatedData)

# Discard all SNVs within 10 bp of another called variant (if one read spans both SNVs, some independence assumptions are violated)
annotatedData <- RangeFilter(annotatedData)
dim(annotatedData)

# NOT AVAILABLE FOR B NAPUS. Discard all SNVs in highly repetitive genomic regions, which we defined as regions with sequence identity of > 95% to another genomic region based on selfChain Link track from UCSC genome browser.

```
```{r Estimate Overdispersion and Reference Allele Bias}
# If our data presents any, we should include these metrics when creating mySNVs. 
# Use the additional functions in MBASED

```




```{r Single Sample Analysis}
mySNVs <- GRanges(
  seqnames = annotatedData$CHROM,
  ranges = IRanges(start = annotatedData$POS, width = 1),
  aseID = as.vector(annotatedData$GeneID),
  allele1 = annotatedData$REF,
  allele2 = annotatedData$ALT)
  
names(mySNVs) <- annotatedData$GeneID

MBASED.F1.414 <- SingleSample(annotatedData, mySNVs, genotype = "F1_414", numSim = 0)
save(MBASED.F1.414, file = "MBASED.F1.414.Rdata")
MBASED.F1.415 <- SingleSample(annotatedData, mySNVs, genotype = "F1_415", numSim = 0)
save(MBASED.F1.414, file = "MBASED.F1.415.Rdata")
MBASED.Ae <- SingleSample(annotatedData, mySNVs, genotype = "Ae", numSim = 0)
save(MBASED.F1.414, file = "MBASED.F1.Ae.Rdata")
MBASED.Ol <- SingleSample(annotatedData, mySNVs, genotype = "Ol", numSim = 0)
save(MBASED.F1.414, file = "MBASED.F1.Ol.Rdata")
```

In Progress: Two Sample analysis  

```{r Two Sample Analysis}
# Estimating two sample run time using smaller datasets
annotatedData.trimmed <- annotatedData[1:500 ,]

mySNVs.trimmed <- GRanges(
  seqnames = annotatedData.trimmed$CHROM,
  ranges = IRanges(start = annotatedData.trimmed$POS, width = 1),
  aseID = as.vector(annotatedData.trimmed$GeneID),
  allele1 = annotatedData.trimmed$REF,
  allele2 = annotatedData.trimmed$ALT)

time100.10 <- system.time(MBASED.F1.414.vs.F1.415 <- TwoSample(annotatedData.trimmed, mySNVs.trimmed, "F1_414", "F1_415"))

time100.10
# 100: 9 seconds

```

```{r Analysis and Summary}
load("MBASED.F1.414.Rdata")
load("MBASED.F1.415.Rdata")
load("MBASED.Ae.Rdata")
load("MBASED.Ol.Rdata")

sig.F1.414 <- ExtractASE(MBASED.F1.414)
dim(sig.F1.414[[1]]) #1946 SNVs found to be in ASE
summary(sig.F1.414[[2]])

sig.F1.415 <- ExtractASE(MBASED.F1.415)
dim(sig.F1.415[[1]]) #2164 SNVs found to be in ASE
head(sig.F1.415[[2]]$GeneID)

sig.Ae <- ExtractASE(MBASED.Ae)
dim(sig.Ae[[1]]) #11995 SNVs found to be in ASE

sig.Ol <- ExtractASE(MBASED.Ol)
dim(sig.Ol[[1]]) #11990 SNVs found to be in ASE
```
