---
title: "F1 ASE Analysis with MBASED"
author: Lynn Ly
output: html_document
---

Author comment
File description comment, including purpose of program, inputs, and outputs
source() and library() statements  
Function definitions
Executed statements, if applicable (e.g., print, plot)

Purpose: This uses the MBASED package to analyze allele specific expression. 
  Single sample analysis: Find genes that display allelic imbalance ie. not expressed 0.5/0.5. Default threshold: 0.7 
  Two sample analysis: Find genes that display different ASE ratios between two samples. Doesn't matter what the actual ratios are, as long as they differ by over 0.2
Inputs: Pre-filtered VCF file, simplified gff file with just ranges and feature name
Outputs: major allele frequencies, frequency differences, list of genes displaying ASE

```{r setup, include=FALSE}
library(MBASED)
library(tidyverse)
```

```{r AnnotateSNPs Function}
AnnotateSNPs <- function(SNPdata, gff.mRNA){
  # Combines SNP/SNV loci with gene names
  #
  # Args:
  #   SNP.data: SNP data containing positions to be matched with genomic features
  #   gff: A gff file containing only CHROM, START, END, GeneID
  #
  # Returns:
  #   SNP.data with a new column, GeneID
  
  colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 
  
  genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),
                   ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), 
                   names = gff.mRNA$name)
  
  SNPs <- GRanges(seqnames = Rle(SNPdata$CHROM), 
                 ranges = IRanges(start = SNPdata$POS, SNPdata$POS), 
                 CHROM = SNPdata$CHROM,
                 POS = SNPdata$POS)
  
  # Overlap SNP position with gene range 
  overlappedGenes <- mergeByOverlaps(SNPs, genes)
  overlappedGenes <- overlappedGenes[, c(2, 3, 5)]
  colnames(overlappedGenes) <- c("CHROM", "POS", "GeneID")
  
  annotatedSNPdata <- SNPdata %>% 
    left_join(as.data.frame(overlappedGenes), by=c("CHROM", "POS")) 
  
  return(annotatedSNPdata)  
}
```

```{r Analyzing Results Functions}
SummarizeASEResults_1s <- function(MBASEDOutput) {
  # Output: geneOutputDF is an easier way to look at MAF and p-values at the same time
  geneOutputDF <- data.frame(
    majorAlleleFrequency = assays(MBASEDOutput)$majorAlleleFrequency[,1],
    pValueASE = assays(MBASEDOutput)$pValueASE[,1],
    pValueHeterogeneity = assays(MBASEDOutput)$pValueHeterogeneity[,1])
  
  lociOutputGR <- rowRanges(metadata(MBASEDOutput)$locusSpecificResults)
  lociOutputGR$allele1IsMajor <-  assays(metadata(MBASEDOutput)$locusSpecificResults)$allele1IsMajor[,1]
  lociOutputGR$MAF <- assays(metadata(MBASEDOutput)$locusSpecificResults)$MAF[,1]
  lociOutputList <- split(lociOutputGR, factor(lociOutputGR$aseID, levels = unique(lociOutputGR$aseID)))
  return(list(geneOutput=geneOutputDF, locusOutput=lociOutputList))
}

ExtractASE <- function(MBASEDOutput) {
  # Extract only desired genes
  # Modify ASEindexes to vary the strictness of selection.
  # TODO: Implement Benjamini-Hochberg correction (punish p-values based on rank)
  # Currently using Bonferroni as a crude correction measure (punishes all p-values equally)
  
  results <- SummarizeASEResults_1s(MBASEDOutput)

  ASEindexes <- results$geneOutput$pValueASE * 46577 < 0.05 & 
    results$geneOutput$majorAlleleFrequency > 0.7
  
  significantResults <- list(results$geneOutput[ASEindexes, ], 
                             results$locusOutput[ASEindexes, ])
  return(significantResults)
}
```

```{r Single Sample Function}
# Please change numSim to at least 1,000,000 for final analysis

SingleSample <- function(annotatedData, mySNVs, genotype){
  # create RangedSummarizedExperiment object as input for runMBASED
  # then runMBASED
  
  RO <- paste(genotype, "RO", sep = "_")
  AO <- paste(genotype, "AO", sep = "_")
  
  mySample <- SummarizedExperiment(
    assays = list(lociAllele1Counts = matrix(annotatedData[, RO], ncol = 1, dimnames = list(names(mySNVs), 'mySample')),
                lociAllele2Counts = matrix(annotatedData[, AO], ncol = 1,  dimnames = list(names(mySNVs), 'mySample'))
                ),
    rowRanges=mySNVs)
  
  MBASEDOutput <- runMBASED(
    ASESummarizedExperiment = mySample,
    numSim = 1000, 
    isPhased = FALSE)  

  return(MBASEDOutput) 
} 

MBASED.F1.414 <- SingleSample(annotatedData, mySNVs, genotype = "F1_414")
save(MBASED.F1.414, file = "MBASED.F1.414.Rdata")
MBASED.F1.415 <- SingleSample(annotatedData, mySNVs, genotype = "F1_415")
save(MBASED.F1.414, file = "MBASED.F1.415.Rdata")
MBASED.Ae <- SingleSample(annotatedData, mySNVs, genotype = "Ae")
save(MBASED.F1.414, file = "MBASED.F1.Ae.Rdata")
MBASED.Ol <- SingleSample(annotatedData, mySNVs, genotype = "Ol")
save(MBASED.F1.414, file = "MBASED.F1.Ol.Rdata") 
```

```{r Data to use}
# Change the following line if not using B. napus
gff.mRNA <- read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/gff.mRNA")

# Data to use
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/mizukikadowaki/project/output/F1.young.GQ.filtered.Rdata")

annotatedData <- AnnotateSNPs(SNPdata = F1.young.GQ.filtered, gff.mRNA = gff.mRNA)
  
# Remove SNVs with no associated genes 
annotatedData <- filter(annotatedData, !is.na(GeneID)) 
  
mySNVs <- GRanges(
  seqnames = annotatedData$CHROM,
  ranges = IRanges(start = annotatedData$POS, width = 1),
  aseID = as.vector(annotatedData$GeneID),
  allele1 = annotatedData$REF,
  allele2 = annotatedData$ALT)
  
names(mySNVs) <- annotatedData$GeneID 
```

In Progress: TwoSample analysis

```{r Two Sample Function}
TwoSample <- function(annotatedData, mySNVs, genotype1, genotype2){
  RO1 <- paste(genotype1, "RO", sep = "_")
  AO1 <- paste(genotype1, "AO", sep = "_")
  RO2 <- paste(genotype2, "RO", sep = "_")
  AO2 <- paste(genotype2, "AO", sep = "_")
  
  mySample <- SummarizedExperiment(
    assays = list(lociAllele1Counts = matrix(c(annotatedData[, RO1], annotatedData[, RO2]), ncol = 2,
                                             dimnames = list(names(mySNVs), c(genotype1, genotype2))),
                                             
                  lociAllele2Counts = matrix(c(annotatedData[, AO1], annotatedData[, AO2]), ncol = 2,
                                            dimnames = list(names(mySNVs), c(genotype1, genotype2)))),
    rowRanges=mySNVs)
  
  MBASEDOutput <- runMBASED(
    ASESummarizedExperiment = mySample,
    isPhased = FALSE,
    numSim = 0 
    #BPPARAM = SerialParam() # Default: No paralellization
  )
  
  return(MBASEDOutput) 
} 
```

```{r Run Two Sample}
MBASED.F1.414.vs.F1.415 <- TwoSample(annotatedData, mySNVs, "F1_414", "F1_415")

```

```{r Analysis and Summary}
load("MBASED.F1.414.Rdata")
load("MBASED.F1.415.Rdata")
load("MBASED.Ae.Rdata")
load("MBASED.Ol.Rdata")

sig.F1.414 <- ExtractASE(MBASED.F1.414)
dim(sig.F1.414[[1]]) #1946 SNVs found to be in ASE
summary(sig.F1.414[[2]])


sig.F1.415 <- ExtractASE(MBASED.F1.415)
dim(sig.F1.415[[1]]) #2164 SNVs found to be in ASE
head(sig.F1.415[[2]]$GeneID)

sig.Ae <- ExtractASE(MBASED.Ae)
dim(sig.Ae[[1]]) #11995 SNVs found to be in ASE

sig.Ol <- ExtractASE(MBASED.Ol)
dim(sig.Ol[[1]]) #11990 SNVs found to be in ASE
```

