---
title: '505'
author: "Ruijuan Li"
date: "2/8/2017"
output: 
  html_document: 
    keep_md: yes 
---

# download, trimming, mapping 
```{r}
# 1) download 
# in cabernet, using download.slurm & download_2.slurm
# also download to whitney 

# 2) trimming on whitney only 
# trimming.sh

# 3) check quality, doing 
# before trimming, QC was done by company, so look at the result directly using /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2017_winter/505/data/bioftp.org/TBD170026_20170202/make_summary_report.sh  
# after trimming, fastqc by QC.sh, then check quality 

# 4) transfer all trimmed data to cabernet for mapping 
# get file name: ls | grep "Sample" | sed 's/\///' | tr '\n' ' '
# /share/malooflab/Ruijuan/505/trimmed_data/TBD170026_20170202/fetch_02_13_2017.slurm & 
# /share/malooflab/Ruijuan/505/trimmed_data/TBD170001_1617/fetch_02_13_2017_2.slurm 

# 5) mapping, one sample takes about 15-20 Mins on 12 CPUs 
# /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2017_winter/505/data/bioftp.org/TBD170026_20170202/mapping_505.sh 

# 6) extract start result using 
# /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2017_winter/505/data/bioftp.org/TBD170026_20170202/star_extract_stats.sh  
```

# mapping result analysis 
```{r, results='hide'}
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")
# sample description 
sample_des_a <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/Sequencing_Statistics_Result_TBD170001_1617.csv", header = T)
sample_des_a$batch <- rep("A", nrow(sample_des_a))
sample_des_a_sub <- sample_des_a[,c("Sample.ID","TotalReads","batch")]
sample_des_b <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/Sequencing_Statistics_Result_TBD170026.csv", header = T)
sample_des_b$batch <- rep("B", nrow(sample_des_b))
sample_des_b_sub <- sample_des_b[,c("Sample.ID","TotalReads","batch")]
head(sample_des_a_sub)

# mapping result
mapping_result <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/505/Star_Stats.tab", header = T)
head(mapping_result)
mapping_result$Sample.ID <- gsub("(Sample_)([[:print:]])", "\\2", mapping_result$Sample) 

# merge sample description & mapping result for batch A and batch B 
batch.a.stats <- merge(sample_des_a_sub, mapping_result, by="Sample.ID")
batch.b.stats <- merge(sample_des_b_sub, mapping_result, by="Sample.ID")
mapping_result_505 <- rbind(batch.a.stats, batch.b.stats)

mapping_result_505$batch
mapping_result_505$Number_Input_Reads/(mapping_result_505$TotalReads/2)

# number of total reads 
library(ggplot2) 
p.total.reads <- ggplot(data=mapping_result_505)
p.total.reads <- p.total.reads + geom_bar(aes(x=reorder(Sample.ID, TotalReads/2), y=TotalReads/2, fill=batch), stat = "identity")
p.total.reads <- p.total.reads + labs(list(title = "", x = "", y = "Number of raw reads"))
p.total.reads <- p.total.reads + theme(axis.text.x = element_text(angle = 90, size = 8)) 
p.total.reads   
# ggsave(p.total.reads, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/figure/total_reads.png", height = 8, width = 11) 

# look at HQ, mapped, multimapped in detail for each sample (memory too low, cannot view histogram of total reads...)
colnames(mapping_result_505)
mapping_result_505_sub <- data.frame(Sample.ID = mapping_result_505$Sample.ID, 
                                     batch=mapping_result_505$batch,
                                    trimmed.off = mapping_result_505$TotalReads/2 - mapping_result_505$Number_Input_Reads,
                                    unmapped = mapping_result_505$Number_Input_Reads * (as.numeric(mapping_result_505$Percent_Unmapped_Mismatches + mapping_result_505$Percent_Unmapped_Other + mapping_result_505$Percent_Unmapped_Too_Short)/100), 
                                    multi_too_many_mapped = mapping_result_505$Number_Multi_Mapped + mapping_result_505$Number_Too_Many_Multi_Mapped, 
                                    unqiuemapped = mapping_result_505$Number_Unique_Mapped
) 

mapping_result_505_sub 

mapping_result_505_sub.long <- melt(mapping_result_505_sub, id.vars = c("Sample.ID", "batch"))
mapping_result_505_sub.long.a <- mapping_result_505_sub.long[mapping_result_505_sub.long$batch=="A",]
mapping_result_505_sub.long.b <- mapping_result_505_sub.long[mapping_result_505_sub.long$batch=="B",]

pl.mapping.indi.a <- ggplot(data = mapping_result_505_sub.long.a)
pl.mapping.indi.a <- pl.mapping.indi.a + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.a <- pl.mapping.indi.a + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_blank()) + ylim(0, 4.5e+7) 
pl.mapping.indi.a

pl.mapping.indi.b <- ggplot(data = mapping_result_505_sub.long.b)
pl.mapping.indi.b <- pl.mapping.indi.b + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.b <- pl.mapping.indi.b + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_blank()) + ylim(0, 4.5e+7)
pl.mapping.indi.b
 
library(cowplot)
pl.mapping.indi <- plot_grid(
  pl.mapping.indi.a +labs(title="mapping result of batch a")+theme(legend.position = "none"),
  pl.mapping.indi.b +labs(title="mapping result of batch b") +theme(legend.position = "none"),
  align = 'vh', ncol=2, nrow = 1, labels=c("",""))

get_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}

legend <- get_legend(pl.mapping.indi.a)

# pl.mapping.indi <- plot_grid(arrangeGrob(pl.mapping.indi.a + theme(legend.position="none") + labs(title="mapping result of batch a"),
#                          pl.mapping.indi.b + theme(legend.position="none") + labs(title="mapping result of batch b"),
#                          nrow=1),
#              legend, nrow=2,heights=c(10, 1))

pl.mapping.indi.final <- plot_grid(pl.mapping.indi, legend, rel_widths = c(3, 1))

pl.mapping.indi.final
save_plot("~/Desktop/Brassica_project/KIAT_RNA_seq/505/figure/pl_mapping_indi.png", pl.mapping.indi.final, base_aspect_ratio = 1, base_width = 12)  


# look at histogram of HQ reads 
colnames(mapping_result_505)
p.HQ.perc <- ggplot(data=mapping_result_505)
p.HQ.perc <- p.HQ.perc + geom_histogram(aes(x=mapping_result_505$Number_Input_Reads/(mapping_result_505$TotalReads/2), fill=batch), binwidth = 0.01) 
p.HQ.perc <- p.HQ.perc + labs(list(title = "histogram of HQ reads percentage", x = "Percent of high qulity reads", y = "number of samples"))
p.HQ.perc 
# ggsave(p.HQ.perc, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/HQ_hist.png", height = 8, width = 11) 

# histogram of unqiue mapped reads 
p.uniq.mapping <- ggplot(data=mapping_result_505)
p.uniq.mapping <- p.uniq.mapping + geom_histogram(aes(x=Percent_Unique_Mapped, fill=batch), binwidth = 2.5) 
p.uniq.mapping <- p.uniq.mapping + labs(list(title = "histogram of uniquely mapped reads percentage", x = "Percent of uniquely mapped reads", y = "number of samples"))
p.uniq.mapping  
# ggsave(p.uniq.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/unique_mapped.png", height = 8, width = 11) 

# hisogram of multimapped & too many reads 
p.multi.too.many.mapping <- ggplot(data=mapping_result_505)
p.multi.too.many.mapping <- p.multi.too.many.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Multi_Mapped+mapping_result_505$Percent_Too_Many_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.multi.too.many.mapping <- p.multi.too.many.mapping + labs(list(title = "", x = "Percent of multi & too many mapped reads", y = "number of samples"))

# histogram of multi mapped 
p.multi.mapping <- ggplot(data=mapping_result_505)
p.multi.mapping <- p.multi.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.multi.mapping <- p.multi.mapping + labs(list(title = "histogram of multi mapped reads percentage", x = "Percent of multi mapped reads", y = "number of samples"))
p.multi.mapping
# ggsave(p.multi.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/multi_mapped.png", height = 8, width = 11) 

# histogram of too many mapped 
p.toomany.mapping <- ggplot(data=mapping_result_505)
p.toomany.mapping <- p.toomany.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Too_Many_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.toomany.mapping <- p.toomany.mapping + labs(list(title = "histogram of too many mapped reads percentage", x = "Percent of toomany mapped reads", y = "number of samples"))
p.toomany.mapping
# ggsave(p.toomany.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/toomany_mapped.png", height = 8, width = 11) 

# histogram of unmapped 
p.un.mapping <- ggplot(data=mapping_result_505)
p.un.mapping <- p.un.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Unmapped_Mismatches+mapping_result_505$Percent_Unmapped_Other+mapping_result_505$Percent_Unmapped_Too_Short, fill=batch), binwidth = 2.5) 
p.un.mapping <- p.un.mapping + labs(list(title = "histogram of unmapped reads percentage", x = "Percent of unmapped reads", y = "number of samples"))
p.un.mapping
# ggsave(p.un.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/un_mapped.png", height = 8, width = 11)

# mapping stats 
colnames(mapping_result_505)
# average percentage of HQ, unique, multi, toomany, unmapped 
average.mapping <- aggregate(mapping_result_505[,c("TotalReads","Number_Input_Reads", "Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short")], list(mapping_result_505$batch), mean)

rownames(average.mapping) <- average.mapping$Group.1
average.mapping$Percent_HQ <- average.mapping$Number_Input_Reads/(average.mapping$TotalReads/2)*100
average.mapping

average.mapping.long <- melt(average.mapping[,c("Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short","Percent_HQ","Group.1")], id.vars = "Group.1")

average.mapping.long

p.average.mapping <- ggplot(data = average.mapping.long)
p.average.mapping <- p.average.mapping + geom_bar(aes(x=variable, y=value, fill=variable), stat = "identity")
p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.mapping <- p.average.mapping + labs(list(title = "mapping stats of 505", x = "", y = "percentage"))
p.average.mapping <- p.average.mapping + theme(axis.text.x = element_text(angle = 90), legend.position = "none")
p.average.mapping <- p.average.mapping + geom_text(data = average.mapping.long, aes(x=variable, y=value, label=factor(round(value,0)))) 
p.average.mapping
# ggsave(p.average.mapping, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/figure/average_mapping.png", height = 6, width = 9) 

### split this figure to HQ ratio & mapping ratio (two plots) 
average.mapping.long.HQ <- 
  average.mapping.long[average.mapping.long$variable=="Percent_HQ",]

average.mapping.long.HQ

p.average.HQ <- ggplot(data = average.mapping.long.HQ)
p.average.HQ <- p.average.HQ + geom_bar(aes(x=Group.1, y=value, fill=variable), stat = "identity")
# p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.HQ <- p.average.HQ + labs(list(title = "", x = "", y = "percentage of total raw reads"))
p.average.HQ <- p.average.HQ + theme(legend.position = "none")
p.average.HQ <- p.average.HQ + geom_text(data = average.mapping.long.HQ, aes(x=Group.1, y=value, label=factor(round(value,0)))) 
p.average.HQ

ggsave(p.average.HQ, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_HQ.png", height = 4, width = 2) 

average.mapping.long.mapping <- 
  average.mapping.long[average.mapping.long$variable!="Percent_HQ",]

p.average.mapping <- ggplot(data = average.mapping.long.mapping)
p.average.mapping <- p.average.mapping + geom_bar(aes(x=variable, y=value, fill=variable), stat = "identity")
p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.mapping <- p.average.mapping + labs(list(title = "", x = "", y = "percentage of HQ reads"))
p.average.mapping <- p.average.mapping + theme(axis.text.x = element_text(angle = 90), legend.position = "none")
p.average.mapping <- p.average.mapping + geom_text(data = average.mapping.long.mapping, aes(x=variable, y=value, label=factor(round(value,0)))) 
p.average.mapping 

ggsave(p.average.mapping, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_mapping.png", height = 6, width = 5) 

```

# SNP calling on cabernet & filtering 
```{r}
# 1) prepare data for SNP calling 
# Prep4Freebayes.sh (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/Prep4Freebayes.sh)

# 2) split bam file by chromosome 
# Bam_Split_By_Chrom.sh (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/Bam_Split_By_Chrom.sh)

# 3) call SNPs 
# Freebayes_Array_2_22_2017.slurm (https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/Freebayes_Array_2_22_2017.slurm)

# 4) compress & index 
# compress_index_vcf.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/compress_index_vcf.sh)

# 5) combine into one big vcf file 
# concat.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/concat.sh)
# output: 505.vcf.gz 

# 6) filter using vcftool to get only biallelic SNPs
# extract_SNP.sh & extract_biallelic.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/extract_SNP.sh)  
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/extract_biallelic.sh) 
# output: 505_SNP_biallelic.gz 

# 7) filter based on MAF and Q 
# filtering_MAF_Q.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/filtering_MAF_Q.sh)
# output: 505_filtered_MAF_Q.gz 

# 8) calcaulate mean read depth across individuals
# calc_depth.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/calc_depth.sh) calculate in R
# output: 505_filtered_MAF_Q.ldepth.mean 

# 9) filter based on read depth (after check mean and distribution of read depth, decide to use 5-500 as the threshold)
# filtering_depth.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/filtering_depth.sh)
# output: 505_filtered_MAF_Q_depth.gz 

# 10) calculate missing rate (SNPs with less than median of missing rate should be filtered out) 
# calc_missingness.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/calc_missingness.sh)
# output: 505_filtered_MAF_Q_depth.imiss; 505_filtered_MAF_Q_depth.lmiss 

# 11) filter based on missing rate 
# filter_missingness.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/filtering_missingrate.sh)
# output: 505_filtered_MAF_Q_depth_missingness.gz 

# 12) calculate MAF, missing rate, depth, and LD... 
# calc.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/calc.sh)
# output: a lot of stats start with 505_filtered_MAF_Q_depth_missingness prefix 
```

# stats mainly for filtered dataset 
```{r}
# mean read depth before read depth trimming 
mean.depth <- read.table("~/505/vcf/combined/505_filtered_MAF_Q_depth_missingness.ldepth.mean", header=T)
head(mean.depth)
mean(mean.depth$MEAN_DEPTH) # 34.53817  

png("~/505/output/figure/read.depth.png", width=4, height=3, units="in", res=300) 
hist(log10(mean.depth$MEAN_DEPTH), xlab = "log10(read depth)", ylab="Number of SNPs", main="read depth", breaks=1000)
abline(v=c(log10(5),log10(500)), col="red", lwd=2)
dev.off()

dim(mean.depth) # 434331      4 
# decide to use 5 to 500 as the threshold for depth 

# MAF calculation before filtering 
MAF <- read.table("~/505/vcf/combined/505_freq.txt")
head(MAF)
dim(MAF) 

class(MAF$V3)

MAF$MAF<- round(pmin(MAF$V2, MAF$V3)/(MAF$V2+MAF$V3), digits = 2) 
head(MAF$MAF)
hist(MAF$MAF)

png("~/505/output/figure/MAF.all.png", width=4, height=3, units="in", res=300) 
hist(MAF$MAF, xlab = "Minor allele frequency", ylab="Number of SNPs", main="Minor allele frequency distribution", breaks=50)
abline(v=c(0.01), col="red", lwd=2)
dev.off()

# stats after filtering 
# MAF (see log for how toget freq.txt from .frq file)
MAF.2 <- read.table("~/505/vcf/combined/505_filtered_MAF_Q_depth_missingness_freq.txt")
head(MAF.2)
dim(MAF.2) # 434331      3 

class(MAF.2$V3)

MAF.2$MAF<- round(pmin(MAF.2$V2, MAF.2$V3)/(MAF.2$V2+MAF.2$V3), digits = 2) 
head(MAF.2$MAF)
hist(MAF.2$MAF)

png("~/505/output/figure/MAF.filterd.png", width=4, height=3, units="in", res=300) 
hist(MAF.2$MAF, xlab = "Minor allele frequency", ylab="Number of SNPs", main="Minor allele frequency distribution", breaks=50)
abline(v=c(0.01), col="red", lwd=2)
dev.off()

# missing rate 
missing.individual <- read.table("~/505/vcf/combined/505_filtered_MAF_Q_depth_missingness.imiss", header=T)

head(missing.individual) 
dim(missing.individual) # 94 
hist(missing.individual$F_MISS, breaks=100)

png("~/505/output/figure/missing.individual.png", width=4, height=3, units="in", res=300) 
hist(missing.individual$F_MISS, xlab = "missing rate at taxa level", ylab="Number of individuals", main="Missing rate at taxa level", breaks=100)
dev.off()

# pull out the individuals with high missing rate 
missing.individual[missing.individual$F_MISS > 0.2,]

# missing rate before filtering 
missing.site <- read.table("~/505/vcf/combined/505_filtered_MAF_Q_depth.lmiss", header=T)
head(missing.site)

hist(missing.site$F_MISS, breaks=20)
median(missing.site$F_MISS) # 0.01  
mean(missing.site$F_MISS) # 0.03 
# based on these values, SNPs with less than 0.1 
missing.site$N_DATA # looks like it consider I have 188 samples, but... do I? no!!! I only have 94 samples... why 188 here? wierd... 

png("~/505/output/figure/missing.locus.png", width=4, height=3, units="in", res=300) 
hist(missing.site$F_MISS, xlab = "missing rate at locus level", ylab="Number of SNPs", main="Missing rate at locus level", breaks=20)
dev.off()

# missing rate after filtering 
missing.site.2 <- read.table("~/505/vcf/combined/505_filtered_MAF_Q_depth_missingness.lmiss", header=T)
head(missing.site.2)

hist(missing.site.2$F_MISS, breaks=20)
median(missing.site.2$F_MISS) # 0.01  
mean(missing.site.2$F_MISS) # 0.03 
min(missing.site.2$F_MISS)
max(missing.site.2$F_MISS) # 89% 
# based on these values, SNPs with less than 0.1 
missing.site.2$N_DATA # 188 because this is diploid 
missing.site.2$N_DATA

png("~/505/output/figure/missing.locus.png", width=4, height=3, units="in", res=300) 
hist(missing.site.2$F_MISS, xlab = "missing rate at locus level", ylab="Number of SNPs", main="Missing rate at locus level", breaks=20)
dev.off()
```

# analyze vcf file (come to add GQ, filter based on GQ... )
```{r}
library("vcfR")
leaf_505 <- read.vcfR("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf/combined/505_filtered_MAF_Q_depth_missingness.gz")

# save(leaf_505, file="~/505/output/leaf_505.Rdata")
load("~/505/output/leaf_505.Rdata")

# reform the orginal data
temp <- leaf_505

  vcfbi <- is.biallelic(temp) # return with logics indicating whehter biallelic or not... 
  vcfref <- subset(getREF(temp), subset = vcfbi) # get ref allele
  vcfalt <- subset(getALT(temp), subset = vcfbi) # get alt allele 
  vcfgts <- subset(extract.gt(temp, element = "GT", return.alleles = T, IDtoRowNames=T), subset = vcfbi) # get gt info  
  vcfpos <- subset(getPOS(temp), subset = vcfbi) # get pos info 
  vcfchrom <- subset(getCHROM(temp), subset = vcfbi) # get chromosome info 
  # vcfgls <- subset(extract.gt(temp, element = "GL"), subset = vcfbi) # get gl 

  temp2 <- data.frame(cbind(vcfchrom,vcfpos,vcfref,vcfalt,vcfgts))
  colnames(temp2)[1:4] <- c("CHROM","POS","REF","ALT")
  rnames <- rownames(temp2) 
#   temp2 <- data.frame(sapply(temp2, function(x) sub("0/0","-1",x)))
#   temp2 <- data.frame(sapply(temp2, function(x) sub("0/1","0",x)))
#   temp2 <- data.frame(sapply(temp2, function(x) sub("1/1","1",x)))
  row.names(temp2) <- rnames 

temp2[1:10, 1:10]

temp2.test <- temp2 
temp2.test[1:10, 1:10]

### convert to letters in diploid mode, use apply... 
# temp2.diploid <- matrix("",nrow=nrow(temp2.test),ncol=ncol(temp2)-4) 
# dim(temp2.diploid)
# temp2.diploid[1:10, 1:10]
# 
#     for (i in 1:nrow(temp2)){
#       for (j in 5:ncol(temp2)){
#         print (i)
#       if (temp2[i,j]=="0/0" & is.na(temp2[i,j])== F){
#         temp2.diploid[i,j-4] <- paste(temp2[i,3], temp2[i,3])
#       } else if (temp2[i,j]=="0/1" & is.na(temp2[i,j])== F){
#         temp2.diploid[i,j-4] <- paste(temp2[i,3], temp2[i,4])
#       } else if (temp2[i,j]=="1/1" & is.na(temp2[i,j])== F){
#         temp2.diploid[i,j-4] <- paste(temp2[i,4], temp2[i,4])
#       } else {
#         temp2.diploid[i,j-4] <- temp2[i,j] 
#       }
#     }
#     } 

# save the data afterwards... 

### test this... !!!!!!!!! 

# temp2.diploid <- matrix("",nrow=nrow(temp2.test),ncol=ncol(temp2)-4) 
# dim(temp2.diploid)
# temp2.diploid[1:10, 1:10]
# 
#     for (i in 1:3){
#       for (j in 5:ncol(temp2)){
#         print (i)
#       if (temp2[i,j]=="0/0" && is.na(temp2[i,j])== F){
#         temp2.diploid[i,j-4] <- paste(temp2[i,3], temp2[i,3], sep="")
#       } else if (temp2[i,j]=="0/1" && is.na(temp2[i,j])== F){
#         temp2.diploid[i,j-4] <- paste(temp2[i,3], temp2[i,4], sep="")
#       } else if (temp2[i,j]=="1/1" && is.na(temp2[i,j])== F){
#         temp2.diploid[i,j-4] <- paste(temp2[i,4], temp2[i,4], sep="")
#       } else {
#         temp2.diploid[i,j-4] <- temp2[i,j] 
#       }
#     }
#     } 
# 
# temp2.diploid[1:10, 1:10]
# temp2.test[1:10, 1:10] 

  
  # transpose to desired format 
  temp2.t <- t(temp2)
  rownames(temp2.t)[1:4] <- c("CHROM","POS","REF","ALT")
  colnames(temp2.t) <- rnames

# population structure 
dim(temp2.t) # 98 434331 
temp2.t.2500 <- temp2.t[5:98,1:2500] 
temp2.t.2500[5:8, 1:20] 

geno.numeric <- data.matrix(temp2.t.2500)
genDist <- as.matrix(dist(geno.numeric))

#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
?cmdscale
head(geno.mds) #now we have 2 dimensions
dim(geno.mds)

plot(geno.mds) 

######## all 
temp2.t.all <- temp2.t[5:98,] 
temp2.t.all[5:8, 1:20]

geno.numeric.all <- data.matrix(temp2.t.all)
genDist.all <- as.matrix(dist(geno.numeric.all))

#perform the multi-dimensional scaling
geno.mds.all <- as.data.frame(cmdscale(genDist.all))

?cmdscale
head(geno.mds.all) #now we have 2 dimensions
dim(geno.mds.all)

plot(geno.mds.all) 

save(geno.mds.all, file="~/505/output/geno.mds.all.Rdata")
save(temp2, file="~/505/output/temp2.Rdata")

## add country of origin, winter or spring type to the figure ... 
sample_des_a_revised <- read.csv("~/505/data/phenotype/batch_a_revised.csv", header=T)
sample_des_a_revised_sub <- sample_des_a_revised[,c("Sample.ID", "Sample.Description.Origin", "Sample.Description.Type")]
sample_des_b_revised <- read.csv("~/505/data/phenotype/batch_b_revised.csv", header=T)
sample_des_b_revised_sub <- sample_des_b_revised[,c("Sample.ID", "Sample.Description.Origin", "Sample.Description.Type")]
head(sample_des_a_revised_sub)
sample_des_a_revised_sub$Sample.ID
sample_des_b_revised_sub$Sample.ID

sample_des_revised <- rbind(sample_des_a_revised_sub, sample_des_b_revised_sub)
sample_des_revised <- sample_des_revised[-c(47:51),]
sample_des_revised$Sample.ID

# MDS result 
head(geno.mds.all)
rownames(geno.mds.all)

# reformat MDS result sample ID (X remove & . to -)
geno.mds.all$Sample.ID <- gsub("\\.","\\-", rownames(geno.mds.all)) 
geno.mds.all$Sample.ID <- gsub("X", "", geno.mds.all$Sample.ID)
class(geno.mds.all$Sample.ID)

geno.mds.all$Sample.ID[94]==sample_des_revised$Sample.ID[75]

# merge 
geno.mds.all.des <- merge(geno.mds.all, sample_des_revised, by="Sample.ID")
head(geno.mds.all.des)

plot(geno.mds.all.des$V1, geno.mds.all.des$V2)
save(geno.mds.all.des, file="~/505/output/geno.mds.all.des.Rdata")

# plot 
p.mds.country <- ggplot(data=geno.mds.all.des)
p.mds.country <- p.mds.country + geom_point(aes(V1, V2, color=factor(Sample.Description.Origin),  size=5)
p.mds.country

p.mds.type <- ggplot(data=geno.mds.all.des)
p.mds.type <- p.mds.type + geom_point(aes(V1, V2, color=factor(Sample.Description.Type),  size=5)
p.mds.type 


##  
p.mds <- ggplot(data = distance_matrix) + theme_gray(base_size = 20)
p.mds <- p.mds + geom_point(aes(x, y, color=factor(gt), shape=factor(tissue)), size=5) 
p.mds <- p.mds + labs(y = "BCV distance 2", x="BCV distance 1")
p.mds <- p.mds + theme()
# p.mds <- p.mds + facet_grid(~gt) 
p.mds         
  
####
``` 

# PSMix: Assigning varieties to populations 
```{r}
# install.packages("~/bin/PSMix/", repos = NULL, type="source")
library(PSMix) 

#Convert to character matrix.  The apply function applies a function (in this case as.character()) either to every column or every row in the data.
temp2.t.all.c <- apply(temp2.t.all,2,as.character)

#Create a new Matrix to hold reformatted data
temp2.t.all.ps <- matrix("",nrow=nrow(temp2.t.all.c)*2,ncol=ncol(temp2.t.all.c)) # this is an empty matrix 
temp2.t.all.ps[1:10, 1:10]

#for each row of genotypes, create 2 rows, one with the first allele and one with the second allele. 
for (i in 1:nrow(temp2.t.all.c)) {
  temp2.t.all.ps[(i-1)*2+1,] <- substr(temp2.t.all.c[i,],1,1)
  temp2.t.all.ps[(i-1)*2+2,] <- substr(temp2.t.all.c[i,],2,2)
  } # do I need varities name here??? 

temp2.t.all.c[1:10, 1:10] 
temp2.t.all.ps[1:10, 1:10] 
dim(data.geno.2500.ps) # 826 2500 

geno.x <- data.geno.2500.ps
geno.y <- rownames(data.geno.2500)

dim(geno.x) # 826 2500
length(geno.y) # 413

# PSMix practice ... 
Geno <- geno.x

# Infer population structure with first 30 markers
a = PSMix(K=4,Geno[,1:300],eps=1e-8,seed=as.integer(1+abs(rcauchy(1,scale=180))), MarkerVar=FALSE )
t( round(a[[1]],3) ) 

# Infer population structure with all markers
a = PSMix( K=4,Geno,eps=1e-8,seed=as.integer(1+abs(rcauchy(1,scale=180))), MarkerVar=FALSE )
t( round(a[[1]],3) ) 

# examine the output 
ps4 <- a
names(ps4)
head(ps4$AmPr)
round(head(ps4$AmPr),3) #round to 3 decimal places to make it easier to see

head(ps4$AmId) # assignment of each individual to each population 
table(ps4$AmId)

### reformat for plot
ps4.df <- as.data.frame(cbind(round(ps4$AmPr,3), ps4$AmId))
head(ps4.df) #look at the new data frame

#Next add useful column names
colnames(ps4.df) <- c(paste("pop",1:(ncol(ps4.df)-1),sep=""),"popID")
head(ps4.df) #look at the new data frame

# calculate the largest proportion of the genome from a single ancestral population for each individual. 
maxGenome <- apply(ps4$AmPr,1,max) 

#now we order the varieties by their prediced population membership and their degree of admixture.
ps4.df <- ps4.df[order(ps4.df$popID,-maxGenome),]
#Add a column for sample index
ps4.df$sampleID <- factor(1:413)
head(ps4.df)

library(reshape2)
ps4.df.melt <- melt(ps4.df,id.vars=c("popID","sampleID"))
head(ps4.df.melt) #look a the melted data set.

library(ggplot2)

pl <- ggplot(aes(x=sampleID, y=value, color=variable, fill=variable), data=ps4.df.melt)
pl <- pl + geom_bar(stat="identity") 
pl <- pl + ylab("genome proportion") + scale_color_brewer(type="div") + scale_fill_brewer(type="div")

pl

geno.mds$popID <- factor(ps4$AmId) #this only works because the order of the varieties is the same in geno.mds and ps4
head(geno.mds$popID)
#Also add the admixture proportions for future use
#first give them some useful names
colnames(ps4$AmPr) <- paste("pr",1:4,sep="")

geno.mds <- cbind(geno.mds,ps4$AmPr)
head(geno.mds)

save(data.pheno,geno.mds,file="data_from_SNP_lab.Rdata")
```




# transfer data to local computer 
```{r}
dim(temp2.t.all)
write.csv(temp2.t.all, file="~/505/output/gt_data.csv") 
```

### import into GAPIT for population structure (do this on local computer)
```{r}
# source("http://www.bioconductor.org/biocLite.R")
# biocLite("multtest")
# install.packages("gplots")
# install.packages("LDheatmap")
# install.packages("genetics")
# install.packages("EMMREML") 
# install.packages("scatterplot3d")

library(multtest)
library(gplots)
library(LDheatmap)
library(genetics)
library(EMMREML)
library(compiler) #this library is already installed in R
library("scatterplot3d")

# install GAPIT package 
source("http://zzlab.net/GAPIT/gapit_functions.txt")
# intall EMMA package 
source("http://zzlab.net/GAPIT/emma.txt")

# set working directory and start analysis 
setwd("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT")






```


```{r}
# chromosome distribution 
distri <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/CHROM_POS")
head(distri)
colnames(distri) <- c("number", "CHROM", "POS")

class(distri)
data <- distri
head(data)

# below the block of code should go into function... 
chr_ID <- unique(data$CHROM)
number.unique <- list() 

for (chr in chr_ID) {
  number.unique[chr] <- sum(sum(data$CHROM==chr))
  print(number.unique[chr])
}

SNP_chr.unique <- as.data.frame(t(as.data.frame(number.unique)))
SNP_chr.unique$ID <- rownames(SNP_chr.unique)
head(SNP_chr.unique)
# split to main & random chromosomes 
# main 
SNP_chr_main.unique <- SNP_chr.unique[!grepl("random", SNP_chr.unique$ID),]  
head(SNP_chr_main.unique)
SNP_chr_main.unique$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", SNP_chr_main.unique$ID)   
SNP_chr_main.unique$chr_ID <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", SNP_chr_main.unique$ID) 

# do a little calculation: how many SNPs on main & random chromosomes? 
sum(SNP_chr_main.unique$V1) # 62958     
head(SNP_chr_main.unique) # 200317    

sum(SNP_chr_main.unique[SNP_chr_main.unique$subgenome=="A",]$V1)/sum(SNP_chr_main.unique$V1)

# make a plot to see how many SNPs on each chromosome 
library(ggplot2)
pl.SNP.main.1.unique <- ggplot(data = SNP_chr_main.unique) 
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + geom_bar(aes(x=factor(chr_ID), y = V1, fill=subgenome), stat = "identity")
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + facet_wrap(~subgenome, ncol = 1)
# pl.SNP.main.1 <- pl.SNP.main.1 + geom_text(aes(x=factor(chr_ID), y = number, label=factor(number), size=1)) 
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + theme(legend.position = "none")
pl.SNP.main.1.unique 
ggsave("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/SNP.main.png", width = 9, height = 7) 

########## Per Mb ... 
head(data)
data <- data[!grepl("random", data$CHROM),]  
data$subgenome <-  gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", data$CHROM) 
data$Chr_ID <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", data$CHROM) 
head(data) 

# write.csv(vcf.Ae.Ol.intersect.df.2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/vcf.Ae.Ol.intersect.df.2.csv") # sort in excel manually 

# cf.Ae.Ol.intersect.df.2.sorted <-  read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/vcf.Ae.Ol.intersect.df.2.csv")
head(vcf.Ae.Ol.intersect.df.2.sorted) 

library(ggplot2)
pl.SNP.main.2.unique <- ggplot(data = data)
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + geom_histogram(aes(x=POS, fill=subgenome), binwidth = 1000000) 
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + facet_grid(Chr_ID ~subgenome)
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + theme(legend.position = "none")
pl.SNP.main.2.unique 

ggsave("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/SNP.main.PerMb.png", width = 9, height = 7) 

```

# late silique data
```{r}
# compare late silique & young leaf samples 
sample_des_late_silique <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/late_silique/505 collection_Sequencing_Statistics_Result.csv", header = T)

sample_des_a <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/Sequencing_Statistics_Result_TBD170001_1617.csv", header = T)

sample_des_b <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/Sequencing_Statistics_Result_TBD170026.csv", header = T)

sample_des_a$X_1 <- as.character(sample_des_a$X_1)
sample_des_b$Name <- as.character(sample_des_b$Name)

name_young_leaf <- c(sample_des_a$X_1, sample_des_b$Name)
sample_des_late_silique$Name <- as.character(sample_des_late_silique$Name)
name_late_silique <- sample_des_late_silique$Name

common <- intersect(name_late_silique, name_young_leaf) # only 29 samples have exactly the same ID between young leaf & late silique data... 
common

# get the unique ones in late silique, are they having slightly differnet names in young leaf??? 
name_late_silique
common

setdiff(name_late_silique, common) # setdiff will discard duplicated values, so 3 DG5 lines were discarded here... 
?setdiff

name_young_leaf # 94
name_late_silique # 47 

unique(name_young_leaf) # 90 unqiue 
unique(name_late_silique) # 45 unique ### DG5 winter type... 

name_late_silique[duplicated(name_late_silique)]
name_young_leaf[duplicated(name_young_leaf)]

### add * to duplicated items
sample_des_c <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/late_silique/batch_C.csv", header = T, na.strings = "")
tail(sample_des_c)

sample_des_b <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/late_silique/batch_B.csv", header = T,na.strings = "")

sample_des_a <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/late_silique/batch_A.csv", header = T,na.strings = "")
tail(sample_des_a)

sample_des_a$feature <- paste(sample_des_a$Name, sample_des_a$Sample.Description.Origin...Type., sample_des_a$X, sep = ";")
sample_des_b$feature <- paste(sample_des_b$Name, sample_des_b$Sample.Description.Origin...Type., sample_des_b$X, sep = ";")
sample_des_c$feature <- paste(sample_des_c$Name, sample_des_c$Sample.Description.Origin...Type., sample_des_c$X, sep = ";")

all.new <- c(sample_des_a$feature, sample_des_b$feature, sample_des_c$feature)
all.new.2 <- all.new[which(!is.na(all.new))]

dup.new <- all.new.2[duplicated(all.new.2)]
dup.new # duplicated lines 
dup

sample_des_a$Name <- as.character(sample_des_a$Name)
sample_des_b$Name <- as.character(sample_des_b$Name)
sample_des_c$Name <- as.character(sample_des_c$Name)

all <- c(sample_des_a$Name, sample_des_b$Name, sample_des_c$Name)
all.2 <- all[which(!is.na(all))]

dup <- all.2[duplicated(all.2)] 
dup # duplicated lines 

sample_des_a$dup <- sample_des_a$Name %in% dup
sample_des_b$dup <- sample_des_b$Name %in% dup
sample_des_c$dup <- sample_des_c$Name %in% dup

write.csv(sample_des_a, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_a.csv")
write.csv(sample_des_b, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_b.csv")
write.csv(sample_des_c, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_c.csv")
```








