---
title: "overlay_sig_loci_Erucic_Oleic"
output: html_document
---

Purpose of this script is to 
1) visualize how fatty acid biosynthesis and metabolic related genes overlay with the significant SNPs that I identified from QTL and GWAS analysis. 
2) visualize how previous identified loci overlay with the significant SNPs and loci that I identified from GWAS and QTL analysis. 

### import lib
```{r}
library(tidyverse)
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures)
library("rtracklayer")
# source("https://bioconductor.org/biocLite.R")
# biocLite("ggbio")
library(ggbio) 
library(dplyr)
```

### import GWAS and QTL analysis CIM result 
```{r}
setwd("~/Desktop/Brassica_project/KIAT_RNA_seq/overlay/")

# QTL result 
load("cim.qtl.43traits.Rdata")
load("cim.perm.43traits.Rdata")
cim.qtl.sig.Erucic <- cim.qtl$Erucic_acid[cim.qtl$Erucic_acid$lod > cim.perm$Erucic_acid,]
cim.qtl.sig.Oleic <- cim.qtl$Oleic_acid[cim.qtl$Oleic_acid$lod > cim.perm$Oleic_acid,]

# GWAS result 
Erucic_acid <- read.csv("GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("GAPIT..Oleic_acid.GWAS.Results.csv", header = T) 

Erucic_Oleic <- rbind(Erucic_acid, Oleic_acid)

SNP.Erucic.Oleic <- data.frame(CHROM = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\2", Erucic_Oleic[Erucic_Oleic$P.value < 0.00001,]$SNP), 
                         POS = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\4", Erucic_Oleic[Erucic_Oleic$P.value < 0.00001,]$SNP)) 

SNP.Erucic.Oleic$CHROM <- paste("chr", SNP.Erucic.Oleic$CHROM, sep = "")
SNP.Erucic.Oleic$POS <- as.numeric(as.character(SNP.Erucic.Oleic$POS))

SNP.Erucic.Oleic.unique <- unique(SNP.Erucic.Oleic)
dim(SNP.Erucic.Oleic.unique) # 235 2 
```

### combine and display Erucic and Oleic acid GWAS and QTL analysis result 
```{r}
SNP.Erucic.Oleic 
```

### demo of simple example on how to use ggbio to draw plot of genomic data 
```{r}
data(hg19IdeogramCyto, package = "biovizBase")
head(hg19IdeogramCyto)
getOption("biovizBase")$cytobandColor
autoplot(hg19IdeogramCyto, layout = "karyogram", cytoband = TRUE)

library(GenomicRanges)
hg19 <- keepSeqlevels(hg19IdeogramCyto, paste0("chr", c(1:22, "X", "Y")))
head(hg19)
autoplot(hg19, layout = "karyogram", cytoband = TRUE) 
gffRangeData <- import.gff("~/Desktop/Brassica_project/IGV_file/Brassica_napus.annotation_v5.gff3")
gffRangeData
gff<-as(gffRangeData, "GRanges")
test <- head(gff, 50000)
autoplot(test, layout = "karyogram", aes(color = strand))  

# more demo 
library(devtools)
install_github("genomicsclass/ERBS")
library(ERBS)
data(HepG2)
library(GenomeInfoDb)  # trim all but autosomal chroms
seqlevels(HepG2, force=TRUE) = paste0("chr", 1:22)
data(GM12878)
seqlevels(GM12878, force=TRUE) = paste0("chr", 1:22)
library(ggbio)
autoplot(HepG2, layout="karyogram", main="ESRRA binding on HepG2")

autoplot(GM12878, layout="karyogram", main="ESRRA binding on GM12878")

data(GM12878)
seqlevels(GM12878, force=TRUE) = paste0("chr", 1:22)
autoplot(GM12878, layout="karyogram", aes(colour=log(peak)))

GM12878

HepG2$cell = "HepG2"
GM12878$cell = "Bcell"
tot = c(GM12878, HepG2)
tot$peak10 = tot$peak/10 # copes with automatic scale of y axis
seqlevels(tot, force=TRUE) = paste0("chr", 1:22)
library(scales)

p = autoplot(seqinfo(tot))
p = p + layout_karyogram(tot, aes(fill=cell, colour=cell), geom="rect") +
    scale_colour_manual(values = alpha(c("green", "red"), .1)) +
    scale_fill_manual(values = alpha(c("green", "red"), .1))
p

p + layout_karyogram(tot, aes(x=start, y=peak10), ylim=c(15,30),
    geom="point", color="blue", size=.2) 
tot  
``` 

### draw similar plot for the overlay with sinificant loci and fatty acid genes 
```{r}
# import Grange data of B.napus 
gff.mRNA <- read.table("~/Desktop/Brassica_project/IGV_file/gff.mRNA")
colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 

# import fatty acid genes info (based on GO and IPR info)
lipid_fatty_acid_genes <- read.table("lipid_fatty_acid_genes", header = F, stringsAsFactors = FALSE)
lipid_fatty_acid_genes %>% dim()  # 7346 
colnames(lipid_fatty_acid_genes) <- "name"

lipid_fatty_acid_genes.2 <- 
gff.mRNA %>% 
  inner_join(lipid_fatty_acid_genes, by = "name") %>% 
  mutate(name = "lipid_fatty_acid_genes")

lipid_fatty_acid_genes.2 %>% dim() # 7346    4 

SNP.Erucic.Oleic.2 <-  
SNP.Erucic.Oleic.unique %>% 
  mutate(start = POS, end = POS, name = "SNP") %>% 
  dplyr::select(-POS)

lipid_fatty_acid_genes.Erucic_Oleic <- 
rbind(lipid_fatty_acid_genes.2, SNP.Erucic.Oleic.2)

gff.lipid_fatty_acid_genes.Erucic_Oleic <- GRanges(seqnames = Rle(lipid_fatty_acid_genes.Erucic_Oleic$CHROM),ranges = IRanges(start = lipid_fatty_acid_genes.Erucic_Oleic$start, end = lipid_fatty_acid_genes.Erucic_Oleic$end), names = lipid_fatty_acid_genes.Erucic_Oleic$name)
 
# autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names))
# this plot shows that if just looking at all lipid and fatty acid genes, there are so many that not meaningful to check the result 

### only focus on several lipid and fatty acid related GO terms
# fatty acid beta-oxidation
# lipid transporter activity
# fatty acid biosynthetic process
# fatty acid metabolic process
# lipid metabolic process
# lipid catabolic process
# acetyl-CoA carboxylase complex
# lipid transport
# lipid biosynthetic process
# fatty-acyl-CoA biosynthetic process
# lipid binding

lipid_fatty_acid_GO <- read.table("~/Desktop/Brassica_project/reference/fatty_acid_lipd_GO_11_12_2017", sep = "\"", stringsAsFactors = F)
lipid_fatty_acid_GO <- 
  lipid_fatty_acid_GO %>% 
  dplyr::select(c(V2, V10)) %>% 
  mutate(GO = V2) 

head(lipid_fatty_acid_GO)

napus_GO <- read.table("~/Desktop/Brassica_project/reference/Brassica_napus_GO", stringsAsFactors = F)

napus_GO <- 
  napus_GO %>% 
  mutate(GO = V2)

head(napus_GO)

lipid_fatty_acid_genes <- 
lipid_fatty_acid_GO %>% 
  left_join(napus_GO, by = "GO") %>% 
  mutate(GOterm = V10, name=V1) %>% 
  dplyr::select(c(GOterm, name, GO)) 

dim(lipid_fatty_acid_genes) # 1298
head(lipid_fatty_acid_genes)
save(lipid_fatty_acid_genes, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/lipid_fatty_acid_genes.Rdata")
```

### plot out the new lipid fatty acid gene list 
```{r}
lipid_fatty_acid_genes.2 <- 
gff.mRNA %>% 
  inner_join(lipid_fatty_acid_genes, by = "name") %>% 
  mutate(name = GOterm) %>% 
  dplyr::select(c(CHROM:name))

lipid_fatty_acid_genes.2 %>% dim() # 1298    4 
head(lipid_fatty_acid_genes.2)

# lipid_fatty_acid_genes.Erucic_Oleic <- 
# rbind(lipid_fatty_acid_genes.2, SNP.Erucic.Oleic.2)

lipid_fatty_acid_genes.Erucic_Oleic <- lipid_fatty_acid_genes.2

gff.lipid_fatty_acid_genes.Erucic_Oleic <- GRanges(seqnames = Rle(lipid_fatty_acid_genes.Erucic_Oleic$CHROM),ranges = IRanges(start = lipid_fatty_acid_genes.Erucic_Oleic$start, end = lipid_fatty_acid_genes.Erucic_Oleic$end), names = lipid_fatty_acid_genes.Erucic_Oleic$name) 
 
#  [1] "lipid metabolic process"             "lipid biosynthetic process"         
#  [3] "fatty acid biosynthetic process"     "lipid catabolic process"            
#  [5] "fatty acid metabolic process"        "fatty acid beta-oxidation"          
#  [7] "lipid transport"                     "lipid binding"                      
#  [9] "fatty-acyl-CoA biosynthetic process" "acetyl-CoA carboxylase complex"     
# [11] "lipid transporter activity"          "SNP"   
autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names)) 
ggsave(filename = "~/Desktop/pl.fatty_acid_genes.png", width = 13, height = 8)

save(lipid_fatty_acid_genes.2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/lipid_fatty_acid_genes.2.Rdata")
save(SNP.Erucic.Oleic.2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/SNP.Erucic.Oleic.2.Rdata") 
```

### continue work on this plot 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/lipid_fatty_acid_genes.2.Rdata")

# rename CHROM from normal name to different name 
gff.lipid_fatty_acid_genes.Erucic_Oleic
autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "SNP"))  

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "lipid metabolic process"), legend = FALSE) 

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "fatty acid biosynthetic process"), alpha = 0.1) 

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "lipid catabolic process"), alpha = 0.1) 

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "fatty acid metabolic process"), alpha = 0.1) 

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "fatty acid beta-oxidation"), alpha = 0.1) 

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "lipid transport"), alpha = 0.1) 

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "lipid binding"), alpha = 0.1) 

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "fatty-acyl-CoA biosynthetic process"), alpha = 0.1) 

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "acetyl-CoA carboxylase complex"), alpha = 0.1)

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "fatty-acyl-CoA biosynthetic process"), alpha = 0.1)

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "acetyl-CoA carboxylase complex"), alpha = 0.1)

autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names == "lipid transporter activity"), alpha = 0.1)  
```

### list fatty acid genes within 1Mb of the SNP loci boundary 
```{r}
# genes under the significant loci 
load("~/Desktop/2017_Nov_KIAT_report/SNP_gene_final_fxn.2.Rdata")
View(SNP_gene_final_fxn.2) 

# need to get genes within 1Mb of the significant loci 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/lipid_fatty_acid_genes.Rdata")
load("~/Desktop/2017_Nov_KIAT_report/SNP.Erucic.Oleic.Unique.Rdata")

lipid_fatty_acid_genes.3 <-
gff.mRNA %>% 
  inner_join(lipid_fatty_acid_genes, by = "name") 

lipid_fatty_acid_genes.3 %>% dim() # 1298    6 
head(lipid_fatty_acid_genes.3) 
gff.lipid_fatty_acid_genes <- GRanges(seqnames = Rle(lipid_fatty_acid_genes.3$CHROM),ranges = IRanges(start = lipid_fatty_acid_genes.3$start, end = lipid_fatty_acid_genes.3$end), names = lipid_fatty_acid_genes.3$name, GO = lipid_fatty_acid_genes.3$GOterm)  
gff.lipid_fatty_acid_genes

dim(SNP.Erucic.Oleic.unique) # 235 
SNP.Erucic.Oleic.unique %>%
  filter(CHROM=="chrA08") %>%
  arrange(desc(POS)) # range 6985514 to 11476346 

SNP.Erucic.Oleic.unique %>%
  filter(CHROM=="chrC03") %>%
  arrange(desc(POS))  # range 52598214 to 56930228   

# on the basis of range, add 1000000 (1Mb)
# get range as an range object 
SNP.range <- GRanges(seqnames = Rle(c("chrA08", "chrC03")), ranges = IRanges(start = c(6985514 - 1000000, 52598214 - 1000000), end = c(11476346 + 1000000, 56930228 + 1000000)))
SNP.range

SNP_range_gene <- mergeByOverlaps(SNP.range, gff.lipid_fatty_acid_genes)
SNP_range_gene %>% as.data.frame()
View(SNP_range_gene)[1:26] 

# also need Arabidopsis homolog name 
napus_vs_ara.non_reciprocal <- read.table("~/Desktop/Brassica_project/reference/napus_vs_ara.non_reciprocal.table")

head(napus_vs_ara.non_reciprocal)
dim(napus_vs_ara.non_reciprocal) # 64949 

colnames(napus_vs_ara.non_reciprocal)[1] <- "names"

SNP_gene_final_fxn <- merge(SNP_range_gene, napus_vs_ara.non_reciprocal, by = "names", all.x=T)
SNP_gene_final_fxn
View(SNP_gene_final_fxn) 

colnames(SNP_gene_final_fxn)

SNP_gene_final_fxn <- SNP_gene_final_fxn[,c("SNP.range.seqnames", "SNP.range.start", "SNP.range.end", "gff.lipid_fatty_acid_genes.start", "gff.lipid_fatty_acid_genes.end", "gff.lipid_fatty_acid_genes.names", "gff.lipid_fatty_acid_genes.GO", "ara_ID", "description")]

View(SNP_gene_final_fxn)
save(SNP_gene_final_fxn, file = "~/Desktop/2017_Nov_KIAT_report/SNP_gene_final_fxn.Rdata")
```

### check expression of these genes between Da-Ae and Da-Ol-1 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/output/parent/DEgene.new.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/output/parent/DEgene.pairwise.gt.Rdata") # pairwise 

dim(DEgene.new.gt) # 15098     9 
dim(DEgene.new.interaction) # 1701    8 
dim(DEgene.new.tissue) # 22580     8 

dim(DEgene.young); dim(DEgene.bolting); dim(DEgene.flowering); dim(DEgene.early.silique); dim(DEgene.late.silique)
# [1] 8976    5
# [1] 4648    5
# [1] 6081    5
# [1] 6695    5
# [1] 7413    5

### make a list 
DE.list <- list(gt = DEgene.new.gt, tissue = DEgene.new.tissue, interaction = DEgene.new.interaction, young = DEgene.young, bolting = DEgene.bolting, flowering = DEgene.flowering, early_silique = DEgene.early.silique, late_silique = DEgene.late.silique)
length(DE.list) # 8 

load("~/Desktop/2017_Nov_KIAT_report/SNP_gene_final_fxn.Rdata")
SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% rownames(DEgene.new.gt)]

test1 <- 
SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% rownames(DEgene.new.gt)] %>% unique()

test2 <- 
SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% rownames(DEgene.new.tissue)] %>% unique()

test3 <- 
SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% rownames(DEgene.new.interaction)] %>% unique()

c(test1, test2, test3) %>% unique() %>% length()

test4 <- 
SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% rownames(DEgene.young)] %>% unique()

test5 <- 
SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% rownames(DEgene.bolting)] %>% unique()

test6 <- 
SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% rownames(DEgene.flowering)] %>% unique()

test7 <- 
SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% rownames(DEgene.early.silique)] %>% unique()

test8 <- 
SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% rownames(DEgene.late.silique)] %>% unique()

DE.gt.lipid_fatty_acid <- c(test4, test5, test6, test7, test8) %>% unique()

### why some genes are significant for gt term but not differentially expressed between any tissue type in pairwise analysis 

genes.GWAS <- DE.gt.lipid_fatty_acid
genes.GWAS.df <- as.data.frame(genes.GWAS)
colnames(genes.GWAS.df) <- "V1"
genes.GWAS.df 

expression.pattern.Bn.parent.bar(vstMat.parent = vstMat.parent, gene = genes.GWAS.df) 
ggsave("~/Desktop/2017_Nov_KIAT_report/DE.gt.lipid_fatty_acid.png", width = 7, height = 5) 

temp.temp <- 
SNP_gene_final_fxn[SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names %in% genes.GWAS.df$V1,][,c("gff.lipid_fatty_acid_genes.names", "gff.lipid_fatty_acid_genes.GO", "description")] 
  
colnames(temp.temp) <- c("gene_ID", "GO_term", "Arabidopsis description")

View(temp.temp)
View(SNP_gene_final_fxn)
``` 

### plot out the distribution of sinigicant SNPs identified from published paper and my study
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/SNP.Erucic.Oleic.2.Rdata")
SNP.Erucic.Oleic.2 %>% dim() # 235 

SNP.Erucic.Oleic.known <- read.csv("~/Desktop/Erucic_Oleic_known_sig_loci.csv")
SNP.Erucic.Oleic.known$seq.method %>% unique()

colnames(SNP.Erucic.Oleic.known)

SNP.Erucic.Oleic.known <- SNP.Erucic.Oleic.known[,c("CHROM", "start", "end")]
dim(SNP.Erucic.Oleic.known) # 43 3 
SNP.Erucic.Oleic.known$CHROM <- paste("chr", SNP.Erucic.Oleic.known$CHROM, sep = "")
SNP.Erucic.Oleic.known$name <- rep("SNP_known", nrow(SNP.Erucic.Oleic.known))

SNP.Erucic.Oleic.known$SNP_ID <- paste(SNP.Erucic.Oleic.known$CHROM, SNP.Erucic.Oleic.known$start, sep = "_")
SNP.Erucic.Oleic.2$SNP_ID <- paste(SNP.Erucic.Oleic.2$CHROM, SNP.Erucic.Oleic.2$start, sep = "_")

head(SNP.Erucic.Oleic.known)
head(SNP.Erucic.Oleic.2)  

# no overlapping single SNPs 
sum(SNP.Erucic.Oleic.2$SNP_ID %in% SNP.Erucic.Oleic.known$SNP_ID)

SNP.Erucic.Oleic.all <- rbind(SNP.Erucic.Oleic.2, SNP.Erucic.Oleic.known)

gff.Erucic_Oleic_all <- GRanges(seqnames = Rle(SNP.Erucic.Oleic.all$CHROM),ranges = IRanges(start = SNP.Erucic.Oleic.all$start, end = SNP.Erucic.Oleic.all$end), names = SNP.Erucic.Oleic.all$name) 

gff.Erucic_Oleic_all
autoplot(gff.Erucic_Oleic_all, layout = "karyogram", aes(color = names)) 
ggsave(filename = "~/Desktop/pl.Erucic_Oleic_all.png", width = 13, height = 8) 
```

# sessioninfo
```{r}
# source("https://bioconductor.org/biocLite.R")
# biocLite("ggbio")
library(ggbio)
data(hg19IdeogramCyto, package = "biovizBase")
head(hg19IdeogramCyto)
# sessionInfo() 
# R version 3.2.4 (2016-03-10)
# Platform: x86_64-apple-darwin13.4.0 (64-bit)
# Running under: OS X 10.12.6 (unknown)
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# attached base packages:
# [1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] dplyr_0.7.4             purrr_0.2.4             readr_1.1.1            
#  [4] tidyr_0.7.2             tibble_1.3.4            ggplot2_2.2.1          
#  [7] tidyverse_1.1.1         devtools_1.13.4         rtracklayer_1.30.4     
# [10] GenomicFeatures_1.22.13 AnnotationDbi_1.32.3    Biobase_2.30.0         
# [13] GenomicRanges_1.22.4    GenomeInfoDb_1.6.3      IRanges_2.4.8          
# [16] S4Vectors_0.8.11        BiocGenerics_0.16.1    
# 
# loaded via a namespace (and not attached):
#  [1] httr_1.3.1                 bit64_0.9-7                jsonlite_1.5              
#  [4] modelr_0.1.1               assertthat_0.2.0           blob_1.1.0                
#  [7] cellranger_1.1.0           Rsamtools_1.22.0           RSQLite_2.0               
# [10] lattice_0.20-35            glue_1.2.0                 digest_0.6.12             
# [13] XVector_0.10.0             rvest_0.3.2                colorspace_1.3-2          
# [16] plyr_1.8.4                 psych_1.7.8                XML_3.98-1.9              
# [19] pkgconfig_2.0.1            broom_0.4.2                biomaRt_2.26.1            
# [22] haven_1.1.0                zlibbioc_1.16.0            scales_0.5.0              
# [25] BiocParallel_1.4.3         withr_2.1.0                SummarizedExperiment_1.0.2
# [28] lazyeval_0.2.1             mnormt_1.5-5               magrittr_1.5              
# [31] crayon_1.3.4               readxl_1.0.0               memoise_1.1.0             
# [34] nlme_3.1-131               forcats_0.2.0              xml2_1.1.1                
# [37] foreign_0.8-69             tools_3.2.4                hms_0.3                   
# [40] stringr_1.2.0              munsell_0.4.3              lambda.r_1.2              
# [43] bindrcpp_0.2               Biostrings_2.38.4          rlang_0.1.4               
# [46] futile.logger_1.4.3        grid_3.2.4                 RCurl_1.95-4.8            
# [49] rstudioapi_0.7             bitops_1.0-6               gtable_0.2.0              
# [52] curl_3.0                   DBI_0.7                    reshape2_1.4.2            
# [55] R6_2.2.2                   GenomicAlignments_1.6.3    lubridate_1.7.1           
# [58] knitr_1.17                 bit_1.1-12                 bindr_0.1                 
# [61] futile.options_1.0.0       stringi_1.1.5              Rcpp_0.12.13  
```

