---
title: "eQTL_batch_corrected"
output: html_document
---

Purpose of this script is to find loci controlling expression of different genes. We will firstly focus on fatty acid related genes. this version does not do scale and center, also batch effect were corrected. 

### import lib
```{r} 
library(tidyverse) 
library(qtl)
library(snowfall) 
library(ggrepel) 
library(Biostrings) 
source("~/KIAT/function_BnRNAseq.R") 
```

### import data 
```{r}
load("~/F2/output/network_analysis/vstMat.f2.batch.corrected.Rdata")
dim(vstMat.f2.batch.corrected) # 56180   166 
vstMat.f2.batch.corrected[1:10, 1:10]
write.csv(vstMat.f2.batch.corrected, file = "~/F2/output/network_analysis/vstMat.f2.batch.corrected.csv")
### modify this expression file to the right format 

cross.F2 <- read.cross("csvsr", genfile =  "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised.csv", 
                         phefile = "~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv",
                         genotypes = c("AA", "AB", "BB"))   # although the sample IDs are not matched in the original phe and gen file, I still get the right result. 

cross.F2 <- sim.geno(cross.F2,step=1,n.draws=32) # imputation?  
cross.F2 <- calc.genoprob(cross.F2,step=1) 

scanone_eQTL.F2 <- scanone(cross.F2, pheno.col = 2:ncol(cross.F2$pheno), 
	         method = "imp", use = "all.obs") 

### permutation...  
set.seed(12345)   
permtest.F2 <- scanone(cross.F2, pheno.col = 2, method = "imp", n.perm = 10000)  

alphas <- seq(0.01, 0.10, by = 0.01) 
lod.thrs <- summary(permtest.F2, alphas) 
lod.thrs   

save(cross.F2, scanone_eQTL.F2, permtest.F2, lod.thrs, file = "~/F2/output/eQTL/scanone-eqtl_F2.RData") 
```

### determine cis- and trans- eQTL 
```{r}
load("~/F2/output/eQTL/scanone-eqtl_F2.RData")
scanone_eQTL.F2 %>% dim() # 4887 56182 

# get threshold 
threshold.95 <- lod.thrs[5,]

# get all eQTL based on this threshold  
eQTL_sign <- 
sapply(colnames(scanone_eQTL.F2), function(gene) {
  sum(scanone_eQTL.F2[,gene] > threshold.95) > 0 
}) 

sum(eQTL_sign) # 20,172 genes with eQTL 
scanone_eQTL.F2 <- scanone_eQTL.F2[,eQTL_sign]  
dim(scanone_eQTL.F2) # 4887 20174 

# get bayesint result for every gene 
scanone.gather <-  
scanone_eQTL.F2 %>% 
  gather(key = trait, value = LOD, -chr, -pos) 

sig.chrs <- scanone.gather %>% filter(LOD > threshold.95) %>%
  group_by(trait,chr) %>% 
  dplyr::summarise(count = n()) # this is to get the significant chr ID for each trait 

sig.chrs 
sig.chrs %>% dim() # 22381        3

bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(scanone_eQTL.F2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE 
  )
  colnames(result)[3] <- "LOD" 
  result
}) 

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x) 
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr))
) 

bayesint.result <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>%  
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T), pos=median(pos, na.rm = T), min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(start=ifelse(start==end, start-20000,start), end=ifelse(start==end,end+20000,end)) 
  # mutate(start=start-1000000, end=end+1000000) 
# the original code (below) doesn't work correctly for some reason 
# mutate(start=ifelse(start==end,max(0,start-20000),start), end=ifelse(start==end,end+20000,end)) 

bayesint.result %>% dim() # 22381     7 

# get genome range 
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures)
library("rtracklayer")

### get gff file with gene chrom & pos info, gff3 file must be sorted 
gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("CHROM", "start", "end", "gene_ID") 
gff.mRNA %>% head()  

# look for cis-eQTL 
bayesint.result.2 <- 
bayesint.result %>% 
  mutate(gene_ID = trait, eQTL_chr = chr, eQTL_start = start, eQTL_end = end) %>% 
  dplyr::select(trait, gene_ID, eQTL_chr, pos, eQTL_start, eQTL_end, min_eQTL_LOD, max_eQTL_LOD) %>% 
  left_join(gff.mRNA, by = "gene_ID") 

bayesint.result.2$eQTL_chr <- paste("chr", bayesint.result.2$eQTL_chr, sep = "")
bayesint.result.2 %>% head()

cis_eQTL <- 
bayesint.result.2 %>% 
  filter(eQTL_chr == CHROM) %>% 
  filter((start < eQTL_start & end > eQTL_start) | # also need SNP pos... 
         (start >= eQTL_start & end <= eQTL_end) |
         (start < eQTL_end & end > eQTL_end)) 

dim(cis_eQTL) # 8512   11 

trans_eQTL <- 
bayesint.result.2 %>% 
  anti_join(cis_eQTL) 

save(cis_eQTL, trans_eQTL, file = "~/F2/output/eQTL/cis_trans_result_new.Rdata")
```

### find overlaps between eQTL & trait QTL  
```{r}
# trait QTL using scanone result 
load("~/F2/output/QTL_analysis/traitQTL.annotated.Rdata")
load("~/F2/output/eQTL/cis_trans_result_new.Rdata") 

cis_eQTL %>% dim() # 8512   11 
trans_eQTL %>% dim() # 13869    11 

cis_eQTL.qtl.combined <- inner_join(cis_eQTL,traitQTL.annotated,by=c("gene_ID"="name")) 
cis_eQTL.qtl.combined %>% head()    
cis_eQTL.qtl.combined %>% dim() # 2484   18 

colnames(cis_eQTL.qtl.combined) 

trans_eQTL.qtl.combined <- inner_join(trans_eQTL,traitQTL.annotated,by=c("gene_ID"="name")) 
trans_eQTL.qtl.combined %>% head()    
trans_eQTL.qtl.combined %>% dim() # 2165    18  
  
napus_GO <- read.table("~/Reference/B.napus/Brassica_napus_GO", stringsAsFactors = F)
colnames(napus_GO)[1] <- "gene_ID"
dim(napus_GO) # 150320 
# add GO description 
GO_annotation <- read_delim("~/Reference/B.napus/GO_annotation.csv", delim = "\"")

GO_annotation <- 
GO_annotation %>% 
  dplyr::select(input, name)

dim(GO_annotation) # 1659 2 
colnames(GO_annotation) <- c("GO", "GO_des")

# add GO annotation 
colnames(napus_GO)[2] <- "GO"
napus_GO <-
napus_GO %>% 
  left_join(GO_annotation, "GO")

napus_GO_combined_sub1 <- 
napus_GO %>% 
  group_by(gene_ID) %>% 
  dplyr::summarise(paste(GO, collapse = ";")) 

napus_GO_combined_sub2 <- 
napus_GO %>% 
  group_by(gene_ID) %>% 
  dplyr::summarise(paste(GO_des, collapse = ";")) 

napus_GO_combined <- 
napus_GO_combined_sub1 %>% 
  left_join(napus_GO_combined_sub2)

dim(napus_GO_combined) # 54264     3 
# napus_GO_combined %>% head() %>% View()

colnames(napus_GO_combined)[2:3] <- c("B.napus_GO_term", "GO_des")
save(napus_GO_combined, file = "~/Reference/B.napus/napus_GO_combined.Rdata")

colnames(cis_eQTL.qtl.combined) 

cis_eQTL.qtl.combined.final <- 
cis_eQTL.qtl.combined %>% 
  left_join(napus_GO_combined) %>% 
  mutate(eQTL_start = start.x, eQTL_end = end.x, QTL_start = start.y, QTL_end = end.y) %>% 
  dplyr::select(-c(start.x, end.x, start.y, end.y)) 
  
trans_eQTL.qtl.combined.final <- 
trans_eQTL.qtl.combined %>% 
  left_join(napus_GO_combined) %>% 
  mutate(eQTL_start = start.x, eQTL_end = end.x, QTL_start = start.y, QTL_end = end.y) %>% 
  dplyr::select(-c(start.x, end.x, start.y, end.y)) 

trans_eQTL.qtl.combined.final$B.napus_GO_term 
write.csv(cis_eQTL.qtl.combined.final, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/eQTL/cis_eQTL.qtl.combined.final.csv")

trans_eQTL.qtl.combined.final 
write.csv(cis_eQTL.qtl.combined.final, file = "~/F2/output/eQTL/cis_eQTL.qtl.combined.final.csv")

write.csv(trans_eQTL.qtl.combined.final, file = "~/F2/output/eQTL/trans_eQTL.qtl.combined.final.csv") 

### make some plot for several important genes 
# the two FAE 

# the two 
### make some table & plot for several important genes 
# 1) Fatty acid/flowering/growth related genes for cis-eQTL final list 
View(cis_eQTL.qtl.combined.final[grep("lipid|fatty", cis_eQTL.qtl.combined.final$GO_des),])  

# 2) transcription factor coding genes for cis-eQTL final list
View(cis_eQTL.qtl.combined.final[grep("DNA|nucleotide|transcription", cis_eQTL.qtl.combined.final$GO_des),])  

# 3) Fatty acid/flowering/growth related genes for trans-eQTL final list  
View(trans_eQTL.qtl.combined.final[grep("lipid|fatty", trans_eQTL.qtl.combined.final$GO_des),])   

# 4) figures for the two cis-eQTL gene 
# BnaA08g11060D & BnaA08g11140D two cis-eQTL gene for fatty acid, need to deal with the color problem... more colors than a single palette can handel... 
save(cis_eQTL.qtl.combined.final, trans_eQTL.qtl.combined.final, file = "~/F2/output/eQTL/eQTL.qtl.combined.final.Rdata")

### check whether FLC is cis-regulator for some the traits 
load("~/F2/output/eQTL/eQTL.qtl.combined.final.Rdata")
cis_eQTL.qtl.combined.final %>% 
  filter(AGI == "AT5G10140") %>% 
  View() # FLC is cis-regulator  
```

### make a plot of cis- and trans- eQTL, x-axis being eQTL physical position, y-axis being transcript start position (omit this part for now because I need to plot genetic position VS physical postion but for some the LG I flipped the physical order, need to discuss this with Julin)

* problem of this part: 1) flipped direction of genetic map VS physical map (flip it back and redo the analysis?); 2) many cis-like genes were identified as trans-eQTL because gene location is a little bit away from QTL loci (add 500kb flanking bp to the QTL loci?); 3) the cis-like distribution of trans-eQTL on homeologous chromosomes (the truth or misassembly?). 

```{r} 
library(ggplot2)
load("~/F2/output/eQTL/cis_trans_result_new.Rdata") 

cis_eQTL %>% dim() # 8512   11 
trans_eQTL %>% dim() # 13869    11 

cis_eQTL$class <- rep("cis", nrow(cis_eQTL)) 
trans_eQTL$class <- rep("trans", nrow(trans_eQTL)) 

# get only main chromosome stuff   
eQTL <- rbind(cis_eQTL, trans_eQTL) 
eQTL <- eQTL[grep("random", eQTL$CHROM, value = FALSE, invert = T),] 
eQTL$class <- factor(eQTL$class, levels = c("cis", "trans"))
eQTL$CHROM <- gsub("(chr)([[:print:]]+)", "\\2", eQTL$CHROM)
eQTL$eQTL_chr <- gsub("(chr)([[:print:]]+)", "\\2", eQTL$eQTL_chr)
eQTL$eQTL_chr <- factor(eQTL$eQTL_chr, levels = c("C09", "C08", "C07", "C06", "C05", "C04", "C03", "C02", "C01", "A10", "A09", "A08", "A07", "A06", "A05", "A04", "A03", "A02", "A01"))

# plot out the stuff 
p.eQTL <- 
eQTL %>% 
  ggplot() + 
  geom_point(aes(x = pos, y = start, color = class), size = 0.5) +
  facet_grid(eQTL_chr ~ CHROM, scales = "free", switch="both") + 
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "eQTL genetic position", y = "gene start")

p.eQTL     

p.cis_eQTL <- 
eQTL %>% 
  dplyr::filter(class == "cis") %>%
  ggplot() + 
  geom_point(aes(x = pos, y = start), size = 0.5) +
  facet_grid(eQTL_chr ~ CHROM, scales = "free", switch="both") + 
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "eQTL genetic position", y = "gene start")

p.cis_eQTL

p.trans_eQTL <- 
eQTL %>% 
  dplyr::filter(class == "trans") %>%
  ggplot() + 
  geom_point(aes(x = pos, y = start), size = 0.5) +
  facet_grid(eQTL_chr ~ CHROM, scales = "free", switch="both") + 
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "eQTL genetic position", y = "gene start") 

p.trans_eQTL 

p.eQTL_2mb # when I add 2Mb flanking bp to the 

ggsave(p.eQTL, filename = "~/F2/output/eQTL/figure/p.eQTL.png", height = 10, width = 10) 
ggsave(p.cis_eQTL, filename = "~/F2/output/eQTL/figure/p.cis_eQTL.png", height = 10, width = 10)
ggsave(p.trans_eQTL, filename = "~/F2/output/eQTL/figure/p.trans_eQTL.png", height = 10, width = 10) 
ggsave(p.eQTL_2mb, filename = "~/F2/output/eQTL/figure/p.eQTL_2mb.png", height = 10, width = 10) 
```

### tran-eQTL hotspot & GO enrichment for trans-eQTL 
* trans-eQTL hotspot better to be plotted from the distribution figure above, but since I cannot get that done correctly due to the problem of my current genetic map. trans-eQTL hotspot, hard to see... 

```{r}
load("~/F2/output/eQTL/cis_trans_result_new.Rdata")
# still need to make histogram to look for trans-eQTL hotspot 

cis_eQTL %>% dim() # 8512   11 
trans_eQTL %>% dim() # 13869    11 

trans_eQTL$sub <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", trans_eQTL$eQTL_chr) 

trans_eQTL$sub_chrom <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", trans_eQTL$eQTL_chr) 

p.transeQTL_hotspot <- 
  trans_eQTL %>% 
  ggplot() + 
  geom_histogram(aes(x = pos, fill = sub), binwidth = 10) + 
  facet_grid(sub_chrom ~ sub) +
  labs(list(title = "", x = "trans-eQTL loci", y = "number of trans-eQTL per cM")) +
  theme(legend.position = "none")

ggsave(p.transeQTL_hotspot, filename = "~/F2/output/eQTL/p.transeQTL_hotspot.png", height = 8, width = 12)
``` 

### GO enrichment of trans-eQTL for different phenotypic traits, do this on MAC 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/eQTL/eQTL.qtl.combined.final.Rdata")
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")

dim(cis_eQTL.qtl.combined.final) # 2484   18 
dim(trans_eQTL.qtl.combined.final) #2165   18

trans_eQTL.qtl.combined.final %>% colnames()
trans_eQTL.qtl.combined.final %>% head()

GO_trans_eQTL <- 
lapply(unique(trans_eQTL.qtl.combined.final$trait.y), function(trait) {
  test <- trans_eQTL.qtl.combined.final$trait.x[trans_eQTL.qtl.combined.final$trait.y == trait]
  tryCatch(
    GOseq.Bn.ORA(test), error= function(e) NA)
})

names(GO_trans_eQTL) <- unique(trans_eQTL.qtl.combined.final$trait.y) # no enriched GO terms, probably due to those cis-like trans-eQTL's existence.    
```










