---
title: "eQTL_batch_corrected_flipped_C05C08"
output: html_document
---

Goal of this script is to re-run eQTL analysis using reconstructed map with C05 and C08 modified 

### import lib
```{r} 
library(tidyverse) 
library(qtl)
library(snowfall) 
library(Biostrings) 
source("~/KIAT/function_BnRNAseq.R")    
```

### scanone 

### with flipped LG for eQTL anlaysis just for the plot 

```{r}
cross.F2 <- read.cross("csvsr", genfile ="~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped_C05C08.csv", 
                         phefile = "~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv",
                         genotypes = c("AA", "AB", "BB"))   # although the sample IDs are not matched in the original phe and gen file, I still get the right result. 

cross.F2 <- sim.geno(cross.F2,step=1,n.draws=32) # imputation?  
cross.F2 <- calc.genoprob(cross.F2,step=1) 

# before real analysis, test one gene 
tmp <- which(colnames(cross.F2$pheno) %in% "BnaA08g11140D")  
test <- scanone(cross.F2, pheno.col = tmp,method = "imp", use = "all.obs")
plot(test) # looks right 

scanone_eQTL.F2 <- scanone(cross.F2, pheno.col = 2:ncol(cross.F2$pheno), 
	         method = "imp", use = "all.obs")  

### permutation...  
set.seed(12345)   
permtest.F2 <- scanone(cross.F2, pheno.col = 2, method = "imp", n.perm = 10000)  

alphas <- seq(0.01, 0.10, by = 0.01) 
lod.thrs <- summary(permtest.F2, alphas) 
lod.thrs   

save(cross.F2, scanone_eQTL.F2, permtest.F2, lod.thrs, file = "~/F2/output/eQTL/scanone-eqtl_F2_flipped_C05C08.RData") 
``` 

### determine cis- and trans- eQTL 
```{r}
load("~/F2/output/eQTL/scanone-eqtl_F2_flipped_C05C08.RData")
scanone_eQTL.F2 %>% dim() # 4944 56182 

scanone_eQTL.F2$chr <- as.character(scanone_eQTL.F2$chr) 

# get threshold 
threshold.95 <- lod.thrs[5,] 
threshold.95 # 4.18 

# get all eQTL based on this threshold   
eQTL_sign <- 
sapply(colnames(scanone_eQTL.F2), function(gene) {
  sum(scanone_eQTL.F2[,gene] > threshold.95) > 0 
}) 

sum(eQTL_sign) # 22,695 genes with eQTL 
scanone_eQTL.F2 <- scanone_eQTL.F2[,eQTL_sign]  
dim(scanone_eQTL.F2) # 4944 22695  

# get bayesint result for every gene 
scanone.gather <-  
scanone_eQTL.F2 %>% 
  gather(key = trait, value = LOD, -chr, -pos) 

sig.chrs <- scanone.gather %>% filter(LOD > threshold.95) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarise(count = n()) # this is to get the significant chr ID for each trait 

sig.chrs 
sig.chrs %>% dim() # 26244         3

bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(scanone_eQTL.F2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE 
  )
  colnames(result)[3] <- "LOD" 
  result
})  

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x) 
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr)) 
)  

bayesint.result <- 
as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
    dplyr::select(trait,chr,pos,markername,LOD) %>%  
    separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>%  
    group_by(trait,chr) %>% 
    dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T), pos=median(pos, na.rm = T), min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD), genetic_start=min(pos, na.rm = T), genetic_end=max(pos, na.rm = T)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  # mutate(start=ifelse(start==end, start-20000,start), end=ifelse(start==end,end+20000,end)) 
  mutate(start_modified=ifelse(start-1000000>0, start-1000000, 0), end_modified=end+1000000) 
  

bayesint.result %>% dim() # 26244    11

# get genome range 
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures) 
library("rtracklayer") 

### get gff file with gene chrom & pos info, gff3 file must be sorted 
gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("gene_CHROM", "gene_start", "gene_end", "gene_ID") 
gff.mRNA %>% head()  

# look for cis-eQTL 
bayesint.result.2 <- 
bayesint.result %>% 
  mutate(gene_ID = trait, eQTL_chr = chr, eQTL_start = start_modified, eQTL_end = end_modified) %>% 
  dplyr::select(trait, gene_ID, eQTL_chr, pos, eQTL_start, eQTL_end, start, end, min_eQTL_LOD, max_eQTL_LOD, genetic_start, genetic_end) %>% 
  left_join(gff.mRNA, by = "gene_ID") 

bayesint.result.2$eQTL_chr <- paste("chr", bayesint.result.2$eQTL_chr, sep = "")
bayesint.result.2 %>% head()

cis_eQTL <- 
bayesint.result.2 %>% 
  filter(eQTL_chr == gene_CHROM) %>% 
  filter((gene_start < eQTL_start & gene_end > eQTL_start) | # also need SNP pos... 
         (gene_start >= eQTL_start & gene_end <= eQTL_end) |
         (gene_start < eQTL_end & gene_end > eQTL_end)) 

dim(cis_eQTL) # 11587    15

trans_eQTL <- 
bayesint.result.2 %>% 
  anti_join(cis_eQTL) 

save(cis_eQTL, trans_eQTL, file = "~/F2/output/eQTL/cis_trans_result_new_flipped_C05C08.Rdata") 
```  

### determine cis- and trans- eQTL using lodint 
Also tried with lodint() to get QTL interval, no big difference compared to bayesint() 

```{r} 
library(ggplot2) 
load("~/F2/output/eQTL/cis_trans_result_new_flipped_C05C08.Rdata") 

cis_eQTL %>% dim() # 11587    15
trans_eQTL %>% dim() # 14657    15  

cis_eQTL$index <- paste(cis_eQTL$eQTL_chr, cis_eQTL$eQTL_start, cis_eQTL$eQTL_end, sep = "_")
trans_eQTL$index <- paste(trans_eQTL$eQTL_chr, trans_eQTL$eQTL_start, trans_eQTL$eQTL_end, sep = "_")
cis_eQTL$index %>% unique() %>% length() # 6515
trans_eQTL$index %>% unique() %>% length() # 7117 
6515/(6515+7117)

c(cis_eQTL$index, trans_eQTL$index) %>% unique() %>% length() 

### FAD6
cis_eQTL %>% 
  filter(gene_ID == "BnaA08g12780D") %>% 
  View()
### 

cis_eQTL$class <- rep("cis", nrow(cis_eQTL))  
trans_eQTL$class <- rep("trans", nrow(trans_eQTL))   

# get only main chromosome stuff   
eQTL <- rbind(cis_eQTL, trans_eQTL) 
eQTL <- eQTL[grep("random", eQTL$gene_CHROM, value = FALSE, invert = T),] # random scaffolds are exluced 
eQTL$class <- factor(eQTL$class, levels = c("cis", "trans"))
eQTL$gene_CHROM <- gsub("(chr)([[:print:]]+)", "\\2", eQTL$gene_CHROM)
eQTL$eQTL_chr <- gsub("(chr)([[:print:]]+)", "\\2", eQTL$eQTL_chr)
eQTL$gene_CHROM <- factor(eQTL$gene_CHROM, levels = c("C09", "C08", "C07", "C06", "C05", "C04", "C03", "C02", "C01", "A10", "A09", "A08", "A07", "A06", "A05", "A04", "A03", "A02", "A01"))
eQTL$eQTL_chr <- factor(eQTL$eQTL_chr, levels = c("A01", "A02", "A03", "A04", "A05", "A06", "A07", "A08", "A09", "A10", "C01", "C02", "C03", "C04", "C05", "C06", "C07", "C08", "C09"))

# plot out the stuff   
p.eQTL <- 
eQTL %>% 
  ggplot() + 
  geom_point(aes(x = pos, y = gene_start, color = class), size = 0.5) +
  facet_grid(gene_CHROM ~ eQTL_chr, switch = "both", scales = "free") + 
  theme_classic() + 
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) + 
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "eQTL genetic position", y = "gene start") + 
  theme(text = element_text(size=8)) + 
  scale_color_manual(values=c("red", "royalblue"))  

p.eQTL    

p.cis_eQTL <-  
eQTL %>% 
  dplyr::filter(class == "cis") %>%
  ggplot() + 
  geom_point(aes(x = pos, y = gene_start), size = 0.5) +
  facet_grid(gene_CHROM ~ eQTL_chr, scales = "free", switch="both") + 
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "eQTL genetic position", y = "gene start")

p.cis_eQTL  

p.trans_eQTL <- 
eQTL %>% 
  dplyr::filter(class == "trans") %>%
  ggplot() + 
  geom_point(aes(x = pos, y = gene_start), size = 0.5) + # x is genetic position of eQTL, y is physical start position of gene 
  facet_grid(gene_CHROM ~ eQTL_chr, scales = "free", switch="both") + 
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "eQTL genetic position", y = "gene start")    

p.trans_eQTL
ggsave(p.eQTL, filename = "~/F2/output/eQTL/figure/p.eQTL.flipped.png", height = 10, width = 7.87) 
ggsave(p.cis_eQTL, filename = "~/F2/output/eQTL/figure/p.cis_eQTL.flipped.png", height = 10, width = 10)
ggsave(p.trans_eQTL, filename = "~/F2/output/eQTL/figure/p.trans_eQTL.flipped.png", height = 10, width = 10) 

### check to see the rate of different vs same subgenome trans-regulators 
eQTL %>% colnames()  
eQTL$eQTL_subgenome <- gsub("(A|C)([[:digit:]]+)", "\\1", eQTL$eQTL_chr)
eQTL$gene_subgenome <-  gsub("(A|C)([[:digit:]]+)", "\\1", eQTL$gene_CHROM)

sum(eQTL$eQTL_subgenome == eQTL$gene_subgenome) /nrow(eQTL) # 72% from the same subgenome 

A_genes <- eQTL[eQTL$eQTL_subgenome == "A",] 
sum(A_genes$eQTL_subgenome == A_genes$gene_subgenome) /nrow(A_genes) # 74% 

C_genes <- eQTL[eQTL$eQTL_subgenome == "C",] 
sum(C_genes$eQTL_subgenome == C_genes$gene_subgenome) /nrow(C_genes) # 70%  

# number of eQTL on A & C subgenome 
eQTL[eQTL$eQTL_subgenome == "A",] %>% nrow() /nrow(eQTL) 
eQTL[eQTL$eQTL_subgenome == "C",] %>% nrow() /nrow(eQTL) 

# plot eQTL which only act in trans- but not cis-, this isn't correct 
trans_eQTL$index2 <- paste(trans_eQTL$eQTL_chr, trans_eQTL$genetic_start, trans_eQTL$gene_end, sep = "_")
cis_eQTL$index2 <- paste(cis_eQTL$eQTL_chr, cis_eQTL$genetic_start, cis_eQTL$gene_end, sep = "_") 
# no identical based on genetic position   

trans_eQTL_only <- 
trans_eQTL %>% 
  anti_join(cis_eQTL, by = "index")    

dim(trans_eQTL_only) # 6919 16 
trans_eQTL_only %>% colnames()

trans_eQTL_only <- trans_eQTL_only[grep("random", trans_eQTL_only$gene_CHROM, value = FALSE, invert = T),] # random scaffolds are exluced 
trans_eQTL_only$gene_CHROM <- gsub("(chr)([[:print:]]+)", "\\2", trans_eQTL_only$gene_CHROM)
trans_eQTL_only$eQTL_chr <- gsub("(chr)([[:print:]]+)", "\\2", trans_eQTL_only$eQTL_chr)
trans_eQTL_only$gene_CHROM <- factor(trans_eQTL_only$gene_CHROM, levels = c("C09", "C08", "C07", "C06", "C05", "C04", "C03", "C02", "C01", "A10", "A09", "A08", "A07", "A06", "A05", "A04", "A03", "A02", "A01"))
trans_eQTL_only$eQTL_chr <- factor(trans_eQTL_only$eQTL_chr, levels = c("A01", "A02", "A03", "A04", "A05", "A06", "A07", "A08", "A09", "A10", "C01", "C02", "C03", "C04", "C05", "C06", "C07", "C08", "C09"))

# plot 
p.trans_eQTL_only <- 
trans_eQTL_only %>% 
  ggplot() + 
  geom_point(aes(x = pos, y = gene_start), size = 0.5) + # x is genetic position of eQTL, y is physical start position of gene 
  facet_grid(gene_CHROM ~ eQTL_chr, scales = "free", switch="both") + 
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "eQTL genetic position", y = "gene start")    

p.trans_eQTL_only 
ggsave(p.trans_eQTL_only, filename = "~/F2/output/eQTL/figure/p.trans_eQTL_only.flipped.png", height = 10, width = 10)  

### check to see the rate of different vs same subgenome trans-regulators 
trans_eQTL_only$eQTL_subgenome <- gsub("(A|C)([[:digit:]]+)", "\\1", trans_eQTL_only$eQTL_chr)
trans_eQTL_only$gene_subgenome <-  gsub("(A|C)([[:digit:]]+)", "\\1", trans_eQTL_only$gene_CHROM)

sum(trans_eQTL_only$eQTL_subgenome != trans_eQTL_only$gene_subgenome) /nrow(trans_eQTL_only) # 56.9% from the same subgenome 

A_genes <- trans_eQTL_only[trans_eQTL_only$eQTL_subgenome == "A",] 
sum(A_genes$eQTL_subgenome != A_genes$gene_subgenome) /nrow(A_genes) # %60 

C_genes <- trans_eQTL_only[trans_eQTL_only$eQTL_subgenome == "C",] 
sum(C_genes$eQTL_subgenome != C_genes$gene_subgenome) /nrow(C_genes) # 51%  

# number of eQTL on A & C subgenome 
trans_eQTL_only[trans_eQTL_only$eQTL_subgenome == "A",] %>% nrow() /nrow(trans_eQTL_only) 
trans_eQTL_only[trans_eQTL_only$eQTL_subgenome == "C",] %>% nrow() /nrow(trans_eQTL_only)   

# use genetic position to look for overlaps 
cis_eQTL_range <- GRanges(seqnames = Rle(cis_eQTL$eQTL_chr),ranges = IRanges(start = cis_eQTL$genetic_start, end = cis_eQTL$genetic_end))
cis_eQTL_range

trans_eQTL_range <- GRanges(seqnames = Rle(trans_eQTL$eQTL_chr), ranges = IRanges(start = trans_eQTL$genetic_start, end = trans_eQTL$genetic_end))
trans_eQTL_range

cis_trans_eQTL <- mergeByOverlaps(cis_eQTL_range, trans_eQTL_range) # warning message here. look out!!!!!! 
cis_trans_eQTL_2 <- 
cis_trans_eQTL$trans_eQTL_range %>% unique() %>% as.data.frame() %>% 
  mutate(eQTL_chr = seqnames, genetic_start = start, genetic_end = end) %>% 
  dplyr::select(eQTL_chr, genetic_start, genetic_end)

trans_eQTL$index2 %>% unique() %>% length() 

# exclude these trans-eQTL overlapped with cis-eQTL 
trans_eQTL_only_genetic <- 
trans_eQTL %>% 
  anti_join(cis_trans_eQTL_2) 

trans_eQTL_only_genetic <- trans_eQTL_only_genetic[grep("random", trans_eQTL_only_genetic$gene_CHROM, value = FALSE, invert = T),] # random scaffolds are exluced 
trans_eQTL_only_genetic$gene_CHROM <- gsub("(chr)([[:print:]]+)", "\\2", trans_eQTL_only_genetic$gene_CHROM)
trans_eQTL_only_genetic$eQTL_chr <- gsub("(chr)([[:print:]]+)", "\\2", trans_eQTL_only_genetic$eQTL_chr)
trans_eQTL_only_genetic$gene_CHROM <- factor(trans_eQTL_only_genetic$gene_CHROM, levels = c("C09", "C08", "C07", "C06", "C05", "C04", "C03", "C02", "C01", "A10", "A09", "A08", "A07", "A06", "A05", "A04", "A03", "A02", "A01"))
trans_eQTL_only_genetic$eQTL_chr <- factor(trans_eQTL_only_genetic$eQTL_chr, levels = c("A01", "A02", "A03", "A04", "A05", "A06", "A07", "A08", "A09", "A10", "C01", "C02", "C03", "C04", "C05", "C06", "C07", "C08", "C09"))

# plot 
p.trans_eQTL_only_genetic <- 
trans_eQTL_only_genetic %>% 
  ggplot() + 
  geom_point(aes(x = pos, y = gene_start), size = 0.5) + # x is genetic position of eQTL, y is physical start position of gene 
  facet_grid(gene_CHROM ~ eQTL_chr, scales = "free", switch="both") + 
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "eQTL genetic position", y = "gene start")    

p.trans_eQTL_only_genetic 

# ggsave(p.trans_eQTL_only, filename = "~/F2/output/eQTL/figure/p.trans_eQTL_only.flipped.png", height = 10, width = 10)  

### check to see the rate of different vs same subgenome trans-regulators 
trans_eQTL_only_genetic$eQTL_subgenome <- gsub("(A|C)([[:digit:]]+)", "\\1", trans_eQTL_only_genetic$eQTL_chr)
trans_eQTL_only_genetic$gene_subgenome <-  gsub("(A|C)([[:digit:]]+)", "\\1", trans_eQTL_only_genetic$gene_CHROM) 

sum(trans_eQTL_only_genetic$eQTL_subgenome != trans_eQTL_only_genetic$gene_subgenome) /nrow(trans_eQTL_only_genetic) # 56.9% from different genome  

A_genes <- trans_eQTL_only_genetic[trans_eQTL_only_genetic$eQTL_subgenome == "A",] 
sum(A_genes$eQTL_subgenome != A_genes$gene_subgenome) /nrow(A_genes) # %60 

C_genes <- trans_eQTL_only_genetic[trans_eQTL_only_genetic$eQTL_subgenome == "C",] 
sum(C_genes$eQTL_subgenome != C_genes$gene_subgenome) /nrow(C_genes) # 56%  

# number of eQTL on A & C subgenome 
trans_eQTL_only_genetic[trans_eQTL_only_genetic$eQTL_subgenome == "A",] %>% nrow() /nrow(trans_eQTL_only_genetic) 
trans_eQTL_only_genetic[trans_eQTL_only_genetic$eQTL_subgenome == "C",] %>% nrow() /nrow(trans_eQTL_only_genetic)   
``` 

### find overlaps between eQTL and traitQTL 
```{r}
# trait QTL using scanone result  
load("~/F2/output/QTL_analysis/traitQTL.annotated.flipped.Rdata")
load("~/F2/output/eQTL/cis_trans_result_new_flipped_C05C08.Rdata") 

cis_eQTL %>% dim() #  11587    15
trans_eQTL %>% dim() # 14657    15  
traitQTL.annotated %>% dim() # 18647    13 

traitQTL.annotated %>% colnames()
traitQTL.annotated %>% 
  group_by(trait, chrom) %>% 
  dplyr::summarise(count = n()) %>% 
  dim()

cis_eQTL.qtl.combined <- inner_join(cis_eQTL,traitQTL.annotated,by="gene_ID") 
cis_eQTL.qtl.combined %>% head()    
cis_eQTL.qtl.combined %>% dim() # 2165   27   

colnames(cis_eQTL.qtl.combined) 

cis_eQTL.qtl.combined.final <- 
cis_eQTL.qtl.combined %>% 
  mutate(eQTL_start = start.x, eQTL_end = end.x, QTL_start = start.y, QTL_end = end.y, eQTL_genetic_start = genetic_start.x, eQTL_genetic_end = genetic_end.x, QTL_genetic_start = genetic_start.y, QTL_genetic_end = genetic_end.y) %>% 
  dplyr::select(-c(start.x, end.x, start.y, end.y, genetic_start.x, genetic_end.x, genetic_start.y, genetic_end.y)) 

# add additive allele effect 
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/for_paper/cis_eQTL_effect_all.Rdata")

names(cis_eQTL_effect_all) <- unique(cis_eQTL.qtl.combined.final$gene_ID)
cis_eQTL_effect_all <- sapply(names(cis_eQTL_effect_all), function(gene) cis_eQTL_effect_all[[gene]]$ests[2,1])
cis_eQTL_effect_all <- cis_eQTL_effect_all %>% as.data.frame() ## all possible genes have positive allele effect from Da-Ol-1 
cis_eQTL_effect_all$names <- rownames(cis_eQTL_effect_all) 
colnames(cis_eQTL_effect_all)[1] <- "additive_allele_effect_Da_Ol"
cis_eQTL_effect_all %>% head()

cis_eQTL.qtl.combined.final <- 
cis_eQTL.qtl.combined.final %>% 
  left_join(cis_eQTL_effect_all, by = c("gene_ID" = "names"))    

write.csv(cis_eQTL.qtl.combined.final, file = "~/F2/output/eQTL/cis_eQTL.qtl.combined.final.csv")    
  
# trans-eQTL overlap with trait QTL, use genetic interval as the interval for both trait QTL and eQTL  

traitQTL <-  
traitQTL.annotated %>%  
  group_by(trait, chrom, genetic_start, genetic_end, model, start, end) %>% 
  dplyr::summarise(count = n())  

trans_eQTL %>% colnames() 
trans_eQTL %>% head()
gff.trans_eQTL <- GRanges(seqnames = Rle(trans_eQTL$eQTL_chr),ranges = IRanges(start = trans_eQTL$genetic_start, end = trans_eQTL$genetic_end), names = trans_eQTL$gene_ID) 

traitQTL %>% colnames() 
gff.trait_qtl<- GRanges(seqnames = Rle(paste("chr", traitQTL$chrom, sep = "")),ranges = IRanges(start = traitQTL$genetic_start, end = traitQTL$genetic_end), trait = traitQTL$trait)   

# cis_eQTL.qtl.combined 
trans_candidate <- mergeByOverlaps(gff.trait_qtl, gff.trans_eQTL) 
trans_candidate %>% dim()  # 4022    4  
trans_candidate 

### add the rest of information from trans_eQTL & trait_QTL 
trans_candidate <- 
trans_candidate %>% 
  as.data.frame() %>% 
  mutate(chrom = gff.trait_qtl.seqnames, trait_genetic_start = gff.trait_qtl.start, trait_genetic_end = gff.trait_qtl.end, gene_ID = names, eQTL_chr = gff.trans_eQTL.seqnames, eQTL_genetic_start = gff.trans_eQTL.start, eQTL_genetic_end = gff.trans_eQTL.end) %>% 
  dplyr::select(chrom, trait_genetic_start, trait_genetic_end, trait, gene_ID, eQTL_chr, eQTL_genetic_start, eQTL_genetic_end) 

### annotate trans_eQTL 
load("~/Reference/B.napus/BnapusAnnotation.Rdata") 
BnapusAnnotation %>% colnames()

trans_candidate.final <- 
trans_candidate %>%  
  left_join(BnapusAnnotation, c("gene_ID" = "name")) %>% 
  mutate(trait_chrom = chrom.x, gene.chrom = chrom.y, gene.start = start, gene.end = end) %>% 
  dplyr::select(-c(chrom.x, chrom.y, start, end))   

## get GO term for each gene
load("~/Reference/B.napus/napus_GO_combined.Rdata") # there is one colomn hiding... 

trans_candidate.final <- 
trans_candidate.final %>% 
  left_join(napus_GO_combined) 

trans_candidate.final %>% dim() # 4022 16 
trans_candidate.final$gene_ID %>% unique() %>% length() # 1972 
write.csv(trans_candidate.final, file = "~/F2/output/eQTL/trans_candidate.final.csv") 

### make some plot for several important genes 
# the two FAE 

# the two 
### make some table & plot for several important genes 
# 1) Fatty acid/flowering/growth related genes for cis-eQTL final list 
cis_eQTL.qtl.combined.final_fatty_acid <- 
cis_eQTL.qtl.combined.final[grep("lipid|fatty", cis_eQTL.qtl.combined.final$GO_des),] 
  
# 2) transcription factor coding genes for cis-eQTL final list
cis_eQTL.qtl.combined.final_TF <- 
cis_eQTL.qtl.combined.final[grep("DNA|nucleotide|transcription", cis_eQTL.qtl.combined.final$GO_des),] %>% 
  dplyr::select(gene_ID, trait.y, AGI, At_symbol, GO_des)  

# 3) Fatty acid/flowering/growth related genes for trans-eQTL final list
trans_candidate.final_fatty_acid <- 
trans_candidate.final[grep("lipid|fatty", trans_candidate.final$GO_des),] %>%
    dplyr::select(gene_ID, trait, AGI, At_symbol, GO_des) 

# 4) figures for the two cis-eQTL gene 
# BnaA08g11060D & BnaA08g11140D two cis-eQTL gene for fatty acid, need to deal with the color problem... more colors than a single palette can handel... 
# save(cis_eQTL.qtl.combined.final, trans_eQTL.qtl.combined.final, file = "~/F2/output/eQTL/eQTL.qtl.combined.final.Rdata") 

### check whether FLC is cis-regulator for some the traits 
# load("~/F2/output/eQTL/eQTL.qtl.combined.final.Rdata") 

cis_eQTL.qtl.combined.final %>%  
  filter(AGI == "AT5G10140") %>% 
  View() # FLC is cis-regulator    

cis_eQTL.qtl.combined.final[grep("acid|aicd|content", cis_eQTL.qtl.combined.final$trait.y),] %>% 
  dplyr::select(gene_ID, trait.y, AGI, At_symbol, GO_des, additive_allele_effect_Da_Ol) %>% 
  # dplyr::select(At_symbol) %>%
  # unique() %>% 
  View()  


# cis_eQTL.qtl.combined.final <- 
test <- 
cis_eQTL.qtl.combined.final[grep("acid|aicd|content", cis_eQTL.qtl.combined.final$trait.y, invert = T),] %>% 
  dplyr::select(gene_ID, trait.y, AGI, At_symbol, GO_des, additive_allele_effect_Da_Ol) 
  # dplyr::select(At_symbol) %>%
  # unique() %>% 

test[grep("DNA|nucleotide|transcription", test$GO_des),] %>% 
  View()

###### for fatty acid trait  
test <- 
cis_eQTL.qtl.combined.final[grep("acid|aicd|content", cis_eQTL.qtl.combined.final$trait.y),] %>% 
  dplyr::select(gene_ID, trait.y, AGI, At_symbol, GO_des, additive_allele_effect_Da_Ol) 
  # dplyr::select(At_symbol) %>%
  # unique() %>% 

test[grep("lipid|fatty", test$GO_des),]  %>% View()

cis_eQTL.qtl.combined.final %>%  
  filter(gene_ID == "BnaA08g11060D" | gene_ID == "BnaA08g11140D" | gene_ID == "BnaA01g03770D"| gene_ID == "BnaA03g37760D" | gene_ID == "BnaC06g28980D" | gene_ID == "BnaA01g07910D") %>% 
  dplyr::select(gene_ID, trait.y, AGI, At_symbol, GO_des, additive_allele_effect_Da_Ol) %>% 
  # dplyr::select(At_symbol) %>%
  # unique() %>% 
  View()    

cis_eQTL.qtl.combined.final %>% 
  filter(gene_ID == "BnaA10g22080D" | gene_ID == "BnaC06g20640D" | gene_ID == "BnaC06g27170D" | gene_ID == "BnaA10g21380D") %>% 
  dplyr::select(gene_ID, trait.y, AGI, At_symbol, GO_des, additive_allele_effect_Da_Ol) %>% 
  View()

#### other important flower genes under C06? 
### important flower related genes 
flowering_time_gene <- read.csv("~/F2/for_paper/flowering_time/flowering_time_genes.csv")
flowering_time_gene$Gene.number <- gsub("\\ ", "", flowering_time_gene$Gene.number)
flowering_time_gene$Gene.number <- gsub("\\\xca", "", flowering_time_gene$Gene.number) 

cis_eQTL.qtl.combined.final_flowering <- 
cis_eQTL.qtl.combined.final[cis_eQTL.qtl.combined.final$AGI %in% flowering_time_gene$Gene.number,] %>% 
  dplyr::select(gene_ID, trait.y, AGI, At_symbol, GO_des) # no cis-flowering regulator under C06 

trans_candidate.final_flowering <- 
trans_candidate.final[trans_candidate.final$AGI %in% flowering_time_gene$Gene.number,] %>%
  dplyr::select(gene_ID, trait, AGI, At_symbol, GO_des)

trans_candidate.final[grep("acid|aicd|content", trans_candidate.final$trait, invert = T),] %>% 
  dplyr::select(gene_ID, trait, AGI, At_symbol, GO_des) %>% 
  dplyr::select(At_symbol) %>%
  unique() %>% 
  View() 
trans_candidate.final[grep("MADS", trans_candidate.final$At_symbol),] 

test <- 
trans_candidate.final[grep("acid|aicd|content", trans_candidate.final$trait, invert = T),] %>% 
  dplyr::select(gene_ID, trait, AGI, At_symbol, GO_des)

test[grep("DNA|nucleotide|transcription", test$GO_des),] %>%
  dplyr::select(trait, gene_ID, AGI, At_symbol, GO_des) %>% View()

test <- 
trans_candidate.final[grep("acid|aicd|content", trans_candidate.final$trait),] %>% 
  dplyr::select(gene_ID, trait, AGI, At_symbol, GO_des)

test[grep("lipid|fatty", test$GO_des),] %>% 
  dplyr::select(trait, gene_ID, AGI, At_symbol, GO_des) %>% View()

# trans effect for BnaA08g27750D
trans_candidate.final %>% 
  filter(gene_ID == "BnaA08g27750D") 


write.csv(cis_eQTL.qtl.combined.final_fatty_acid, file = "~/F2/output/eQTL/cis_eQTL.qtl.combined.final_fatty_acid.csv")
write.csv(cis_eQTL.qtl.combined.final_flowering, file = "~/F2/output/eQTL/cis_eQTL.qtl.combined.final_flowering.csv")
write.csv(cis_eQTL.qtl.combined.final_TF, file = "~/F2/output/eQTL/cis_eQTL.qtl.combined.final_TF.csv")

write.csv(trans_candidate.final_fatty_acid, file = "~/F2/output/eQTL/trans_candidate.final_fatty_acid.csv")
write.csv(trans_candidate.final_flowering, file =  "~/F2/output/eQTL/trans_candidate.final_flowering.csv") 
```

### promoter motif enrichment analysis on trans-eQTL to look for potential up-stream genes in the cis-eQTL site
```{r}
trans_candidate.final <- read.csv(file = "~/F2/output/eQTL/trans_candidate.final.csv", row.names = 1) 
save(trans_candidate.final, file = "~/F2/output/eQTL/trans_candidate.final.Rdata")

trans_candidate.final %>% dim() # 4022 16 
trans_candidate.final$gene_ID %>% unique() %>% length() # 1972 

trans_candidate.final$gene_ID %>% unique() %>% length() 

# get trans-candidate gene fasta file 
# perl extract-utr.pl --gff_file=/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus.annotation_v5.gff3 --genome_fa_file=/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa --cds_fa_file=/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa --fiveprime --utr_length=1000 --gene_length=0 --output_fa_file=Brassica_napus.upstream.fa
# cat ~/F2/output/eQTL/trans_candidate.final.csv | awk 'BEGIN{FS= ","}{print $5}' | sort | uniq | grep "Bna" | sed 's/"//g' > trans_candidate.final.ID
# cat trans_candidate.final.ID | grep -v "Unn" | grep -v "Ann" | grep -v "Cnn"  > trans_candidate.final.main.ID
# perl -ne 'if(/^>(\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' trans_candidate.final.ID ~/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa | grep ">" > trans_candidate_upstream.fa # all genes including random chromosome 
# perl -ne 'if(/^>(\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' trans_candidate.final.main.ID ~/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa | grep ">" > trans_candidate_upstream_main.fa 

# promoter motif enrichment for trans-eQTL for each trait 

# get the promoter sequences 
library(Biostrings) #R package for handling DNA and protein data
promoters <- readDNAStringSet("~/F2/for_paper/Brassica_napus.upstream.fa")

#convert "N" to "-" in promoters.  otherwise motifs will match strings of "N"s
promoters <- DNAStringSet(gsub("N","-",promoters))
promoters 

# motif 
motifs <- read.delim("~/Reference/promoter/element_name_and_motif_IUPACsupp.txt",header=FALSE,as.is=TRUE)
head(motifs)
motifsV <- as.character(motifs[,2])
names(motifsV) <- motifs[,1]
motifsSS <- DNAStringSet(motifsV)
motifsSS 

# subset "universe" & "target" genes's motif 
#get names to match...there are are few names in the DEgene list not in the promoter set
# only get genes on main chromosomes 
expressed_gene <- read.csv("~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv", row.names = 1)
expressed_gene_main <- unique(grep("nn",  rownames(expressed_gene), value = T, invert = T)) 
expressed_gene_main.match <- expressed_gene_main[expressed_gene_main %in% names(promoters)] 
universe.promoters <- promoters[expressed_gene_main.match]  

motif.results <- 
lapply(unique(trans_candidate.final$trait), function(trait) {
  trans_eQTL_main <- unique(grep("nn",  trans_candidate.final[trans_candidate.final$trait == trait,]$gene_ID, value = T, invert = T)) 
  trans_eQTL_main.match <- trans_eQTL_main[trans_eQTL_main %in% names(promoters)] 

  #subset promoter files 
  target.promoters <- promoters[trans_eQTL_main.match]   

  # enrichment analysis 
  motif.results <- motifEnrichment(target.promoters,universe.promoters) 
}) 

<<<<<<< HEAD
### demo 




=======
names(motif.results) <- unique(trans_candidate.final$trait)
motif.results.2 <- 
lapply(names(motif.results), function(trait) {
  motif.results[[trait]][motif.results[[trait]]$p.value < 0.05,]
})

names(motif.results.2) <- unique(trans_candidate.final$trait)
motif.results.2 
motif.results.final <- do.call("rbind", motif.results.2) 
motif.results.final$trait <- gsub("([[:print:]]+)(\\.)(+[[:print:]]+)", "\\1", rownames(motif.results.final))
motif.results.final %>% View()
write.csv(motif.results.final, file = "~/F2/for_paper/trans_regulator_motif_results.csv")  

motif.results.final[grep("acid|aicd|content", motif.results.final$trait),]
motif.results.final[grep("acid|aicd|content", motif.results.final$trait, invert = T),] %>% View() 
>>>>>>> c5edaab06ec02da01b8889bd27552411ca8538ab
```

### GO enrichment analysis 
```{r}
# trans-eQTL 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/eQTL/trans_candidate.final.Rdata")
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")

dim(trans_candidate.final) # 4022   16 

trans_candidate.final %>% colnames() 
trans_candidate.final %>% head()

GO_trans_eQTL <- 
lapply(unique(trans_candidate.final$trait), function(trait) {
  test <- trans_candidate.final$gene_ID[trans_candidate.final$trait == trait]
  tryCatch(
    GOseq.Bn.ORA(test), error= function(e) NA)
}) 

names(GO_trans_eQTL) <- unique(trans_candidate.final$trait) 
# make a dataframe from this result 
GO_trans_eQTL$Behenic_acid
GO_trans_eQTL <- do.call("rbind", GO_trans_eQTL) 
GO_trans_eQTL %>% View()
load("~/Desktop/F2_paper/data/GO_trans_eQTL.Rdata")
save(GO_trans_eQTL, file = "~/Desktop/F2_paper/data/GO_trans_eQTL.Rdata") 

# how many trans-eQTL gene for each trait 
number_trans_eQTL <- 
trans_candidate.final %>% 
  group_by(trait) %>% 
  summarise(count = n()) 

number_trans_eQTL[grep("acid|content", number_trans_eQTL$trait),]

# cis-coding candidate 
traitQTL.annotated.SNP <- read.csv("~/Desktop/F2_paper/supplementary/table/traitQTL.annotated.SNP.csv", row.names = 1)

dim(traitQTL.annotated.SNP) # 4062    9  

traitQTL.annotated.SNP %>% colnames() 
traitQTL.annotated.SNP %>% head()

GO_cis_coding <- 
lapply(unique(traitQTL.annotated.SNP$trait), function(trait) {
  test <- traitQTL.annotated.SNP$gene_ID[traitQTL.annotated.SNP$trait == trait]
  tryCatch(
    GOseq.Bn.ORA(test), error= function(e) NA)
})

GO_cis_coding 
names(GO_cis_coding) <- unique(traitQTL.annotated.SNP$trait) 
# make a dataframe from this result 
GO_cis_coding <- do.call("rbind", GO_cis_coding) 
GO_cis_coding %>% View() 

# cis-regulator  



```

### examine the correlation between expression of FAD, FAE, and SQS with Erucic acid content level 
```{r}
phefile.expression <- read.csv("~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv", row.names = 1)
phefile.expression 

phefile.trait <- read.csv("~/F2/data/QTL_analysis/F2.pheno.csv", row.names = 1)    
phefile.trait  

trait <- as.data.frame(t(phefile.trait)) 
ID <- c("BnaA08g11140D", "BnaA08g11060D", "BnaA08g12780D")
expression <- as.data.frame(t(phefile.expression[rownames(phefile.expression) %in% ID,]))
expression$id <- rownames(expression)

test_gene <- 
trait %>% 
  dplyr::select(Erucic_acid, id) %>% 
  left_join(expression) 

colnames(test_gene)

png("~/F2/output/expression_vs_Erucic.png", width=9, height=3, units="in", res=300)
par(mfrow=c(1,3)) 
for (i in colnames(test_gene)[3:5]) {
  title <- round(cor(as.numeric(test_gene$Erucic_acid), test_gene[,i]), digits = 2)
  plot(test_gene[,i], test_gene$Erucic_acid, main = title, xlab = i)
} 
dev.off() 

# expression between Da-Ae and Da-Ol-1 
### plot the expression of the two genes 
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/vstMat.parent.Rdata")
genes.GWAS <- c("BnaA08g11140D", "BnaA08g11060D", "BnaA08g12780D")
genes.GWAS.df <- as.data.frame(genes.GWAS)
colnames(genes.GWAS.df) <- "V1"
genes.GWAS.df 

expression.pattern.Bn.parent.bar(vstMat.parent = vstMat.parent, gene = genes.GWAS.df) 
ggsave("/Users/ruijuanli/Desktop/F2_paper/three_genes.png", width = 7, height = 5)   
```

### appendix 
```{r}
# function for promoter motif enrichment analysis 
#create a function to summarize the results and test for significance
motifEnrichment <- function(target.promoters,universe.promoters,all.counts=F,motifs=motifsSS) {
  
  #use vcountPDict to count the occurrences of each motif in each promoter
  target.counts <- vcountPDict(motifs,target.promoters,fixed=F) + 
    vcountPDict(motifsSS,reverseComplement(target.promoters),fixed=F)
  universe.counts <- vcountPDict(motifs,universe.promoters,fixed=F) + 
    vcountPDict(motifsSS,reverseComplement(universe.promoters),fixed=F)
  
  if (all.counts) { 
    #count all occurrences of a motif instead of the number of promoters that it occurs in
    target.counts.sum <- apply(target.counts,1,sum)
    universe.counts.sum <- apply(universe.counts,1,sum)
  } else {
    target.counts.sum <- apply(ifelse(target.counts > 0,1,0),1,sum)
    universe.counts.sum <- apply(ifelse(universe.counts > 0 , 1, 0),1,sum)
  }
  n.motifs <- length(target.counts.sum)
  results <- vector(mode="numeric",length=n.motifs)
  for (i in 1:n.motifs) {
    if (all.counts) { #the contigency tables are different depending on whether we are looking at promoters or overall occurrences
      #test if ratio of occurrences to promoters is the same in the target and the universe
      m <- matrix(c(
        target.counts.sum[i],                       #number of occurrences within target
        dim(target.counts)[2],                      #number of promoters in target
        universe.counts.sum[i],                  #number of occurrences within universe
        dim(universe.counts)[2]                  #number of promoters in universe
      ),ncol=2)
    } else { #looking at promoters with and without hits
      m <- matrix(c(
        target.counts.sum[i],                        #number of promoters in target with hit
        dim(target.counts)[2]-target.counts.sum[i],            #number of promoters in target with no hit
        universe.counts.sum[i],                   #number of promoters in universe with hit
        dim(universe.counts)[2]-universe.counts.sum[i]   #number of promoters in universe with no hit
      ),ncol=2)
    } #else
    results[i] <- fisher.test(m,alternative="greater")$p.value
  } #for loop
  results.table <- data.frame(
    motif=names(motifs),
    universe.percent = round(universe.counts.sum/dim(universe.counts)[2],3)*100,
    target.percent = round(target.counts.sum/dim(target.counts)[2],3)*100,
    p.value =  results)
  results.table <- results.table[order(results.table$p.value),]
  results.table
} 


allele.effect.trans <- function(gene_ID){
  
  gene_ID <- "BnaA08g27750D" 
  load("~/F2/output/eQTL/cis_trans_result_new_flipped_C05C08.Rdata")
  cis_eQTL %>% dim()
  trans_eQTL %>% dim()
  
  cross.F2 <- read.cross("csvsr",
                         genfile ="~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped_C05C08.csv",
                         phefile = "~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv",
                         genotypes = c("AA", "AB", "BB"))   # although the sample IDs are not matched in the original phe and gen file, I still get the right result.
  cross.F2$pheno <- as.data.frame(cross.F2$pheno[,gene_ID])
  cross.F2 <- sim.geno(cross.F2,step=1,n.draws=32) # imputation? 
  cross.F2 <- calc.genoprob(cross.F2,step=1)

  chr = trans_eQTL[trans_eQTL$gene_ID == gene_ID,]$eQTL_chr
  chr = gsub("chr", "", chr)
  pos = trans_eQTL[trans_eQTL$gene_ID == gene_ID,]$pos
  qtlm <- makeqtl(cross.F2,chr=chr,pos=pos)

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
  qtl.fit <- fitqtl(cross=cross.F2,qtl=qtlm,formula = y ~ Q1, get.ests=T)

#examine our model
  summary(qtl.fit) 
} 
```


