---
title: "F2_map_construction"
author: "Ruijuan Li"
date: "4/18/2017"
output: html_document 
---

### Goal
The goal of this analysis is to build genetic map using F2 genotyping information, F2 genotyping was done based on the SNPs we identified from its parents 

### install & load package 
```{r}
# install.packages("onemap")  
# install.packages("tkrplot", type="source") 
library(tcltk)
library(tkrplot)
library(onemap)  
library(ggplot2)
library(reshape) 
library(tidyverse) 
library("qtl") 
library(snowfall) 
```

### import & format data
```{r} 
F2_geno_data <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) # 18226   172 

rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)] 
dim(F2_geno_data) # 18226   166  
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data)) 
```

### check missing rate & remove data with missing rate above 0.10
```{r}
F2_geno_data_with_missing_rate <- mutate(F2_geno_data, 
       missing_rate = round(apply(F2_geno_data, 1, function(x) sum(is.na(x)))/166, 2)) 

rownames(F2_geno_data_with_missing_rate) <- rownames(F2_geno_data) 
hist(F2_geno_data_with_missing_rate$missing_rate)
max(F2_geno_data_with_missing_rate$missing_rate) # 0.47, the highest missing rate is 0.47 

# remove markers with missing rate greater than 0.90
F2_geno_data_with_missing_rate_filtered <-  F2_geno_data_with_missing_rate[F2_geno_data_with_missing_rate$missing_rate < 0.10,]

dim(F2_geno_data_with_missing_rate_filtered) # 3608 167  
hist(F2_geno_data_with_missing_rate_filtered$missing_rate)

F2_geno_data_with_missing_rate_filtered.final <- subset(F2_geno_data_with_missing_rate_filtered, select=-missing_rate)

dim(F2_geno_data_with_missing_rate_filtered.final) # 3608 166 
```

### calculate pairwise corrlelation and remove SNPs with corrlelation greater or equal to 0.9
some SNPs should have exactly the same genotype across all individuals, they are redundant in terms of genetic map construction, so they should be removed. find those SNPs by doing correlation test. 
```{r}
F2_geno_data_t <- as.data.frame(t(F2_geno_data_with_missing_rate_filtered.final))
F2_geno_data_t.numeric <- data.matrix(F2_geno_data_t)

# delete markers with all same genotypes across individuals 
test <- apply(F2_geno_data_t.numeric, 2, function(x) length(unique(x[!is.na(x)])))
filter.polymorphsm <- test != 1 

F2_geno_data_t.numeric.2 <- F2_geno_data_t.numeric[,filter.polymorphsm]
dim(F2_geno_data_t.numeric.2) # 166 3606

# also output save markers with same genotype across individuals
polymorphism <- test == 1
SNP.not.poly <- colnames(F2_geno_data_t.numeric)[polymorphism]
length(SNP.not.poly) # two markers are not polymorphism across individuals 

# correlation test
options(warn=-1) # suppress warning message
F2_SNP_correlation <- cor(F2_geno_data_t.numeric.2, use = "pairwise.complete.obs")
save(F2_SNP_correlation, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_SNP_correlation.Rdata")
options(warn=0) # unsuppress warning message
dim(F2_SNP_correlation) # 3606 3606 

# calcualte number of SNPs with correlation of 1
nrow(which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = T)) # 172 
# find SNPs with correlation of 1 and remove them 
dup.cordinate <- which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = F)
length(dup.cordinate) # 344 
dup.cordinate.df <- as.data.frame(dup.cordinate)
sample.ID <- colnames(F2_SNP_correlation)

# extract duplicate pair information based on their coordicate
dup.pair <- data.frame(matrix(nrow = nrow(dup.cordinate), ncol = 2))
for (i in 1:nrow(dup.cordinate)){
 dup.pair[i,1] <- sample.ID[dup.cordinate[i,1]]
 dup.pair[i,2] <- sample.ID[dup.cordinate[i,2]]
}

length(dup.pair[,2]) # 171 
rem <- unique(dup.pair[,2])
length(rem) # 163 markers are removed due to pairwise correlation of 1 

### 
F2_geno_data_1 <- F2_geno_data_with_missing_rate_filtered.final[!(rownames(F2_geno_data_with_missing_rate_filtered.final) %in% rem),]  

# remove non polymorphism SNPs 
F2_geno_data_2 <- F2_geno_data_1[!(rownames(F2_geno_data_1) %in% SNP.not.poly),]

dim(F2_geno_data_2) # 3443  166, end up with 2932 SNPs for map construction 

save(F2_geno_data_2, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata") # this is non-redundant SNPs with missing data less than 0.1
``` 

### format data for onemap, polarize using parental genotype data 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 3443 166

########### below should be corrected
# read parent data 
Ae_Ol <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/vcf.Ae.Ol.intersect.final.csv", stringsAsFactors = F)

# left join to filter Ae_Ol SNP based on F2 genotypes
Ae_Ol$index <- paste(Ae_Ol$CHROM, Ae_Ol$POS, sep = "_")
F2_geno_data_2$index <- rownames(F2_geno_data_2)
F2_geno_data_2_Ae_Ol <- 
left_join(F2_geno_data_2, Ae_Ol, by="index") %>% 
  dplyr::select(-(X:ALT)) 

F2_geno_data_2_Ae_Ol <- as.matrix(F2_geno_data_2_Ae_Ol) 

# reassign genotype according to parents genotypes
F2_geno_data_2_Ae_Ol_new <- data.frame()

for (i in colnames(F2_geno_data_2_Ae_Ol)[1:166]) {
  print(i)
  for (j in 1:nrow(F2_geno_data_2_Ae_Ol)){
    if (is.na(F2_geno_data_2_Ae_Ol[j,i])){
    F2_geno_data_2_Ae_Ol_new[j,i] = "-"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == "0/1"){
      F2_geno_data_2_Ae_Ol_new[j,i] = "H"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == F2_geno_data_2_Ae_Ol[j,"Ae.gt"]){
      F2_geno_data_2_Ae_Ol_new[j,i] = "A"
    } else {
      F2_geno_data_2_Ae_Ol_new[j,i] = "B"
    }
  }
}

rownames(F2_geno_data_2_Ae_Ol_new) <- F2_geno_data_2_Ae_Ol[,"index"]
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166  

save(F2_geno_data_2_Ae_Ol_new, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2_Ae_Ol_new.Rdata")
write.table(F2_geno_data_2_Ae_Ol_new, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2_Ae_Ol_new_5556.txt")

# change file format in linux 
# cat F2_geno_data_2_Ae_Ol_new.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' | sed 's/ //g' > tmp
# tail -3443 tmp > tmp.1  

write.table(rownames(F2_geno_data_2_Ae_Ol_new), file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/marker_info.txt" )
# cat marker_info.txt | awk '{print "*"$2}' | tail -3334 | sed 's/"//g' > marker_info_reform.txt
# paste marker_info_reform.txt tmp.1 | awk '{print $1 " " $2}' > F2_geno_for_one_map.txt 
# cat header_one_map_input F2_geno_for_one_map.txt > F2_geno_for_one_map_final.txt 
# change header info: maker number to the right marker number 
```

### calculate two point rf 
F2_map_construction_LOD3_rf0.5_missing_0.10.R  

### use madmapper to assign markers to clusters/LGs 
[http://cgpdb.ucdavis.edu/XLinkage/MadMapper/Genetic_Map_MadMapper_Arabidopsis.html]

### running madmapper 
### 1) format marker data to madmapper accepted format (on MAC)
```{r}
# cat F2_geno_data_2_Ae_Ol_new.txt | sed 's/"//g' | tr " " "\t" > F2_geno_data_m.txt 
# vim edit add ";" to the 1st line
# mv F2_geno_data_m.txt F2_geno_data_m.loc 
# wc -l F2_geno_data_m.loc 3444 
# cat F2_geno_data_m.loc | awk '{print NF}' | sort | uniq # 167 
# scp to whitney 
```

* madmapper (on whitney)
http://cgpdb.ucdavis.edu/XLinkage/MadMapper/#Part_2
```{r} 
# https://github.com/leejimmy93/KIAT/blob/master/madmapper.sh
# the output file we should check is tree.clust file 
```

check madmapper output (export *.tree.clust file to Excel, and decide which SNP belong to which LG based on tree value) /Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/F2_geno_data_m_loc_mmout.x_tree_clust.xlsx 

```{r}
madmapper.result <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/F2_geno_data_m_loc_mmout.x_tree_clust.csv", header = F)

LG.A01 <- madmapper.result$V26[1:212] # 212 
LG.C01 <- madmapper.result$V26[225:355] # 131
LG.A03 <- madmapper.result$V26[356:621] # 266
LG.C03 <- madmapper.result$V26[659:844] # 186
LG.A09 <- madmapper.result$V26[845:1186] # 342
LG.C08 <- madmapper.result$V26[1187:1248] # 62
LG.A06 <- madmapper.result$V26[1261:1699] # 439
LG.A05 <- madmapper.result$V26[1717:1871] # 155
LG.C05 <- madmapper.result$V26[1872:1909] # 38
LG.A02 <- madmapper.result$V26[1924:2015] # 92
LG.A04 <- madmapper.result$V26[2016:2178] # 163
LG.C04 <- madmapper.result$V26[2179:2335] # 157
LG.A07 <- madmapper.result$V26[2336:2626] # 291
LG.A10 <- madmapper.result$V26[2627:2921] # 295
LG.C06 <- madmapper.result$V26[2922:3026] # 105
LG.C09 <- madmapper.result$V26[3027:3062] # 36
LG.A08 <- madmapper.result$V26[3075:3251] # 177
LG.C02 <- madmapper.result$V26[3252:3319] # 68
LG.C07 <- madmapper.result$V26[3320:3442] # 123

LG.madmapper <- list(LG.A01, LG.A02, LG.A03, LG.A04, LG.A05, LG.A06, LG.A07, LG.A08, LG.A09, LG.A10, LG.C01, LG.C02, LG.C03, LG.C04, LG.C05, LG.C06, LG.C07, LG.C08, LG.C09) 
length(LG.madmapper) # 19 
save(LG.madmapper, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/LG.madmapper.R")
```  

### get index for the above markers in different linkage group 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata")
twopts.f2.LOD3_rf0.5$marnames

group.AC <- list()

for (i in 1:length(LG.madmapper)) {
  group.AC[[i]] <-  which(twopts.f2.LOD3_rf0.5$marnames %in% LG.madmapper[[i]])
}
group.AC 
save(group.AC, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/group.AC.Rdata") 
```

#### order within each chromosome 
https://github.com/leejimmy93/KIAT/blob/master/F2/map_AC.R 
also tried with different LOD and rf (those are ending with too few markers ~1000 SNPs)
https://github.com/leejimmy93/KIAT/blob/master/F2/map_AC_LOD4_rf0.5.R
https://github.com/leejimmy93/KIAT/blob/master/F2/map_AC_LOD3_rf0.4.R
https://github.com/leejimmy93/KIAT/blob/master/F2/map_AC_LOD4_rf0.4.R

### combine A & C subgenome 
```{r}
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata")

# order data 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.A.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.C.Rdata")

# get order for A sub
LG.f2 <- list() 

for (i in c(1, 3, 4, 6, 7, 8, 9, 10, 11, 13)){
  LG.f2[[i]] <- make.seq(LG.f2.ord.A[[i]], "safe") 
}

for (i in c(1, 2, 3, 4, 6, 8, 9, 10, 12)){
  LG.f2[[13+i]] <- make.seq(LG.f2.ord.C[[i]], "safe") 
} 

# remove null LGs 
LG.f2 = LG.f2[-which(sapply(LG.f2, is.null))] 
draw.map(LG.f2, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.map")  
```

# R/qtl package to plot LOD & rf headmap to check marker order & assignment to LG correctness 
The reason I need to get it to R/qtl format is because double crossover function works well with that format. 
```{r}
## all markers on A & C 
LG.f2.madmapper <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.map") 

summary(LG.f2.madmapper) # 2013  
plotMap(LG.f2.madmapper) 
totmar(LG.f2.madmapper)  

# iterate over all LGs 
# rename to the right name 
names(LG.f2.madmapper$geno)[1:10] <- paste("A", c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10"), sep = "")
names(LG.f2.madmapper$geno)[11:19] <- paste("C", c("01", "02", "03", "04", "05", "06", "07", "08", "09"), sep = "") 
png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /map_before_crossover.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper, alternate.chrid = T) # the old genetic map 
dev.off()

# write.csv(pull.geno(LG.f2.madmapper),file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.before.doublecrossover.csv") 
```

### refine map 
### ripple (this part was done on whitney)
ripple uses two methods for comparing orders: maximium likelihood and minimal obligate crossovers. It is simpler to count the number of obligate crossover in the data, for a given order, and then choose the order for which the number of obligate crossover is minimized. It is generally recommended using the crossover count method with a large window, followed by maximum likelihood with a smaller window. 

### ripple with number of double crossover 
```{r} 
LG.f2.madmapper <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.map")

LG.f2.madmapper.before.crossover <- LG.f2.madmapper
save(LG.f2.madmapper.before.crossover, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.before.crossover.Rdata")

sfInit(parallel = TRUE, cpus = 12)
sfLibrary(qtl)
sfExport("LG.f2.madmapper.before.crossover")

rip <- 
  sfLapply(names(LG.f2.madmapper.before.crossover$geno), function(LG) {
    ripple(cross = LG.f2.madmapper.before.crossover, window = 7, chr = LG, method = "countxo")
  })

sfStop()

save(rip, file = "~/F2/output/missing_rate_0.10/rip.Rdata")
 
### pick a better order based on double crossover 
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[2]][2,], chr = 2)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[4]][2,], chr = 4)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[5]][2,], chr = 5)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[6]][2,], chr = 6)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[7]][2,], chr = 7)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[8]][2,], chr = 8)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[9]][2,], chr = 9)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[10]][2,], chr = 10)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[12]][2,], chr = 12)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[13]][2,], chr = 13)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[14]][2,], chr = 14)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[16]][2,], chr = 16)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[17]][2,], chr = 17)
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, rip[[18]][2,], chr = 18)

save(LG.f2.madmapper.before.crossover, file = "~/F2/output/LG.f2.madmapper.before.crossover.ripple1.Rdata")
```

# ripple with maximum likelihood, bigger LOD score and less chrlen suggest a better order
```{r}
load(file = "~/F2/output/LG.f2.madmapper.before.crossover.ripple1.Rdata")
# ripple using maximum likelihood method, window size 3: run for each LG seperately... 
# https://github.com/leejimmy93/KIAT/blob/master/F2/ripple_ML_LG1.R ...

load("~/F2/output/missing_rate_0.10/rip2.2.Rdata")
summary(rip2.2) # 2nd is the best
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 2, rip2.2[2,])
load("~/F2/output/missing_rate_0.10/rip2.3.Rdata")
summary(rip2.3) # 2nd is the best 
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 3, rip2.3[2,]) 
load("~/F2/output/missing_rate_0.10/rip2.4.Rdata")
summary(rip2.4) # 2nd is the best 
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 4, rip2.4[2,]) 
load("~/F2/output/missing_rate_0.10/rip2.5.Rdata")
summary(rip2.5) # 2nd is the best 
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 5, rip2.5[2,])
load("~/F2/output/missing_rate_0.10/rip2.6.Rdata")
summary(rip2.6) # questionable 
load("~/F2/output/missing_rate_0.10/rip2.7.Rdata")
summary(rip2.7) # questionqable 
load("~/F2/output/missing_rate_0.10/rip2.8.Rdata")
summary(rip2.8) # 2nd is the best 
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 8, rip2.8[2,])
load("~/F2/output/missing_rate_0.10/rip2.9.Rdata")
summary(rip2.9) # 2nd is the best 
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 9, rip2.9[2,])

load("~/F2/output/missing_rate_0.10/rip2.11.Rdata")
summary(rip2.11) # default
load("~/F2/output/missing_rate_0.10/rip2.12.Rdata")
summary(rip2.12) # second is the best
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 12, rip2.12[2,])
load("~/F2/output/missing_rate_0.10/rip2.13.Rdata")
summary(rip2.13) # second is the best 
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 13, rip2.13[2,])
load("~/F2/output/missing_rate_0.10/rip2.14.Rdata")
summary(rip2.14) # second is the best 
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 14, rip2.14[2,])
load("~/F2/output/missing_rate_0.10/rip2.15.Rdata")
summary(rip2.15) # default 
load("~/F2/output/missing_rate_0.10/rip2.16.Rdata")
summary(rip2.16) # default 
load("~/F2/output/missing_rate_0.10/rip2.17.Rdata")
summary(rip2.17) # default 
load("~/F2/output/missing_rate_0.10/rip2.18.Rdata")
summary(rip2.18) # second is the best
LG.f2.madmapper.before.crossover <- switch.order(LG.f2.madmapper.before.crossover, chr = 18, rip2.18[2,])
load("~/F2/output/missing_rate_0.10/rip2.19.Rdata")
summary(rip2.19) # default is the best 

# after ripple with maximum likelihood, check the result and select orders which gives me the largest LOD and shortest chrlen   

# reestimate map 
f2.map.new1 <- est.map(LG.f2.madmapper.before.crossover,verbose=T,error.prob=.01)  
LG.f2.madmapper.before.crossover <- replace.map(LG.f2.madmapper.before.crossover, f2.map.new1)
summaryMap(LG.f2.madmapper.before.crossover)
save(LG.f2.madmapper.before.crossover, file = "~/F2/output/missing_rate_0.10/LG.f2.madmapper.before.crossover.ripple2.Rdata")
```

### remove double crossover after ripple 
```{r} 
# remove double crossover 
load("~/F2/output/missing_rate_0.10/LG.f2.madmapper.before.crossover.ripple2.Rdata")
summaryMap(LG.f2.madmapper.before.crossover) # 3586.9 
LG.f2.madmapper <- LG.f2.madmapper.before.crossover

for (chr in names(LG.f2.madmapper$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2.madmapper$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2.madmapper$geno <- within(LG.f2.madmapper$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2.madmapper$geno)$data))))
}  

# restimate map with double crossover removed 
f2.map.new <- est.map(LG.f2.madmapper,verbose=T,error.prob=.01)  
LG.f2.madmapper <- replace.map(LG.f2.madmapper, f2.map.new)
LG.f2.madmapper.after.crossover <- LG.f2.madmapper
plot.map(LG.f2.madmapper.after.crossover, alternate.chrid = T) # the old genetic map

png("~/F2/output/missing_rate_0.10/map_after_ripple_rmdoublecrossover.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper.after.crossover, col.scheme = "redblue", alternate.chrid = T) 
dev.off()

LG.f2.madmapper.after.crossover.ripple <- LG.f2.madmapper.after.crossover 
save(LG.f2.madmapper.after.crossover.ripple, file = "~/F2/output/LG.f2.madmapper.after.crossover.ripple.Rdata") 
```

### just remove double crossover 
```{r}
# remove double crossover 
load("~/F2/output/LG.f2.madmapper.before.crossover.Rdata")
summaryMap(LG.f2.madmapper.before.crossover) # 3586.9 
LG.f2.madmapper <- LG.f2.madmapper.before.crossover

for (chr in names(LG.f2.madmapper$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2.madmapper$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2.madmapper$geno <- within(LG.f2.madmapper$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2.madmapper$geno)$data))))
}  

# restimate map with double crossover removed 
f2.map.new <- est.map(LG.f2.madmapper,verbose=T,error.prob=.01)  
LG.f2.madmapper <- replace.map(LG.f2.madmapper, f2.map.new)
LG.f2.madmapper.after.crossover <- LG.f2.madmapper
```

```{r}
load("~/F2/output/LG.f2.madmapper.after.crossover.Rdata")
plot.map(LG.f2.madmapper.after.crossover, alternate.chrid = T) # the old genetic map
summaryMap(LG.f2.madmapper.after.crossover.ripple) # this method gives me short length 
```

### compare LG.f2.madmapper.after.crossover, LG.f2.madmapper.after.crossover.ripple
the map with fewer missing data and shorter length should be used for QTL analysis 
```{r}
load("~/F2/output/LG.f2.madmapper.after.crossover.Rdata")
load("~/F2/output/LG.f2.madmapper.after.crossover.ripple.Rdata")

# comapre marker order 
plot.map(LG.f2.madmapper.after.crossover, LG.f2.madmapper.after.crossover.ripple)

# compare genetic length in total 
summaryMap(LG.f2.madmapper.after.crossover) # 2853.2  
summaryMap(LG.f2.madmapper.after.crossover.ripple) # 2847.2

# compare the amount of missing data 
crossover_only <- pull.geno(LG.f2.madmapper.after.crossover)
crossover_ripple <- pull.geno(LG.f2.madmapper.after.crossover.ripple)

sum(sapply(colnames(crossover_only), function(marker) sum(is.na(crossover_only[,marker])))) # 21009

sum(sapply(colnames(crossover_ripple), function(marker) sum(is.na(crossover_ripple[,marker])))) # 21015 
# so keep the original one 
```

I also tried to drop marker with high double crossover, eg above 10, using drop.marker(), then use est.map() to reestimate map, however, map length was even longer. please refer to the unused.Rmd script to see how I did that. Since that step didn't help, so it is not included here.

### continue refine map, ripple on LG10 
```{r}
load("~/F2/output/LG.f2.madmapper.after.crossover.Rdata")

# work on A10 
plot.rf(LG.f2.madmapper.after.crossover, chr = "10") # this 
plotMap(LG.f2.madmapper.after.crossover, chr = '10') 

set.seed(16)
LG.f2.madmapper.after.crossover <- orderMarkers(LG.f2.madmapper.after.crossover, chr = c(10), 
	                        window = 5, use.ripple = TRUE, maxit = 4000, 
	                        error.prob = 0.0001, verbose = T)

plotMap(LG.f2.madmapper.after.crossover, chr = '10') 
plot.rf(LG.f2.madmapper.after.crossover, chr = '10') # plot appears better but length increased... still decided to use this new order 
plotMap(LG.f2.madmapper.after.crossover)
save(LG.f2.madmapper.after.crossover, file = "~/F2/output/missing_rate_0.10/LG.f2.madmapper.after.crossover.final.Rdata")
```

### plot new rf plot 
```{r}
load("~/F2/output/missing_rate_0.10/LG.f2.madmapper.after.crossover.final.Rdata")
summaryMap(LG.f2.madmapper.after.crossover) # 2884 cM 

names(LG.f2.madmapper.after.crossover$geno)[1:10] <- paste("A", c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10"), sep = "")
names(LG.f2.madmapper.after.crossover$geno)[11:19] <- paste("C", c("01", "02", "03", "04", "05", "06", "07", "08", "09"), sep = "")  

write.cross(LG.f2.madmapper.after.crossover, format = "csvsr", filestem = "~/F2/output/missing_rate_0.10/LG.f2.madmapper.final") 
```

### graphing for report: check rf graph for map 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.before.crossover.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.Rdata")

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /rf_chr10.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper.before.crossover, chr = '10', main = "RF and LOD scores for chrom A10")
dev.off()

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /rf_madmapper.png", width=10, height=10, units="in", res=300)
plot.rf(LG.f2.madmapper.before.crossover, main = "")
dev.off() 

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /rf_chr10.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper.after.crossover, chr = 'A10', main = "RF and LOD scores for chrom A10")
dev.off() 

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /map_before_crossover.png", width=8, height=8, units="in", res=300)
plot.map(LG.f2.madmapper.before.crossover, alternate.chrid = T) # the old genetic map
dev.off()   

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /map_after_crossover.png", width=8, height=8, units="in", res=300)
plot.map(LG.f2.madmapper.after.crossover, alternate.chrid = T) # genetic map comparison
dev.off() 

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /map_comparison_1.png", width=8, height=8, units="in", res=300)
plot.map(LG.f2.madmapper.before.crossover,LG.f2.madmapper.after.crossover, alternate.chrid = T) # genetic map comparison
dev.off()  
``` 

### get line ID for the genetic map file, format pheno data  
```{r}
F2_geno_data_2_Ae_Ol_new <- 
read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2_Ae_Ol_new.txt")
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166 

F2_map <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/LG.f2.madmapper.final_gen.csv")
colnames(F2_map)[4:169]  <- gsub("X", "Sample_F2_", colnames(F2_geno_data_2_Ae_Ol_new))
colnames(F2_map) 
write.csv(F2_map, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/LG.f2.madmapper.final_gen_revised.csv")  

### format phenotype data to the right format for read.cross 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/pheno.t.Rdata")
F2.pheno <- as.data.frame(t(pheno.t))
colnames(F2.pheno)
colnames(F2.pheno) <- gsub("ID", "Sample_F2", colnames(F2.pheno))
colnames(F2.pheno)

F2.pheno <- F2.pheno[match(colnames(F2_map)[4:169], colnames(F2.pheno))] 
write.csv(F2.pheno, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2.pheno.csv")
```

<<<<<<< HEAD
### flip order for several LGs, for eQTL plotting purpose 
=======
### flip order for several LGs
```{r}
load("~/F2/output/missing_rate_0.10/LG.f2.madmapper.after.crossover.final.Rdata")
summaryMap(LG.f2.madmapper.after.crossover) # 2884 cM 

names(LG.f2.madmapper.after.crossover$geno)[1:10] <- paste("A", c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10"), sep = "")
names(LG.f2.madmapper.after.crossover$geno)[11:19] <- paste("C", c("01", "02", "03", "04", "05", "06", "07", "08", "09"), sep = "")  

plotMap(LG.f2.madmapper.after.crossover)
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A02")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A03")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A04")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A06")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A08")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A09")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "C01")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "C03")

plotMap(LG.f2.madmapper.after.crossover)
save(LG.f2.madmapper.after.crossover, file = "~/F2/output/missing_rate_0.10/LG.f2.madmapper.after.crossover.flipped.Rdata")  

### write out gen file 
write.cross(LG.f2.madmapper.after.crossover, format = "csvsr", filestem = "~/F2/output/missing_rate_0.10/LG.f2.madmapper.final.flipped") 

F2_geno_data_2_Ae_Ol_new <- 
read.table("~/F2/output/missing_rate_0.10/F2_geno_data_2_Ae_Ol_new.txt")
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166 

F2_map <- read.csv("~/F2/output/missing_rate_0.10/LG.f2.madmapper.final.flipped_gen.csv")
colnames(F2_map)[4:169]  <- gsub("X", "Sample_F2_", colnames(F2_geno_data_2_Ae_Ol_new))
colnames(F2_map) 
write.csv(F2_map, file = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv")   
``` 

### draw plot for paper 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.flipped.Rdata")

png("~/Desktop/F2_paper/rf.png", width=12, height=12, units="in", res=300)
par(mar = c(5, 5, 5, 5))  
plot.rf(LG.f2.madmapper.after.crossover, alternate.chrid = T, col.scheme = "redblue", main = "") # the old genetic map  
dev.off() 

png("~/Desktop/F2_paper/map_after_crossover.png", width=8, height=8, units="in", res=300)
plot.map(LG.f2.madmapper.after.crossover, alternate.chrid = T, main = "", xlab = "linkage groups") # genetic map comparison
dev.off()   

## map summary 
map.summary <- summaryMap(LG.f2.madmapper.after.crossover)  
map.summary %>% dim()  
map.summary[21, 1] <- sum(map.summary[1:10,1])
map.summary[22, 1] <- sum(map.summary[11:19,1])
map.summary[21, 2] <- sum(map.summary[1:10,2])
map.summary[22, 2] <- sum(map.summary[11:19,2])
map.summary[21, 3] <- mean(map.summary[1:10,3])
map.summary[22, 3] <- mean(map.summary[11:19,3])
map.summary[21, 4] <- max(map.summary[1:10,4])
map.summary[22, 4] <- max(map.summary[11:19,4])
map.summary

map_geno <- read.csv("~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv") 

map_geno <-
map_geno %>% 
  dplyr::select(id, X, X.1)  

map_geno$genetic_chr <- map_geno$X
map_geno$genetic_pos <- map_geno$X.1
map_geno$physical_chr <- gsub("(chr)([[:print:]]+)(_)([[:print:]]+)", "\\2", map_geno$id)
map_geno$genetic_chr <- factor(map_geno$physical_chr, levels = c("C09", "C08", "C07", "C06", "C05", "C04", "C03", "C02", "C01", "A10", "A09", "A08", "A07", "A06", "A05", "A04", "A03", "A02", "A01"))
map_geno$physical_pos <- gsub("(chr)([[:print:]]+)(_)([[:print:]]+)", "\\4", map_geno$id)

p.genetic_vs_physical <- 
map_geno %>% 
  ggplot() + 
  geom_point(aes(x = genetic_pos, y = as.numeric(physical_pos)), size = 0.5) +
  facet_grid(genetic_chr ~ physical_chr, scales = "free", switch="both") + 
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) + 
  labs(x = "genetic position", y = "physical position") 

p.genetic_vs_physical_A05 <- 
map_geno %>% 
  filter(map_geno$genetic_chr == "A05") %>% 
  ggplot() + 
  geom_point(aes(x = genetic_pos, y = as.numeric(physical_pos)), size = 0.5) +
  # facet_grid(genetic_chr ~ physical_chr, scales = "free", switch="both") + 
  # theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  # theme(#axis.text.x=element_blank(),
        #axis.text.y=element_blank(), 
        # axis.ticks.x=element_blank(),
        # axis.ticks.y=element_blank()) + 
  labs(x = "genetic position", y = "physical position") 

ggsave(p.genetic_vs_physical_A05, filename = "~/F2/output/p.genetic_vs_physical_A05.png", height = 5, width = 5) 
``` 

### refine the map again, see whether removing markers with segregation distortion helps shorten the length
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.flipped.Rdata")
summary(LG.f2.madmapper.after.crossover) 

gt <- geno.table(LG.f2.madmapper.after.crossover)

# remove markers with a lot of missing data, these missing data are from double crossover  
mar2drop <- rownames(gt[ gt$missing > 27,]) 
length(mar2drop)/2013 
LG.f2.madmapper.after.crossover.2 <- drop.markers(LG.f2.madmapper.after.crossover, mar2drop)
summaryMap(LG.f2.madmapper.after.crossover.2) 
plotMap(LG.f2.madmapper.after.crossover.2)   
# doesn't help 

# segregation distortion 
mar2drop <- rownames(gt[ gt$P.value < 1e-2,]) # there are quite several markers showing segregation distortion

length(mar2drop) # 397 
length(mar2drop)/2013 

mar2drop.count <- 
gt[ gt$P.value < 1e-2,] %>% 
  group_by(chr) %>% 
  dplyr::summarise(count =n()) %>% 
  mutate(count_drop = count) %>% 
  dplyr::select(-count)

# total.count <-
gt %>% 
  group_by(chr) %>% 
  dplyr::summarise(count = n()) %>% 
  left_join(mar2drop.count) %>% 
  mutate(ratio = count_drop / count) %>% 
  ggplot(aes(chr, ratio)) +
  geom_col() 

LG.f2.madmapper.after.crossover.2 <- drop.markers(LG.f2.madmapper.after.crossover, mar2drop)
summaryMap(LG.f2.madmapper.after.crossover.2) 
plotMap(LG.f2.madmapper.after.crossover.2)      

test <- LG.f2.madmapper.after.crossover.2$geno$A05$map %>% as.data.frame() 
test$physical <- gsub("([[:print:]]+)(_)([[:print:]]+)","\\3",rownames(test))
colnames(test)[1] <- "genetic"

test %>% 
  ggplot() + 
  geom_point(aes(x = genetic, y = as.numeric(as.character(physical))))  
```

check SNP distribution before filtered by F2 taxa coverage 
```{r}
F2_geno_data <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) # 18226   172 

rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)]
dim(F2_geno_data) # 18226   166 
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data))
F2_geno_data$chrom <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)(_)([[:print:]]+)", "\\3", rownames(F2_geno_data))
F2_geno_data$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)(_)([[:print:]]+)", "\\2", rownames(F2_geno_data))
F2_geno_data$pos <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)(_)([[:print:]]+)", "\\5", rownames(F2_geno_data))

# gene_density <- 
F2_geno_data %>% 
  mutate(pos1 = as.numeric(pos)) %>% 
  dplyr::select(chrom, subgenome, pos1) %>% 
  ggplot() + 
  geom_histogram(aes(x=pos1, fill=subgenome), binwidth = 1000000) + 
  facet_grid(chrom ~subgenome) + 
  theme_bw() + 
  labs(x = "gene start position", y = "number of genes") 

# after filter by F2 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 3443 166
F2_geno_data_2 %>% rownames()
F2_geno_data_2$chrom <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)(_)([[:print:]]+)", "\\3", rownames(F2_geno_data_2))
F2_geno_data_2$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)(_)([[:print:]]+)", "\\2", rownames(F2_geno_data_2))
F2_geno_data_2$pos <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)(_)([[:print:]]+)", "\\5", rownames(F2_geno_data_2))

F2_geno_data_2 %>% 
  mutate(pos1 = as.numeric(pos)) %>% 
  dplyr::select(chrom, subgenome, pos1) %>% 
  ggplot() + 
  geom_histogram(aes(x=pos1, fill=subgenome), binwidth = 1000000) + 
  facet_grid(chrom ~subgenome) + 
  theme_bw() + 
  labs(x = "gene start position", y = "number of genes")  

test <- 
sapply(1:length(LG.f2), function(i) 
 LG.f2[[i]]$seq.num %>% length)
sum(test) 
```

