---
title: "F2_map_construction_Apr_2018"
author: "Ruijuan Li"
date: "4/4/2018"
output: html_document
---

* C05 and C08 genetic map only capture half of the physical map, this was caused by the clustering process when a subset of the markers were grouped into another group, here I will order all of them no matter they group together or not in the .tree file 
* several markers with big segregation distortion at the end of LG, should be removed especially if they don't show good genetic and physical map colinearity 

```{r}
library(tidyverse)
library(qtl)
```

### 1) 
order all markers on C05 and C08   
```{r}
madmapper.result <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/F2_geno_data_m_loc_mmout.x_tree_clust.csv", header = F)

gff.mRNA.main <- gff.mRNA[grep("random", gff.mRNA$chrom, invert = T),]

LG.C05 <- grep("chrC05", madmapper.result$V26, value = T) 
LG.C08 <- grep("chrC08", madmapper.result$V26, value = T) 

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/LG.madmapper.R")
LG.madmapper[[15]] <- LG.C05
LG.madmapper[[18]] <- LG.C08

### get index for the above markers in different linkage group 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata")
twopts.f2.LOD3_rf0.5$marnames

group.AC <- list()

for (i in 1:length(LG.madmapper)) {
  group.AC[[i]] <-  which(twopts.f2.LOD3_rf0.5$marnames %in% LG.madmapper[[i]])
}
group.AC.C05_C08 <- list(group.AC[[15]], group.AC[[18]]) 
save(group.AC.C05_C08, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/group.AC.C05C08.Rdata") 

#### order within each chromosome 
# /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/KIAT/F2/map_C05_C08.R
```

```{r}
library(onemap)
library(qtl)
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata")

# order data 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.A.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.C.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.AC.C05C08.Rdata")

LG.f2.ord.C[[6]] <- LG.f2.ord.C05C08[[1]]
LG.f2.ord.C[[10]] <- LG.f2.ord.C05C08[[2]]

# get order for A sub
LG.f2 <- list() 

for (i in c(1, 3, 4, 6, 7, 8, 9, 10, 11, 13)){
  LG.f2[[i]] <- make.seq(LG.f2.ord.A[[i]], "safe") 
}

for (i in c(1, 2, 3, 4, 6, 8, 9, 10, 12)){ # 6 & 10 are C05 and C08, which only captured half of the genetic map, use "force" to extract all markers 
  LG.f2[[13+i]] <- make.seq(LG.f2.ord.C[[i]], "safe") 
} 

# C05 and C08 
# for (i in c(6, 10)){ # 6 & 10 are C05 and C08, which only captured half of the genetic map, use "force" to extract all markers 
#   LG.f2[[13+i]] <- make.seq(LG.f2.ord.C[[i]], "force") 
# } 

# remove null LGs 
LG.f2 = LG.f2[-which(sapply(LG.f2, is.null))] 
draw.map(LG.f2, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.C05C08.map")  
```

### 2) 
replace markers with double crossover (indication of mis-genotyping) with missing data, and remove markers with too many missgenotyping 
```{r}
## all markers on A & C 
LG.f2.madmapper <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.C05C08.map") 

summary(LG.f2.madmapper) # 2029  
plotMap(LG.f2.madmapper) 
totmar(LG.f2.madmapper)  

# iterate over all LGs 
# rename to the right name 
names(LG.f2.madmapper$geno)[1:10] <- paste("A", c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10"), sep = "")
names(LG.f2.madmapper$geno)[11:19] <- paste("C", c("01", "02", "03", "04", "05", "06", "07", "08", "09"), sep = "") 

# remove double crossover 
summaryMap(LG.f2.madmapper) # 6835
LG.f2.madmapper.before.crossover <- LG.f2.madmapper 

for (chr in names(LG.f2.madmapper$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2.madmapper$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2.madmapper$geno <- within(LG.f2.madmapper$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2.madmapper$geno)$data))))
}  

# restimate map with double crossover removed 
f2.map.new <- est.map(LG.f2.madmapper,verbose=T,error.prob=.01)  
LG.f2.madmapper <- replace.map(LG.f2.madmapper, f2.map.new)
LG.f2.madmapper.after.crossover <- LG.f2.madmapper
plot.map(LG.f2.madmapper.after.crossover, alternate.chrid = T) # the old genetic map

summary(LG.f2.madmapper.after.crossover) 
summaryMap(LG.f2.madmapper.after.crossover) # 2045

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/map_after_ripple_rmdoublecrossover_force.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper.after.crossover, col.scheme = "redblue", alternate.chrid = T) 
dev.off()

### deal with A10 
plot.rf(LG.f2.madmapper.after.crossover, chr = "A10")  
plotMap(LG.f2.madmapper.after.crossover, chr = 'A10') 

set.seed(16)
LG.f2.madmapper.after.crossover <- orderMarkers(LG.f2.madmapper.after.crossover, chr = "A10", 
	                        window = 5, use.ripple = TRUE, maxit = 4000, 
	                        error.prob = 0.0001, verbose = T)

plotMap(LG.f2.madmapper.after.crossover, chr = 'A10') 
plot.rf(LG.f2.madmapper.after.crossover, chr = 'A10') # plot appears better but length increased... still decided to use this new order 
plotMap(LG.f2.madmapper.after.crossover)
summaryMap(LG.f2.madmapper.after.crossover) # 3063
save(LG.f2.madmapper.after.crossover, file =  "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.C05C08.Rdata")
```

### 3) 
check segregation distortion and genetic VS physical colinearity, remove markers at the end of LG with big gaps, and showing segregation distortion, also seem to be misplaced based on genetic VS physical map colinearity  

```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.C05C08.Rdata")
summaryMap(LG.f2.madmapper.after.crossover) 
gt <- geno.table(LG.f2.madmapper.after.crossover)

# remove markers with a lot of missing data, these missing data are from double crossover, means mis-genotyping   
mar2drop <- rownames(gt[ gt$missing > 27,]) 
length(mar2drop)
LG.f2.madmapper.after.crossover.2 <- drop.markers(LG.f2.madmapper.after.crossover, mar2drop)
summaryMap(LG.f2.madmapper.after.crossover.2) # 2009 
markers_large_gap_end <- c("chrA04_6268533", "chrA05_21794646", "chrA09_30078310", "chrC05_40797215", "chrC06_35741582", "chrC08_34141406") 
plotMap(LG.f2.madmapper.after.crossover.2)

# segregation distortion 
mar_seg_dis <- rownames(gt[ gt$P.value < 1e-2,]) # there are quite several markers showing segregation distortion
mar_seg_dis
mar2drop <- markers_large_gap_end[markers_large_gap_end %in% mar2drop]
mar2drop

length(mar2drop) # 5 

LG.f2.madmapper.after.crossover.2 <- drop.markers(LG.f2.madmapper.after.crossover, mar2drop)
summaryMap(LG.f2.madmapper.after.crossover.2) # 2929.9 
plotMap(LG.f2.madmapper.after.crossover.2)      

# check total rf graph 
png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/map_after_ripple_rmdoublecrossover_force.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper.after.crossover.2, col.scheme = "redblue", alternate.chrid = T) 
dev.off() # looks good  
save(LG.f2.madmapper.after.crossover.2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/LG.f2.madmapper.after.crossover.2.Rdata")
```

### flip order for several LGs
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/LG.f2.madmapper.after.crossover.2.Rdata")
LG.f2.madmapper.after.crossover <- LG.f2.madmapper.after.crossover.2
summaryMap(LG.f2.madmapper.after.crossover) # 2929 cM, 2024 markers 

plotMap(LG.f2.madmapper.after.crossover)
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A02")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A03")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A04")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A06")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A08")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "A09")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "C01")
LG.f2.madmapper.after.crossover <- flip.order(LG.f2.madmapper.after.crossover, chr = "C03")

# three markers are not on their physical chromosome, take them out from the genetic map, this might be caused by mis-mapping
markers_miss_mapping <- c("chrA05_20283387", "chrA04_10350184", "chrA09_9478710") 
LG.f2.madmapper.after.crossover <- drop.markers(LG.f2.madmapper.after.crossover, markers_miss_mapping)
summaryMap(LG.f2.madmapper.after.crossover) # 2021 2929.9 

plotMap(LG.f2.madmapper.after.crossover) 
save(LG.f2.madmapper.after.crossover, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/LG.f2.madmapper.after.crossover.flipped.C05C08.Rdata")  

### write out gen file 
write.cross(LG.f2.madmapper.after.crossover, format = "csvsr", filestem = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/LG.f2.madmapper.final.flipped_gen_C05C08.csv")

F2_geno_data_2_Ae_Ol_new <- 
read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2_Ae_Ol_new.txt")
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166 

F2_map <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/LG.f2.madmapper.final.flipped_gen_C05C08.csv_gen.csv") 
colnames(F2_map)[4:169]  <- gsub("X", "Sample_F2_", colnames(F2_geno_data_2_Ae_Ol_new))
colnames(F2_map) 
write.csv(F2_map, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped_C05C08.csv")   # manually modify for qtl format   
```  

### make plot 
```{r}
map_geno <- read.csv("~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped_C05C08.csv") 

map_geno <-
map_geno %>% 
  dplyr::select(id, X, X.1)  

map_geno$genetic_chr <- map_geno$X
map_geno$genetic_pos <- map_geno$X.1
map_geno$physical_chr <- gsub("(chr)([[:print:]]+)(_)([[:print:]]+)", "\\2", map_geno$id)
map_geno$genetic_chr <- factor(map_geno$X, levels = c("C09", "C08", "C07", "C06", "C05", "C04", "C03", "C02", "C01", "A10", "A09", "A08", "A07", "A06", "A05", "A04", "A03", "A02", "A01"))
map_geno$physical_pos <- gsub("(chr)([[:print:]]+)(_)([[:print:]]+)", "\\4", map_geno$id)

p.genetic_vs_physical <- 
map_geno %>% 
  ggplot() + 
  geom_point(aes(x = as.numeric(genetic_pos), y = as.numeric(physical_pos)), size = 0.5) +
  facet_grid(genetic_chr ~ physical_chr, scales = "free", switch="both") + 
  theme_classic() +
  theme(panel.spacing = unit(0, "lines"), panel.border = element_rect(colour = "black", fill=NA, size=0.1)) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.ticks.y=element_blank()) +  
  labs(x = "genetic position", y = "physical position") 
p.genetic_vs_physical  

ggsave(p.genetic_vs_physical, filename = "~/F2/for_paper/p.genetic_vs_physical.png", height = , width = )

setEPS() 
postscript(file = "~/F2/output/QTL_analysis/figure/P.genetic_vs_physical.eps", width=6.8, height=9.2) 
par(mfrow=c(5,4), mai = c(0.3, 0.3, 0.3, 0.3), ps = 8, cex = 1, cex.main = 1, cex.axis = 0.8, tcl = -0.3)  

for (i in unique(map_geno$genetic_chr)){
  plot(map_geno[map_geno$genetic_chr == i,]$genetic_pos, map_geno[map_geno$genetic_chr == i,]$physical_pos, 
<<<<<<< HEAD
       xlab = "", ylab = "", main = i, type = "p", mgp=c(.1,.5,.0)) 
=======
       xlab = "genetic pos", ylab = "physical pos", main = i, type = "p", mgp=c(.1,.5,.0)) 
>>>>>>> 6c08ebd02d931dc7e6eacc741e2d1bc306ab6e8e
}

dev.off()       

# genetic map plot 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/LG.f2.madmapper.after.crossover.flipped.C05C08.Rdata")
png("~/Desktop/F2_paper/Fig4.png", width=6.7, height=3.5, units = "in", res = 300)
# setEPS()
# postscript(file = "~/Desktop/F2_paper/Fig4.eps", width=6.7, height=3.5)
par(mfrow=c(1,2), ps = 8, cex = 0.8, cex.main = 0.8, cex.axis = 0.8)

plot.map(LG.f2.madmapper.after.crossover, alternate.chrid = T, main = "") # genetic map comparison
plot.rf(LG.f2.madmapper.after.crossover, col.scheme = "redblue", alternate.chrid = T, main = "") 

dev.off()  
# summary map 
map_summary <- summaryMap(LG.f2.madmapper.after.crossover) 
write.csv(map_summary, file = "~/Desktop/F2_paper/map_summary.csv")  
``` 

# test for A04  
```{r}
map_geno %>% 
  filter(X == "A04") %>% 
  ggplot(aes(genetic_pos, as.numeric(physical_pos))) + 
  geom_point() 

map_geno %>% 
  filter(X == "A04") %>% 
  ggplot() + 
  geom_point(aes(x = as.numeric(genetic_pos), y = as.numeric(physical_pos)), size = 0.5) +
  labs(x = "genetic position", y = "physical position")  

plot(map_geno[map_geno$genetic_chr == "A04",]$genetic_pos, map_geno[map_geno$genetic_chr == i,]$physical_pos, 
       xlab = "genetic pos", ylab = "physical pos", main = i) 
```

