---
title: "gene_density"
output: html_document
---

Purpose of this script is to check several things that needs to be answered for the F2 paper 

1) gene density across the genome 

https://bernatgel.github.io/karyoploter_tutorial//Examples/GeneDensity/GeneDensity.html

practice 

```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("TxDb.Hsapiens.UCSC.hg19.knownGene")
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
all.genes <- genes(txdb)
head(all.genes)

library(karyoploteR)
kp <- plotKaryotype(genome="hg19")
kpPlotDensity(kp, all.genes)
kp <- plotKaryotype(genome="hg19", plot.type=4, ideogram.plotter = NULL, labels.plotter = NULL)
kpAddCytobandsAsLine(kp)
kpAddChromosomeNames(kp, srt=45)
kpPlotDensity(kp, all.genes)
pp <- getDefaultPlotParams(plot.type = 4)
pp$data1inmargin <- 0
pp$bottommargin <- 20

kp <- plotKaryotype(genome="hg19", plot.type=4, ideogram.plotter = NULL,
                    labels.plotter = NULL, plot.params = pp,
                    main="Gene Density")
kpAddCytobandsAsLine(kp)
kpAddChromosomeNames(kp, srt=45)
kpPlotDensity(kp, all.genes, window.size = 10e6, col="#ddaacc")

### sushi is another package for geome wide data visualization 
library('Sushi')
```

my data using a simple histogram for this 

```{r}
gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
colnames(gff.mRNA) <- c("chrom", "start", "end", "gene_ID")
gff.mRNA$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", gff.mRNA$chrom)

gff.mRNA.main <- gff.mRNA[grep("random", gff.mRNA$chrom, invert = T),]
gff.mRNA.main$chrom <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", gff.mRNA.main$chrom)
dim(gff.mRNA.main) 

gene_density <- 
gff.mRNA.main %>% 
  ggplot() + 
  geom_histogram(aes(x=start, fill=subgenome), binwidth = 1000000) + 
  facet_grid(chrom ~subgenome) + 
  theme_bw() + 
  labs(x = "gene start position", y = "number of genes") 

ggsave(gene_density, filename = "~/F2/for_paper/gene_density.png", height = 8, width = 12)  
```

###

distribution of SNP and genes using circular plot 
```{r}
library(tidyverse)
# demo 
library(ggbio)
data("CRC", package = "biovizBase")
head(hg19sub)
autoplot(hg19sub, layout = "circle", fill = "gray70")

p <- ggbio() + circle(hg19sub, geom = "ideo", fill = "gray70") +
circle(hg19sub, geom = "scale", size = 2) +
circle(hg19sub, geom = "text", aes(label = seqnames), vjust = 0, size = 3)
p 

p <- ggbio(trackWidth = 10, buffer = 0, radius = 10) + circle(hg19sub, geom = "ideo", fill = "gray70") +
circle(hg19sub, geom = "scale", size = 2) +
circle(hg19sub, geom = "text", aes(label = seqnames), vjust = 0, size = 3)
p

head(mut.gr)
p <- ggbio() + circle(mut.gr, geom = "rect", color = "steelblue") +
circle(hg19sub, geom = "ideo", fill = "gray70") +
circle(hg19sub, geom = "scale", size = 2) +
circle(hg19sub, geom = "text", aes(label = seqnames), vjust = 0, size = 3)
p

gr.crc1 <- crc.gr[values(crc.gr)$individual == "CRC-1"]
p <- p + circle(gr.crc1, geom = "point", aes(y = score, size = tumreads),
color = "red", grid = TRUE, radius = 30) + scale_size(range = c(1, 2.5))
p

p <- p + circle(gr.crc1, geom = "link", linked.to = "to.gr", aes(color = rearrangements),
radius = 23)
p

p <- ggbio() +
circle(gr.crc1, geom = "link", linked.to = "to.gr", aes(color = rearrangements)) +
circle(gr.crc1, geom = "point", aes(y = score, size = tumreads),
color = "red", grid = TRUE) + scale_size(range = c(1, 2.5)) +
circle(mut.gr, geom = "rect", color = "steelblue") +
circle(hg19sub, geom = "ideo", fill = "gray70") +
circle(hg19sub, geom = "scale", size = 2) +
circle(hg19sub, geom = "text", aes(label = seqnames), vjust = 0, size = 3)
p

# my data 
genetic_map <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped_C05C08.csv") 

genetic_map <- 
genetic_map %>% 
  mutate(SNP_ID = id, CHROM = X, genetic_pos = X.1) %>% 
  dplyr::select(SNP_ID, CHROM, genetic_pos) 

genetic_map <- GRanges(seqnames = Rle(genetic_map$CHROM),ranges = IRanges(start = genetic_map$genetic_pos, end = genetic_map$genetic_pos))
genetic_map

p <- ggbio() + 
circle(genetic_map, geom = "scale", size = 2) 
p   
```

### use circlize to draw a circos plot 

A circular layout is composed of sectors and tracks. For data in different categories, they are allocated into different sectors and for multiple measurements of the same category, they are represented as stacked tracks from outside of the circle to the inside. The intersection of a sector and a track is called a cell (or a grid, a panel), which is the basic unit in a circular layout. 

```{r}
# still doesn't look right, try circlize 
library(circlize) 

# demo
set.seed(999)
n = 1000
df = data.frame(factors = sample(letters[1:8], n, replace = TRUE),
    x = rnorm(n), y = runif(n))

## split into sectors based on factors
circos.par("track.height" = 0.1)
circos.initialize(factors = df$factors, x = df$x) 


circos.track(factors = df$factors, y = df$y,
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, CELL_META$cell.ylim[2] + uy(5, "mm"), 
            CELL_META$sector.index)
        circos.axis(labels.cex = 0.6)
})
col = rep(c("#FF0000", "#00FF00"), 4)
circos.trackPoints(df$factors, df$x, df$y, col = col, pch = 16, cex = 0.5)
circos.text(-1, 0.5, "text", sector.index = "a", track.index = 1)

bgcol = rep(c("#EFEFEF", "#CCCCCC"), 4)
circos.trackHist(df$factors, df$x, bin.size = 0.2, bg.col = bgcol, col = NA)
circos.track(factors = df$factors, x = df$x, y = df$y,
    panel.fun = function(x, y) {
        ind = sample(length(x), 10)
        x2 = x[ind]
        y2 = y[ind]
        od = order(x2)
        circos.lines(x2[od], y2[od])
})

circos.update(sector.index = "d", track.index = 2, 
    bg.col = "#FF8080", bg.border = "black")
circos.points(x = -2:2, y = rep(0.5, 5), col = "white")
circos.text(CELL_META$xcenter, CELL_META$ycenter, "updated", col = "white")

circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    xlim = CELL_META$xlim
    ylim = CELL_META$ylim
    breaks = seq(xlim[1], xlim[2], by = 0.1)
    n_breaks = length(breaks)
    circos.rect(breaks[-n_breaks], rep(ylim[1], n_breaks - 1),
                breaks[-1], rep(ylim[2], n_breaks - 1),
                col = rand_color(n_breaks), border = NA)
})

circos.link("a", 0, "b", 0, h = 0.3)
circos.link("c", c(-0.5, 0.5), "d", c(-0.5,0.5), col = "red",
    border = "blue", h = 0.2)
circos.link("e", 0, "g", c(-1,1), col = "green", border = "black", lwd = 2, lty = 2)

circos.clear() 

### more practice 
set.seed(123)
bed1 = generateRandomBed(nr = 100)
bed1 = bed1[sample(nrow(bed1), 20), ]
bed2 = generateRandomBed(nr = 100)
bed2 = bed2[sample(nrow(bed2), 20), ]

circos.initializeWithIdeogram()
circos.genomicLink(bed1, bed2, col = rand_color(nrow(bed1), transparency = 0.5), 
    border = NA) 

``` 

### try my own data 

```{r}
library(circlize)

# png("~/Desktop/circos_density.png", width=7, height=7, units="in", res=300)
# par(mfrow=c(1,1)) 
setEPS() 
postscript(file = "~/Desktop/F2_paper/density.eps", width=5.7, height=5.7)
par(mai = c(0.4, 0.4, 0.4, 0.4), ps = 8, cex = 1, cex.main = 1, cex.axis = 0.8, tcl = -0.3) 

# add SNP data for the marker, cytoband format 
F2_map <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/LG.f2.madmapper.final.flipped_gen_C05C08.csv_gen.csv")
F2_map$chrom <- F2_map$X
F2_map$start <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)(_)([[:print:]]+)", "\\5", F2_map$id) %>% as.numeric()

F2_map_cytoband <- data.frame(V1 = F2_map$chrom, 
                              V2 = F2_map$start,
                              V3 = F2_map$start + 100000, 
                              V4 = F2_map$id, 
                              V5 = c("stalk"))

circos.initializeWithIdeogram(F2_map_cytoband)    

# data 
# add SNP data 
vcf.Ae.Ol.intersect.df.2.sorted <-  read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/vcf.Ae.Ol.intersect.df.2.csv")
vcf.Ae.Ol.intersect.df.2.sorted$CHROM <- gsub("chr","",vcf.Ae.Ol.intersect.df.2.sorted$CHROM)
dim(vcf.Ae.Ol.intersect.df.2.sorted)  #  62958     9 

circos.trackHist(factors = vcf.Ae.Ol.intersect.df.2.sorted$CHROM, x = vcf.Ae.Ol.intersect.df.2.sorted$POS, bin.size = 1000000, col = "blue", border = NA, bg.col = "#EFEFEF")  

gff.mRNA <- read.table("~/Desktop/F2_paper/data/gff.mRNA")
colnames(gff.mRNA) <- c("chrom", "start", "end", "gene_ID")
gff.mRNA$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", gff.mRNA$chrom)

gff.mRNA.main <- gff.mRNA[grep("random", gff.mRNA$chrom, invert = T),]
gff.mRNA.main$chrom <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", gff.mRNA.main$chrom)
dim(gff.mRNA.main)  

gff.mRNA.main$chrom <- paste(gff.mRNA.main$subgenome, gff.mRNA.main$chrom, sep = "")

# circos.par("track.height" = 0.1)
# circos.initialize(factors = gff.mRNA.main$chrom, x = gff.mRNA.main$start)

circos.trackHist(factors = gff.mRNA.main$chrom, x = gff.mRNA.main$start, bin.size = 1000000, 
    col = "orange", border = NA, bg.col = "#EFEFEF") 

dev.off()    
```


