---
title: "linolenic_acid_genes_analysis"
author: "Ruijuan Li"
date: "2/22/2018"
output: html_document 
---

Purpose of this script is to find genes (or maybe fatty acid related genes) whose allelic pattern match the linolenic acid content level (the interaction plot). 

based on the information from the interaction plot, I am looking for genes 
1) fall into the two major QTL interval on A08 and C03 
2) have SNPs in them 
3) when both genes are homozygous identical as either parent, the linolenic acid level is very low; when both genes are heterzygous, the linolenic acid level is medium, when one gene is homozygous identical as one parent, whereas the other gene is homozygous identical as the other parent, the linolenic acid level is the highest. 

To do this, I will 
1) get John's F2 genotyping data where there are allelic information for many more SNPs
2) polarize them based on the parental genotype information 
3) only extract SNPs in the QTL interval 
4) get annotation 
5) check the pattern for several fatty acid genes firslty (plot the linolenic acid level based on the genotype for the two genes, maybe check the erucic and oleic acid etc. level as well if such gene is found for linolenic acid) 
6) more maybe afterwards if no genes found follow this pattern. 

### load data and polarize according to parental information 
```{r}
library(tidyverse) 
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures)
library("rtracklayer")

# F2_geno_data <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Final_SNP_Calls", header = T)
F2_geno_data <- read.table("~/F2/data/linolenic_acid/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) # 18226   172 

rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)]
dim(F2_geno_data) # 18226   166 
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data))

# read parent data 
# Ae_Ol <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/vcf.Ae.Ol.intersect.final.csv", stringsAsFactors = F)
Ae_Ol <- read.csv("~/F2/data/linolenic_acid/vcf.Ae.Ol.intersect.final.csv", stringsAsFactors = F)
Ae_Ol %>% dim() #  59343     7 

# left join to filter Ae_Ol SNP based on F2 genotypes
Ae_Ol$index <- paste(Ae_Ol$CHROM, Ae_Ol$POS, sep = "_")
F2_geno_data$index <- rownames(F2_geno_data)
F2_geno_data_2_Ae_Ol <- 
left_join(F2_geno_data, Ae_Ol, by="index") %>% 
  dplyr::select(-(X:ALT)) 

F2_geno_data_2_Ae_Ol <- as.matrix(F2_geno_data_2_Ae_Ol)  

# reassign genotype according to parents genotypes
F2_geno_data_2_Ae_Ol_new <- data.frame()

for (i in colnames(F2_geno_data_2_Ae_Ol)[1:166]) {
  print(i)
  for (j in 1:nrow(F2_geno_data_2_Ae_Ol)){
    if (is.na(F2_geno_data_2_Ae_Ol[j,i])){
    F2_geno_data_2_Ae_Ol_new[j,i] = "-"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == "0/1"){
      F2_geno_data_2_Ae_Ol_new[j,i] = "H"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == F2_geno_data_2_Ae_Ol[j,"Ae.gt"]){
      F2_geno_data_2_Ae_Ol_new[j,i] = "A"
    } else {
      F2_geno_data_2_Ae_Ol_new[j,i] = "B"
    }
  }
} 

dim(F2_geno_data_2_Ae_Ol_new) # 18226   166
rownames(F2_geno_data_2_Ae_Ol_new) <- F2_geno_data_2_Ae_Ol[,"index"]
F2_geno_data_2_Ae_Ol_new[1:10, 1:10] 
```

### extract SNP in the QTL interval 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/linolenic_acid_analysis/bayesint.result.physical.Rdata")
bayesint.result %>% dim() # 35 7 

bayesint.result.Linolenic <- 
bayesint.result %>% 
  filter(trait == "Linolenic_acid") 

A08.interval <- 
bayesint.result.Linolenic %>% 
  filter(chr == "A08") 

C03.interval <- 
bayesint.result.Linolenic %>% 
  filter(chr == "C03") 

dim(F2_geno_data_2_Ae_Ol_new) # 18226   166
F2_geno_data_2_Ae_Ol_new$chr = gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", rownames(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new$pos = gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\3", rownames(F2_geno_data_2_Ae_Ol_new))

F2_geno_data_2_Ae_Ol_new_interval <- 
F2_geno_data_2_Ae_Ol_new %>% 
  filter((chr == "chrA08" & as.numeric(pos) >=  A08.interval$start & as.numeric(pos) <= A08.interval$end) | (chr == "chrC03" & as.numeric(pos) >=  C03.interval$start & as.numeric(pos) <= C03.interval$end))
  
dim(F2_geno_data_2_Ae_Ol_new_interval) # 256 168 

### get gene annotation, including AGI & GO description 
gff.mRNA <- read.table("~/Desktop/Brassica_project/IGV_file/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 

colnames(F2_geno_data_2_Ae_Ol_new_interval)[167] <- "CHROM"
F2_geno_data_2_Ae_Ol_new_interval$pos <- as.numeric(F2_geno_data_2_Ae_Ol_new_interval$pos)

genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), names = gff.mRNA$name)
SNP <- GRanges(seqnames = Rle(F2_geno_data_2_Ae_Ol_new_interval$CHROM), ranges = IRanges(start = F2_geno_data_2_Ae_Ol_new_interval$pos, end = F2_geno_data_2_Ae_Ol_new_interval$pos), ID = paste(F2_geno_data_2_Ae_Ol_new_interval$CHROM, F2_geno_data_2_Ae_Ol_new_interval$pos, sep = "_")) 

SNP_gene <- mergeByOverlaps(SNP, genes)
SNP_gene <- as.data.frame(SNP_gene)

SNP_gene %>% dim() # 241 14 

load("~/Desktop/Brassica_project/reference/BnapusAnnotation.Rdata")
load("~/Desktop/Brassica_project/reference/napus_GO_combined.Rdata")
BnapusAnnotation %>% dim()
napus_GO_combined %>% dim()

SNP_gene.annotated <- 
SNP_gene %>% 
  left_join(BnapusAnnotation, c("names" = "name")) %>% 
  left_join(napus_GO_combined, c("names" = "gene_ID")) %>% 
  dplyr::select(-c(SNP.width, SNP.strand, ID, genes.width, genes.strand, names, chrom, start, end))

colnames(F2_geno_data_2_Ae_Ol_new_interval)
colnames(SNP_gene.annotated)

SNP_gene.annotated.linolenic <- 
F2_geno_data_2_Ae_Ol_new_interval %>% 
  left_join(SNP_gene.annotated, c("CHROM" = "SNP.seqnames", "pos" = "SNP.start" ))  
```

### check the pattern for several fatty acid genes firstly  
(plot the linolenic acid level based on the genotype for the two genes, maybe check the erucic and oleic acid etc. level as well if such gene is found for linolenic acid) 
```{r}
SNP_gene.annotated.linolenic %>% dim() # 256 179 
SNP_gene.annotated.linolenic %>% colnames()

phefile = read.csv("~/Desktop/F2.pheno.csv", stringsAsFactors = F, row.names = 1, header = F) 
phefile %>% dim() # 44 166 

sum(SNP_gene.annotated.linolenic$AGI == "AT4G34520")
sum(SNP_gene.annotated.linolenic$AGI == "AT4G34510")
# no fatty acid genes have SNPs identified ... 
sum(SNP_gene.annotated.linolenic$genes.names == "BnaA08g11130D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaA08g11140D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaA08g12780D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaA08g12800D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaC03g65980D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaC03g66040D")

SNP_gene.annotated.linolenic$AGI
SNP_gene.annotated.linolenic$GO_des

# do I really have SNPs in these genes between Da-Ae and Da-Ol-1? 
Ae_Ol <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/vcf.Ae.Ol.intersect.final.csv", stringsAsFactors = F)
Ae_Ol %>% dim()
Ae_Ol %>% head()

SNP_all <- GRanges(seqnames = Rle(Ae_Ol$CHROM), ranges = IRanges(start = Ae_Ol$POS, end = Ae_Ol$POS), ID = paste(Ae_Ol$CHROM, Ae_Ol$POS, sep = "_")) 

SNP_gene_all <- mergeByOverlaps(SNP_all, genes)
SNP_gene_all <- as.data.frame(SNP_gene_all)

SNP_gene_all %>% dim() # 55520    14 

SNP_gene.annotated.all <- 
SNP_gene_all %>% 
  left_join(BnapusAnnotation, c("names" = "name")) %>% 
  left_join(napus_GO_combined, c("names" = "gene_ID")) %>% 
  dplyr::select(-c(SNP_all.width, SNP_all.strand, ID, genes.width, genes.strand, names, chrom, start, end)) 

SNP_gene.annotated.all %>% colnames()

F2_geno_data_2_Ae_Ol_new_interval.all <- 
SNP_gene.annotated.all %>% 
  filter((SNP_all.seqnames == "chrA08" & as.numeric(SNP_all.start) >=  A08.interval$start & as.numeric(SNP_all.start) <= A08.interval$end) | (SNP_all.seqnames == "chrC03" & as.numeric(SNP_all.start) >=  C03.interval$start & as.numeric(SNP_all.start) <= C03.interval$end)) 

sum(SNP_gene.annotated.linolenic$genes.names == "BnaA08g11140D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaA08g11130D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaA08g11140D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaA08g12780D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaA08g12800D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaC03g65980D")
sum(SNP_gene.annotated.linolenic$genes.names == "BnaC03g66040D")

sum(SNP_gene.annotated.linolenic$genes.names == "BnaC03g64390D")

fatty_acid_genes <- c("BnaA08g11130D", "BnaA08g11140D", "BnaA08g12780D", "BnaA08g12800D", "BnaC03g65980D", "BnaC03g66040D", "BnaC03g67820D")

fatty_acid_genes %in% as.character(SNP_gene.annotated.linolenic$genes.names)

fatty_acid_genes[fatty_acid_genes %in% test$genes.names,]
test[which(test$genes.names %in% fatty_acid_genes),] 


load("~/Desktop/Brassica_project/KIAT_RNA_seq/2017_Nov_KIAT_report/SNP_gene_final_fxn.Rdata")
unique(SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names)[unique(SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names) %in% as.character(SNP_gene.annotated.linolenic$genes.names)] 

unique(SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names)[unique(SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names) %in% F2_geno_data_2_Ae_Ol_new_interval.all$genes.names]

### we don't have allele information for the 3 FAD & 4 FAE genes in the two interval. can be that: 
# 1) there is SNP in its promoter region that we could not capture from RNA-seq
# 2) four genes on A08 interval that have SNPs between Da-Ae & Da-Ol-1, but not enough coverage in the F2 population, filtered due to too much missing data across taxa  
# 3) structural variation (indel) that were not captured by our analysis (can redo the filtering step for parents and F2 population to capture indel variations to look for potential candidates).   
```

All the above thoughts are incorrect, because of genetic linkage, all genes in those two intervals, their allele combination should be the same. the jpg figure attached on github 

I am looking for genes which 
1) fall into the two QTL interval for linolenic acid 
2) has cis-eQTL 
3) two homeologous pairs with opposite direction of allele effect 

```{r}
### genes which are cis-regulators for linolenic acid fall into the two loci and have cis-eQTL 

cis_eQTL.qtl.combined.final <- read.csv("~/F2/output/eQTL/cis_eQTL.qtl.combined.final.csv", row.names = 1) 
cis_eQTL.qtl.combined.final %>% dim() # 2180 20 

cis_eQTL.qtl.combined.final.linolenic <- 
cis_eQTL.qtl.combined.final %>% 
  dplyr::filter(trait.y == "Linolenic_acid") 

dim(cis_eQTL.qtl.combined.final.linolenic) # 217 20 

### find homeologous by reprocal blast 
linolenic_acid_test_gene_eQTL <- cis_eQTL.qtl.combined.final.linolenic$gene_ID %>% unique()
write.csv(linolenic_acid_test_gene_eQTL, file = "~/F2/for_paper/linolenic_acid/linolenic_acid_test_gene_eQTL.csv")

# below on linux 
# cat linolenic_acid_test_gene_eQTL.csv | awk 'BEGIN{FS=","}{print $2}' | sed 's/"//g' | grep "BnaA" > A08_test_genes_ID_2
# cat linolenic_acid_test_gene_eQTL.csv | awk 'BEGIN{FS=","}{print $2}' | sed 's/"//g' | grep "BnaC" > C03_test_genes_ID_2

# perl -ne 'if(/^>(\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' A08_test_genes_ID_2 /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa > A08_test_genes_2.fa

# perl -ne 'if(/^>(\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' C03_test_genes_ID_2 /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa > C03_test_genes_2.fa

# makeblastdb -in A08_test_genes_2.fa -input_type fasta -dbtype nucl -out A08_test_genes_2
# makeblastdb -in C03_test_genes_2.fa -input_type fasta -dbtype nucl -out C03_test_genes_2

# ./reciprocal_blast.sh get reciprocal blast result, no new gene pairs compared to using just AGI blast hit 

### find homeologous by extracting genes which have the same Arabidopsis hit in the two intervals 
cis_eQTL.qtl.combined.final.linolenic %>% colnames() 

AGI_homeologous_hit <- 
cis_eQTL.qtl.combined.final.linolenic %>% 
  group_by(AGI, At_symbol) %>% 
  dplyr::summarise(count = n()) %>% 
  filter(count > 1) %>% 
  dplyr::select(AGI) %>% 
  unlist() %>% 
  as.character() 

AGI_homeologous_hit <- AGI_homeologous_hit[!is.na(AGI_homeologous_hit)]
 
cis_eQTL.qtl.combined.final.linolenic %>% colnames()

homolog_pair <- 
cis_eQTL.qtl.combined.final.linolenic[cis_eQTL.qtl.combined.final.linolenic$AGI %in% AGI_homeologous_hit,] %>% 
  dplyr::select(gene_ID, AGI, At_symbol, GO_des) 

### check allele effect for the genes which satisfy the model requirement 
Ara_ID_homolog <- c("AT4G18570", "AT4G19110", "AT4G19120", "AT4G19160", "AT4G19185", "AT4G20890", "AT4G21450") 

test <-   
lapply(homolog_pair[homolog_pair$AGI %in% Ara_ID_homolog,]$gene_ID %>% as.character(), function(gene) {
  allele.effect(gene)
}
)   

names(test) <- homolog_pair[homolog_pair$AGI %in% Ara_ID_homolog,]$gene_ID

allele_effect_result <- sapply(names(test), function(gene) test[[gene]]$ests[2,1])
allele_effect_result ## all possible genes have positive allele effect from Da-Ol-1 

# check the allele effect for all genes? not filtering for homeologous pair, using /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/KIAT/F2/linolenic_cis_eQTL.R 

load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/for_paper/linolenic_acid/cis_eQTL_effect.Rdata")
cis_eQTL_effect %>% length()
# next, check result to see whether there are negative effect  
names(cis_eQTL_effect) <- cis_eQTL.qtl.combined.final.linolenic$gene_ID

allele_effect_result <- sapply(names(cis_eQTL_effect), function(gene) cis_eQTL_effect[[gene]]$ests[2,1])
allele_effect_result <- allele_effect_result %>% as.data.frame() ## all possible genes have positive allele effect from Da-Ol-1 
allele_effect_result$names <- rownames(allele_effect_result) 
colnames(allele_effect_result)[1] <- "cis_effect"
allele_effect_result 

# add annotation for these genes, Ara homolog & B.napus GO annotation 
load("~/Reference/B.napus/BnapusAnnotation.Rdata") 
load("~/Reference/B.napus/napus_GO_combined.Rdata")

allele_effect_result %>% 
  left_join(BnapusAnnotation, c("names" = "name")) %>% 
  left_join(napus_GO_combined, c("names" = "gene_ID")) %>%
  View()  

colnames(napus_GO_combined)  # dosen't seem to have good candidates 
save(allele_effect_result, file = "~/F2/for_paper/linolenic_acid/allele_effect_result.Rdata")

### further exploration... 

# do FA genes have QTL at all? the answer is only two FA related genes on A08 have eQTL 
load("~/F2/for_paper/linolenic_acid/SNP_gene_final_fxn.Rdata")
SNP_gene_final_fxn %>% dim() # 26 9 fatty acid genes under QTL interval 

cis_eQTL.qtl.combined.final.linolenic[cis_eQTL.qtl.combined.final.linolenic$gene_ID %in% SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names,] 

# do these FA genes have detectable expression at all? almost all fatty acid related genes have detactable expression except for one FAE on C03.  
load("~/F2/output/eQTL/scanone-eqtl_F2_flipped.RData")
scanone_eQTL.F2 %>% dim() # 4887 56182 
colnames(scanone_eQTL.F2)[colnames(scanone_eQTL.F2) %in% SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names] 

# how about genes with trans-eQTL 
load("~/F2/output/eQTL/cis_trans_result_new_flipped.Rdata") 

cis_eQTL %>% dim() # 11384    12
trans_eQTL %>% dim() # 15197    12  

trans_eQTL %>% colnames()

trans_eQTL[trans_eQTL$gene_ID %in% SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names,] %>% View() 
cis_eQTL[cis_eQTL$gene_ID %in% SNP_gene_final_fxn$gff.lipid_fatty_acid_genes.names,] 

### check SNP would be impossible either because none of the fatty acid related genes have SNPs in them. as I remmeber, only one gene has SNPs inside the RNA sequence.  

### allele effect of the three fatty acid synthesis genes and one FAD genes on A08 
```

### can also check SNP effect, see whether there are genes fit the homeologous gene model Julin suggested 
```{r}
F2_geno_data <- read.table("~/F2/data/linolenic_acid/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) # 18226   172  

load("~/F2/for_paper/synteny_analysis/bayesint.result.physical.Rdata")
bayesint.result %>% dim() # 35 7 

bayesint.result.Linolenic <- 
bayesint.result %>% 
  filter(trait == "Linolenic_acid") 

A08.interval <- 
bayesint.result.Linolenic %>% 
  filter(chr == "A08") 

C03.interval <- 
bayesint.result.Linolenic %>% 
  filter(chr == "C03") 

dim(F2_geno_data) # 18226   172

F2_geno_data_interval <- 
F2_geno_data %>% 
  filter((CHROM == "chrA08" & as.numeric(POS) >=  (A08.interval$start - 1000000) & as.numeric(POS) <= (A08.interval$end + 1000000)) | (CHROM == "chrC03" & as.numeric(POS) >=  (C03.interval$start - 1000000) & as.numeric(POS) <= (C03.interval$end + 1000000)))
  
dim(F2_geno_data_interval) # 404 168  
F2_geno_data_interval %>% colnames()

F2_geno_data_interval_linolenic <- 
F2_geno_data_interval %>% 
  dplyr::select(CHROM, POS, REF, ALT, Ae.gt, Ol.gt) 

write.csv(F2_geno_data_interval_linolenic, file = "~/F2/for_paper/linolenic_acid/F2_geno_data_interval_linolenic.csv")

# cat F2_geno_data_interval_linolenic.csv | sed 's/"//g' | sed 's/,/ /g' | awk '{print $2 "_" $3}' | sed 's/chr/S/g' | tail -404 > significant_SNP

# cat Ae_Ol_modified_SNPonly_biallelic.recode.ann.vcf | grep -v "#" | awk '{print $0,  $1 "_" $2}' > Ae_Ol_modified_SNPonly_biallelic.recode.ann.modified.vcf

# grep -f significant_SNP /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/SNP_parent/result/freebayes/Ae_Ol_modified_SNPonly_biallelic.recode.ann.modified.vcf | awk '{print $1 "|" $2 "|" $4 "|" $5 "|" $8}' | awk 'BEGIN{FS="|"}{print $1, $2, $3, $4, $6}' > significant_SNP_annotation

sig_SNP_ann <- read_table2("~/F2/for_paper/linolenic_acid/significant_SNP_annotation", col_names=F)
head(sig_SNP_ann)  
colnames(sig_SNP_ann) <- c("CHROM", "POS", "REF", "ALT", "mutation_type")

unique(sig_SNP_ann$mutation_type)                          

F2_geno_data_interval_linolenic <- 
F2_geno_data_interval_linolenic %>% 
  left_join(sig_SNP_ann)  

gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 

genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), names = gff.mRNA$name)
genes 

SNP <- GRanges(seqnames = Rle(F2_geno_data_interval_linolenic$CHROM), ranges = IRanges(start = F2_geno_data_interval_linolenic$POS, end = F2_geno_data_interval_linolenic$POS), ID = paste(F2_geno_data_interval_linolenic$CHROM, F2_geno_data_interval_linolenic$POS, sep = "_"))
SNP

SNP_gene <- mergeByOverlaps(SNP, genes) # warning message here. look out!!!!!! 
SNP_gene

SNP_gene_df <- as.data.frame(SNP_gene)
SNP_gene_df %>% dim()

SNP_gene_final <- SNP_gene_df[,c("SNP.ID", "genes.seqnames", "SNP.start", "names")]

F2_geno_data_interval_linolenic$SNP.ID <- paste(F2_geno_data_interval_linolenic$CHROM, F2_geno_data_interval_linolenic$POS, sep = "_")

SNP_interval_linolenic <- 
F2_geno_data_interval_linolenic %>% 
  left_join(SNP_gene_final) 

write.csv(SNP_interval_linolenic, file = "~/F2/for_paper/linolenic_acid/SNP_interval_linolenic.csv")  

# send this SNP_interval_linolenic file to John, so that John can do provean analysis on these genes to see whether there are deleterious mutation on homeologous genes 

# deleterious_genes.txt is the names of the 22 genes which had a variation that scored -2.0 or less. On Provean's site they consider for human anything less than -2.5 to be a deleterious change

deleterious_genes <- read.table("~/F2/for_paper/linolenic_acid/deleterious_genes.txt")
colnames(deleterious_genes) <- "names"

# add annotation to these genes 
load("~/Reference/B.napus/BnapusAnnotation.Rdata") 
load("~/Reference/B.napus/napus_GO_combined.Rdata")

deleterious_genes %>% 
  left_join(BnapusAnnotation, c("names" = "name")) %>% 
  left_join(napus_GO_combined, c("names" = "gene_ID")) %>%
  View() 

colnames(napus_GO_combined)  # dosen't seem to have good candidates 
```

While there was no good candidate from either expression or SNP analysis, I think I might should have used the SNPs between Da-Ae and Da-Ol-1, dosn't consider whether they have good coverage in the F2s.
```{r}
Ae_Ol <- read.csv("~/F2/for_paper/linolenic_acid/vcf.Ae.Ol.intersect.2.csv", header = T, row.names = 1)
head(Ae_Ol) 
dim(Ae_Ol) # 63981     6  

load("~/F2/for_paper/synteny_analysis/bayesint.result.physical.Rdata")
bayesint.result %>% dim() # 35 7 

bayesint.result.Linolenic <- 
bayesint.result %>% 
  filter(trait == "Linolenic_acid") 

A08.interval <- 
bayesint.result.Linolenic %>% 
  filter(chr == "A08") 

C03.interval <- 
bayesint.result.Linolenic %>% 
  filter(chr == "C03") 

F2_geno_data_interval <- 
Ae_Ol %>% 
  filter((CHROM == "chrA08" & as.numeric(POS) >=  (A08.interval$start - 1000000) & as.numeric(POS) <= (A08.interval$end + 1000000)) | (CHROM == "chrC03" & as.numeric(POS) >=  (C03.interval$start - 1000000) & as.numeric(POS) <= (C03.interval$end + 1000000)))
  
dim(F2_geno_data_interval) # 1472 6  
F2_geno_data_interval %>% colnames()

F2_geno_data_interval_linolenic <- 
F2_geno_data_interval %>% 
  dplyr::select(CHROM, POS, REF, ALT, Ae.gt, Ol.gt) 

write.csv(F2_geno_data_interval_linolenic, file = "~/F2/for_paper/linolenic_acid/F2_geno_data_interval_linolenic_2.csv")

# cat F2_geno_data_interval_linolenic_2.csv | sed 's/"//g' | sed 's/,/ /g' | awk '{print $2 "_" $3}' | sed 's/chr/S/g' | tail -1472 > significant_SNP_2

# cat Ae_Ol_modified_SNPonly_biallelic.recode.ann.vcf | grep -v "#" | awk '{print $0,  $1 "_" $2}' > Ae_Ol_modified_SNPonly_biallelic.recode.ann.modified.vcf

# cat significant_SNP_2 | sed 's/S//g' | awk '{print "chr" $0}' > significant_SNP

# grep -f significant_SNP /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/SNP_parent/result/freebayes/Ae_Ol_modified_SNPonly_biallelic.recode.ann.modified.vcf | awk '{print $1 "|" $2 "|" $4 "|" $5 "|" $8}' | awk 'BEGIN{FS="|"}{print $1, $2, $3, $4, $6}' > significant_SNP_annotation

sig_SNP_ann <- read_table2("~/F2/for_paper/linolenic_acid/significant_SNP_annotation", col_names=F)
head(sig_SNP_ann)  
dim(sig_SNP_ann)
colnames(sig_SNP_ann) <- c("CHROM", "POS", "REF", "ALT", "mutation_type")

unique(sig_SNP_ann$mutation_type)                          

F2_geno_data_interval_linolenic <- 
F2_geno_data_interval_linolenic %>% 
  left_join(sig_SNP_ann)  

gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 

genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), names = gff.mRNA$name)
genes 

SNP <- GRanges(seqnames = Rle(F2_geno_data_interval_linolenic$CHROM), ranges = IRanges(start = F2_geno_data_interval_linolenic$POS, end = F2_geno_data_interval_linolenic$POS), ID = paste(F2_geno_data_interval_linolenic$CHROM, F2_geno_data_interval_linolenic$POS, sep = "_"))
SNP

SNP_gene <- mergeByOverlaps(SNP, genes) # warning message here. look out!!!!!! 
SNP_gene

SNP_gene_df <- as.data.frame(SNP_gene)
SNP_gene_df %>% dim()

SNP_gene_final <- SNP_gene_df[,c("SNP.ID", "genes.seqnames", "SNP.start", "names")]

F2_geno_data_interval_linolenic$SNP.ID <- paste(F2_geno_data_interval_linolenic$CHROM, F2_geno_data_interval_linolenic$POS, sep = "_")

SNP_interval_linolenic <- 
F2_geno_data_interval_linolenic %>% 
  left_join(SNP_gene_final) 
dim(SNP_interval_linolenic) # 1472 

save(SNP_interval_linolenic, file = "~/F2/for_paper/linolenic_acid/SNP_interval_linolenic.Rdata")
load("~/Reference/B.napus/BnapusAnnotation.Rdata") 
load("~/Reference/B.napus/napus_GO_combined.Rdata")

homeo_AGI <- 
SNP_interval_linolenic %>% 
  left_join(BnapusAnnotation, c("names" = "name")) %>% 
  left_join(napus_GO_combined, c("names" = "gene_ID")) %>%
  dplyr::select(CHROM, AGI) %>% 
  unique() %>% 
  group_by(AGI) %>% 
  count(count = n()) %>% 
  filter(count == 2) %>%
  dplyr::select(AGI) %>% 
  unlist() %>% 
  as.character()

homeo_AGI <- homeo_AGI[!is.na(homeo_AGI)]

SNP_interval_linolenic_annotated <-   
SNP_interval_linolenic %>% 
  left_join(BnapusAnnotation, c("names" = "name")) %>% 
  left_join(napus_GO_combined, c("names" = "gene_ID")) 

candidate_linolenic <- 
SNP_interval_linolenic_annotated[SNP_interval_linolenic_annotated$AGI %in% as.character(unlist(homeo_AGI)),] 

write.csv(candidate_linolenic, file = "~/F2/for_paper/linolenic_acid/candidate_linolenic.csv")
write.csv(SNP_interval_linolenic_annotated, file = "~/F2/for_paper/linolenic_acid/SNP_interval_linolenic_2_annotated.csv")   

SNP_interval_linolenic_annotated[SNP_interval_linolenic_annotated$AGI %in% as.character(unlist(homeo_AGI)),] %>% 
  dplyr::select(CHROM, mutation_type, SNP.ID, names, AGI, At_symbol) %>% 
  View()

SNP_interval_linolenic_annotated %>% 
  filter(names == "BnaA08g11130D") # FAE1

SNP_interval_linolenic_annotated %>% 
  filter(names == "BnaC03g65980D") # FAE1

SNP_interval_linolenic_annotated %>% 
  filter(names == "BnaA08g11060D") # SQS

SNP_interval_linolenic_annotated %>% 
  filter(names == "BnaA08g11140D") # KCS17

SNP_interval_linolenic_annotated %>% 
  filter(names == "BnaA08g12780D") # FAD6 

SNP_interval_linolenic_annotated %>% 
  filter(names == "BnaA08g27750D") # KCS2

SNP_interval_linolenic_annotated %>% 
  filter(names == "BnaA08g12800D") # FAD6 

# John run PROVEAN on the list of SNPs that I sent to him, 
provean.result <- read.table("~/F2/for_paper/linolenic_acid/provean_results.txt")
candidate_linolenic %>% 
  filter(names == "BnaA08g10160D")

provean.result <- read.csv("~/Desktop/F2_paper/data/candidate_linolenic.csv")
provean.result %>% colnames()
provean.result %>% 
  dplyr::select(names, AGI, At_symbol, mutation_type, GO_des:X.4) %>% 
  View()  

# we are worried that using Arabidopsis blast hit to find homeologou might not be the best method, so I will use reciprocal blast to look for homeologs on A08 and C03 for linolenic acid QTL region 

load("~/F2/for_paper/linolenic_acid/SNP_interval_linolenic.Rdata")
linolenic_acid_test_gene <- SNP_interval_linolenic$names %>% unique() 
write.csv(linolenic_acid_test_gene, file = "~/F2/for_paper/linolenic_acid/linolenic_acid_test_gene.csv")

# below on linux 
# cat linolenic_acid_test_gene.csv | awk 'BEGIN{FS=","}{print $2}' | sed 's/"//g' | grep "BnaA" > A08_test_genes_ID
# cat linolenic_acid_test_gene.csv | awk 'BEGIN{FS=","}{print $2}' | sed 's/"//g' | grep "BnaC" > C03_test_genes_ID

# perl -ne 'if(/^>(\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' A08_test_genes_ID /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa > A08_test_genes.fa

# perl -ne 'if(/^>(\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' C03_test_genes_ID /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa > C03_test_genes.fa

# makeblastdb -in A08_test_genes.fa -input_type fasta -dbtype nucl -out A08_test_genes
# makeblastdb -in C03_test_genes.fa -input_type fasta -dbtype nucl -out C03_test_genes

# ./reciprocal_blast.sh get reciprocal blast result 
blast_result <- read.delim(file = "~/F2/for_paper/linolenic_acid/reciprocal.blast", header = F)
blast_result

homeo_blast <- 
  c(as.character(blast_result$V1), as.character(blast_result$V2))

homeo_blast <- 
  homeo_blast[!(homeo_blast %in% (candidate_linolenic$names %>% unique()))] 

candidate_linolenic_2genes <- 
SNP_interval_linolenic_annotated[SNP_interval_linolenic_annotated$names %in% homeo_blast,] 

candidate_linolenic_2genes %>% 
  dplyr::select(Ae.gt, Ol.gt, names, mutation_type, AGI, At_symbol) %>% 
  View()
  
write.csv(candidate_linolenic_2genes, file = "~/F2/for_paper/linolenic_acid/candidate_linolenic_2genes.csv") 
```

# look for linolenic acid candidate genes by combining expression/eQTL and SNP/provean data 

```{r}
load("~/F2/for_paper/linolenic_acid/allele_effect_result.Rdata")
allele_effect_result

blast_result_eQTL <- read.delim(file = "~/F2/for_paper/linolenic_acid/reciprocal.blast_eQTL", header = F)
blast_result_eQTL

homeo_blast_eQTL <- 
  c(as.character(blast_result_eQTL$V1), as.character(blast_result_eQTL$V2))

homeo_blast_eQTL
cis_eQTL_linolenic <- allele_effect_result[allele_effect_result$names %in% homeo_blast_eQTL,]
cis_eQTL_linolenic

# add annotation 
  
cis_eQTL_linolenic %>% 
  left_join(BnapusAnnotation, c("names" = "name")) %>% 
  left_join(napus_GO_combined, c("names" = "gene_ID"))%>% 
  View() 

# combine with SNP result to look for potential candidates ... 

blast_result_SNP <- read.delim(file = "~/F2/for_paper/linolenic_acid/reciprocal.blast", header = F)
blast_result_SNP 
```

### appendix function  
```{r}
### need a function to get allele effect 
load("~/F2/output/eQTL/cis_trans_result_new_flipped.Rdata") 

allele.effect <- function(gene_ID){
  cross.F2 <- read.cross("csvsr", 
                         genfile ="~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv", 
                         phefile = "~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv",
                         genotypes = c("AA", "AB", "BB"))   # although the sample IDs are not matched in the original phe and gen file, I still get the right result. 
  cross.F2$pheno <- as.data.frame(cross.F2$pheno[,gene_ID]) 
  cross.F2 <- sim.geno(cross.F2,step=1,n.draws=32) # imputation?  
  cross.F2 <- calc.genoprob(cross.F2,step=1) 

  chr = cis_eQTL[cis_eQTL$gene_ID == gene_ID,]$eQTL_chr
  chr = gsub("chr", "", chr)
  pos = cis_eQTL[cis_eQTL$gene_ID == gene_ID,]$pos
  qtlm <- makeqtl(cross.F2,chr=chr,pos=pos)  

  qtl.fit <- fitqtl(cross=cross.F2,qtl=qtlm,formula = y ~ Q1, get.ests=T)

  summary(qtl.fit)    
} 
```



