---
title: "QTL_analysis_flipped_C05C08"
output: html_document
---

# Goal of this script is to use the new genetic map to conduct QTL analysis, re-make all the figures and tables... 

### get genes under QTL for different traits, use cim result, if there is no cim result, using scanone result 
```{r} 
load("~/F2/output/QTL_analysis/QTL_result_all.C05C08.Rdata")

scanone.perm.imp.all %>% length() # 51 
scanone.imp.all %>% length() 
cim.perm.all %>% length()
cim.qtl.all %>% length()

### scanone 
threshold.95 <- tibble(perm.threshold = bind_rows(scanone.perm.imp.all) %>% as.numeric(), 
                       trait = colnames(bind_rows(scanone.perm.imp.all)))
threshold.95

scanone.qtl.2 <-  
bind_cols(scanone.imp.all) %>% 
  dplyr::select(chr, pos, starts_with("lod"))
rownames(scanone.qtl.2) <- rownames(scanone.imp.all$Crude_oil_contents)
colnames(scanone.qtl.2)[3:ncol(scanone.qtl.2)] <- names(scanone.imp.all)

scanone.gather <- scanone.qtl.2 %>%
  gather(key = trait, value = LOD, -chr, -pos) %>%
  left_join(threshold.95)

scanone.gather %>% head() 

# look for overlap, for each trait, find QTL border and look for genes under QTL peaks 
sig.chrs <- scanone.gather %>% dplyr::filter(LOD > perm.threshold) %>%
  group_by(trait,chr) %>%
  dplyr::summarise(count = n())
sig.chrs 

# now for each significant chromosome/trait combo run bayesint
bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(scanone.qtl.2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE 
  )
  colnames(result)[3] <- "LOD" 
  result
})  
 

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x)  
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr))
) 

bayesint.list 
bayesint.list %>% length() # 33 --> 32
bayesint.list.scanone <- bayesint.list
length(bayesint.list.scanone) 

bayesint.result.scanone <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD), genetic_start=min(pos, na.rm = T), genetic_end=max(pos, na.rm = T)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(start=ifelse(start==end,max(0,start-20000),start), end=ifelse(start==end,end+20000,end))

bayesint.result.scanone %>% dim() # 33 8 --> 32 8 

### cim
threshold.95 <- tibble(perm.threshold = bind_rows(cim.perm.all) %>% as.numeric(), 
                       trait = colnames(bind_rows(cim.perm.all)))
threshold.95

cim.qtl.2 <-  
bind_cols(cim.qtl.all) %>% 
  dplyr::select(chr, pos, starts_with("lod"))
rownames(cim.qtl.2) <- rownames(cim.qtl.all$Crude_oil_contents)
colnames(cim.qtl.2)[3:ncol(cim.qtl.2)] <- names(cim.qtl.all)

cim.gather <- cim.qtl.2 %>%
  gather(key = trait, value = LOD, -chr, -pos) %>%
  left_join(threshold.95)

cim.gather %>% head() 

# look for overlap, for each trait, find QTL border and look for genes under QTL peaks 
sig.chrs <- cim.gather %>% dplyr::filter(LOD > perm.threshold) %>%
  group_by(trait,chr) %>%
  dplyr::summarise(count = n()) 
sig.chrs 

# now for each significant chromosome/trait combo run bayesint
sig.chrs <- 
sig.chrs %>% 
  filter(trait != "root_weight_2016_05_13")

##
bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(cim.qtl.2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE 
  )
  colnames(result)[3] <- "LOD" 
  result
})   # problem with root weight 

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x)  
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr))
) 

bayesint.list 
bayesint.list %>% length() # 26 --> 24 

# save bayesint result for later 
bayesint.list.cim <- bayesint.list
save(bayesint.list.scanone, bayesint.list.cim, file = "~/F2/output/QTL_analysis/bayesint.list.scanone.cim.C05C08.Rdata")

bayesint.result.cim <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD), genetic_start=min(pos, na.rm = T), genetic_end=max(pos, na.rm = T)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(start=ifelse(start==end,max(0,start-20000),start), end=ifelse(start==end,end+20000,end))

bayesint.result.cim %>% dim() # 26 8

bayesint.result.cim$model <- rep("cim", nrow(bayesint.result.cim))
bayesint.result.scanone$model <- rep("scanone", nrow(bayesint.result.scanone))

bayesint.result <- 
bayesint.result.cim %>% 
  full_join(bayesint.result.scanone, by = c("trait", "chr")) 

bayesint.result %>% dim() # 35 16 
colnames(bayesint.result)

for (i in 1:nrow(bayesint.result)){
  if(is.na(bayesint.result[i, "start.x"])){
    bayesint.result[i, c(3:9)] <- bayesint.result[i, c(10:16)]
  }
}

bayesint.result <- bayesint.result[,1:9] 
colnames(bayesint.result) <- gsub("\\.x$", "", colnames(bayesint.result)) 

bayesint.result %>% dim() # 35 9 

# annotate QTL   
# load("~/Reference/B.napus/BnapusAnnotation.Rdata") 

## where are the genes fkf for flowering time (Long et al. 2007) 
BnapusAnnotation %>% colnames()
BnapusAnnotation %>% 
  filter(AGI == "AT1G68050") 
###

traitQTL.annotated <- lapply(1:nrow(bayesint.result),function(row) { # for each trait/module 
  qtl <- bayesint.result[row,]  
  results <- subset(BnapusAnnotation, chrom==qtl$chr &
                    start >= qtl$start & # genes which fall into the QTL interval 
                    end <= qtl$end)
} 
)  
names(traitQTL.annotated) <- bayesint.result$trait 

traitQTL.annotated <- bind_rows(traitQTL.annotated,.id="trait") %>% # combine list into data.frame 
  mutate(chrom=as.character(chrom)) %>%
  left_join(bayesint.result,by=c("trait","chrom"="chr")) #get eQTL LOD

traitQTL.annotated <- 
traitQTL.annotated %>% 
  mutate(start = start.y, end = end.y) %>% 
  dplyr::select(-start.x, -end.x, -start.y, -end.y, -min_eQTL_LOD, -max_eQTL_LOD) 

traitQTL.annotated %>% dim() # 18647     11

## get GO term for each gene
load("~/Reference/B.napus/napus_GO_combined.Rdata") # there is one colomn hiding... 

colnames(traitQTL.annotated)[3] <- "gene_ID"

traitQTL.annotated <- 
traitQTL.annotated %>% 
  left_join(napus_GO_combined) 

traitQTL.annotated %>% dim() #  18647    13 
# save(traitQTL.annotated, file =  "~/F2/output/QTL_analysis/traitQTL.annotated.flipped.Rdata") 
# write.csv(traitQTL.annotated, file = "~/F2/for_paper/traitQTL.annotated.csv") 
``` 

### table to summarize QTL mapping result, make R square, additive effect... 
```{r}
load("~/F2/output/QTL_analysis/bayesint.list.scanone.cim.C05C08.Rdata")
length(bayesint.list.scanone) # 33
length(bayesint.list.cim) # 26  

bayesint.result.scanone <- as.tibble(bind_rows(bayesint.list.scanone,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(pos, na.rm = T),end=max(pos, na.rm = T),pos = median(pos, na.rm = T), LOD=max(LOD))  

bayesint.result.scanone %>% dim() # 33 6 

bayesint.result.tmp <- as.tibble(bind_rows(bayesint.list.scanone,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  # separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) 

bayesint.result.tmp$index <- paste(bayesint.result.tmp$trait, bayesint.result.tmp$chr, bayesint.result.tmp$LOD)
bayesint.result.scanone$index <- paste(bayesint.result.scanone$trait, bayesint.result.scanone$chr, bayesint.result.scanone$LOD)

test <- 
bayesint.result.tmp %>% 
  anti_join(bayesint.result.scanone) %>% 
  dplyr::select(trait, chr, markername) %>% 
  mutate(index = paste(trait, chr, sep = "_")) 

tmp <- c()
tmp  

for (i in seq_along(1:(nrow(test)/2))){
  tmp[i] <- paste(test$markername[i*2-1], test$markername[i*2-0], sep = "-") 
}

bayesint.result.scanone$index <- tmp

bayesint.result.scanone <- 
bayesint.result.scanone %>%
  mutate(start = round(start, 2), end = round(end, 2), pos = round(pos, 2), LOD = round(LOD, 2)) %>%
  unite(confidence_interval, start, end, sep = "-") %>% 
  mutate(flanking_marker = index) %>% 
  dplyr::select(-index)  

bayesint.result.scanone %>% dim() # 33 6 

### cim result 
bayesint.list.cim %>% length() # 26

bayesint.result <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(pos, na.rm = T),end=max(pos, na.rm = T),pos = median(pos, na.rm = T), LOD=max(LOD)) 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.

bayesint.result %>% dim() # 26 6 
bayesint.result 

bayesint.result.tmp <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  # separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) 

bayesint.result.tmp$index <- paste(bayesint.result.tmp$trait, bayesint.result.tmp$chr, bayesint.result.tmp$LOD)
bayesint.result$index <- paste(bayesint.result$trait, bayesint.result$chr, bayesint.result$LOD)

test <- 
bayesint.result.tmp %>% 
  anti_join(bayesint.result) %>% 
  dplyr::select(trait, chr, markername) %>% 
  mutate(index = paste(trait, chr, sep = "_")) 

tmp <- c()
tmp  

for (i in seq_along(1:(nrow(test)/2))){
  tmp[i] <- paste(test$markername[i*2-1], test$markername[i*2-0], sep = "-") 
}

bayesint.result$index <- tmp ##### 

bayesint.result.cim <- 
bayesint.result %>% 
  mutate(start = round(start, 2), end = round(end, 2), pos = round(pos, 2), LOD = round(LOD, 2)) %>%
  unite(confidence_interval, start, end, sep = "-") %>% 
  mutate(flanking_marker = index) %>% 
  dplyr::select(-index)  

bayesint.result.cim %>% dim() # 26 6  

### combine cim & scanone result 
bayesint.result.scanone$model <- rep("scanone", nrow(bayesint.result.scanone))
bayesint.result.cim$model <- rep("cim", nrow(bayesint.result.cim))

bayesint.result.paper <- 
bayesint.result.cim %>% 
  full_join(bayesint.result.scanone, by = c("trait", "chr")) 

bayesint.result.paper %>% dim() # 32 12
colnames(bayesint.result.paper)

for (i in 1:nrow(bayesint.result.paper)){
  if(is.na(bayesint.result.paper[i, "pos.x"])){
    bayesint.result.paper[i, c(3:7)] <- bayesint.result.paper[i, c(8:12)]
  }
}

bayesint.result.paper <- bayesint.result.paper[,1:7] 
colnames(bayesint.result.paper) <- gsub("\\.x$", "", colnames(bayesint.result.paper)) 

bayesint.result.paper %>% dim() # 35 7 
bayesint.result.paper %>% View()

save(bayesint.result.paper, file = "~/F2/output/QTL_analysis/bayesint.result.paper.Rdata") 
write.csv(bayesint.result.paper, file = "~/F2/output/QTL_analysis/bayesint.result.paper.csv") 

### re-edit the position according to scantwo result 

### 1) days_to_flower 
# 		fitqtl summary
# 
# Method: multiple imputation 
# Model:  normal phenotype
# Number of observations : 165 
# 
# Full model result
# ----------------------------------  
# Model formula: y ~ Q1 + Q2 
# 
#        df       SS        MS      LOD     %var Pvalue(Chi2)    Pvalue(F)
# Model   4 10800.02 2700.0041 12.45336 29.36008 1.044798e-11 2.055756e-11
# Error 160 25984.69  162.4043                                            
# Total 164 36784.70                                                      
# 
# 
# Drop one QTL at a time ANOVA table: 
# ----------------------------------  
#           df Type III SS   LOD  %var F value Pvalue(Chi2) Pvalue(F)    
# A10@183.9  2        6080 7.533 16.53   18.72            0  4.96e-08 ***
# C06@67.3   2        4264 5.444 11.59   13.13            0  5.26e-06 *** ####### need to change this pos 
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# 
# Estimated effects:
# -----------------
#                  est        SE       t
# Intercept  151.55170   1.01004 150.046 
# A10@183.9a  -7.35767   1.33487  -5.512
# A10@183.9d   4.71660   2.03236   2.321
# C06@67.3a   -7.63332   1.49116  -5.119
# C06@67.3d   -0.09159   2.02693  -0.045

### linolenic acid 
# Full model result
# ----------------------------------  
# Model formula: y ~ Q1 + Q2 + Q1:Q2 
# 
#        df        SS       MS      LOD     %var Pvalue(Chi2) Pvalue(F)
# Model   8 1335.8103 166.9763 66.82275 84.33586            0         0
# Error 157  248.1071   1.5803                                         
# Total 165 1583.9174                                                  
# 
# 
# Drop one QTL at a time ANOVA table: 
# ----------------------------------  
#                    df Type III SS   LOD  %var F value Pvalue(Chi2) Pvalue(F)    
# A08@39.0            6      1183.1 63.17 74.70   124.8            0    <2e-16 ***
# C03@181.0           6      1124.7 61.67 71.01   118.6            0    <2e-16 ***
# A08@39.0:C03@181.0  4       982.3 57.72 62.02   155.4            0    <2e-16 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# 
# Estimated effects:
# -----------------
#                          est      SE       t
# Intercept            10.8127  0.1293  83.606
# A08@39.0a            -0.3259  0.1892  -1.722
# A08@39.0d             1.7490  0.2591   6.752
# C03@181.0a           -0.8086  0.1862  -4.343
# C03@181.0d            0.9781  0.2588   3.779 
# A08@39.0a:C03@181.0a -4.3295  0.2648 -16.349
# A08@39.0d:C03@181.0a  1.3157  0.3739   3.519
# A08@39.0a:C03@181.0d  1.2020  0.3776   3.183
# A08@39.0d:C03@181.0d -0.7604  0.5193  -1.464

### days_to_bolt ### the additional marker is maginal, can add or not, so no need to care about it too much 

# Full model result
# ----------------------------------  
# Model formula: y ~ Q1 + Q2 + Q3 
# 
#        df       SS        MS      LOD    %var Pvalue(Chi2)    Pvalue(F)
# Model   6 11634.05 1939.0081 15.47209 35.0679 2.263745e-13 6.379341e-13
# Error 158 21541.73  136.3401                                           
# Total 164 33175.78                                                     
# 
# 
# Drop one QTL at a time ANOVA table: 
# ----------------------------------  
#           df Type III SS   LOD  %var F value Pvalue(Chi2) Pvalue(F)    
# A10@176.0  2        4731 7.113 14.26  17.350            0  1.54e-07 ***
# C06@85.7   2        3833 5.868 11.55  14.057            0  2.41e-06 ***
# A05@114.0  2        2697 4.227  8.13   9.891            0  8.97e-05 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# 
# Estimated effects:
# -----------------
#                 est       SE       t
# Intercept  119.4954   0.9383 127.351
# A10@176.0a  -7.0704   1.2035  -5.875
# A10@176.0d   1.0081   1.8948   0.532
# C06@85.7a   -7.2351   1.4103  -5.130
# C06@85.7d    2.4429   1.8949   1.289
# A05@114.0a   5.7853   1.3349   4.334
# A05@114.0d  -0.6941   1.9010  -0.365

### linolenic acid, need to be more careful to analyze 

### combine with allele effect information 
bayesint.result.allele_effect <- read.csv("~/F2/output/QTL_analysis/bayesint.result.paper_allele_effect.csv")
bayesint.result.allele_effect %>% dim()
bayesint.result.paper %>% dim()

bayesint.result.paper.final <- 
bayesint.result.paper %>%
  left_join(bayesint.result.allele_effect, by = c("trait", "chr")) %>% 
  mutate(model = model.x) %>%
  dplyr::select(-model.x, -model.y, -X) 

bayesint.result.paper.final %>% View()
write.csv(bayesint.result.paper.final, file = "~/F2/output/QTL_analysis/bayesint.result.paper.final.csv")  

### add physical position interval  
load("~/F2/output/QTL_analysis/bayesint.list.scanone.cim.Rdata")
length(bayesint.list.scanone) # 33
length(bayesint.list.cim) # 26 

bayesint.result.scanone <- as.tibble(bind_rows(bayesint.list.scanone,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD))

bayesint.result.cim <- as.tibble(bind_rows(bayesint.list.cim,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD))

bayesint.result.cim$model <- rep("cim", nrow(bayesint.result.cim))
bayesint.result.scanone$model <- rep("scanone", nrow(bayesint.result.scanone))

bayesint.result <- 
bayesint.result.cim %>% 
  full_join(bayesint.result.scanone, by = c("trait", "chr")) 

bayesint.result %>% dim()
colnames(bayesint.result)

for (i in 1:nrow(bayesint.result)){
  if(is.na(bayesint.result[i, "start.x"])){
    bayesint.result[i, c(3:7)] <- bayesint.result[i, c(8:12)]
  }
}

bayesint.result <- bayesint.result[,1:7] 
colnames(bayesint.result) <- gsub("\\.x$", "", colnames(bayesint.result)) 

bayesint.result %>% dim() # 35 7 
bayesint.result %>% View()
save(bayesint.result, file = "~/F2/for_paper/synteny_analysis/bayesint.result.physical.Rdata") 

bayesint.result.physical <-
bayesint.result.paper.final %>% 
  left_join(bayesint.result, by= c("trait", "chr")) %>% 
  mutate(model = model.x) %>% 
  dplyr::select(-min_eQTL_LOD, -max_eQTL_LOD, -model.y, -model.x) 
  
write.csv(bayesint.result.physical, file = "~/F2/output/QTL_analysis/bayesint.result.paper.physical.csv") 
``` 