---
title: "scantwo_flipped"
output: html_document
---

* Goal of this script is to analyze scantwo results for all phenotypes, including growth model result 

```{r}
library(qtl) 
```

### load data 
```{r}
# scantwo analysis on cluster 
# /share/malooflab/Ruijuan/KIAT_cabernet/F2/scantwo_flipped 

# scp result to whitney 
# QTL result 
load("~/F2/output/scantwo/flipped/scantwo.imp.v2.flipped.1.Rdata") 
scantwo.imp.1 <- scantwo.imp
load("~/F2/output/scantwo/flipped/scantwo.imp.v2.flipped.2.Rdata")
scantwo.imp.2 <- scantwo.imp
load("~/F2/output/scantwo/flipped/scantwo.imp.v2.flipped.3.Rdata")
scantwo.imp.3 <- scantwo.imp
load("~/F2/output/scantwo/flipped/scantwo.imp.v2.flipped.4.Rdata")
scantwo.imp.4 <- scantwo.imp

scantwo.imp <- c(scantwo.imp.1, scantwo.imp.2, scantwo.imp.3, scantwo.imp.4)
length(scantwo.imp) # 15 

# scanone & CIM result 
load("~/F2/output/QTL_analysis/QTL_result_all.Rdata")
cim.perm.all %>% length() # 51
cim.qtl.all %>% length() # 51
scanone.perm.imp.all %>% length() # 51
scanone.imp.all %>% length() # 51

names(scantwo.imp) 
```

### Crude oil content 
```{r}
# load data 
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Crude_oil_contents_Caprylic_acid/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Crude_oil_contents_Caprylic_acid <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
scantwo_perm_Crude_oil_contents <- summary(do.call(c, scantwo_perm_Crude_oil_contents_Caprylic_acid)[,1])

# oil content 
scantwo_perm_Crude_oil_contents
summary(scantwo.imp[["Crude_oil_contents"]], thresholds = c(11.7, 9.19, 7.72, 7.47, 5.51)) # signal for just additive    

summary(scanone.imp.all[["Crude_oil_contents"]]) # 98.5
summary(cim.qtl.all[["Crude_oil_contents"]]) # 98.5

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Crude_oil_contents)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.oil_content <- makeqtl(LG.f2.after.crossover,chr=c("A03"),pos=c(98)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.oil_content # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.oil_content <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.oil_content,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.oil_content)   
effectplot(cross=LG.f2.after.crossover, mname1 = "A03@98")  
```

### Palmitoliec_aicd 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Palmitoliec_Heptadecanoic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Palmitoliec_Heptadecanoic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
summary(do.call(c, scantwo_perm_Palmitoliec_Heptadecanoic)[,1])
# Palmitoliec_aicd (160 permutations)
#      full  fv1  int  add  av1  one
# 5%  100.0 98.8 93.2 91.7 90.6 5.11
# 10%  99.5 98.2 92.0 91.4 90.3 4.64 

summary(scantwo.imp[["Palmitoliec_acid"]], thresholds = c(100.0, 98.8, 93.2, 91.7, 90.6)) # signal for just additive 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Palmitoliec_aicd)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["Palmitoliec_aicd"]])
summary(cim.qtl.all[["Palmitoliec_aicd"]]) # 47

qtlm.Palmitoliec <- makeqtl(LG.f2.after.crossover,chr=c("C09"),pos=c(47)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.Palmitoliec # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.Palmitoliec <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Palmitoliec,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.Palmitoliec)    

effectplot(cross=LG.f2.after.crossover, mname1 = "C09@47")   
``` 

### Palmitic_acid & Myristic acid 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Myristic_Palmitic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Myristic_Palmitic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Palmitic <- do.call(c, scantwo_perm_Myristic_Palmitic)[,2]
summary(scantwo_perm_Palmitic)
# Palmitic_acid (128 permutations)
#     full  fv1  int  add  av1  one
# 5%  13.9 11.5 9.48 9.31 7.35 4.13
# 10% 13.0 11.3 9.07 8.75 6.85 3.86

summary(scantwo.imp[["Palmitic_acid"]],perm=scantwo_perm_Palmitic,alpha=.05,pval=T)
# no two loci 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Palmitic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["Palmitic_acid"]]) 
summary(cim.qtl.all[["Palmitic_acid"]]) # 38, 174 

# object hold QTL of interest 
qtlm.Palmitic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(38, 174))  
qtlm.Palmitic
qtl.fit.Palmitic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Palmitic,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.Palmitic) 

# refine position
qtl.refine.Palmitic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Palmitic,
                        formula= y ~ Q1 + Q2)

addint(LG.f2.after.crossover,qtl=qtl.refine.Palmitic) # significant 

# scan the genome for additional QTL 
qtl.add.Palmitic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Palmitic)
summary(qtl.add.Palmitic) # no additional loci 

qtl.fit.final.Palmitic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine.Palmitic,formula = y ~ Q1 * Q2 ,get.ests=T) 
summary(qtl.fit.final.Palmitic)  

# Myristic
scantwo_perm_Myristic <- do.call(c, scantwo_perm_Myristic_Palmitic)[,1]
summary(scantwo_perm_Myristic)
# Myristic_acid (128 permutations)
#     full  fv1  int  add  av1  one
# 5%  11.3 8.61 7.14 7.34 4.22 4.03
# 10% 10.7 8.09 6.97 6.63 3.94 3.69

summary(scantwo.imp[["Myristic_acid"]],perm=scantwo_perm_Myristic,alpha=.05,pval=T)
# no two loci  

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Myristic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["Myristic_acid"]]) 
summary(cim.qtl.all[["Myristic_acid"]])

qtlm.Myristic <- makeqtl(LG.f2.after.crossover,chr=c("C06"),pos=c(86)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.Myristic # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.Myristic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Myristic,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.Myristic)    
effectplot(cross=LG.f2.after.crossover, mname1 = "A06@86")    
``` 

### Stearic_acid & Vaccenic_acid 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Stearic_vaccenic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Stearic_vaccenic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Stearic <- do.call(c, scantwo_perm_Stearic_vaccenic)[,1]
summary(scantwo_perm_Stearic)
# Stearic_acid (112 permutations)
#     full  fv1  int  add  av1  one
# 5%  11.4 9.32 7.39 7.74 5.38 4.11
# 10% 10.8 8.75 7.05 7.21 5.13 3.86

summary(scantwo.imp[["Stearic_acid"]],perm=scantwo_perm_Stearic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Stearic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Stearic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(38, 174))  
qtlm.Stearic
qtl.fit.Stearic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Stearic,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.Stearic) 

# refine position
qtl.refine.Stearic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Stearic,
                        formula= y ~ Q1 + Q2)

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.Stearic) # no interaction 

# test for additional loci 
qtl.add.Stearic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Stearic)
summary(qtl.add.Stearic) # chrC02_25271659 C02  59.1 6.016 

qtlm2.Stearic <- addtoqtl(LG.f2.after.crossover,qtl.refine.Stearic,chr="C02", pos=58)
summary(qtlm2.Stearic)

# scan the genome for additional QTL 
qtl.refine2.Stearic <- refineqtl(LG.f2.after.crossover,qtl=qtlm2.Stearic,
                              formula = y ~ Q1 + Q2 + Q3)

summary(qtl.refine2.Stearic) # no additional loci  

qtl.fit2.Stearic <- fitqtl(LG.f2.after.crossover,qtl=qtl.refine2.Stearic,get.ests=T)
summary(qtl.fit2.Stearic) 

### Vaccenic_acid
scantwo_perm_vaccenic <- do.call(c, scantwo_perm_Stearic_vaccenic)[,2]
summary(scantwo_perm_vaccenic)
# vaccenic_acid (112 permutations)
#     full  fv1  int  add  av1  one
# 5%  29.7 28.2 25.6 25.3 24.1 4.32
# 10% 29.5 27.9 25.1 24.7 23.5 3.65

summary(scantwo.imp[["vaccenic_acid"]],perm=scantwo_perm_vaccenic,alpha=.05,pval=T)

## vaccenic has two QTL fit the model, however they are highly linked, so I will just examine the allele effect using fitqtl() 

# load map and make Oleic acid as the only phenotype 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$vaccenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["vaccenic_acid"]])
summary(cim.qtl.all[["vaccenic_acid"]]) # vaccenic acid 

# object hold QTL of interest 
qtlm.vaccenic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(40, 156))  
qtlm.vaccenic
qtl.fit.vaccenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.vaccenic,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.vaccenic) 

# refine position
qtl.refine.vaccenic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.vaccenic,
                        formula= y ~ Q1 + Q2) 

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.vaccenic) # 

# test for additional loci 
qtl.add.vaccenic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.vaccenic)
summary(qtl.add.vaccenic) 

qtl.fit.final.vaccenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine.vaccenic, get.ests=T, formula = y ~ Q1 + Q2)
summary(qtl.fit.final.vaccenic) 

``` 

### Linoleic_acid 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Linoleic_Arachidic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Linoleic_Arachidic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Linoleic <- do.call(c, scantwo_perm_Linoleic_Arachidic)[,1]
summary(scantwo_perm_Linoleic)
# Linoleic_acid (144 permutations)
#     full  fv1   int  add  av1  one
# 5%  14.6 13.1 10.20 10.4 8.91 4.02
# 10% 13.9 12.4  9.52  9.6 8.23 3.67

summary(scantwo.imp[["Linoleic_acid"]],perm=scantwo_perm_Linoleic,alpha=.05,pval=T)
# There were no pairs of loci meeting the criteria.   

# in order to get allele effect 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linoleic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["Linoleic_acid"]])
summary(cim.qtl.all[["Linoleic_acid"]])

# object hold QTL of interest 
qtlm.Linoleic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(38, 172))  
qtlm.Linoleic
qtl.fit.Linoleic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linoleic,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.Linoleic)  

#### scantwo result gave me different QTL pairs, so I will just consider the QTL get from one dimentional scan
# in order to get allele effect  
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Arachidic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["Arachidic_acid"]]) # 32
summary(cim.qtl.all[["Arachidic_acid"]])

# object hold QTL of interest 
qtlm.Arachidic <- makeqtl(LG.f2.after.crossover,chr=c("A08"),pos=c(32))  
qtlm.Arachidic
qtl.fit.Arachidic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Arachidic,formula = y ~ Q1 ,get.ests=T) 
summary(qtl.fit.Arachidic)    
```

### days_to_flower & days_to_bolt 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/days_to_flower_days_to_bolt_bolting_to_flowering/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_days_to_flower_days_to_bolt_bolting_to_flowering <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_days_to_flower <- do.call(c, scantwo_perm_days_to_flower_days_to_bolt_bolting_to_flowering)[,1]
summary(scantwo_perm_days_to_flower)
# days_to_flower (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  10.3 8.13 7.11 7.31 4.11 4.04
# 10%  9.9 7.81 6.43 6.69 3.72 3.78

summary(scantwo.imp[["days_to_flower"]],perm=scantwo_perm_days_to_flower,alpha=.05,pval=T)  

# load map and make Oleic acid as the only phenotype 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$days_to_flower)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.days_to_flower <- makeqtl(LG.f2.after.crossover,chr=c("A10", "C06"),pos=c(184, 81))  
qtlm.days_to_flower
qtl.fit.days_to_flower <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.days_to_flower,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.days_to_flower) 

# refine position
qtl.refine.days_to_flower <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.days_to_flower,
                        formula= y ~ Q1 + Q2)

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.days_to_flower) # no interaction 

# test for additional loci 
qtl.add.days_to_flower <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.days_to_flower)
summary(qtl.add.days_to_flower) # no additional loci 

# final model 
qtl.fit.final.days_to_flower <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine.days_to_flower,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.final.days_to_flower)

effectplot(LG.f2.after.crossover, mname1 = "A10@183.9", mname2 = "C06@67.3") 

### days_to_bolt
scantwo_perm_days_to_bolt <- do.call(c, scantwo_perm_days_to_flower_days_to_bolt_bolting_to_flowering)[,2]
summary(scantwo_perm_days_to_bolt)
# days_to_bolt (160 permutations)
#      full  fv1  int  add  av1  one
# 5%  10.17 7.76 6.72 7.18 4.04 3.87
# 10%  9.67 7.54 6.29 6.64 3.76 3.77

summary(scantwo.imp[["days_to_bolt"]],perm=scantwo_perm_days_to_bolt,alpha=.05,pval=T)  

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$days_to_bolt)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.days_to_bolt <- makeqtl(LG.f2.after.crossover,chr=c("A10", "C06"),pos=c(155, 83))  
qtlm.days_to_bolt
qtl.fit.days_to_bolt <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.days_to_bolt,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.days_to_bolt) 

# refine position
qtl.refine.days_to_bolt <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.days_to_bolt,
                        formula= y ~ Q1 + Q2)

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.days_to_bolt) # no interaction 

# test for additional loci 
qtl.add.days_to_bolt <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.days_to_bolt)
summary(qtl.add.days_to_bolt) # # A05 112.030 4.099   

qtlm2.days_to_bolt <- addtoqtl(LG.f2.after.crossover,qtl.refine.days_to_bolt,chr="A05", pos=113.030)
summary(qtlm2.days_to_bolt) 

# scan the genome for additional QTL 
qtl.refine2.days_to_bolt <- refineqtl(LG.f2.after.crossover,qtl=qtlm2.days_to_bolt,
                              formula = y ~ Q1 + Q2 + Q3)

addint(LG.f2.after.crossover,qtl=qtl.refine2.days_to_bolt) # no interaction 
summary(qtl.refine2.days_to_bolt) # no additional loci  

qtl.fit2.days_to_bolt <- fitqtl(LG.f2.after.crossover,qtl=qtl.refine2.days_to_bolt,get.ests=T)
summary(qtl.fit2.days_to_bolt)  

effectplot(LG.f2.after.crossover, mname1 = "A10@176", mname2 = "C06@83")  
``` 

### cis_11_Eicosenoic_acid & Linolenic_acid 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/cis_11_Eicosenoic_Linolenic/" 
files <- list.files(path=path)
setwd(path)
scantwo_perm_cis_11_Eicosenoic_Linolenic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
summary(do.call(c, scantwo_perm_cis_11_Eicosenoic_Linolenic)[,1])
# cis_11_Eicosenoic_acid (160 permutations) 
#     full  fv1  int  add  av1  one
# 5%  67.1 65.9 48.2 31.8 30.4 4.91
# 10% 52.9 51.9 42.6 28.7 27.6 4.20

summary(scantwo.imp[["cis_11_Eicosenoic_acid"]], thresholds = c(67.1, 65.9, 48.2, 31.8, 30.4)) # signal for just additive 
# summary: no QTL 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$cis_11_Eicosenoic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["cis_11_Eicosenoic_acid"]]) # 35
summary(cim.qtl.all[["cis_11_Eicosenoic_acid"]])

# object hold QTL of interest 
qtlm.cis_11_Eicosenoic <- makeqtl(LG.f2.after.crossover,chr=c("A08"),pos=c(35))  
qtlm.cis_11_Eicosenoic
qtl.fit.cis_11_Eicosenoic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.cis_11_Eicosenoic,formula = y ~ Q1 ,get.ests=T) 
summary(qtl.fit.cis_11_Eicosenoic)     

################ Linolenic acid ######################  
scantwo_perm_Linolenic <- do.call(c, scantwo_perm_cis_11_Eicosenoic_Linolenic)[,2]
summary(scantwo_perm_Linolenic)
# Linolenic_acid (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  13.3 10.9 9.00 8.36 5.96 3.96
# 10% 12.5 10.4 8.43 7.53 5.57 3.84

summary(scantwo.imp[["Linolenic_acid"]],thresholds = c(13.3, 10.9, 9.00, 8.36, 5.96))

# load map and make Oleic acid as the only phenotype 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linolenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Linolenic <- makeqtl(LG.f2.after.crossover,chr=c("A01", "C03", "A08", "C03"),pos=c(99, 171, 38, 174))  
qtlm.Linolenic
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 + Q2 + Q3 + Q4 + Q3:Q4 ,get.ests=T) 
summary(qtl.fit.Linolenic) 

# drop one marker 
qtlm.Linolenic <- dropfromqtl(qtlm.Linolenic,qtl.name="Q1")
qtlm.Linolenic
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 + Q2 + Q3 + Q2:Q3) 
qtlm.Linolenic <- dropfromqtl(qtlm.Linolenic,qtl.name="Q1")
qtlm.Linolenic
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 + Q2 + Q1:Q2) 
summary(qtl.fit.Linolenic)

# refine position 
qtl.refine.Linolenic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Linolenic,
                        formula= y ~ Q1 + Q2 + Q1:Q2)

# scan the genome for additional QTL 
qtl.add.Linolenic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Linolenic)
summary(qtl.add.Linolenic) # no additional loci

qtl.fit.final.Linolenic <- fitqtl(LG.f2.after.crossover,qtl=qtl.refine.Linolenic, formula = y ~Q1 + Q2 + Q1:Q2, get.ests = T)
summary(qtl.fit.final.Linolenic) 
``` 

### Behenic 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Behenic_Weight_of_survey/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Behenic_Weight_of_survey <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Behenic <- do.call(c, scantwo_perm_Behenic_Weight_of_survey)[,1]
summary(scantwo_perm_Behenic)
# Behenic_acid (160 permutations)
#     full fv1 int add av1  one
# 5%   132 131 127 127 126 6.08
# 10%  132 131 127 127 126 5.17

summary(scantwo.imp[["Behenic_acid"]],perm=scantwo_perm_Behenic,alpha=.05,pval=T)
### no two loci satisify the criteria  
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Behenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["Behenic_acid"]]) # A01: 60; A08: 38; C03: 158 
summary(cim.qtl.all[["Behenic_acid"]]) 

# object hold QTL of interest 
qtlm.Behenic <- makeqtl(LG.f2.after.crossover,chr=c("A01", "A08", "C03"),pos=c(60, 33, 175))  
qtlm.Behenic
qtl.fit.Behenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Behenic,formula = y ~ Q1 + Q2 + Q3,get.ests=T) 
summary(qtl.fit.Behenic)   
``` 

### just estimate allele effect for growth model result 
```{r}
# height_Hmax
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/output/pheno/growth_model_trait.2.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$height_Hmax)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["height_Hmax"]]) # A10 179 
summary(cim.qtl.all[["height_Hmax"]])

qtlm.height_Hmax <- makeqtl(LG.f2.after.crossover,chr=c("A10"),pos=c(179)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.height_Hmax # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.heigth_Hmax <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.height_Hmax,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.heigth_Hmax)   
effectplot(cross=LG.f2.after.crossover, mname1 = "A10@179")   

## leaf_number_I
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/output/pheno/growth_model_trait.2.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$leaf_number_I)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["leaf_number_I"]]) # A10 176 
summary(cim.qtl.all[["leaf_number_I"]])

qtlm.leaf_number_I <- makeqtl(LG.f2.after.crossover,chr=c("A10"),pos=c(176)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.leaf_number_I # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.leaf_number_I <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.leaf_number_I,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.leaf_number_I)   
effectplot(cross=LG.f2.after.crossover, mname1 = "A10@176")

## lobe_number_I 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/output/pheno/growth_model_trait.2.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$lobe_number_I)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

summary(scanone.imp.all[["lobe_number_I"]]) # C06 83.5 
summary(cim.qtl.all[["lobe_number_I"]])

qtlm.lobe_number_I <- makeqtl(LG.f2.after.crossover,chr=c("C06"),pos=c(83.5)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.lobe_number_I # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno().  

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.lobe_number_I <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.lobe_number_I,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.lobe_number_I)    
effectplot(cross=LG.f2.after.crossover, mname1 = "A10@176")   
```

### effect plot for flower traits 
```{r}
png("~/F2/output/QTL_analysis/figure/QTL_effect_flower.png", width=10, height=5, units="in", res=300)
par(mfrow=c(1,2)) 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$days_to_flower)
effectplot(LG.f2.after.crossover, mname1 = "A10@184", mname2 = "C06@82", ylab = "days_to_flower", main = "days_to_flower")  

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$days_to_bolt)
effectplot(LG.f2.after.crossover, mname1 = "A10@176", mname2 = "C06@86", ylab = "days_to_bolt", main = "days_to_bolt") 

dev.off() 
```

### effect plot 
```{r}
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

png("~/F2/output/QTL_analysis/figure/QTL_effect.png", width=10, height=15, units="in", res=300)
par(mfrow=c(5,2)) 

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Palmitic_acid)
effectplot(LG.f2.after.crossover, mname1 = "A08@38.6", mname2 = "C03@173.9", ylab = "Palmitic_acid", main = "Palmitic_acid") 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Oleic_acid)
effectplot(LG.f2.after.crossover, mname1 = "A08@37.6", mname2 = "C03@174.0", ylab = "Oleic_acid", main = "Oleic_acid") 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linoleic_acid)
effectplot(LG.f2.after.crossover, mname1 = "A08@38.6", mname2 = "C03@158.0", ylab = "Linoleic_acid", main = "Linoleic_acid") 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linolenic_acid) 
effectplot(LG.f2.after.crossover, mname1 = "A08@37.6", mname2 = "C03@174.0", ylab = "Linolenic_acid", main = "Linolenic_acid") 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$cis_11_Eicosenoic_acid)
effectplot(LG.f2.after.crossover, mname1 = "A08@35.3", ylab = "cis_11_Eicosenoic_acid", main = "cis_11_Eicosenoic_acid") 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Stearic_acid)
effectplot(LG.f2.after.crossover, mname1 = "A08@38.6", mname2 = "C03@172.0", ylab = "Stearic_acid", main = "Stearic_acid") 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$vaccenic_acid)
effectplot(LG.f2.after.crossover, mname1 = "A08@40.0", mname2 = "C03@156.0", ylab = "vaccenic_acid", main = "vaccenic_acid") 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Arachidic_acid)
effectplot(LG.f2.after.crossover, mname1 = "A08@31.9", ylab = "Arachidic_acid", main = "Arachidic_acid") 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Erucic_acid)
effectplot(LG.f2.after.crossover, mname1 = "A08@37.6", mname2 = "C03@174.0", ylab = "Erucic_acid", main = "Erucic_acid") 

LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Behenic_acid)
effectplot(LG.f2.after.crossover, mname1 = "A08@38.0", mname2 = "C03@158.0", ylab = "Behenic_acid", main = "Behenic_acid") 

dev.off() 
``` 

# root weight & plant width  
```{r}
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$root_weight_2016_05_13)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.root_weight <- makeqtl(LG.f2.after.crossover,chr=c("A10"),pos=c(171)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.root_weight # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.root_weight <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.root_weight,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.root_weight)   

## plant width 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$plant_width_2016_02_17)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.plant_width <- makeqtl(LG.f2.after.crossover,chr=c("C02"),pos=c(76)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.plant_width # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.plant_width <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.plant_width,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.plant_width)    
``` 

### other single day growth QTL result 
```{r}
### leaf_number_2015_12_28 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$leaf_number_2015_12_28)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.leaf_number_2015_12_28 <- makeqtl(LG.f2.after.crossover,chr=c("A10"),pos=c(154)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.leaf_number_2015_12_28 # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.leaf_number_2015_12_28 <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.leaf_number_2015_12_28,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.leaf_number_2015_12_28)   

## leaf_number_2016_03_21 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$leaf_number_2016_03_21)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.leaf_number_2016_03_21 <- makeqtl(LG.f2.after.crossover,chr=c("A10"),pos=c(179)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.leaf_number_2016_03_21 # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.leaf_number_2016_03_21 <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.leaf_number_2016_03_21,formula = y ~ Q1, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.leaf_number_2016_03_21)    
```  

### Erucic & Oleic 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Erucic_Oleic"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Erucic_Oleic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
scantwo_perm_Erucic_Oleic.t <- transpose(scantwo_perm_Erucic_Oleic) 
scantwo_perm_Erucic <- do.call(c, scantwo_perm_Erucic_Oleic.t[[2]])
summary(scantwo_perm_Erucic)

summary(scantwo.imp[["Erucic_acid"]], thresholds = c(10.28, 8.08, 6.99, 7.30, 4.33)) # signal for just additive 

# review previous result 
### load scanone/cim/scantwo result
summary(scanone.imp.all[["Erucic_acid"]]) # 37.6 & 174
summary(cim.qtl.all[["Erucic_acid"]]) # 38.9 & 173
summary(scantwo.imp[["Erucic_acid"]],perm=scantwo_perm_Erucic,alpha=.05,pval=T) # 38 & 179 

# load LG 
#create an object to hold QTL of interest
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Erucic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.Erucic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(38, 179))  
### position from additive model if no interaction identified 

qtlm.Erucic # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.Erucic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Erucic,formula = y ~ Q1 + Q2, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.Erucic)  

#we can now ask if all qtls are at their most likely positions
#this also provides a nice map
qtl.refine.Erucic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Erucic,
                        formula= y ~ Q1 + Q2)
#you can see from the LOD scores that the new model is not much better than the old model.  Nevertheless we can keep this as our new model

#we can test for interactions specifically among these loci
addint(LG.f2.after.crossover,qtl=qtl.refine.Erucic)
qtl.add.Erucic <- addint(LG.f2.after.crossover,qtl=qtl.refine.Erucic) # evidence for interactions
summary(qtl.add.Erucic)

# scan the genome for additional QTL to add to this model 
qtl.add.Erucic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Erucic)
summary(qtl.add.Erucic) # no additional QTL 

# refine position
qtl.refine2.Erucic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Erucic,
                        formula= y ~ Q1 * Q2) 

qtl.refine2.Erucic
# final
qtl.fit.final.Erucic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine2.Erucic, get.ests=T, formula = y ~ Q1 * Q2)
summary(qtl.fit.final.Erucic)    

### more check 
# while other poeple all find FAE to be the gene responsible for erucic acid content, I excluded thoese two genes, and found them to be controlled by a trans-eQTL hotspot on A03. Now I want to see 
# 1) whether adding genotype info of A03 loci info can increase the R2 of erucic acid 
# 2) whether adding expression of the two FAE1 genes can increase the R2 of erucic acid

# based on eQTL analysis result, that loci is A03@61  
qtl.fit.final.Erucic.FAE <- addtoqtl(LG.f2.after.crossover, qtl.refine2.Erucic, chr = "A03", pos = 61)
summary(qtl.fit.final.Erucic.FAE)

qtl.refine2.Erucic.FAE <- refineqtl(LG.f2.after.crossover,qtl=qtl.fit.final.Erucic.FAE,
                              formula = y ~ Q1 + Q2 + Q3 + Q1:Q3 + Q1:Q2 + Q2:Q3)


summary(qtl.refine2.Erucic.FAE)

qtl.fit.final.Erucic.FAE <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine2.Erucic.FAE, get.ests=T, formula = y ~ Q1 + Q2 + Q3 + Q1:Q2 + Q2:Q3 + Q3:Q1) 
summary(qtl.fit.final.Erucic.FAE)   # somewhat suggested FAE is also important? 

# what if I don't refine the position 
qtl.fit.final.Erucic.FAE <- addtoqtl(LG.f2.after.crossover, qtl.refine2.Erucic, chr = "A03", pos = 61)
summary(qtl.fit.final.Erucic.FAE)

qtl.fit.final.Erucic.FAE.2 <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.fit.final.Erucic.FAE, get.ests=T, formula = y ~ Q1 + Q2 + Q3 + Q1:Q2 + Q2:Q3 + Q3:Q1)  
summary(qtl.fit.final.Erucic.FAE.2)  


### Oleic acid 
# permutation result 
scantwo_perm_Oleic <- do.call(c, scantwo_perm_Erucic_Oleic.t[[1]])
summary(scantwo_perm_Oleic) 

# look for siginificant loci 
summary(scantwo.imp[["Oleic_acid"]], thresholds = c(12.3, 9.45, 7.95, 7.37, 4.98)) 
summary(scantwo.imp[["Oleic_acid"]],perm=scantwo_perm_Oleic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Oleic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Oleic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(38, 179))  
qtlm.Oleic
qtl.fit.Oleic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Oleic,formula = y ~ Q1 * Q2 ,get.ests=T) 
summary(qtl.fit.Oleic) 

# refine position
qtl.refine.Oleic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Oleic,
                        formula= y ~ Q1 * Q2)

# scan the genome for additional QTL 
qtl.add.Oleic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Oleic)
summary(qtl.add.Oleic) # no additional loci 

qtl.fit.final.Oleic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine.Oleic,formula = y ~ Q1 * Q2 ,get.ests=T) 
summary(qtl.fit.final.Oleic) 
```

### more check 
while other poeple all find FAE to be the gene responsible for erucic acid content, I found them to be controlled by a trans-eQTL hotspot on A03. Now I want to see 
1) whether adding genotype info of A03 loci info can increase the R2 of erucic acid 
2) whether adding expression of the two FAE1 genes can increase the R2 of erucic acid

using brms() package for this purpose 

```{r}
library(brms)
library(qtl)
library(tidyverse)
# need the genotype info at the two major QTL and A03@61 as well as the expression data of the two FAE1
# chrC03_5779676 is the nearest marker to C03@38 and A08_10120555 is the marker for A08@179; chrA03_12781099 is the marker nearest to A03@61 

# geno 
load("~/F2/for_paper/F2_geno_data_2_Ae_Ol_new.Rdata")
ID <- c("chrC03_53442609", "chrA08_10120555", "chrA03_12781099")
geno <- as.data.frame(t(F2_geno_data_2_Ae_Ol_new[rownames(F2_geno_data_2_Ae_Ol_new) %in% ID,])) 
geno$id <- paste("Sample_F2_", rownames(geno), sep = "")

# expression 
cross.F2 <- read.cross("csvsr", genfile ="~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv", 
                         phefile = "~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv",
                         genotypes = c("AA", "AB", "BB"))   # although the sample IDs are not matched in the original phe and gen file, I still get the right result. 
ID_gene <- c("BnaA08g11130D", "BnaC03g65980D")

expression <- cross.F2$pheno[,colnames(cross.F2$pheno) %in% c(ID_gene, "id")] 
expression

# pheno 
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv", 
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

colnames(LG.f2.after.crossover$pheno)
pheno <- 
LG.f2.after.crossover$pheno %>% 
  dplyr::select(ends_with("acid"), id)

# combine 
FAE1_data <- 
geno %>% 
  left_join(expression) %>%
  left_join(pheno) 

FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("-",NA,x)))
FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("A",-1,x)))
FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("B",1,x)))
FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("H",0,x)))
FAE1_data$BnaC03g65980D <- as.numeric(FAE1_data$BnaC03g65980D)
FAE1_data$BnaA08g11130D <- as.numeric(FAE1_data$BnaA08g11130D)
FAE1_data$chrA03_12781099 <- as.numeric(as.character(FAE1_data$chrA03_12781099))
FAE1_data$chrA08_10120555 <- as.numeric(as.character(FAE1_data$chrA08_10120555))
FAE1_data$chrC03_53442609 <- as.numeric(as.character(FAE1_data$chrC03_53442609))

FAE1_data$Erucic_acid <- as.numeric(FAE1_data$Erucic_acid)
summary(FAE1_data)
id <- FAE1_data$id

FAE1_data <- 
sapply(colnames(FAE1_data), function(i) 
  as.numeric(as.character(FAE1_data[,i]))) 
FAE1_data <- as.data.frame(FAE1_data)
FAE1_data$id <- id

pheno <- colnames(FAE1_data)[19] %>% as.symbol()
pheno

colnames(FAE1_data) 
# I wanted to test A03's significance for every FA, however seems not that easy, so determined to only look at erucic acid, because FAE1 genes are regualted by this loci, which control FA elongation 

  # just the two loci, with interaction 
  FAE1 <- brm(Erucic_acid ~ chrC03_53442609 * chrA08_10120555,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))  # mean of 0 and sd of 50

  # add genotype from A03, pairwise interaction 
  FAE1.2 <- brm(Erucic_acid ~ chrA03_12781099 + chrA08_10120555 + chrC03_53442609 + chrA03_12781099:chrA08_10120555 + chrA03_12781099:chrC03_53442609 + chrC03_53442609:chrA08_10120555,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))

  # no interaction at all, add genotype from A03 
  FAE1.2.2 <- brm(Erucic_acid ~ chrA03_12781099 + chrA08_10120555 + chrC03_53442609,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))

  # just add interaction between A08 and C03
  FAE1.2.3 <- brm(Erucic_acid ~ chrA03_12781099 + chrA08_10120555 + chrC03_53442609 + chrA08_10120555:chrC03_53442609,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))
  
  ### no interaction, just the two major QTL 
  FAE0 <- brm(Erucic_acid ~ chrC03_53442609 + chrA08_10120555,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))  # mean of 0 and sd of 10

  
# Erucic acid A03 is signficant with positive effect from Da-Ol-1 
# Oleic acid A03 is not significant 
WAIC(FAE0,FAE1, FAE1.2, FAE1.2.2, FAE1.2.3)
loo(FAE0,FAE1, FAE1.2, FAE1.2.2, FAE1.2.3) 

loo(FAE0, FAE1) 
loo(FAE1.2, FAE1.2.2, FAE1.2.3) 

# [1] "chrA03_12781099" "chrA08_10120555" "chrC03_5779676"  "id"              "BnaC03g65980D"   "BnaA08g11130D"  [7] "Erucic_acid"   

### hypothesis testing 
# A08 & C03, no interaction 
summary(FAE0) 
hypothesis(FAE0, "chrA08_10120555 < 0")
hypothesis(FAE0, "chrC03_53442609 < 0")  

# only A08 & C03
summary(FAE1)
hypothesis(FAE1, "chrA08_10120555 < 0")
hypothesis(FAE1, "chrC03_53442609 < 0")  

loo(FAE1,FAE0) # 

# add A03, pairwise interaction 
summary(FAE1.2)
hypothesis(FAE1.2, "chrA08_10120555 < 0") 
hypothesis(FAE1.2, "chrC03_53442609 < 0") 
hypothesis(FAE1.2, "chrA03_12781099 > 0")

# add A03, no interaction  
summary(FAE1.2.2) 
hypothesis(FAE1.2.2, "chrA08_10120555 < 0")  
hypothesis(FAE1.2.2, "chrC03_53442609 < 0") 
hypothesis(FAE1.2.2, "chrA03_12781099 > 0")  ### allele from Da-Ol-1 increase Erucic acid level. 
plot(hypothesis(FAE1.2.2, "chrA03_12781099 > 0"))   

# add A03, and include interaction between A08 and C03
summary(FAE1.2.3)
hypothesis(FAE1.2.3, "chrA08_10120555 < 0") 
hypothesis(FAE1.2.3, "chrC03_53442609 < 0") 
hypothesis(FAE1.2.3, "chrA03_12781099 > 0")  ### allele from Da-Ol-1 increase Erucic acid level. 
plot(hypothesis(FAE1.2.2, "chrA03_12781099 > 0")) 

# cool, so A03 genotype is significant for Erucic acid, how about other FA? I will just test oleic acid and linolenic acid 

FAE1_data <- 
geno %>% 
  left_join(expression) %>%
  left_join(pheno) 

FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("-",NA,x)))
FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("A",-1,x)))
FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("B",1,x)))
FAE1_data <- data.frame(sapply(FAE1_data, function(x) sub("H",0,x)))
FAE1_data$BnaC03g65980D <- as.numeric(FAE1_data$BnaC03g65980D)
FAE1_data$BnaA08g11130D <- as.numeric(FAE1_data$BnaA08g11130D)
FAE1_data$chrA03_12781099 <- as.numeric(as.character(FAE1_data$chrA03_12781099))
FAE1_data$chrA08_10120555 <- as.numeric(as.character(FAE1_data$chrA08_10120555))
FAE1_data$chrC03_53442609 <- as.numeric(as.character(FAE1_data$chrC03_53442609))

  FAE1_data$Stearic_acid <- as.numeric(FAE1_data$Stearic_acid)
  id <- FAE1_data$id

  FAE1_data <- 
  sapply(colnames(FAE1_data), function(i) 
    as.numeric(as.character(FAE1_data[,i]))) 
  FAE1_data <- as.data.frame(FAE1_data)
  FAE1_data$id <- id

  # just the two loci, with interaction 
  FAE1 <- brm(Stearic_acid ~ chrC03_53442609 * chrA08_10120555,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))  # mean of 0 and sd of 50

  # add genotype from A03, pairwise interaction 
  FAE1.2 <- brm(Stearic_acid ~ chrA03_12781099 + chrA08_10120555 + chrC03_53442609 + chrA03_12781099:chrA08_10120555 + chrA03_12781099:chrC03_53442609 + chrC03_53442609:chrA08_10120555,
            data = FAE1_data,
            prior = set_prior("normal(0,50)",class="b"))

  # no interaction at all, add genotype from A03 
  FAE1.2.2 <- brm(Stearic_acid ~ chrA03_12781099 + chrA08_10120555 + chrC03_53442609,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b")) 

  # just add interaction between A08 and C03
  FAE1.2.3 <- brm(Stearic_acid ~ chrA03_12781099 + chrA08_10120555 + chrC03_53442609 + chrA08_10120555:chrC03_53442609,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))
  
  ### no interaction, just the two major QTL 
  FAE0 <- brm(Stearic_acid ~ chrC03_53442609 + chrA08_10120555,
            data = FAE1_data, 
            prior = set_prior("normal(0,50)",class="b"))  # mean of 0 and sd of 10

### hypothesis testing 
# A08 & C03, no interaction 
summary(FAE0) 
hypothesis(FAE0, "chrA08_10120555 > 0")
hypothesis(FAE0, "chrC03_53442609 > 0")  

# only A08 & C03
summary(FAE1)
hypothesis(FAE1, "chrA08_10120555 > 0")
hypothesis(FAE1, "chrC03_53442609 > 0")  

loo(FAE1,FAE0) # 

# add A03, pairwise interaction 
summary(FAE1.2)
hypothesis(FAE1.2, "chrA08_10120555 > 0") 
hypothesis(FAE1.2, "chrC03_53442609 > 0") 
hypothesis(FAE1.2, "chrA03_12781099 < 0") 

# add A03, no interaction  
summary(FAE1.2.2) 
hypothesis(FAE1.2.2, "chrA08_10120555 > 0")  
hypothesis(FAE1.2.2, "chrC03_53442609 > 0") 
hypothesis(FAE1.2.2, "chrA03_12781099 < 0")  ### allele from Da-Ol-1 increase Erucic acid level. 

# add A03, and include interaction between A08 and C03
summary(FAE1.2.3)
hypothesis(FAE1.2.3, "chrA08_10120555 > 0") 
hypothesis(FAE1.2.3, "chrC03_53442609 > 0") 
hypothesis(FAE1.2.3, "chrA03_12781099 < 0")  ### allele from Da-Ol-1 increase Erucic acid level.    
  
Oleic_result <- c(FAE0, FAE1, FAE1.2, FAE1.2.2, FAE1.2.3) 
linoleic_result <- c(FAE0, FAE1, FAE1.2, FAE1.2.2, FAE1.2.3) 
###### adding the two gene's expression doesn't help, so not included here 
# add two FAE gene, the two QTL remain significant  
summary(FAE1.3)
hypothesis(FAE1.3, "chrA08_10120555B < 0")
hypothesis(FAE1.3, "chrC03_53442609B < 0")
hypothesis(FAE1.3, "BnaA08g11130D > 0")
hypothesis(FAE1.3, "BnaC03g65980D > 0") 

# add two FAE gene with pairwise interaction 
summary(FAE1.4)
hypothesis(FAE1.4, "chrA08_10120555B < 0") 
hypothesis(FAE1.4, "chrC03_53442609B < 0") 
hypothesis(FAE1.4, "BnaA08g11130D > 0")
hypothesis(FAE1.4, "BnaC03g65980D > 0")   

# test just the two genes 
FAE1.5 <- brm(Erucic_acid ~ BnaA08g11130D * BnaC03g65980D,
            data = FAE1_data, 
            prior = set_prior("normal(0,10)",class="b"))

summary(FAE1.5)

hypothesis(FAE1.4, "BnaA08g11130D > 0")
hypothesis(FAE1.4, "BnaC03g65980D > 0")   

# add expression of the two FAE1, with interaction between two QTL 
FAE1.3 <- brm(Erucic_acid ~ BnaA08g11130D + BnaC03g65980D + chrA08_10120555 + chrC03_53442609 + chrC03_53442609:chrA08_10120555,
            data = FAE1_data, 
            prior = set_prior("normal(0,10)",class="b"))

summary(FAE1.3) 
# plot(FAE1.2,ask=FALSE) 

# 
FAE1.4 <- brm(Erucic_acid ~ BnaA08g11130D + BnaC03g65980D + chrA08_10120555 + chrC03_53442609 + BnaA08g11130D:BnaC03g65980D + BnaA08g11130D:chrA08_10120555 + BnaA08g11130D:chrC03_53442609 + BnaC03g65980D:chrC03_53442609 + chrC03_53442609:chrA08_10120555 + chrA08_10120555:chrC03_53442609,
            data = FAE1_data, 
            prior = set_prior("normal(0,10)",class="b"))

summary(FAE1.4) 
```

### plot for paper
```{r}
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped_C05C08.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linolenic_acid) 

setEPS()
postscript(file = "~/F2/for_paper/Fig6.eps", width=5, height=7)
par(ps = 8, cex = 1, cex.main = 1, cex.axis = 1, tcl = -0.3) 

effectplot(LG.f2.after.crossover, mname1 = "A08@37.6", mname2 = "C03@174.0", ylab = "Linolenic_acid", main = "") 
dev.off()
```



