---
title: "scantwo_43traits"
output: html_document
---

```{r}
library(qtl)
library(tidyverse)
```

### 2D scan
scanone and cim only evaluate the evidence for a QTL one location a time.  It is possible to instead scan two locations at once. This can be particularly advantageous in identifying QTL if there is epistasis. 

### load data 
```{r}
# running scantwo analysis and permutation seprately
# 1) scantwo analysis on cabernet 
# https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scantwo_v2.R  

# scantwo permutation 100 times for 43 traits, split traits file so that each file has 2 traits, in total 20 files, and each run is only 10 permutation, combine permutation result after finish running. 
# https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scantwo_perm_1.R ... scantwo_perm_10.R, submit to slurm queue using https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scantwo_perm_1_10.slurm 

# scp result to whitney
# QTL result 
load("~/F2/output/scantwo/scantwo.imp.v2.Rdata") 
scantwo.imp.42traits <- scantwo.imp 
load("~/F2/output/scantwo/scantwo.imp.v2.bolting_to_flowering.Rdata")

scantwo.imp.43traits <- c(scantwo.imp.42traits, list(scantwo.imp))
names(scantwo.imp.43traits)[[43]] <- "bolting_to_flowering"

length(scantwo.imp.43traits) # 43 
save(scantwo.imp.43traits, file = "~/F2/output/scantwo/scantwo.imp.43traits.Rdata")

# also load scanone & CIM result 
load("~/F2/output/QTL_analysis/scanone.imp.43traits")
load("~/F2/output/QTL_analysis/cim.qtl.43traits.Rdata") 

load("~/F2/output/scantwo/LG.f2.after.crossover.final.Rdata")
names(LG.f2.after.crossover$geno)[1:10] <- paste("A", c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10"), sep = "")
names(LG.f2.after.crossover$geno)[11:19] <- paste("C", c("01", "02", "03", "04", "05", "06", "07", "08", "09"), sep = "") 
save(LG.f2.after.crossover, file = "~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
```

### oil content & Caprylic acid 
```{r}
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
# load data 
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Crude_oil_contents_Caprylic_acid/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Crude_oil_contents_Caprylic_acid <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
scantwo_perm_Crude_oil_contents <- summary(do.call(c, scantwo_perm_Crude_oil_contents_Caprylic_acid)[,1])

# oil content 
scantwo_perm_Crude_oil_contents
# Crude_oil_contents (144 permutations)
#     full  fv1  int  add  av1  one
# 5%  11.7 9.19 7.72 7.47 5.51 4.08
# 10% 11.0 8.88 7.37 7.25 5.24 3.86 

summary(scantwo.imp.42traits[["Crude_oil_contents"]], thresholds = c(11.7, 9.19, 7.72, 7.47, 5.51)) # signal for just additive   

# caprylic acid 
# permutation data check 
scantwo_perm_Caprylic_acid <- summary(do.call(c, scantwo_perm_Crude_oil_contents_Caprylic_acid)[,2])

# oil content 
scantwo_perm_Caprylic_acid
# Caprylic_acid (144 permutations)
#     full  fv1  int  add  av1  one
# 5%  10.7 8.11 6.82 6.96 3.64 4.02
# 10% 10.0 7.66 6.33 6.49 3.36 3.75

summary(scantwo.imp.42traits[["Caprylic_acid"]], thresholds = c(10.7, 8.11, 6.82, 6.96, 3.64)) # signal for just additive   
#         pos1f pos2f lod.full lod.fv1 lod.int     pos1a pos2a lod.add lod.av1
# c11:c11    23    51     7.63    5.01    0.53        23    51     7.1    4.48

# review previous result 
### load scanone/cim/scantwo result
summary(scanone.imp[["Caprylic_acid"]])
summary(cim.qtl[["Caprylic_acid"]]) 
summary(scantwo.imp.42traits[["Caprylic_acid"]], thresholds = c(10.7, 8.11, 6.82, 6.96, 3.64)) # 

# load LG 
#create an object to hold QTL of interest
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Caprylic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.Caprylic <- makeqtl(LG.f2.after.crossover,chr=c("C01", "C01"),pos=c(23, 51)) 

qtlm.Caprylic # 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.Caprylic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Caprylic,formula = y ~ Q1 + Q2, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.Caprylic)  

#we can now ask if all qtls are at their most likely positions
#this also provides a nice map
qtl.refine.Caprylic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Caprylic,
                        formula= y ~ Q1 + Q2)
#you can see from the LOD scores that the new model is not much better than the old model.  Nevertheless we can keep this as our new model

#we can test for interactions specifically among these loci
addint(LG.f2.after.crossover,qtl=qtl.refine.Caprylic) # not significant 

# scan the genome for additional QTL to add to this model 
qtl.add.Caprylic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Caprylic)
summary(qtl.add.Caprylic) # no additional QTL 

qtl.refine.Caprylic
# final
qtl.fit.final.Caprylic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine.Caprylic, get.ests=T, formula = y ~ Q1 + Q2)
summary(qtl.fit.final.Caprylic)  

# plot out 
png("~/F2/output/scantwo/figure/Caprylic_scantwo_result.png", width=20, height=10, units="in", res=300)
par(mfrow=c(1,2)) 
#positions can be plotted on the map
plot(qtl.refine.Caprylic)
#or showing lod scores
plotLodProfile(qtl.refine.Caprylic) 
dev.off()
save(qtl.fit.final.Caprylic, file = "~/F2/output/scantwo/qtl.fit.final.Caprylic.Rdata")   
```

### Erucic acid & Oleic acid 
```{r}
######## start from here tonight...
# Erucic acid 
# permutation result for Erucic acid and Oleic acid import 
# load data 
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Erucic_Oleic"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Erucic_Oleic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
scantwo_perm_Erucic_Oleic.t <- transpose(scantwo_perm_Erucic_Oleic) 
scantwo_perm_Erucic <- do.call(c, scantwo_perm_Erucic_Oleic.t[[2]])
summary(scantwo_perm_Erucic)

png("~/F2/output/scantwo/figure/Erucic_scantwo_int_full.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Erucic_acid"]], main = "Erucic_acid_int_full")
dev.off()

png("~/F2/output/scantwo/figure/Erucic_scantwo_int_fv1.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Erucic_acid"]], main = "Erucic_acid_int_fv1", lower = "fv1") 
dev.off()

png("~/F2/output/scantwo/figure/Erucic_scantwo_add_fv1.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Erucic_acid"]], main = "Erucic_acid_add_fv1", upper = "add", lower = "fv1") 
dev.off()

summary(scantwo.imp.42traits[["Erucic_acid"]], thresholds = c(10.28, 8.08, 6.99, 7.30, 4.33)) # signal for just additive 

# review previous result 
### load scanone/cim/scantwo result
summary(scanone.imp[["Erucic_acid"]])
summary(cim.qtl[["Erucic_acid"]]) 
summary(scantwo.imp.42traits[["Erucic_acid"]],perm=scantwo_perm_Erucic,alpha=.05,pval=T)

# load LG 
#create an object to hold QTL of interest
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Erucic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.Erucic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(73, 30)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.Erucic # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.Erucic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Erucic,formula = y ~ Q1 + Q2, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.Erucic)  

#we can now ask if all qtls are at their most likely positions
#this also provides a nice map
qtl.refine.Erucic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Erucic,
                        formula= y ~ Q1 + Q2)
#you can see from the LOD scores that the new model is not much better than the old model.  Nevertheless we can keep this as our new model

#we can test for interactions specifically among these loci
addint(LG.f2.after.crossover,qtl=qtl.refine.Erucic)
qtl.add.Erucic <- addint(LG.f2.after.crossover,qtl=qtl.refine.Erucic) # evidence for interactions
summary(qtl.add.Erucic)

# scan the genome for additional QTL to add to this model 
qtl.add.Erucic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Erucic)
summary(qtl.add.Erucic) # no additional QTL 

# refine position
qtl.refine2.Erucic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Erucic,
                        formula= y ~ Q1 * Q2) 

qtl.refine2.Erucic
# final
qtl.fit.final.Erucic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine2.Erucic, get.ests=T, formula = y ~ Q1 * Q2)
summary(qtl.fit.final.Erucic)  

# plot out 
png("~/F2/output/scantwo/figure/Erucic_scantwo_result.png", width=20, height=10, units="in", res=300)
par(mfrow=c(1,2)) 
#positions can be plotted on the map
plot(qtl.refine2.Erucic)
#or showing lod scores
plotLodProfile(qtl.refine.Erucic) 
dev.off()
save(qtl.fit.final.Erucic, file = "~/F2/output/scantwo/qtl.fit.final.Erucic.Rdata") 

# summary: two loci A08@72.9 and C03@30.0 with interaction 

################ Oleic acid ######################  
# permutation result 
scantwo_perm_Oleic <- do.call(c, scantwo_perm_Erucic_Oleic.t[[1]])
summary(scantwo_perm_Oleic)

# plot 
png("~/F2/output/scantwo/figure/Oleic_scantwo_int_full.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Oleic_acid"]], main = "Oleic_acid_int_full")
dev.off()

png("~/F2/output/scantwo/figure/Oleic_scantwo_int_fv1.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Oleic_acid"]], main = "Oleic_acid_int_fv1", lower = "fv1") 
dev.off()

png("~/F2/output/scantwo/figure/Oleic_scantwo_add_fv1.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Oleic_acid"]], main = "Oleic_acid_add_fv1", upper = "add", lower = "fv1") 
dev.off() 

# look for siginificant loci 
summary(scantwo.imp.42traits[["Oleic_acid"]], thresholds = c(12.3, 9.45, 7.95, 7.37, 4.98)) 
summary(scantwo.imp.42traits[["Oleic_acid"]],perm=scantwo_perm_Oleic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Oleic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Oleic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(73, 31))  
qtlm.Oleic
qtl.fit.Oleic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Oleic,formula = y ~ Q1 * Q2 ,get.ests=T) 
summary(qtl.fit.Oleic) 

# refine position
qtl.refine.Oleic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Oleic,
                        formula= y ~ Q1 * Q2)

# scan the genome for additional QTL 
qtl.add.Oleic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Oleic)
summary(qtl.add.Oleic) # no additional loci 

qtl.fit.final.Oleic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine.Oleic,formula = y ~ Q1 * Q2 ,get.ests=T) 
summary(qtl.fit.final.Oleic)

# plot result 
png("~/F2/output/scantwo/figure/Oleic_scantwo_result.png", width=20, height=10, units="in", res=300)
par(mfrow=c(1,2)) 
plot(qtl.refine.Oleic)
plotLodProfile(qtl.refine.Oleic) 
dev.off() 
save(qtl.fit.final.Oleic, file = "~/F2/output/scantwo/qtl.fit.final.Oleic.Rdata")
# summary: A08@72.3 and C03@31.0 with interaction  
```  

### Palmitoliec_aicd & Heptadecanoic_acid  
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Palmitoliec_Heptadecanoic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Palmitoliec_Heptadecanoic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
summary(do.call(c, scantwo_perm_Palmitoliec_Heptadecanoic)[,1])
# Palmitoliec_aicd (160 permutations)
#      full  fv1  int  add  av1  one
# 5%  100.0 98.8 93.2 91.7 90.6 5.11
# 10%  99.5 98.2 92.0 91.4 90.3 4.64 

summary(scantwo.imp.42traits[["Palmitoliec_acid"]], thresholds = c(100.0, 98.8, 93.2, 91.7, 90.6)) # signal for just additive 

# summary: no QTL 

################ Heptadecanoic acid ######################  
summary(do.call(c, scantwo_perm_Palmitoliec_Heptadecanoic)[,2])
# Heptadecanoic_acid (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  38.3 36.7 29.7 23.9 22.2 5.21
# 10% 36.2 34.1 27.4 21.4 20.1 4.53

summary(scantwo.imp.42traits[["Heptadecanoic_acid"]], thresholds = c(38.3, 36.7, 29.7, 23.9, 22.2)) # signal for just additive 
```

### cis_11_Eicosenoic_acid & Linolenic_acid 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/cis_11_Eicosenoic_Linolenic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_cis_11_Eicosenoic_Linolenic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
summary(do.call(c, scantwo_perm_cis_11_Eicosenoic_Linolenic)[,1])
# cis_11_Eicosenoic_acid (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  67.1 65.9 48.2 31.8 30.4 4.91
# 10% 52.9 51.9 42.6 28.7 27.6 4.20

summary(scantwo.imp.42traits[["cis_11_Eicosenoic_acid"]], thresholds = c(67.1, 65.9, 48.2, 31.8, 30.4)) # signal for just additive 

# summary: no QTL 
### problematic dataset? 

################ Linolenic acid ######################  
scantwo_perm_Linolenic <- do.call(c, scantwo_perm_cis_11_Eicosenoic_Linolenic)[,2]
summary(scantwo_perm_Linolenic)
# Linolenic_acid (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  13.3 10.9 9.00 8.36 5.96 3.96
# 10% 12.5 10.4 8.43 7.53 5.57 3.84

summary(scantwo.imp.42traits[["Linolenic_acid"]],perm=scantwo_perm_Linolenic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linolenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Linolenic <- makeqtl(LG.f2.after.crossover,chr=c("A01", "C03", "A08", "C03"),pos=c(86, 39, 73, 36))  
qtlm.Linolenic
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 + Q2 + Q3 + Q4 + Q3:Q4 ,get.ests=T) 
summary(qtl.fit.Linolenic) 

# drop one marker 
qtlm.Linolenic <- dropfromqtl(qtlm.Linolenic,qtl.name="Q2")
qtlm.Linolenic
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 + Q2 + Q3 + Q2:Q3) 
summary(qtl.fit.Linolenic)

# drop again 
qtlm.Linolenic <- dropfromqtl(qtlm.Linolenic,qtl.name="Q1")
qtlm.Linolenic
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 * Q2, get.ests = T) 
summary(qtl.fit.Linolenic) 

# refine position 
qtl.refine.Linolenic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Linolenic,
                        formula= y ~ Q1 * Q2)

# scan the genome for additional QTL 
qtl.add.Linolenic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Linolenic)
summary(qtl.add.Linolenic) # no additional loci

qtl.fit.final.Linolenic <- fitqtl(LG.f2.after.crossover,qtl=qtl.refine.Linolenic, formula = y ~Q1 * Q2, get.ests = T)
summary(qtl.fit.final.Linolenic)

# plot result 
png("~/F2/output/scantwo/figure/Linolenic_scantwo_result.png", width=20, height=8, units="in", res=300)
par(mfrow=c(1,2)) 
plot(qtl.refine.Linolenic)
plotLodProfile(qtl.refine.Linolenic) 
dev.off() 
save(qtl.fit.final.Linolenic, file = "~/F2/output/scantwo/qtl.fit.final.Linolenic.Rdata")
# summary: A08@72.3 and C03@31.0 with interaction 
```

#### only work on traits with significant loci 
### Palmnitic_acid & Myristic acid 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Myristic_Palmitic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Myristic_Palmitic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Palmitic <- do.call(c, scantwo_perm_Myristic_Palmitic)[,2]
summary(scantwo_perm_Palmitic)
# Palmitic_acid (128 permutations)
#     full  fv1  int  add  av1  one
# 5%  13.9 11.5 9.48 9.31 7.35 4.13
# 10% 13.0 11.3 9.07 8.75 6.85 3.86

summary(scantwo.imp.42traits[["Palmitic_acid"]],perm=scantwo_perm_Palmitic,alpha=.05,pval=T)
# no two loci 

# Myristic
scantwo_perm_Myristic <- do.call(c, scantwo_perm_Myristic_Palmitic)[,1]
summary(scantwo_perm_Myristic)
# Myristic_acid (128 permutations)
#     full  fv1  int  add  av1  one
# 5%  11.3 8.61 7.14 7.34 4.22 4.03
# 10% 10.7 8.09 6.97 6.63 3.94 3.69

summary(scantwo.imp.42traits[["Myristic_acid"]],perm=scantwo_perm_Myristic,alpha=.05,pval=T)
# no two loci  
```

### Stearic_acid & Vaccenic_acid 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Stearic_vaccenic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Stearic_vaccenic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Stearic <- do.call(c, scantwo_perm_Stearic_vaccenic)[,1]
summary(scantwo_perm_Stearic)
# Stearic_acid (112 permutations)
#     full  fv1  int  add  av1  one
# 5%  11.4 9.32 7.39 7.74 5.38 4.11
# 10% 10.8 8.75 7.05 7.21 5.13 3.86

summary(scantwo.imp.42traits[["Stearic_acid"]],perm=scantwo_perm_Stearic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Stearic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Stearic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(71, 29))  
qtlm.Stearic
qtl.fit.Stearic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Stearic,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.Stearic) 

# refine position
qtl.refine.Stearic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Stearic,
                        formula= y ~ Q1 + Q2)

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.Stearic) # no interaction 

# test for additional loci 
qtl.add.Stearic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Stearic)
summary(qtl.add.Stearic) # chrC02_25271659 C02  59.1 6.016 

qtlm2.Stearic <- addtoqtl(LG.f2.after.crossover,qtl.refine.Stearic,chr="C02", pos=59.1)
summary(qtlm2.Stearic)

# scan the genome for additional QTL 
qtl.refine2.Stearic <- refineqtl(LG.f2.after.crossover,qtl=qtlm2.Stearic,
                              formula = y ~ Q1 + Q2 + Q3)

summary(qtl.refine2.Stearic) # no additional loci  

qtl.fit2.Stearic <- fitqtl(LG.f2.after.crossover,qtl=qtl.refine2.Stearic,get.ests=T)
summary(qtl.fit2.Stearic) 

# plot result 
png("~/F2/output/scantwo/figure/Stearic_scantwo_result.png", width=20, height=10, units="in", res=300)
par(mfrow=c(1,2)) 
plot(qtl.refine2.Stearic)
plotLodProfile(qtl.refine2.Stearic) 
dev.off() 
save(qtl.fit2.Stearic, file = "~/F2/output/scantwo/qtl.fit2.Stearic.Rdata")
# summary: A08@72.3 and C03@31.0 with interaction  

### Vaccenic_acid
scantwo_perm_vaccenic <- do.call(c, scantwo_perm_Stearic_vaccenic)[,2]
summary(scantwo_perm_vaccenic)
# vaccenic_acid (112 permutations)
#     full  fv1  int  add  av1  one
# 5%  29.7 28.2 25.6 25.3 24.1 4.32
# 10% 29.5 27.9 25.1 24.7 23.5 3.65

summary(scantwo.imp.42traits[["vaccenic_acid"]],perm=scantwo_perm_vaccenic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$vaccenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.vaccenic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "A08"),pos=c(70, 73))  
qtlm.vaccenic
qtl.fit.vaccenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.vaccenic,formula = y ~ Q1 * Q2 ,get.ests=T) # wierd, cannot add interaction 
# Error in solve.default(t(Z) %*% Z, t(Z) %*% X) : 
#   system is computationally singular: reciprocal condition number = 4.20223e-18
summary(qtl.fit.vaccenic) 

# refine position
qtl.refine.vaccenic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.vaccenic,
                        formula= y ~ Q1 + Q2) 

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.vaccenic) # 

# test for additional loci 
qtl.add.vaccenic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.vaccenic)
summary(qtl.add.vaccenic) 

# I guess this is the same loci ... leave it as it is now... same two linked loci 
```

### Linoleic_acid & 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Linoleic_Arachidic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Linoleic_Arachidic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Linoleic <- do.call(c, scantwo_perm_Linoleic_Arachidic)[,1]
summary(scantwo_perm_Linoleic)
# Linoleic_acid (144 permutations)
#     full  fv1   int  add  av1  one
# 5%  14.6 13.1 10.20 10.4 8.91 4.02
# 10% 13.9 12.4  9.52  9.6 8.23 3.67

summary(scantwo.imp.42traits[["Linoleic_acid"]],perm=scantwo_perm_Linoleic,alpha=.05,pval=T)
# There were no pairs of loci meeting the criteria.   

### Arachidic_acid
scantwo_perm_Arachidic <- do.call(c, scantwo_perm_Linoleic_Arachidic)[,2]
summary(scantwo_perm_Arachidic)
# Arachidic_acid (144 permutations)
#     full   fv1  int  add  av1  one
# 5%  11.8 10.03 8.10 8.05 6.58 3.91
# 10% 11.5  9.75 7.77 7.75 6.33 3.62

summary(scantwo.imp.42traits[["Arachidic_acid"]],perm=scantwo_perm_Arachidic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Arachidic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Arachidic <- makeqtl(LG.f2.after.crossover,chr=c("A03", "A03"),pos=c(116, 118))  
qtlm.Arachidic
qtl.fit.Arachidic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Arachidic,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.Arachidic) 

# refine position
qtl.refine.Arachidic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Arachidic,
                        formula= y ~ Q1 + Q2)

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.Arachidic) # no interaction 

# test for additional loci 
qtl.add.Arachidic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Arachidic)
summary(qtl.add.Arachidic) # chrA08_2980704  A08  94.4 7.327 

qtlm2.Arachidic <- addtoqtl(LG.f2.after.crossover,qtl.refine.Arachidic,chr="A08", pos=94.4)
summary(qtlm2.Arachidic) 

# scan the genome for additional QTL 
qtl.refine2.Arachidic <- refineqtl(LG.f2.after.crossover,qtl=qtlm2.Arachidic,
                              formula = y ~ Q1 + Q2 + Q3)

addint(LG.f2.after.crossover,qtl=qtl.refine2.Arachidic) # no interaction 
summary(qtl.refine2.Arachidic) # no additional loci  

qtl.fit2.Arachidic <- fitqtl(LG.f2.after.crossover,qtl=qtl.refine2.Arachidic,get.ests=T)
summary(qtl.fit2.Arachidic) 

# plot result 
png("~/F2/output/scantwo/figure/Arachidic_scantwo_result.png", width=20, height=10, units="in", res=300)
par(mfrow=c(1,2)) 
plot(qtl.refine2.Arachidic)
plotLodProfile(qtl.refine2.Arachidic) 
dev.off() 
save(qtl.fit2.Arachidic, file = "~/F2/output/scantwo/qtl.fit2.Arachidic.Rdata")
```

### Behenic_acid (wait for new result because this phenotype dataset has problem)
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Behenic_Weight_of_survey/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Behenic_Weight_of_survey <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Behenic <- do.call(c, scantwo_perm_Behenic_Weight_of_survey)[,1]
summary(scantwo_perm_Behenic)
# Behenic_acid (160 permutations)
#     full fv1 int add av1  one
# 5%   132 131 127 127 126 6.08
# 10%  132 131 127 127 126 5.17

summary(scantwo.imp.42traits[["Behenic_acid"]],perm=scantwo_perm_Behenic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Behenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Behenic <- makeqtl(LG.f2.after.crossover,chr=c("A03", "A03"),pos=c(59, 60))  
qtlm.Behenic
qtl.fit.Behenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Behenic,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.Behenic) 
# two nearby loci with contrasting effect 

# refine position
qtl.refine.Behenic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Behenic,
                        formula= y ~ Q1 + Q2)

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.Behenic) # no interaction 

# test for additional loci 
qtl.add.Behenic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Behenic)
summary(qtl.add.Behenic) # chrA08_2980704  A08  94.4 7.327 

###### problematic dataset???  
```

### Weight_of_one_thousand_seeds  
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Weight_of_one_thousand_seeds_Number_of_survey_seeds/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Weight_of_one_thousand_seeds_Number_of_survey_seeds <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Weight_of_one_thousand_seeds <- do.call(c, scantwo_perm_Weight_of_one_thousand_seeds_Number_of_survey_seeds)[,1]
summary(scantwo_perm_Weight_of_one_thousand_seeds)
# Weight_of_one_thousand_seeds (128 permutations)
#      full  fv1  int  add  av1  one
# 5%  10.05 8.15 7.07 7.02 4.49 3.75
# 10%  9.87 7.77 6.56 6.65 4.14 3.61

summary(scantwo.imp.42traits[["Weight_of_one_thousand_seeds"]],perm=scantwo_perm_Weight_of_one_thousand_seeds,alpha=.05,pval=T)  
```

### leaf_number_2016_03_21 & plant_height_2016_05_13 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/leaf_number_2016_03_21_plant_height_2016_05_13/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_leaf_number_2016_03_21_plant_height_2016_05_13 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_leaf_number_2016_03_21 <- do.call(c, scantwo_perm_leaf_number_2016_03_21_plant_height_2016_05_13)[,1]
summary(scantwo_perm_leaf_number_2016_03_21)
# leaf_number_2016_03_21 (128 permutations)
#     full  fv1  int  add  av1  one
# 5%  10.5 8.55 7.43 7.67 5.05 4.15
# 10% 10.2 8.26 6.89 7.26 4.58 3.85

summary(scantwo.imp.42traits[["leaf_number_2016_03_21"]],perm=scantwo_perm_leaf_number_2016_03_21,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria. 

### plant_height_2016_05_13
scantwo_perm_plant_height_2016_05_13 <- do.call(c, scantwo_perm_leaf_number_2016_03_21_plant_height_2016_05_13)[,2]
summary(scantwo_perm_plant_height_2016_05_13)
# plant_height_2016_05_13 (128 permutations)
#     full  fv1  int  add  av1  one
# 5%  11.0 8.47 7.07 7.52 4.59 3.95
# 10% 10.3 8.06 6.66 6.96 4.35 3.70

summary(scantwo.imp.42traits[["plant_height_2016_05_13"]],perm=scantwo_perm_plant_height_2016_05_13,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria. 
```

### plant_width_2016_02_17 & plant_width_2016_01_20 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/plant_width_2016_01_20_plant_width_2016_02_17/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_plant_width_2016_01_20_plant_width_2016_02_17 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_plant_width_2016_02_17 <- do.call(c, scantwo_perm_plant_width_2016_01_20_plant_width_2016_02_17)[,2]
summary(scantwo_perm_plant_width_2016_02_17)
# plant_width_2016_02_17 (128 permutations)
#     full  fv1  int  add  av1  one
# 5%  11.6 9.16 7.48 8.09 5.74 4.43
# 10% 11.0 8.91 7.11 7.27 5.49 3.86

summary(scantwo.imp.42traits[["plant_width_2016_02_17"]],perm=scantwo_perm_plant_width_2016_02_17,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  

scantwo_perm_plant_width_2016_01_20 <- do.call(c, scantwo_perm_plant_width_2016_01_20_plant_width_2016_02_17)[,1]
summary(scantwo_perm_plant_width_2016_01_20)
# plant_width_2016_01_20 (128 permutations)
#      full  fv1  int  add  av1  one
# 5%  10.14 8.13 6.99 7.32 4.60 3.76
# 10%  9.89 7.89 6.56 6.96 4.41 3.62

summary(scantwo.imp.42traits[["plant_width_2016_01_20"]],perm=scantwo_perm_plant_width_2016_01_20,alpha=.05,pval=T)   
# There were no pairs of loci meeting the criteria. 
```

### root_weight_2016_05_13 & plant_height_2016_03_16 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/plant_height_2016_03_16_root_weight_2016_05_13/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_plant_height_2016_03_16_root_weight_2016_05_13 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_root_weight_2016_05_13 <- do.call(c, scantwo_perm_plant_height_2016_03_16_root_weight_2016_05_13)[,2]
summary(scantwo_perm_root_weight_2016_05_13)
# root_weight_2016_05_13 (96 permutations)
#     full  fv1  int  add  av1  one
# 5%  13.1 11.5 9.49 8.83 6.69 3.81
# 10% 12.8 11.1 8.97 7.81 6.30 3.66

summary(scantwo.imp.42traits[["root_weight_2016_05_13"]],perm=scantwo_perm_root_weight_2016_05_13,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  

# plant_height_2016_03_16 
scantwo_perm_plant_height_2016_03_16 <- do.call(c, scantwo_perm_plant_height_2016_03_16_root_weight_2016_05_13)[,1]
summary(scantwo_perm_plant_height_2016_03_16)
# plant_height_2016_03_16 (96 permutations)
#     full  fv1  int  add  av1  one
# 5%  11.5 9.34 7.86 7.53 5.92 4.10
# 10% 11.3 8.90 7.45 7.17 5.52 3.84

summary(scantwo.imp.42traits[["plant_height_2016_03_16"]],perm=scantwo_perm_plant_height_2016_03_16,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  
```

### days_to_flower & days_to_bolt & bolting_to_flower 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/days_to_flower_days_to_bolt_bolting_to_flowering/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_days_to_flower_days_to_bolt_bolting_to_flowering <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_days_to_flower <- do.call(c, scantwo_perm_days_to_flower_days_to_bolt_bolting_to_flowering)[,1]
summary(scantwo_perm_days_to_flower)
# days_to_flower (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  10.3 8.13 7.11 7.31 4.11 4.04
# 10%  9.9 7.81 6.43 6.69 3.72 3.78

summary(scantwo.imp.42traits[["days_to_flower"]],perm=scantwo_perm_days_to_flower,alpha=.05,pval=T)  

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$days_to_flower)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.days_to_flower <- makeqtl(LG.f2.after.crossover,chr=c("A10", "C06"),pos=c(184, 81))  
qtlm.days_to_flower
qtl.fit.days_to_flower <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.days_to_flower,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.days_to_flower) 

# refine position
qtl.refine.days_to_flower <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.days_to_flower,
                        formula= y ~ Q1 + Q2)

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.days_to_flower) # no interaction 

# test for additional loci 
qtl.add.days_to_flower <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.days_to_flower)
summary(qtl.add.days_to_flower) # no additional loci 

# final model 
qtl.fit.final.days_to_flower <- fitqtl(cross=LG.f2.after.crossover,qtl=qtl.refine.days_to_flower,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.final.days_to_flower)

# plot result 
png("~/F2/output/scantwo/figure/days_to_flower_scantwo_result.png", width=20, height=10, units="in", res=300)
par(mfrow=c(1,2)) 
plot(qtl.refine.days_to_flower)
plotLodProfile(qtl.refine.days_to_flower) 
dev.off() 
save(qtl.fit.final.days_to_flower, file = "~/F2/output/scantwo/qtl.fit.final.days_to_flower.Rdata")

### days_to_bolt
scantwo_perm_days_to_bolt <- do.call(c, scantwo_perm_days_to_flower_days_to_bolt_bolting_to_flowering)[,2]
summary(scantwo_perm_days_to_bolt)
# days_to_bolt (160 permutations)
#      full  fv1  int  add  av1  one
# 5%  10.17 7.76 6.72 7.18 4.04 3.87
# 10%  9.67 7.54 6.29 6.64 3.76 3.77

summary(scantwo.imp.42traits[["days_to_bolt"]],perm=scantwo_perm_days_to_bolt,alpha=.05,pval=T)  

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$days_to_bolt)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.days_to_bolt <- makeqtl(LG.f2.after.crossover,chr=c("A10", "C06"),pos=c(176, 83))  
qtlm.days_to_bolt
qtl.fit.days_to_bolt <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.days_to_bolt,formula = y ~ Q1 + Q2 ,get.ests=T) 
summary(qtl.fit.days_to_bolt) 

# refine position
qtl.refine.days_to_bolt <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.days_to_bolt,
                        formula= y ~ Q1 + Q2)

# test for interaction 
addint(LG.f2.after.crossover,qtl=qtl.refine.days_to_bolt) # no interaction 

# test for additional loci 
qtl.add.days_to_bolt <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.days_to_bolt)
summary(qtl.add.days_to_bolt) # # cA05.loc116     A05 116.0 4.332   

qtlm2.days_to_bolt <- addtoqtl(LG.f2.after.crossover,qtl.refine.days_to_bolt,chr="A05", pos=116.0)
summary(qtlm2.days_to_bolt) 

# scan the genome for additional QTL 
qtl.refine2.days_to_bolt <- refineqtl(LG.f2.after.crossover,qtl=qtlm2.days_to_bolt,
                              formula = y ~ Q1 + Q2 + Q3)

addint(LG.f2.after.crossover,qtl=qtl.refine2.days_to_bolt) # no interaction 
summary(qtl.refine2.days_to_bolt) # no additional loci  

qtl.fit2.days_to_bolt <- fitqtl(LG.f2.after.crossover,qtl=qtl.refine2.days_to_bolt,get.ests=T)
summary(qtl.fit2.days_to_bolt) 

addint(LG.f2.after.crossover,qtl=qtl.refine2.days_to_bolt) # no interaction 

# plot result 
png("~/F2/output/scantwo/figure/days_to_bolt_scantwo_result.png", width=20, height=10, units="in", res=300)
par(mfrow=c(1,2)) 
plot(qtl.refine2.days_to_bolt)
plotLodProfile(qtl.refine2.days_to_bolt) 
dev.off() 
save(qtl.fit2.days_to_bolt, file = "~/F2/output/scantwo/qtl.fit2.days_to_bolt.Rdata") 

# bolting to flower 
scantwo_perm_bolting_to_flowering <- do.call(c, scantwo_perm_days_to_flower_days_to_bolt_bolting_to_flowering)[,3]
summary(scantwo_perm_bolting_to_flowering)
# bolting_to_flowering (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  14.3 12.7 9.53 9.17 7.64 3.92
# 10% 13.8 12.0 9.12 8.77 7.21 3.62

load("~/F2/output/scantwo/scantwo.imp.43traits.Rdata")
summary(scantwo.imp.43traits[["bolting_to_flowering"]],perm=scantwo_perm_bolting_to_flowering,alpha=.05,pval=T)  
#  There were no pairs of loci meeting the criteria. 
```

### No__of transplanting_plants & No.of survival plants (two invalid pheno data)
### Survival rate & lobe number 2015_12_28     21,22
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Survival_rate_lobe_number_2015_12_28/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Survival_rate_lobe_number_2015_12_28 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_Survival_rate <- do.call(c, scantwo_perm_Survival_rate_lobe_number_2015_12_28)[,1]
summary(scantwo_perm_Survival_rate)
# Survival_rate (112 permutations)
#     full   fv1  int  add  av1  one
# 5%  12.4 10.43 8.39 7.59 5.33 4.28
# 10% 11.6  9.51 7.86 7.15 4.89 3.91

summary(scantwo.imp.42traits[["Survival_rate"]],perm=scantwo_perm_Survival_rate,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  

scantwo_perm_lobe_number_2015_12_28 <- do.call(c, scantwo_perm_Survival_rate_lobe_number_2015_12_28)[,2]
summary(scantwo_perm_lobe_number_2015_12_28)
# lobe_number_2015_12_28 (112 permutations)
#     full  fv1  int  add   av1  one
# 5%  17.5 16.1 13.1 12.1 10.52 4.39
# 10% 17.0 15.6 12.7 11.4  9.98 3.96

summary(scantwo.imp.42traits[["lobe_number_2015_12_28"]],perm=scantwo_perm_lobe_number_2015_12_28,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  
```

### lobe_number_2016_01_21 & lobe_number_2016_02_18   23,24
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/lobe_number_2016_01_21_lobe_number_2016_02_18/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_lobe_number_2016_01_21_lobe_number_2016_02_18 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_lobe_number_2016_01_21 <- do.call(c, scantwo_perm_lobe_number_2016_01_21_lobe_number_2016_02_18)[,1]
summary(scantwo_perm_lobe_number_2016_01_21)
# lobe_number_2016_01_21 (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  10.8 8.21 7.12 7.15 4.28 4.30
# 10% 10.5 8.01 6.82 6.90 4.07 3.92

summary(scantwo.imp.42traits[["lobe_number_2016_01_21"]],perm=scantwo_perm_lobe,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  

scantwo_perm_lobe_number_2016_02_18 <- do.call(c, scantwo_perm_lobe_number_2016_01_21_lobe_number_2016_02_18)[,2]
summary(scantwo_perm_lobe_number_2016_02_18)

summary(scantwo.imp.42traits[["lobe_number_2016_02_18"]],perm=scantwo_perm_lobe_number_2016_02_18,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.   
```

### lobe_number_03_21 & leaf_number_2015_12_28   25,26
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/lobe_number_03_21_leaf_number_2015_12_28/"
files <- list.files(path=path)
setwd(path) 
scantwo_perm_lobe_number_03_21_leaf_number_2015_12_28 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_lobe_number_03_21 <- do.call(c, scantwo_perm_lobe_number_03_21_leaf_number_2015_12_28)[,1]
summary(scantwo_perm_lobe_number_03_21) 

summary(scantwo.imp.42traits[["lobe_number_2016_03_21"]],perm=scantwo_perm_lobe_number_03_21,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  

scantwo_perm_leaf_number_2015_12_28 <- do.call(c, scantwo_perm_lobe_number_03_21_leaf_number_2015_12_28)[,2]
summary(scantwo_perm_leaf_number_2015_12_28)

summary(scantwo.imp.42traits[["leaf_number_2015_12_28"]],perm=scantwo_perm_leaf_number_2015_12_28,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  
```

### leaf_number_2016_01_21 & leaf_number_2016_02_18   27,28
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/leaf_number_2016_01_21_leaf_number_2016_02_18/"
files <- list.files(path=path)
setwd(path) 
scantwo_perm_leaf_number_2016_01_21_leaf_number_2016_02_18 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_leaf_number_2016_01_21 <- do.call(c, scantwo_perm_leaf_number_2016_01_21_leaf_number_2016_02_18)[,1]
summary(scantwo_perm_leaf_number_2016_01_21) 

summary(scantwo.imp.42traits[["leaf_number_2016_01_21"]],perm=scantwo_perm_leaf_number_2016_01_21,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  

scantwo_perm_leaf_number_2016_02_18 <- do.call(c, scantwo_perm_leaf_number_2016_01_21_leaf_number_2016_02_18)[,2]
summary(scantwo_perm_leaf_number_2016_02_18)

summary(scantwo.imp.42traits[["leaf_number_2016_02_18"]],perm=scantwo_perm_leaf_number_2016_02_18,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.   
```

### plant_weight_2016_05_13 & plant_width_2015_12_18   31,32
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/plant_weight_2016_05_13_plant_width_2015_12_18/"
files <- list.files(path=path)
setwd(path) 
scantwo_perm_plant_weight_2016_05_13_plant_width_2015_12_18 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_plant_weight_2016_05_13 <- do.call(c, scantwo_perm_plant_weight_2016_05_13_plant_width_2015_12_18)[,1]
summary(scantwo_perm_plant_weight_2016_05_13) 

summary(scantwo.imp.42traits[["plant_weight_2016_05_13"]],perm=scantwo_perm_plant_weight_2016_05_13,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  

scantwo_perm_plant_width_2015_12_18 <- do.call(c, scantwo_perm_plant_weight_2016_05_13_plant_width_2015_12_18)[,2]
summary(scantwo_perm_plant_width_2015_12_18)

summary(scantwo.imp.42traits[["plant_width_2015_12_18"]],perm=scantwo_perm_plant_width_2015_12_18,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  
```

### plant_width_2016_03_16 & plant_height_2015_12_18   35,36
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/plant_width_2016_03_16_plant_height_2015_12_18/"
files <- list.files(path=path)
setwd(path) 
scantwo_perm_plant_width_2016_03_16_plant_height_2015_12_18 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_plant_width_2016_03_16 <- do.call(c, scantwo_perm_plant_width_2016_03_16_plant_height_2015_12_18)[,1]
summary(scantwo_perm_plant_width_2016_03_16) 

summary(scantwo.imp.42traits[["plant_width_2016_03_16"]],perm=scantwo_perm_plant_width_2016_03_16,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  

scantwo_perm_plant_height_2015_12_18 <- do.call(c, scantwo_perm_plant_width_2016_03_16_plant_height_2015_12_18)[,2]
summary(scantwo_perm_plant_height_2015_12_18)

summary(scantwo.imp.42traits[["plant_height_2015_12_18"]],perm=scantwo_perm_plant_height_2015_12_18,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  
```

### plant_height_2016_01_20 & plant_height_2016_02_17   37,38
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/plant_height_2016_01_20_plant_height_2016_02_17/"
files <- list.files(path=path)
setwd(path) 
scantwo_perm_plant_height_2016_01_20_plant_height_2016_02_17 <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

scantwo_perm_plant_height_2016_01_20 <- do.call(c, scantwo_perm_plant_height_2016_01_20_plant_height_2016_02_17)[,1]
summary(scantwo_perm_plant_height_2016_01_20) 

summary(scantwo.imp.42traits[["plant_height_2016_01_20"]],perm=scantwo_perm_plant_height_2016_01_20,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  

scantwo_perm_plant_height_2016_02_17 <- do.call(c, scantwo_perm_plant_height_2016_01_20_plant_height_2016_02_17)[,2]
summary(scantwo_perm_plant_height_2016_02_17)

summary(scantwo.imp.42traits[["plant_height_2016_02_17"]],perm=scantwo_perm_plant_height_2016_02_17,alpha=.05,pval=T)  
# There were no pairs of loci meeting the criteria.  
``` 

### check result and make effectplot 
```{r}
# Erucic acid 
load("~/F2/output/scantwo/qtl.fit.final.Erucic.Rdata")
summary(qtl.fit.final.Erucic) 
# Drop one QTL at a time ANOVA table: 
# ----------------------------------  
#                   df Type III SS    LOD   %var F value Pvalue(Chi2) Pvalue(F)    
# A08@72.3           6     13772.7 46.371 51.859  68.552            0   < 2e-16 ***
# C03@31.0           6      5857.0 26.985 22.053  29.152            0   < 2e-16 ***
# A08@72.3:C03@31.0  4       701.5  4.515  2.641   5.237            0   0.00055 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# 
# Estimated effects:
# -----------------
#                           est        SE       t
# Intercept            27.24744   0.46372  58.758
# A08@72.3a           -13.30794   0.68323 -19.478
# A08@72.3d             0.69096   0.94627   0.730
# C03@31.0a            -7.88981   0.65820 -11.987
# C03@31.0d            -0.09388   0.95188  -0.099
# A08@72.3a:C03@31.0a  -2.37504   0.92388  -2.571
# A08@72.3d:C03@31.0a   2.96093   1.32524   2.234
# A08@72.3a:C03@31.0d  -2.42409   1.36404  -1.777
# A08@72.3d:C03@31.0d   1.51147   1.87515   0.806

summary(qtl.fit.final.Erucic, dropone = FALSE) 

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Erucic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

find.marker(LG.f2.after.crossover, "A08", 72.3)   
find.marker(LG.f2.after.crossover, "C03", 31.0) 

mar <- find.marker(LG.f2.after.crossover, c("A08", "C03"), c(72.3, 31.0)) 

png("~/F2/output/QTL_analysis/figure/scantwo_Erucic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2))
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@72.3", mname2 = "C03@31.0")
dev.off()

test <- effectplot(LG.f2.after.crossover, mname1 = "chrA08_10120555", mname2 = "chrC03_56259623")
test

# Oleic acid 
load("~/F2/output/scantwo/qtl.fit.final.Oleic.Rdata")
summary(qtl.fit.final.Oleic)

# Drop one QTL at a time ANOVA table: 
# ----------------------------------  
#                   df Type III SS   LOD   %var F value Pvalue(Chi2) Pvalue(F)    
# A08@72.3           6        8593 39.07 44.952   51.18            0   < 2e-16 ***
# C03@31.0           6        5972 30.94 31.242   35.57            0   < 2e-16 ***
# A08@72.3:C03@31.0  4        1411 10.04  7.381   12.61            0  6.42e-09 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# 
# Estimated effects:
# -----------------
#                         est      SE      t
# Intercept           28.4146  0.4299 66.099
# A08@72.3a            9.8861  0.6485 15.244
# A08@72.3d           -2.1516  0.8709 -2.470
# C03@31.0a            7.2126  0.6160 11.709
# C03@31.0d           -0.6224  0.8932 -0.697
# A08@72.3a:C03@31.0a  4.9585  0.8509  5.827
# A08@72.3d:C03@31.0a -3.5777  1.2135 -2.948
# A08@72.3a:C03@31.0d -0.6806  1.2485 -0.545
# A08@72.3d:C03@31.0d -0.5451  1.7150 -0.318

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Oleic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, c("A08", "C03"), c(72.3, 31.0)) 
png("~/F2/output/QTL_analysis/figure/scantwo_Oleic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2))
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@72.3", mname2 = "C03@31.0")
dev.off()

test <- effectplot(LG.f2.after.crossover, mname1 = "chrA08_10120555", mname2 = "chrC03_56259623")
test

### Linolenic 
load("~/F2/output/scantwo/qtl.fit.Linolenic.Rdata") 
summary(qtl.fit.Linolenic) 
# Drop one QTL at a time ANOVA table: 
# ----------------------------------  
#                   df Type III SS   LOD  %var F value Pvalue(Chi2) Pvalue(F)    
# A08@73.0           6      1049.3 52.92 66.25   87.41            0    <2e-16 ***
# C03@36.0           6      1050.0 52.93 66.29   87.46            0    <2e-16 ***
# A08@73.0:C03@36.0  4       857.1 47.44 54.12  107.10            0    <2e-16 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# 
# Estimated effects:
# -----------------
#                         est      SE       t
# Intercept           10.9442  0.1211  90.346 
# A08@73.0a           -0.1025  0.1756  -0.584
# A08@73.0d            1.5387  0.2423   6.350
# C03@36.0a           -0.8820  0.1740  -5.069
# C03@36.0d            0.9355  0.2436   3.840
# A08@73.0a:C03@36.0a -4.2130  0.2395 -17.594
# A08@73.0d:C03@36.0a  1.2538  0.3479   3.603
# A08@73.0a:C03@36.0d  1.3749  0.3512   3.915
# A08@73.0d:C03@36.0d -0.6091  0.4877  -1.249

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linolenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, c("A08", "C03"), c(73, 36)) 
png("~/F2/output/QTL_analysis/figure/scantwo_Linolenic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2))
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@73", mname2 = "C03@36")
dev.off()

test <- effectplot(LG.f2.after.crossover, mname1 = "chrA08_10120555", mname2 = "chrC03_56259623")
test

### Stearic 
load("~/F2/output/scantwo/qtl.fit2.Stearic.Rdata")
summary(qtl.fit2.Stearic)
# Drop one QTL at a time ANOVA table: 
# ----------------------------------  
#          df Type III SS    LOD   %var F value Pvalue(Chi2) Pvalue(F)    
# A08@71.0  2      5.1313 27.731 36.553   92.08            0   < 2e-16 ***
# C03@29.0  2      2.6552 16.928 18.915   47.65            0   < 2e-16 ***
# C02@59.1  2      0.8046  6.016  5.732   14.44            0  1.73e-06 ***
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# 
# Estimated effects:
# -----------------
#                est       SE       t
# Intercept  1.40447  0.01330 105.562
# A08@71.0a  0.25348  0.01911  13.262
# A08@71.0d -0.04099  0.02671  -1.535
# C03@29.0a  0.18243  0.01939   9.407
# C03@29.0d -0.04361  0.02683  -1.626
# C02@59.1a  0.05947  0.02011   2.957
# C02@59.1d -0.11989  0.02651  -4.523

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Stearic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, c("A08", "C03"), c(71, 29)) 
png("~/F2/output/QTL_analysis/figure/scantwo_Stearic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2))
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@73", mname2 = "C03@36")
dev.off()

test <- effectplot(LG.f2.after.crossover, mname1 = "chrA08_10120555", mname2 = "chrC03_56259623")
test

### Palmitic acid 
load("~/F2/output/QTL_analysis/cim.qtl.43traits.corrrected.Rdata")
bayesint(cim.qtl$Palmitic_acid, chr = "A08", expandtomarkers = T) # 72
bayesint(cim.qtl$Palmitic_acid, chr = "C03", expandtomarkers = T) # 36 

load("LG.f2.after.crossover.43traits.corrected.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Palmitic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, c("A08", "C03"), c(72, 36)) 
png("~/F2/output/QTL_analysis/figure/loci_effect_Palmitic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2))
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@72", mname2 = "C03@36")
dev.off()  

### Arachidic acid 
bayesint(cim.qtl$Arachidic_acid, chr = "A08", expandtomarkers = T) # 78

load("LG.f2.after.crossover.43traits.corrected.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Arachidic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, "A08", 78) 
png("~/F2/output/QTL_analysis/figure/loci_effect_Arachidic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2))
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@78")
dev.off() 

### Behenic acid 
bayesint(cim.qtl$Behenic_acid, chr = "A08", expandtomarkers = T) # 73
bayesint(cim.qtl$Behenic_acid, chr = "C03", expandtomarkers = T) # 35

load("LG.f2.after.crossover.43traits.corrected.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Behenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, c("A08", "C03"), c(73, 35)) 
png("~/F2/output/QTL_analysis/figure/loci_effect_Behenic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2))
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@73", mname2 = "A03@35")
dev.off() 

### vaccennic acid 
bayesint(cim.qtl$vaccenic_acid, chr = "A08", expandtomarkers = T) # 72

load("LG.f2.after.crossover.43traits.corrected.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$vaccenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, "A08", 72) 
png("~/F2/output/QTL_analysis/figure/loci_effect_vaccenic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2))
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@72") 
dev.off()  

### Escosinic acid 
bayesint(cim.qtl$cis_11_Eicosenoic_acid, chr = "A08", expandtomarkers = T) # 76

load("LG.f2.after.crossover.43traits.corrected.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$cis_11_Eicosenoic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, "A08", 76) 
png("~/F2/output/QTL_analysis/figure/loci_effect_escosinic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2))
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@76")
dev.off()  

### Linoieic acid 
bayesint(cim.qtl$Linoleic_acid, chr = "A08", expandtomarkers = T) # 72.6
bayesint(cim.qtl$Linoleic_acid, chr = "C03", expandtomarkers = T) # 39 

load("~/F2/output/QTL_analysis/LG.f2.after.crossover.43traits.corrected.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linoleic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, c("A08", "C03"), c(72.6, 39)) 
png("~/F2/output/QTL_analysis/figure/loci_effect_linoleic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2)) 
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "A08@72.6", mname2 = "C03@39")
dev.off()    

### Caprylic acid 
load("~/F2/output/QTL_analysis/LG.f2.after.crossover.43traits.corrected.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Caprylic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 

mar <- find.marker(LG.f2.after.crossover, c("C01", "C01"), c(22.0, 50.1)) 
png("~/F2/output/QTL_analysis/figure/loci_effect_Caprylic.png", width=15, height=7, units="in", res=300)
par(mfrow = c(1,2)) 
plot.pxg(LG.f2.after.crossover, marker = mar)
effectplot(LG.f2.after.crossover, mname1 = "C01@22.0", mname2 = "C01@50.1")
dev.off()    

### 
load("~/F2/output/scantwo/qtl.fit2.days_to_bolt.Rdata")
summary(qtl.fit2.days_to_bolt)

load("~/F2/output/scantwo/qtl.fit.days_to_flower.Rdata")
summary(qtl.fit.days_to_flower)  
```

### correct 3 phenotype data and re-run scantwo 
three traits don't seem to have correct phenotype data, so I corrected the data and re-run scantwo on them. cis_11_Eicosenoic_acid, Palmitoliec_aicd, and Behenic_acid were corrected and new data saved as "~/F2/output/QTL_analysis/LG.f2.after.crossover.43traits.corrected.Rdata".  

running scantwo analysis and permutation seprately
1) scantwo analysis on cabernet 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scantwo_v2.R  

scantwo permutation 100 times for 43 traits, split traits file so that each file has 2 traits, in total 20 files, and each run is only 10 permutation, combine permutation result after finish running. 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scantwo_perm_1.R ... scantwo_perm_10.R, submit to slurm queue using https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scantwo_perm_1_10.slurm 


