---
title: "trans_eQTL_hostpot"
output: html_document
---

Purpose of this script is to identify trans-eQTL hotspot 

### practice  
```{r}
library(qtlhot)

ncross1 <- sim.null.cross(chr.len = rep(100, 4), 
n.mar = 51,
n.ind = 100,
type = "bc",
n.pheno = 1000,
latent.eff = 3,
res.var = 1, 
init.seed = 123457) 

ncross1
cross1 <- include.hotspots(cross = ncross1,
hchr = c(2, 3, 4), 
hpos = c(25, 75, 50),
hsize = c(100, 50, 20),
Q.eff = 2,
latent.eff = 3,
lod.range.1 = c(2.5, 2.5),
lod.range.2 = c(5, 8),
lod.range.3 = c(10, 15),
res.var = 1,
n.phe = 1000,
init.seed = 12345)

cross1

ncor1 <- cor(cross1$pheno)
summary(ncor1[lower.tri(ncor1)])

rm(ncor1)
?rm

set.seed(123)
pt <- scanone(ncross1, method = "hk", n.perm = 1000)
alphas <- seq(0.01, 0.10, by=0.01)
lod.thrs <- summary(pt, alphas) # get lod thrs from scanone result for different alpha 
lod.thrs

lod.thr <- lod.thrs[5]
lod.thr # 95% as the threhold 

scan1 <- scanone(cross1, pheno.col = 1:1000, method = "hk")
scan1 %>% dim() # 204 1002 

high1 <- highlod(scan1, lod.thr = min(lod.thrs), drop.lod = 1.5)
high1 %>% names()
high1[[1]] %>% head(100) # LOD support interval for significant peaks for each trait 
high1[[2]] %>% dim() # marker information 
high1[[3]] # number of permutations 

max(high1, lod.thr = lod.thrs) #  maximum hotspot size for each possible single trait threshold, max.N is the number of traits out of 1000 pheno traits  

hots1 <- hotsize(high1, lod.thr = lod.thr) # for each genomic position, we count the number of traits that map to it with a LOD score equal or higher than the threshold in lod.thr.
summary(hots1) # 

plot(hots1, cex.lab = 1.5, cex.axis = 1.5) # plot the hotspot architecture inferred using the single trait permutation threshold 2.44 
lod.thr # 2.44 

# perform permutation tests to assess the statistical significance of the hotspots detected, hotperm uses NL & N method for permutation, Q method is implemented in ww.perm().  
set.seed(12345) 
hotperm1 <- hotperm(cross = cross1,
n.quant = 300,  # The parameter n.quant sets the maximum hotspot size to be analyzed 
n.perm = 100, # number of permutation 
lod.thrs = lod.thrs, # vector of LOD thresholds 
alpha.levels = alphas, # vector of significance levels  
drop.lod = 1.5, # LOD drop amount for support intervals 
verbose = FALSE) 
?hotperm # conduct NL & N permutation test,  
alphas 

hotperm1 
names(hotperm1)
summary(hotperm1)

hotperm1$max.N %>% dim() 
# max.N element of the hotperm1 object stores the output of the N-methodâ€™s permutations and is given by a matrix with 100 rows representing the permutations, and 10 columns representing the QTL mapping thresholds
hotperm1$max.N %>% head()

hotperm1$max.lod.quant %>% dim() # 100 300 # 100 rows representing the permutations, and 300 columns representing the hotspot sizes analyzed
hotperm1$max.lod.quant[1:10, 1:10] 

quant1 <- quantile(hotperm1, 0.05, lod.thr = lod.thr)
plot(high1, quant.level = quant1, sliding = TRUE)
summary(hotperm1)

hotsq1 <- hotsize(high1, lod = lod.thr, window = 5, quant.level = quant1)
plot(hotsq1)
summary(hotsq1)

### example with uncorrelated phenotypes 
ncross2 <- sim.null.cross(chr.len = rep(100,4),
n.mar = 51,
n.ind = 100,
type = "bc",
n.phe = 1000,
latent.eff = 0,
res.var = 1,
init.seed = 123457)

cross2 <- include.hotspots(cross = ncross2,
hchr = c(2, 3, 4),
hpos = c(25, 75, 50),
hsize = c(100, 50, 20),
Q.eff = 2,
latent.eff = 0,
lod.range.1 = c(2.5, 2.5),
lod.range.2 = c(5, 8),
lod.range.3 = c(10, 15),
res.var = 1,
n.phe = 1000,
init.seed = 12345)

ncor2 <- cor(cross2$pheno)
summary(ncor2[lower.tri(ncor2)])

rm(ncor2)
scan2 <- scanone(cross2, pheno.col = 1:1000, method = "hk")
high2 <- highlod(scan2, lod.thr = lod.thr, drop.lod = 1.5)

hots2 <- hotsize(high2)
plot(hots2, cex.lab = 1.5, cex.axis = 1.5)

set.seed(12345)

hotperm2 <- hotperm(cross = cross2,
n.quant = 300,
n.perm = 100,
lod.thrs = lod.thrs,
alpha.levels = alphas,
drop.lod = 1.5,
verbose = FALSE)

quant2 <- quantile(hotperm2, 0.05, lod.thr = lod.thr)
quant2

plot(high2, lod.thr = lod.thr, quant.level = quant2, sliding = TRUE)
plot(high2, quant.level = quant2)
### still several points not very clear, especially how permutation works here... 
``` 

### using N method for trans-eQTL hotspot identification 
```{r}
library(tidyverse)
library(qtl)
library(qtlhot)

load("~/F2/output/eQTL/scanone-eqtl_F2_flipped.RData") 
scanone_eQTL.F2 %>% dim() # 4887 56182 

scanone_eQTL.F2[1:10, 1:10]

### extract genes with trans-eQTL only 
load("~/F2/output/eQTL/cis_trans_result_new_flipped.Rdata") 

cis_eQTL %>% dim() # 11384    12
trans_eQTL %>% dim() # 15197    12  

# include genes with trans-eQTL only  
cis_eQTL %>% head  
cis_eQTL$gene_ID %>% unique() %>% length()
trans_eQTL$gene_ID %>% unique() %>% length() # 13296 
length(cis_eQTL$gene_ID[cis_eQTL$gene_ID %in% trans_eQTL$gene_ID]) # 1814 genes not only have cis-eQTL, but also have trans-eQTL 

scanone_eQTL.F2.trans <- scanone_eQTL.F2[,(colnames(scanone_eQTL.F2) %in% c("chr", "pos", trans_eQTL$gene_ID))]
cbind(pos, scanone_eQTL.F2.trans) %>% dim() 

dim(scanone_eQTL.F2.trans) # 4887 13298
scanone_eQTL.F2.trans[1:10, 1:10]
scanone_eQTL.F2[1:10, 1:10]
### check correlation, later... 
cross.F2
lod.thrs
alphas <- seq(0.01, 0.10, by=0.01)  

### LOD threshold level for e-trait 
lod.thr <- lod.thrs[5]
lod.thr 

### get cross.F2 for trans-eQTL only 
cross.F2$pheno <- cross.F2$pheno[,(colnames(cross.F2$pheno) %in% c(trans_eQTL$gene_ID))]

### get only significant intervals for each e-trait, using LOD drop method, understand the threhsold here 
high1 <- highlod(scanone_eQTL.F2.trans, lod.thr = min(lod.thrs), drop.lod = 1.5) 
max(high1, lod.thr = lod.thrs) # max number of e-trait fall into loci with different lod threshold   

hots1 <- hotsize(high1, lod.thr = lod.thr)  
summary(hots1) # for each genomic position 
save(hots1, file = "~/F2/for_paper/hots1.Rdata")
### permutation to get the statistical significance # permutation takes time, did this in screen 
# trans_eQTL_hotspot_perm_trans.R 

load("~/F2/output/eQTL/hotperm1.trans.Rdata")
hotperm1.trans   

png("~/F2/output/QTL_analysis/figure/trans_eQTL_hotspot.png", width=188, height=80, units="mm", res=300) 
par(ps = 8, cex = 1, cex.main = 1, cex.axis = 0.8, tcl = -0.3) 

plot(hots1, ylab="trans-eQTL counts") 
abline(h = 129) 
dev.off()  

### how to get all the genetic pos with trans-eQTL hotspot? 
hot_trans <- data.frame(chr = hots1$chr,
                        pos = hots1$pos,
                        max.N = hots1$max.N)

hot_trans <- 
hot_trans %>% 
  filter(max.N > 129) %>% 
  mutate(eQTL_chr = chr) 

trans_eQTL %>% colnames()
hot_trans

### hotspot 
# A01: 128-129
# A01: 160-164
# A02: 0-9
# A03: 17-25
# A03: 36-38
# A03: 43
# A03: 49-64
# A06: 122-124
# A09: 17-19 
# A10: 173-185
# A10: 217-226
# C01: 156-166
# C02: 1-13
# C02: 43-62
# C03: 10-24
# C04: 97
# C08: 40-41
# C08: 47-62
# C09: 0-4
# C09: 57-60 

hotspot <- data.frame(chr = c("A01", "A01", "A02", "A03", "A03", "A03", "A03", "A06", "A09", "A10", "A10", "C01", "C02", "C02", "C03", "C04", "C08", "C08", "C09", "C09"), 
                      pos_min = c(128, 160, 0, 17, 36, 43, 49, 122, 17, 173, 217, 156, 1, 43, 10, 97, 40, 47, 0, 57), 
                      pos_max = c(129, 164, 9, 25, 38, 43, 64, 124, 19, 185, 226, 166, 13, 62, 24, 97, 41, 62, 4, 60))

hotspot$chr <- paste("chr", hotspot$chr, sep = "")  
hotspot$hotspot_id <- rownames(hotspot)

### GO enrichment of trans-eQTL hotspot regulated genes 
hot_trans$eQTL_chr <- paste("chr", hot_trans$eQTL_chr, sep = "")

hot_trans %>% colnames()
hot_trans

# use the flipped one 
load("~/F2/output/eQTL/cis_trans_result_new_flipped_C05C08.Rdata") 

cis_eQTL %>% dim() # 11587    15
trans_eQTL %>% dim() # 14657    15  

trans_eQTL_hot <- 
trans_eQTL %>% 
  semi_join(hot_trans, by = c("eQTL_chr", "pos")) 

dim(trans_eQTL_hot) # 4221 12 

trans_eQTL_genes <- GRanges(seqnames = Rle(trans_eQTL_hot$eQTL_chr),ranges = IRanges(start = trans_eQTL_hot$pos, end = trans_eQTL_hot$pos), names = trans_eQTL_hot$gene_ID)
trans_eQTL_genes 

hotspots <- 
lapply(1:nrow(hotspot), function(i) {
  GRanges(seqnames = Rle(hotspot$chr[i]), ranges = IRanges(start = hotspot$pos_min[i], end = hotspot$pos_max[i]), ID = hotspot$hotspot_id[i])
})

trans_eQTL_genes_list <- 
lapply(1:length(hotspots), function(i) 
  mergeByOverlaps(trans_eQTL_genes, hotspots[[i]])$names
)

length(trans_eQTL_genes_list) # 20  
names(trans_eQTL_genes_list) <- paste(hotspot$chr, hotspot$pos_min, hotspot$pos_max, sep = "_")

trans_eQTL_genes_list$chrA03_49_64 

BnapusAnnotation[BnapusAnnotation$name %in% trans_eQTL_genes_list$chrA03_49_64,] %>% View()

trans_eQTL %>% 
  filter(gene_ID == "BnaA08g11130D" | gene_ID == "BnaC03g65980D") %>%
  View()

### BnaA08g11130D (FAE) & BnaC03g65980D (FAE) are regulated by A03 trans-eQTL hostpot, the two BnaA.FAE1 reported by other people  

# also add B.napus GO annotation 
load("~/Reference/B.napus/napus_GO_combined.Rdata") # there is one colomn hiding... 
load("~/Reference/B.napus/BnapusAnnotation.Rdata") 
BnapusAnnotation %>% colnames()

trans_eQTL_genes_chrA03_49_64 <- as.data.frame(trans_eQTL_genes_list$chrA03_49_64) 
colnames(trans_eQTL_genes_chrA03_49_64) <- "name"

trans_eQTL_genes_chrA03_49_64 <- 
trans_eQTL_genes_chrA03_49_64 %>% 
  left_join(napus_GO_combined, c("name" = "gene_ID")) %>% 
  left_join(BnapusAnnotation) 

trans_eQTL_genes_chrA03_49_64_FA <- 
trans_eQTL_genes_chrA03_49_64[grep("lipid|fatty", trans_eQTL_genes_chrA03_49_64$GO_des),] %>% 
  dplyr::select(-chrom, -start, -end)

tmp <- test[grep("GO:0006633", test$B.napus_GO_term),]$name
tmp 

trans_eQTL_genes_chrA03_49_64_FA$note <- ifelse(trans_eQTL_genes_chrA03_49_64_FA$name %in% tmp, "enriched", "not_enriched")
trans_eQTL_genes_chrA03_49_64_FA %>% View()
trans_eQTL_genes_chrA03_49_64_FA %>% dim()

write.csv(trans_eQTL_genes_chrA03_49_64_FA, file = "~/F2/for_paper/trans_eQTL_genes_chrA03_49_64_FA.csv")

load("~/F2/output/eQTL/cis_trans_result_new_flipped_C05C08.Rdata")  

tmp[!(tmp %in% trans_eQTL$gene_ID)]
"BnaC09g41580D" %in% cis_eQTL$gene_ID   

colnames(napus_GO_combined) 
colnames(BnapusAnnotation)
nrow(trans_eQTL_genes_chrA03_49_64)  
```  

### GO enrichment of trans-eQTL hotspot regulated genes  
```{r}
load("~/Desktop/F2_paper/data/trans_eQTL_genes_list.Rdata")
source("~/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")  

GO_trans_eQTL <- 
lapply(1:length(trans_eQTL_genes_list), function(i) {
  tryCatch(
    GOseq.Bn.ORA(trans_eQTL_genes_list[[i]]), error= function(e) NA)
}) 

GO_trans_eQTL 
save(GO_trans_eQTL, file = "~/Desktop/F2_paper/data/GO_trans_eQTL.Rdata")

# check GO enrichment result 
hotspot <- data.frame(chr = c("A01", "A01", "A02", "A03", "A03", "A03", "A03", "A06", "A09", "A10", "A10", "C01", "C02", "C02", "C03", "C04", "C08", "C08", "C09", "C09"), 
                      pos_min = c(128, 160, 0, 17, 36, 43, 49, 122, 17, 173, 217, 156, 1, 43, 10, 97, 40, 47, 0, 57), 
                      pos_max = c(129, 164, 9, 25, 38, 43, 64, 124, 19, 185, 226, 166, 13, 62, 24, 97, 41, 62, 4, 60))

hotspot$chr <- paste("chr", hotspot$chr, sep = "")  
hotspot$hotspot_id <- rownames(hotspot)

### GO enrichment of trans-eQTL hotspot regulated genes 
hot_trans$eQTL_chr <- paste("chr", hot_trans$eQTL_chr, sep = "") 

load("~/F2/for_paper/GO_trans_eQTL.Rdata")
names(GO_trans_eQTL) <- paste(hotspot$chr, hotspot$pos_min, hotspot$pos_max, sep = "_")
GO_trans_eQTL$chrA03_49_64
<<<<<<< HEAD 
=======

# GO_trans_eQTL <- GO_trans_eQTL[!is.na(GO_trans_eQTL)]   
GO_trans_eQTL <- do.call("rbind", GO_trans_eQTL) 
write.csv(GO_trans_eQTL, file = "~/F2/for_paper/GO_trans_eQTL.csv")  
>>>>>>> a5efc319082df97479d180622f1a67544ada6069
```

### since the two FAE's expression is not correlated with Erucic acid level, how about other targets of A03 trans-eQTL hotspot? what is their expression correlation level with Erucic acid content? 

```{r}
vstMat.F2 <- read.csv("~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv", as.is = T, row.names = 1)
vstMat.F2[1:10, 1:10] 
colnames(vstMat.F2) 
vstMat.F2 <- vstMat.F2 %>% t() %>% as.data.frame()
vstMat.F2[1:10, 1:10]
vstMat.F2$id <- rownames(vstMat.F2)

F2_pheno <- read.csv("~/F2/data/QTL_analysis/F2.pheno.csv", as.is = T, row.names = 1, header = F)
F2_pheno <- t(F2_pheno) %>% as.data.frame()
F2_pheno

# genes regulated by A03 
A03_FA_gene <- read.csv("~/F2/for_paper/trans_eQTL_genes_chrA03_49_64_FA.csv", row.names = 1)
A03_FA_gene[1:10,] 

expression_A03_FA_gene <- vstMat.F2[,colnames(vstMat.F2) %in% A03_FA_gene$name] 
expression_A03_FA_gene$id <- rownames(expression_A03_FA_gene)

A03_FA_gene_data <- 
expression_A03_FA_gene %>% 
  left_join(F2_pheno[,c("Crude_oil_contents", "id")]) 

colnames(A03_FA_gene_data)  
str(A03_FA_gene_data)

test <- 
sapply(colnames(A03_FA_gene_data)[1:45], function(i) {
  cor(A03_FA_gene_data[,i], as.numeric(as.character(A03_FA_gene_data$Crude_oil_contents)))
}) 

max(test) 

colnames(F2_pheno)
```




















