---
title: "505_network_WGCNA_2"
output: html_document
---

purpose of this script is to 
1) relate modules to phenotypic traits
2) GO enrichment of different modules 
3) Export of networks to cytoscape for visualization 

### related modules to trait and identify important genes
```{r}
load("~/505/network_analysis/output/MEs_505_signed.Rdata")
load("~/505/network_analysis/output/pheno_data_505.Rdata")

# include triat value  
dim(MEs) # 39 moduels 
rownames(vstMat.505.WGCNA.t) 
rownames(MEs) 
rownames(MEs) <- rownames(vstMat.505.WGCNA.t)
MEs$ID <- rownames(MEs)

# merge trait and module values 
MEs$ID <- gsub("\\_|\\-", "", MEs$ID)
MEs$ID <- gsub("(505)(K)([[:digit:]]+)", "ID_\\1_\\2\\3", MEs$ID)

# There are two K200 sequencing info (this line was sequenced twice), modify the name, change the 1st K200 to K200-1, the 2nd to K200-2, for both module and trait info  
MEs$ID[15] <- "ID_505_K200-1"
MEs$ID[16] <- "ID_505_K200-2"

pheno_data_505$ID[16] <- "ID_505_K200-1"
pheno_data_505$ID[17] <- "ID_505_K200-2"

ME.blups.505 <- 
MEs %>% 
  left_join(pheno_data_505, by = "ID") 

dim(ME.blups.505) # 39 module + 56 traits + 1 "line"
colnames(ME.blups.505)

blups.cor <- cor(ME.blups.505[,41:96], ME.blups.505[,1:39], use = "pairwise.complete.obs") # the 1st part is phenotype, 2nd part is the module info, calculate correlation between eigengene and trait value
blups.cor.P <- corPvalueStudent(blups.cor,nrow(vstMat.505.WGCNA.t))
blups.cor.sig <- blups.cor
blups.cor.sig[blups.cor.P>0.05] <- NA 
blups.cor.sig 

MEs <- MEs %>% dplyr::select(-ID) 
ME.blups.505 <- ME.blups.505 %>% dplyr::select(-ID)  
blups.505 <- pheno_data_505 %>% dplyr::select(-ID)  
save(blups.505, file = "~/505/network_analysis/output/blups.505.Rdata")

# plot it 
# Will display correlations and their p-values
png(filename = "~/505/network_analysis/output/Module-trait_heatmap_505.png",width=20, height=16, units="in", res=300)
par(mar = c(8, 18, 3, 3)); 
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = blups.cor,
               yLabels = names(blups.505),
               xLabels = names(MEs), 
               xSymbols = names(MEs),
               ySymbols = names(blups.505),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = signif(blups.cor.sig,2),
               setStdMargins = FALSE,
               cex.text = 0.5,
               zlim = c(-1,1),
               main = paste("505 Module-trait relationships")) # this figure shows the correlation between traits and eigene gene of different modules, only correlation with P-value smaller than 0.05 were shown with correlation value 

dev.off()    

# again follow Julin's method 
# Arbitrary, but let's take the max and min for each trait (so long as they are significant)

blups.cor.5 <- blups.cor.sig  
dim(blups.cor.5) # 56 39

cor.top <- t(apply(blups.cor.5,1,function(x) { # for each trait 
  maxx = max(x,na.rm=TRUE) 
  minx = min(x,na.rm=TRUE)
  ifelse(x == maxx | x == minx, x, NA) # only keep the largest and smallest correlation for each trait 
}
)) 
dim(cor.top) # 56 39

cor.top <- cor.top[,apply(cor.top,2,function(x) !all(is.na(x)))] # keep modules which have significant correlations with traits
cor.top %>% dim() # 56 28
write.csv(cor.top,"~/505/network_analysis/output/Eigen_trait_cor_505_threshold.csv") 

png(filename = "~/505/network_analysis/output/Module-trait_heatmap_505_cortop.png",width=20, height=16, units="in", res=300)
par(mar = c(8, 18, 3, 3)); 
# make plot for this new cor.top object 
labeledHeatmap(Matrix = cor.top, # blups.cor
               yLabels = names(blups.505),
               xLabels = colnames(cor.top), 
               xSymbols = colnames(cor.top),
               ySymbols = names(blups.505),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               # textMatrix = signif(blups.cor.sig,2),
               setStdMargins = FALSE,
               cex.text = 0.5,
               zlim = c(-1,1),
               main = paste("505 Module-trait relationships")) 
dev.off()   

# write the Eigen genes
head(MEs[,colnames(cor.top)]) 
MEs[,colnames(cor.top)] %>% dim # get 28 modules which are significantly correlated with traits 
write.csv(MEs[,colnames(cor.top)], file = "~/505/network_analysis/output/Top_Eigen_genes_505.csv") 
```

Based on the above result, for oil related traits, looks like oil and pink modules are more important than other modules because they are have the highest correlation with multiple FA composition traits. 

### GO enrichment of different modules 
```{r}
source("~/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/MEs_505_signed.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/vstMat.505.batch.corrected.largeVar.Rdata")  

MEs %>% dim() # 131 39 
moduleColors %>% length() # 13454
vstMat.505.batch.corrected.largeVar %>% dim()  # 13454   131

# split genes into list based on their module name 

GO.module <- 
lapply(unique(moduleColors), function(i)
{ print(i)
  genes <- vstMat.505.batch.corrected.largeVar[moduleColors == i,] %>% rownames()
  print(length(genes))
  tryCatch(
    GOseq.Bn.ORA(genes), error = function(e) NA
  )
  })

names(GO.module) <- unique(moduleColors) 
GO.module 

# look at what genes do I have for module brown and module pink 
gene.module <- 
lapply(unique(moduleColors), function(i)
{ print(i)
  genes <- vstMat.505.batch.corrected.largeVar[moduleColors == i,] %>% rownames()
  })

names(gene.module) <- unique(moduleColors) 

save(gene.module, GO.module, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/module.gene.GO.Rdata")
```

### check modules with significant correlation with traits, combined with GO enrichment analysis result 
```{r}
load("~/505/network_analysis/input/module.gene.GO.Rdata")
GO.module %>% length()
gene.module %>% length()

# where are the two FAE1 genes? "BnaA08g11130D" & "BnaC03g65980D"

"BnaA08g11130D" %in% gene.module[["brown"]]
"BnaC03g65980D" %in% gene.module[["brown"]] # although "brown" is enriched for FA, it is not significantly correlated with FA traits. 

"BnaA08g11060D" %in% rownames(vstMat.505.batch.corrected.largeVar) # not detected 
"BnaA08g11140D" %in% rownames(vstMat.505.batch.corrected.largeVar) # not detected 
"BnaA08g12780D" %in% rownames(vstMat.505.batch.corrected.largeVar) # not detected
"BnaC04g05920D" %in% rownames(vstMat.505.batch.corrected.largeVar) 

rownames(vstMat.505.batch.corrected) 

# FA trait
GO.module[["salmon"]]
GO.module[["yellow"]]
GO.module[["turquoise"]]
GO.module[["darkturquoise"]]  # not that high association 

# flower 
GO.module[["darkorange"]] 

# leaf, bolt, flower, width 
GO.module[["darkred"]] # not very high association with traits 
GO.module[["green"]]
GO.module[["ivory"]]  

GO.module[!is.na(GO.module)] %>% names()   
```

### Gene significance and module membership

genes with high GS and MM are more correlated with certain trait through expression, eQTL? GS is defined as the correlation between the gene and the trait, MM as the correlation beteween the module eigengene and the module gene expression profile. 

```{r}
load("~/505/network_analysis/output/blups.505.Rdata")
load("~/505/network_analysis/output/vstMat.505.WGCNA.t.Rdata")
load("~/505/network_analysis/output/MEs_505_signed.Rdata") 

modNames = substring(names(MEs), 3)

# calculate gene membership and membership significance  
geneModuleMembership = as.data.frame(cor(vstMat.505.WGCNA.t, MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples = nrow(vstMat.505.WGCNA.t)))
names(geneModuleMembership) = paste("MM", modNames, sep="") 
names(MMPvalue) = paste("p.MM", modNames, sep="")

# calculate gene significance and significance P-value 
geneTraitSignificance <- 
  lapply(colnames(blups.505), function(i) {
  tmp <- as.data.frame(cor(vstMat.505.WGCNA.t, blups.505[,i], use = "p"))
  names(tmp) = paste("GS.", i, sep="")
  tmp
}) 

names(geneTraitSignificance) <- colnames(blups.505)

GSPvalue <- 
  lapply(names(geneTraitSignificance), function(i) {
  tmp <- as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance[[i]]), nSamples = nrow(vstMat.505.WGCNA.t)))
  names(tmp) = paste("p.GS.", i, sep="")   
  tmp 
})    

# identify genes with high GS and MM for specific modules, which are highly correlated with certain trait, these genes might be key player for specific trait, eQTL...  

load("~/Reference/B.napus/BnapusAnnotation.Rdata")  
load("~/Reference/B.napus/napus_GO_combined.Rdata") 
BnapusAnnotation <- 
BnapusAnnotation %>% 
  left_join(napus_GO_combined, by=c("name"="gene_ID")) # use annotation file from B.napus 

########################### down to here, start from here tomorrow. 
### summary output of genes with high GS and MM: combine module correlation information and gene annotation to check the result 

# import B.napus annotation file 
probes = colnames(vstMat.505.WGCNA.t)
probes2annot = match(probes, BnapusAnnotation$name)
# The following is the number or probes without annotation:
sum(is.na(probes2annot))
# Should return 0. 


# Create the starting data frame
geneInfo0 = data.frame(substanceBXH = probes,
                        geneSymbol = BnapusAnnotation$name[probes2annot],
                        LocusLinkID = BnapusAnnotation$At_symbol[probes2annot],
                        moduleColor = moduleColors,
                        geneTraitSignificance,
                        GSPvalue) 
# Order modules by their significance for weight
modOrder = order(-abs(cor(MEs, as.numeric(blups.505$Erucic_acid_year1), use = "p"))); 
# Add module membership information in the chosen order

for (mod in 1:ncol(geneModuleMembership))
{
  oldNames = names(geneInfo0)
  geneInfo0 = data.frame(geneInfo0, geneModuleMembership[, modOrder[mod]],
  MMPvalue[, modOrder[mod]]);
  names(geneInfo0) = c(oldNames, paste("MM.", modNames[modOrder[mod]], sep=""),
  paste("p.MM.", modNames[modOrder[mod]], sep=""))
}

# Order the genes in the geneInfo variable first by module color, then by geneTraitSignificance
geneOrder = order(geneInfo0$moduleColor, -abs(geneInfo0$GS.Erucic_acid));
geneInfo = geneInfo0[geneOrder, ] 
geneInfo %>% colnames()    
```


