---
title: "505_network_WGCNA_3"
output: html_document
---

Purpose of this script is to do eigene network analysis 

### 0) 

check expression profile of genes in important modules as well as their eigengenes
```{r}
load("~/505/network_analysis/input/module.gene.GO.Rdata")
GO.module %>% length() # 39 
gene.module %>% length() # 39 

# FA trait
GO.module[["salmon"]]
GO.module[["yellow"]]
GO.module[["turquoise"]]
GO.module[["darkturquoise"]]  # not that high association 

# flower 
GO.module[["darkorange"]] 

# leaf, bolt, flower, width 
GO.module[["darkred"]] # not very high association with traits 
GO.module[["green"]]
GO.module[["ivory"]]  

load("~/505/network_analysis/output/vstMat.505.batch.corrected.largeVar.Rdata")  
vstMat.505.batch.corrected.largeVar %>% dim() # 13454   131

load("~/505/network_analysis/output/MEs_505_signed.Rdata")
MEs$ID <-colnames(vstMat.505.batch.corrected.largeVar)
save(MEs, file = "~/505/network_analysis/output/MEs_with_ID.Rdata")

sig_modules <- c("salmon", "yellow", "turquoise", "darkturquoise", "darkorange","darkred", "green", "ivory")

for(i in sig_modules){ 

filename <- paste("~/505/network_analysis/output/module_gene_", i, ".png",  sep = "")
png(filename = filename, width=16, height=9, units="in", res=300)

vstMat.505.batch.corrected.largeVar[moduleColors == i,] %>% 
  pheatmap(  
     cluster_rows = F, 
     cluster_cols = F,
     clustering_method = "average", 
     # cellheight = 10,
     # cellwidth = 5,
     border_color=NA, 
     fontsize_col = 8, 
     legend = T, 
     main = i, 
     show_rownames = F)

dev.off()
}

sig_modules.eigene <- paste("ME", sig_modules, sep = "")

expression.eigengene <- 
lapply(sig_modules.eigene, function(i){
MEs %>% melt() %>% 
  filter(variable == i) %>% 
  ggplot() + 
  geom_bar(aes(x = ID, y = value, fill = value < 0), stat = "identity") + 
  scale_fill_manual(values = c("TRUE" = "lightblue", "FALSE" = "orange")) + 
  theme(legend.position = "none") + 
  labs(x = "", title = i)
} 
)

plots <- expression.eigengene

paths <- stringr::str_c(sig_modules.eigene, ".pdf") 
paths 
pwalk(list(paths, plots), ggsave, path = "~/505/network_analysis/output/", width = 15, height = 7) # save on whitney   
```

### 1) 

GWAS on eigenegenes genes of important modules 
```{r}
load(file = "~/505/network_analysis/output/MEs_with_ID.Rdata")

MEs <- 
MEs %>% 
  mutate(Taxa = ID) %>% 
  dplyr::select(Taxa, MEgreen:MEturquoise) 

library(multtest)
library(gplots)
library(genetics)
library(EMMREML)
library(compiler) #this library is already installed in R
library("scatterplot3d")

source("http://zzlab.net/GAPIT/gapit_functions.txt")
source("http://zzlab.net/GAPIT/emma.txt")

setwd("~/505/network_analysis/output/GWAS_eigengene/")

# genotype data
geno_505_hmp <- read.table("~/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.hmp.txt", head=FALSE)
geno_505_hmp$V3 <- as.numeric(as.factor(geno_505_hmp$V3))
geno_505_hmp[1,3] <- "chrom"

MEs[,c("Taxa")]
colnames(MEs)[40] <- "Taxa"
pheno_data_505 <- MEs

myY <- pheno_data_505 
myG <- geno_505_hmp 

# run GAPIT
myGAPIT <- GAPIT(
Y=myY,
G=myG,
PCA.total=0,
Geno.View.output=FALSE,
PCA.View.output=FALSE,
Model.selection = TRUE
)  
```

### 2) 

GWAS on all traits 
```{r}
load("~/505/network_analysis/output/pheno_data_505.Rdata")

library(multtest)
library(gplots)
library(genetics)
library(EMMREML)
library(compiler) #this library is already installed in R
library("scatterplot3d")

source("http://zzlab.net/GAPIT/gapit_functions.txt")
source("http://zzlab.net/GAPIT/emma.txt")

setwd("~/505/network_analysis/output/GWAS_trait/")

# genotype data
geno_505_hmp <- read.table("~/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.hmp.txt", head=FALSE)
geno_505_hmp$V3 <- as.numeric(as.factor(geno_505_hmp$V3))
geno_505_hmp[1,3] <- "chrom"

colnames(pheno_data_505) 
geno_505_hmp[1,] 

geno_505_hmp[1,] <- gsub("\\_|\\-", "", geno_505_hmp[1,])
geno_505_hmp[1,] <- gsub("(505)(K)([[:digit:]]+)", "ID_\\1_\\2\\3", geno_505_hmp[1,])

colnames(pheno_data_505)[1] <- "Taxa"

myY <- pheno_data_505 
myG <- geno_505_hmp 

# run GAPIT
myGAPIT <- GAPIT(
Y=myY,
G=myG,
PCA.total=0,
Geno.View.output=FALSE,
PCA.View.output=FALSE,
Model.selection = TRUE
)    
```

3) look for overlaps between GWAS signals for eigene genes and traits 
```{r}

```

