---
title: '505_late_silique'
author: "Ruijuan Li"
date: "4/7/2017"
output: 
  html_document: 
    keep_md: yes 
--- 

### data description 
```{r}
# RNA seq data from late silique, batch C: 47 lines; batch D: 41 lines, batch E: 43 lines
# 505 collection in total has 130 lines, two lines are duplicated in batch D & batch E 
```

# preprocess 
```{r}
# 1) download 

# 2) trimming 
# batch C: https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/trimming.slurm
# batch D: https://github.com/leejimmy93/KIAT_whitney/blob/master/505/trimming_batchD.sh

# 3) mapping using STAR 
# batch C: https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/mapping_array.slurm (sometimes there is error: could not load STAR module, for this case, load STAR module in the environment before sbatch) 
```

# analyze mapping result (QC & mapping rate)
```{r}
# star_extract_stats.sh get star mapping stats 
```

# mapping result analysis 
```{r, results='hide'}
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")

# sample description 
sample_des_c <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_c.csv", header = T) 
sample_des_c$batch <- rep("C", nrow(sample_des_c))
sample_des_c_sub <- sample_des_c[,c("Sample.ID","TotalReads","batch")]
sample_des_d <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_d.csv", header = T)
sample_des_d$batch <- rep("D", nrow(sample_des_d))
sample_des_d_sub <- sample_des_d[,c("Sample.ID","TotalReads","batch")]
head(sample_des_d_sub)

# mapping result
mapping_result <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/Star_Stats.tab", header = T)
head(mapping_result)
mapping_result$Sample.ID <- gsub("(Sample_)([[:print:]])", "\\2", mapping_result$Sample) 

# merge sample description & mapping result for batch A and batch B 
batch.c.stats <- merge(sample_des_c_sub, mapping_result, by="Sample.ID")
batch.d.stats <- merge(sample_des_d_sub, mapping_result, by="Sample.ID")
mapping_result_505 <- rbind(batch.c.stats, batch.d.stats)

mapping_result_505$batch
mapping_result_505$Number_Input_Reads/(mapping_result_505$TotalReads/2)

### calculate average 
mean(mapping_result_505[mapping_result_505$batch=="C",]$TotalReads/2) # 29598998
mean(mapping_result_505[mapping_result_505$batch=="D",]$TotalReads/2) # 27072598 

# number of total reads 
library(ggplot2) 
p.total.reads <- ggplot(data=mapping_result_505)
p.total.reads <- p.total.reads + geom_bar(aes(x=reorder(Sample.ID, TotalReads/2), y=TotalReads/2, fill=batch), stat = "identity")
p.total.reads <- p.total.reads + labs(list(title = "", x = "", y = "Number of raw reads"))
p.total.reads <- p.total.reads + theme(axis.text.x = element_text(angle = 90, size = 8)) 
p.total.reads   
# ggsave(p.total.reads, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/total_reads.png", height = 8, width = 11) 

# look at HQ, mapped, multimapped in detail for each sample (memory too low, cannot view histogram of total reads...)
colnames(mapping_result_505)
mapping_result_505_sub <- data.frame(Sample.ID = mapping_result_505$Sample.ID, 
                                     batch=mapping_result_505$batch,
                                    trimmed.off = mapping_result_505$TotalReads/2 - mapping_result_505$Number_Input_Reads,
                                    unmapped = mapping_result_505$Number_Input_Reads * (as.numeric(mapping_result_505$Percent_Unmapped_Mismatches + mapping_result_505$Percent_Unmapped_Other + mapping_result_505$Percent_Unmapped_Too_Short)/100), 
                                    multi_too_many_mapped = mapping_result_505$Number_Multi_Mapped + mapping_result_505$Number_Too_Many_Multi_Mapped, 
                                    unqiuemapped = mapping_result_505$Number_Unique_Mapped
) 

mapping_result_505_sub 
dim(mapping_result_505_sub) # 88 6 

mapping_result_505_sub.long <- melt(mapping_result_505_sub, id.vars = c("Sample.ID", "batch"))
mapping_result_505_sub.long.c <- mapping_result_505_sub.long[mapping_result_505_sub.long$batch=="C",]
mapping_result_505_sub.long.d <- mapping_result_505_sub.long[mapping_result_505_sub.long$batch=="D",]

pl.mapping.indi.c <- ggplot(data = mapping_result_505_sub.long.c)
pl.mapping.indi.c <- pl.mapping.indi.c + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.c <- pl.mapping.indi.c + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_text(angle = 90, size = 8)) + ylim(0, 4.5e+7) 
# p.total.reads <- p.total.reads + theme(axis.text.x = element_text(angle = 90, size = 8)) 
pl.mapping.indi.c

pl.mapping.indi.d <- ggplot(data = mapping_result_505_sub.long.d)
pl.mapping.indi.d <- pl.mapping.indi.d + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.d <- pl.mapping.indi.d + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_text(angle = 90, size = 8)) + ylim(0, 4.5e+7)
pl.mapping.indi.d 
 
library(cowplot)
pl.mapping.indi <- plot_grid(
  pl.mapping.indi.c +labs(title="mapping result of batch c")+theme(legend.position = "none"),
  pl.mapping.indi.d +labs(title="mapping result of batch d") +theme(legend.position = "none"),
  align = 'vh', ncol=2, nrow = 1, labels=c("",""))

get_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}

legend <- get_legend(pl.mapping.indi.c)

# pl.mapping.indi <- plot_grid(arrangeGrob(pl.mapping.indi.a + theme(legend.position="none") + labs(title="mapping result of batch a"),
#                          pl.mapping.indi.b + theme(legend.position="none") + labs(title="mapping result of batch b"),
#                          nrow=1),
#              legend, nrow=2,heights=c(10, 1))

pl.mapping.indi.final <- plot_grid(pl.mapping.indi, legend, rel_widths = c(3, 1))

pl.mapping.indi.final 
save_plot("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/pl_mapping_indi.png", pl.mapping.indi.final, base_aspect_ratio = 1, base_width = 12)  


# look at histogram of HQ reads 
colnames(mapping_result_505)
p.HQ.perc <- ggplot(data=mapping_result_505)
p.HQ.perc <- p.HQ.perc + geom_histogram(aes(x=mapping_result_505$Number_Input_Reads/(mapping_result_505$TotalReads/2), fill=batch), binwidth = 0.01) 
p.HQ.perc <- p.HQ.perc + labs(list(title = "histogram of HQ reads percentage", x = "Percent of high qulity reads", y = "number of samples"))
p.HQ.perc 
# ggsave(p.HQ.perc, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/HQ_hist.png", height = 8, width = 11) 

# histogram of unqiue mapped reads 
p.uniq.mapping <- ggplot(data=mapping_result_505)
p.uniq.mapping <- p.uniq.mapping + geom_histogram(aes(x=Percent_Unique_Mapped, fill=batch), binwidth = 2.5) 
p.uniq.mapping <- p.uniq.mapping + labs(list(title = "histogram of uniquely mapped reads percentage", x = "Percent of uniquely mapped reads", y = "number of samples"))
p.uniq.mapping  
# ggsave(p.uniq.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/unique_mapped.png", height = 8, width = 11) 

# hisogram of multimapped & too many reads 
p.multi.too.many.mapping <- ggplot(data=mapping_result_505)
p.multi.too.many.mapping <- p.multi.too.many.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Multi_Mapped+mapping_result_505$Percent_Too_Many_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.multi.too.many.mapping <- p.multi.too.many.mapping + labs(list(title = "", x = "Percent of multi & too many mapped reads", y = "number of samples")) 

# histogram of multi mapped 
p.multi.mapping <- ggplot(data=mapping_result_505)
p.multi.mapping <- p.multi.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.multi.mapping <- p.multi.mapping + labs(list(title = "histogram of multi mapped reads percentage", x = "Percent of multi mapped reads", y = "number of samples"))
p.multi.mapping
# ggsave(p.multi.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/multi_mapped.png", height = 8, width = 11) 

# histogram of too many mapped 
p.toomany.mapping <- ggplot(data=mapping_result_505)
p.toomany.mapping <- p.toomany.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Too_Many_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.toomany.mapping <- p.toomany.mapping + labs(list(title = "histogram of too many mapped reads percentage", x = "Percent of toomany mapped reads", y = "number of samples"))
p.toomany.mapping
# ggsave(p.toomany.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/toomany_mapped.png", height = 8, width = 11) 

# histogram of unmapped 
p.un.mapping <- ggplot(data=mapping_result_505)
p.un.mapping <- p.un.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Unmapped_Mismatches+mapping_result_505$Percent_Unmapped_Other+mapping_result_505$Percent_Unmapped_Too_Short, fill=batch), binwidth = 2.5) 
p.un.mapping <- p.un.mapping + labs(list(title = "histogram of unmapped reads percentage", x = "Percent of unmapped reads", y = "number of samples"))
p.un.mapping
# ggsave(p.un.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/un_mapped.png", height = 8, width = 11)

# mapping stats 
colnames(mapping_result_505)
# average percentage of HQ, unique, multi, toomany, unmapped 
average.mapping <- aggregate(mapping_result_505[,c("TotalReads","Number_Input_Reads", "Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short")], list(mapping_result_505$batch), mean)

rownames(average.mapping) <- average.mapping$Group.1
average.mapping$Percent_HQ <- average.mapping$Number_Input_Reads/(average.mapping$TotalReads/2)*100
average.mapping

average.mapping.long <- melt(average.mapping[,c("Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short","Percent_HQ","Group.1")], id.vars = "Group.1")

average.mapping.long

p.average.mapping <- ggplot(data = average.mapping.long)
p.average.mapping <- p.average.mapping + geom_bar(aes(x=variable, y=value, fill=variable), stat = "identity")
p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.mapping <- p.average.mapping + labs(list(title = "mapping stats of 505", x = "", y = "percentage"))
p.average.mapping <- p.average.mapping + theme(axis.text.x = element_text(angle = 90), legend.position = "none")
p.average.mapping <- p.average.mapping + geom_text(data = average.mapping.long, aes(x=variable, y=value, label=factor(round(value,0)))) 
p.average.mapping
# ggsave(p.average.mapping, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_mapping.png", height = 6, width = 9) 

### split this figure to HQ ratio & mapping ratio (two plots)
average.mapping.long.HQ <- 
  average.mapping.long[average.mapping.long$variable=="Percent_HQ",]

average.mapping.long.HQ

p.average.HQ <- ggplot(data = average.mapping.long.HQ)
p.average.HQ <- p.average.HQ + geom_bar(aes(x=Group.1, y=value, fill=variable), stat = "identity")
# p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.HQ <- p.average.HQ + labs(list(title = "", x = "", y = "percentage of total raw reads"))
p.average.HQ <- p.average.HQ + theme(legend.position = "none")
p.average.HQ <- p.average.HQ + geom_text(data = average.mapping.long.HQ, aes(x=Group.1, y=value, label=factor(round(value,0)))) 
p.average.HQ

ggsave(p.average.HQ, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_HQ_late_silique.png", height = 4, width = 2) 

average.mapping.long.mapping <- 
  average.mapping.long[average.mapping.long$variable!="Percent_HQ",]

p.average.mapping <- ggplot(data = average.mapping.long.mapping)
p.average.mapping <- p.average.mapping + geom_bar(aes(x=variable, y=value, fill=variable), stat = "identity")
p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.mapping <- p.average.mapping + labs(list(title = "", x = "", y = "percentage of HQ reads"))
p.average.mapping <- p.average.mapping + theme(axis.text.x = element_text(angle = 90), legend.position = "none")
p.average.mapping <- p.average.mapping + geom_text(data = average.mapping.long.mapping, aes(x=variable, y=value, label=factor(round(value,0)))) 
p.average.mapping 

ggsave(p.average.mapping, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_mapping_late_silique.png", height = 6, width = 5) 

```
 
# SNP calling (wait untill I get QC and mapping rate)
```{r} 
# 1) prepare data for SNP calling 
# Prep4Freebayes.sh (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/Prep4Freebayes.sh)

# 2) split bam file by chromosome 
# Bam_Split_By_Chrom.sh (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/Bam_Split_By_Chrom.sh) 

# 3) SNP calling 
# Freebayes_Array_no_pop_late_silique.slurm (https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/Freebayes_Array_no_pop_late_silique.slurm) 

# scp to whitney 
# scp * ruijuanli@whitney.plb.ucdavis.edu:/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_no_pop 

# 4) compress & index 
# compress_index_vcf.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/compress_index_vcf.sh)

# 5) combine into one big vcf file 
# concat.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/concat.sh)
# output: 505.vcf.gz  

# 6) filter 
# filter.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/filter.sh)
# output: 505.filtered.vcf.gz 

# 7) calculate a bunch of stats (haven't done this step yet...)
# ./calc.sh 505_filtered.vcf.gz 505_filtered 
# https://github.com/leejimmy93/KIAT_whitney/blob/master/505/calc.sh 
# output: many stats start with 505_filtered
```

# stats, wait for later 
```{r}
# mean read depth before read depth trimming 
mean.depth <- read.table("~/505/vcf_late_silique_131_sample/combined/505.ldepth.mean", header=T)
head(mean.depth)
mean(mean.depth$MEAN_DEPTH) # 15.58821 

png("~/505/output/figure/read.depth.png", width=4, height=3, units="in", res=300) 
hist(log10(mean.depth$MEAN_DEPTH), xlab = "log10(read depth)", ylab="Number of SNPs", main="read depth", breaks=50)
abline(v=c(log10(5),log10(500)), col="red", lwd=2)
dev.off()

dim(mean.depth) # 434331      4 
# decide to use 5 to 500 as the threshold for depth

# output: many stats start with 505_filtered 
```

```{r}
# chromosome distribution 
geno_505_hmp.late_silique <- read.table("~/", head=T)
head(geno_505_hmp.late_silique)
colnames(geno_505_hmp.late_silique)
nrow(geno_505_hmp.late_silique) # 242314  

distri <- geno_505_hmp.late_silique[,c("chrom", "pos")] 
head(distri)
colnames(distri) <- c("CHROM", "POS")

class(distri)
data <- distri
head(data)

# below the block of code should go into function... 
chr_ID <- unique(data$CHROM)
number.unique <- list() 

for (chr in chr_ID) {
  number.unique[chr] <- sum(sum(data$CHROM==chr))
  print(number.unique[chr])
}

SNP_chr.unique <- as.data.frame(t(as.data.frame(number.unique)))
SNP_chr.unique$ID <- rownames(SNP_chr.unique)
head(SNP_chr.unique)
# split to main & random chromosomes 
# main 
SNP_chr_main.unique <- SNP_chr.unique[!grepl("RANDOM", SNP_chr.unique$ID),]  
head(SNP_chr_main.unique)
SNP_chr_main.unique$subgenome <- gsub("(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\1", SNP_chr_main.unique$ID)   
SNP_chr_main.unique$chr_ID <- gsub("(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", SNP_chr_main.unique$ID) 

# do a little calculation: how many SNPs on main & random chromosomes? 
sum(SNP_chr_main.unique$V1) # 200897        
head(SNP_chr_main.unique)     

sum(SNP_chr_main.unique[SNP_chr_main.unique$subgenome=="A",]$V1)/sum(SNP_chr_main.unique$V1) # 56% 

# make a plot to see how many SNPs on each chromosome 
library(ggplot2)
pl.SNP.main.1.unique <- ggplot(data = SNP_chr_main.unique) 
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + geom_bar(aes(x=factor(chr_ID), y = V1, fill=subgenome), stat = "identity")
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + facet_wrap(~subgenome, ncol = 1)
# pl.SNP.main.1 <- pl.SNP.main.1 + geom_text(aes(x=factor(chr_ID), y = number, label=factor(number), size=1)) 
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + theme(legend.position = "none")
pl.SNP.main.1.unique 
ggsave("~/505/output/figure/late_silique/SNP.main.png", width = 9, height = 7) 

########## Per Mb ... 
head(data)
data <- data[!grepl("RANDOM", data$CHROM),]  
data$subgenome <-  gsub("(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\1", data$CHROM) 
data$Chr_ID <- gsub("(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", data$CHROM) 
head(data) 

# write.csv(vcf.Ae.Ol.intersect.df.2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/vcf.Ae.Ol.intersect.df.2.csv") # sort in excel manually 

# cf.Ae.Ol.intersect.df.2.sorted <-  read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/vcf.Ae.Ol.intersect.df.2.csv")
# head(vcf.Ae.Ol.intersect.df.2.sorted) 

library(ggplot2)
pl.SNP.main.2.unique <- ggplot(data = data)
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + geom_histogram(aes(x=POS, fill=subgenome), binwidth = 1000000) 
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + facet_grid(Chr_ID ~subgenome)
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + theme(legend.position = "none")
pl.SNP.main.2.unique 

ggsave("~/505/output/figure/late_silique/SNP.main.PerMb.png", width = 9, height = 7) 
 
# calculate SNPs per Mb number 
chrom.length <- read.table("~/Reference/B.napus/Brassica_napus_v4.1.length", header=F)
chrom.length
nrow(geno_505_hmp.late_silique)/sum(chrom.length$V2)*1000000 

```

# cluster analysis   
```{r} 
# reform_505_late_silique.R to reform vcf file for GWAS 
# output: late_silique_505.2 

# cluster analysis 
load("~/505/output/late_silique/late_silique_505.2.Rdata")
class(late_silique_505.2.t)
dim(late_silique_505.2.t) #  88 347282 
late_silique_505.2.t[1:10, 1:10]
# late_silique_505.2.t <- as.data.frame(late_silique_505.2.t)
# geno.numeric <- data.matrix(late_silique_505.2.t)
# geno.numeric[1:10, 1:10]

geno.numeric <- apply(late_silique_505.2.t, 2, function(x) as.integer(x))
rnames <- rownames(late_silique_505.2.t)
rownames(geno.numeric) <- rnames

geno.numeric.df <- as.data.frame(geno.numeric)

geno.numeric.df.2 <- geno.numeric.df[which(!rownames(geno.numeric.df) %in% "X505.K88"),]
### 
geno.numeric.2 <- as.matrix(geno.numeric.df.2)

genDist <- as.matrix(dist(geno.numeric.2))

#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions
dim(geno.mds)
plot(geno.mds)  

############  
save(geno.numeric, file="~/505/output/late_silique_505_geno_numeric.Rdata") 

## add country of origin, winter or spring type to the figure ... 
sample_des_c_revised <- read.csv("~/505/data/phenotype/batch_c_revised.csv", header=T)
sample_des_c_sub <- sample_des_c_revised[,c("Sample.ID", "Name","Sample.Description.Origin", "Sample.Description.Type")]
sample_des_d_revised <- read.csv("~/505/data/phenotype/batch_d_revised.csv", header=T)
sample_des_d_sub <- sample_des_d_revised[,c("Sample.ID", "Name", "Sample.Description.Origin", "Sample.Description.Type")]
head(sample_des_c_sub)
sample_des_c_sub$Sample.ID
sample_des_d_sub$Sample.ID

sample_des_revised <- rbind(sample_des_c_sub, sample_des_d_sub)
sample_des_revised <- sample_des_revised[-c(48, 90:93),]
sample_des_revised$Sample.ID
save(sample_des_revised, file="~/505/output/late_silique/sample_des_revised.Rdata")

# MDS result 
# reformat MDS result sample ID (X remove & . to -)
geno.mds$Sample.ID <- gsub("\\.","\\-", rownames(geno.mds)) 
geno.mds$Sample.ID <- gsub("X", "", geno.mds$Sample.ID)
class(geno.mds$Sample.ID)

geno.mds$Sample.ID
sample_des_revised$Sample.ID

# merge 
geno.mds.all.des <- merge(geno.mds, sample_des_revised, by="Sample.ID") 
colnames(geno.mds.all.des)[1] <- "Taxa"
head(SN_Taxa_131_sample)
geno.mds.all.des.2 <- merge(geno.mds.all.des, SN_Taxa_131_sample, by="Taxa")
head(geno.mds.all.des.2)
colnames(geno.mds.all.des.2)[5] <- "location_of_origin"

library(ggrepel)
set.seed(111) 

p.mds.country <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country <- p.mds.country + geom_point(size=1) 
p.mds.country <- p.mds.country + scale_color_brewer(type="qual",palette="Set1")
p.mds.country <- p.mds.country + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country 
ggsave(p.mds.country, filename="~/505/output/figure/late_silique/p.mds.country.png", height=20, width=20)

p.mds.country.2 <- ggplot(data=geno.mds.all.des.2, aes(V1, V2, color=location_of_origin)) + theme_bw()
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
# p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + scale_color_manual(values = c("lightblue", "orange", "red", "green", "blue", "grey", "purple", "pink", "darkred", "black"))
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(SN)))
p.mds.country.2 <- p.mds.country.2 + theme(legend.position = c(0.9, 0.2), legend.background = element_rect(fill="lightgrey", size=0.5, linetype="solid"), axis.title.x=element_blank(), axis.title.y =element_blank())
p.mds.country.2 <- p.mds.country.2 + labs(title="MDS of 505 collection") + theme(plot.title = element_text(hjust = 0.5, size = 16, face="bold")) 
p.mds.country.2
ggsave(p.mds.country.2, filename="~/505/output/figure/leaf/p.mds.country.2.png", height=10, width=10) 
# ggsave(p.mds.country.2, filename="~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/late_silique/p.mds.country.2.png", height=10, width=10)

p.mds.country.2 <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country.2
ggsave(p.mds.country.2, filename="~/505/output/figure/late_silique/p.mds.country.2.png", height=20, width=20)
 
# cluster for only the common lines 
load("~/505/output/common.lines.Rdata")
common.lines$Taxa.y

rownames(geno.numeric.df.2) 
# change rownames 
rownames(geno.numeric.df.2) <- gsub("\\.","\\-", rownames(geno.numeric.df.2))
rownames(geno.numeric.df.2) <- gsub("X", "", rownames(geno.numeric.df.2))

common.lines$Taxa.y <- as.character(common.lines$Taxa.y)
rownames(geno.numeric.df.2) 
common.lines$Taxa.y <- gsub("\\_","\\-", common.lines$Taxa.y)

geno.numeric.df.common <- geno.numeric.df.2[rownames(geno.numeric.df.2) %in% common.lines$Taxa.y,]
dim(geno.numeric.df.common)  
# check to make sure I get the right subset 

identical(rownames(geno.numeric.df.common)[order(rownames(geno.numeric.df.common))], common.lines$Taxa.y[order(common.lines$Taxa.y)])
geno.numeric.common <- as.matrix(geno.numeric.df.common)

genDist <- as.matrix(dist(geno.numeric.common))

#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions
dim(geno.mds)
plot(geno.mds)  

## add country of origin, winter or spring type to the figure ... 
sample_des_c_revised <- read.csv("~/505/data/phenotype/batch_c_revised.csv", header=T)
sample_des_c_sub <- sample_des_c_revised[,c("Sample.ID", "Name","Sample.Description.Origin", "Sample.Description.Type")]
sample_des_d_revised <- read.csv("~/505/data/phenotype/batch_d_revised.csv", header=T)
sample_des_d_sub <- sample_des_d_revised[,c("Sample.ID", "Name", "Sample.Description.Origin", "Sample.Description.Type")]
head(sample_des_c_sub)
sample_des_c_sub$Sample.ID
sample_des_d_sub$Sample.ID

sample_des_revised <- rbind(sample_des_c_sub, sample_des_d_sub)
sample_des_revised <- sample_des_revised[-c(48, 90:93),]
sample_des_revised$Sample.ID
save(sample_des_revised, file="~/505/output/late_silique/sample_des_revised.Rdata")

# MDS result 
# reformat MDS result sample ID (X remove & . to -)
# geno.mds$Sample.ID <- gsub("\\.","\\-", rownames(geno.mds)) 
# geno.mds$Sample.ID <- gsub("X", "", geno.mds$Sample.ID)
class(geno.mds$Sample.ID) 
geno.mds

geno.mds$Sample.ID
sample_des_revised$Sample.ID <- gsub("\\_", "\\-", sample_des_revised$Sample.ID)

# merge 
geno.mds.all.des <- merge(geno.mds, sample_des_revised, by="Sample.ID")
head(geno.mds.all.des)  
dim(geno.mds.all.des)

library(ggrepel)
set.seed(111) 

# p.mds.country <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
# p.mds.country <- p.mds.country + geom_point(size=1) 
# p.mds.country <- p.mds.country + scale_color_brewer(type="qual",palette="Set1")
# p.mds.country <- p.mds.country + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
# p.mds.country 
# ggsave(p.mds.country, filename="~/505/output/figure/late_silique/p.mds.country.png", height=20, width=20)

p.mds.country.2 <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country.2
ggsave(p.mds.country.2, filename="~/505/output/figure/late_silique/p.mds.country.2.common.png", height=20, width=20)

p.mds.country.2 <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country.2
ggsave(p.mds.country.2, filename="~/505/output/figure/late_silique/p.mds.country.2.png", height=20, width=20)


``` 

# GWAS
```{r}
# reformat for GWAS 
# get hapmap file using tassel 

######## use tassel to get hapmap format ########### 
# run_pipeline.pl -SortGenotypeFilePlugin -inputFile 505_filtered_GQ.vcf.gz -outputFile 505_filtered_GQ_sorted -fileType VCF
# run_pipeline.pl -Xmx5g -fork1 -vcf 505_filtered_GQ_sorted.vcf -export -exportType Hapmap -runfork1 
# output: 505_filtered_GQ_sorted.hmp.txt  
# edit output file to remove "#" after rs & assembly  
####################################################  
```

# GWAS run
```{r}
# GWAS_505_late_silique.R  
```

# GWAS result plot 
```{r} 
library("qqman")
Erucic_acid <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT/late_silique/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT/late_silique/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT/late_silique/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/leaf/Erucic_acid.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
    ) 
dev.off() 
```

# histogram of phenotype data 
```{r}
load("~/505/output/pheno_data_505_late_silique.R")
pheno_data_505

pheno_data_505_melt <- melt(pheno_data_505[,2:4])

# plot 
library(ggplot2)

P.oil.late.silique <- ggplot(data=pheno_data_505_melt)
P.oil.late.silique <- P.oil.late.silique + geom_histogram(aes(value, fill=variable))
P.oil.late.silique <- P.oil.late.silique + facet_wrap(~variable, ncol=1)
P.oil.late.silique <- P.oil.late.silique + labs(list(x="value", y="Number of varieties"))
P.oil.late.silique <- P.oil.late.silique + theme(legend.position = "none") 

P.oil.late.silique 
ggsave(P.oil, filename="~/505/output/figure/late_silique/P.oil.png", width=6, height=10)

# do seperate graph for unique lines 
common.lines
pheno_data_505[pheno_data_505$Taxa %in% common.lines$Taxa.y,]

pheno.unique.late.silique <- pheno_data_505[which(!(pheno_data_505$Taxa %in% common.lines$Taxa.y)),]

pheno_data_505_melt_unique <- melt(pheno.unique.late.silique[,2:4])

P.oil <- ggplot(data=pheno_data_505_melt_unique)
P.oil <- P.oil + geom_histogram(aes(value, fill=variable))
P.oil <- P.oil + facet_wrap(~variable, ncol=1)
P.oil <- P.oil + labs(list(x="value", y="Number of varieties"))
P.oil <- P.oil + theme(legend.position = "none") 
P.oil

ggsave(P.oil, filename="~/505/output/figure/leaf/P.oil.late.silique.unique.png", width=6, height=10)

```

# check whether there are genes that take most of the reads, so that later when make libraries, we can get rid of these highly expressed genes and genes with lower expression can be brought up in lower depth sequencing.    
```{r}
# get read count for all the libs, read count generated by STAR 
read_count_505 <- read.table("~/505/data/late_silique/read.count.505.tsv", header=T, row.names==1)
head(read_count_505)

dim(read_count_505)

# top 20 highly expressed genes in each 
most_highly_expressed_genes <- 
sapply(colnames(read_count_505), function(RNAlib){
  rownames(head(read_count_505[order(read_count_505[,RNAlib],decreasing = TRUE),], 20))
}
)

# get the percentage of reads for the 20 highly expressed genes from each lib 
most_highly_expressed_genes 

most_highly_expressed_genes.percentage <- 
sapply(colnames(most_highly_expressed_genes), function(RNAlib){
  head(read_count_505[order(read_count_505[,RNAlib],decreasing = TRUE),], 20)[,RNAlib]/sum(read_count_505[,RNAlib]) 
}
)

# sum of the percentage for the top 20 
png("~/505/output/figure/late_silique/highly_express_genes.png", width=8, height=6, units="in", res=300)
plot(colSums(most_highly_expressed_genes.percentage), type="h", xlab="library index") 
dev.off()

# how many genes are overlapped for all of the top 20 here 
most_highly_expressed_genes

Reduce(intersect,list(most_highly_expressed_genes[,1], most_highly_expressed_genes[,2], most_highly_expressed_genes[,3]), most_highly_expressed_genes[,4])

Reduce(intersect,list(most_highly_expressed_genes[,7], most_highly_expressed_genes[,8], most_highly_expressed_genes[,6]), most_highly_expressed_genes[,5])

library("reshape")
most_highly_expressed_genes.melt <- melt(most_highly_expressed_genes)
most_highly_expressed_genes.melt

write.table(most_highly_expressed_genes.melt, file="~/505/data/tmp")
# cat tmp | awk '{print $4}' | sort | uniq -c | sort -n > occurrence_gene.txt
occurrence.gene <- read.table("~/505/data/occurrence_gene.txt")
occurrence.gene
colnames(occurrence.gene) <- c("lib#", "geneID") 
```

# missing rate for the significant SNPs that present in late silique data, we didn't find them because they have high missing rate???  
```{r}
load("~/505/output/common.SNPs.Rdata")
common.SNPs

geno_505_hmp.late_silique <- read.table("~/505/vcf_late_silique_no_pop/combined/505_filtered_sorted.hmp.txt", head=T)
geno_505_hmp.late_silique[1:10, 1:10] 
colnames(geno_505_hmp.late_silique) 

geno_505_hmp.late_silique.18SNP <- geno_505_hmp.late_silique[which(geno_505_hmp.late_silique$rs %in% common.SNPs),]

 geno_505_hmp.leaf <- read.table("~/505/vcf_leaf_no_pop/combined/missing_0.5/505_filtered_sorted.hmp.txt", head=T)
geno_505_hmp.leaf[1:10, 1:10] 
colnames(geno_505_hmp.leaf) 
geno_505_hmp.leaf.18SNP <- geno_505_hmp.leaf[which(geno_505_hmp.leaf$rs %in% common.SNPs),]

dim(geno_505_hmp.leaf.18SNP)

missing.silique.18SNP <- c() 
  
for (i in 1:nrow(geno_505_hmp.late_silique.18SNP)){
  missing.silique.18SNP[i] <- length(which(geno_505_hmp.late_silique.18SNP[i,] == "N"))/88
}
missing.silique.18SNP.final <- data.frame(snpID = geno_505_hmp.late_silique.18SNP$rs, 
                                          silique.missing.rate = missing.silique.18SNP)

missing.leaf.18SNP <- c()

for (i in 1:nrow(geno_505_hmp.leaf.18SNP)){
  missing.leaf.18SNP[i] <- length(which(geno_505_hmp.leaf.18SNP[i,] == "N"))/94
}

missing.leaf.18SNP
missing.leaf.18SNP.final <- data.frame(snpID = geno_505_hmp.leaf.18SNP$rs, 
                                       leaf.missing_rate = missing.leaf.18SNP)

missing.leaf.18SNP.final
missing.silique.18SNP.final

tmp <- merge(missing.leaf.18SNP.final, missing.silique.18SNP.final, by = "snpID")  
tmp.melt <- melt(tmp)

p <- ggplot(data=tmp.melt)
p <- p + geom_bar(aes(x=snpID, y=value), stat="identity")
p <- p + facet_wrap(~variable) 
p

p <- p + geom_col(aes(x=snpID, y=value))
p <- p + facet_wrap(~variable)  
p 
```

# coverage for the two samples that are not clustering together for late silique & leaf 
```{r}
# two samples: 774-6C(S20)  NA259-DH02 

# save(sample_des_a_revised, file="~/505/output/sample_des_a_revised.Rdata")
# save(sample_des_b_revised, file="~/505/output/sample_des_b_revised.Rdata")
# save(sample_des_c_revised, file="~/505/output/sample_des_c_revised.Rdata")
# save(sample_des_d_revised, file="~/505/output/sample_des_a_revised.Rdata")

tmp1 <- sample_des_revised[sample_des_revised$Name == "774-6C(S20)",] 
tmp2 <- sample_des_revised[sample_des_revised$Name == "NA259-DH02",] 
tmp <- rbind(tmp1, tmp2)
tmp

sample_des_a_revised.sub2 <- sample_des_a_revised[,c("Name", "TotalReads")]
sample_des_b_revised.sub2 <- sample_des_b_revised[,c("Name", "TotalReads")]

sample_des_c_revised.sub2 <- sample_des_c_revised[,c("Name", "TotalReads")]
sample_des_d_revised.sub2 <- sample_des_d_revised[,c("Name", "TotalReads")]

leaf_sample <- rbind(sample_des_a_revised.sub2, sample_des_b_revised.sub2)
silique_sample <- rbind(sample_des_c_revised.sub2, sample_des_d_revised.sub2)

colnames(tmp3)
colnames(tmp)

merge(tmp, leaf_sample, by="Name")[c(1,3),]
merge(tmp, silique_sample, by="Name")[c(2,4),]
```

# batch E 
```{r}
# 1) download using download_arry.slurm
# 2) trimming https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/trimming.slurm
# pay attention to number of cores and time, sometimes jobs can be cancled due to limited memory or time 
# 3) mapping 
# https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/mapping_array.slurm
# 4) Prep4Freebayes 
# https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/mapping_array.slurm
# remember to import bamaddrg to the environment before running the script 
# 5) Split by chromosome
# https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/Bam_Split_By_Chrom.slurm

# 6) move file to the same directory where the previous 88 sample resides, however, because I sorted the bam file after extracting uniquely mapped reads for the 43 samples + sample-K88, Freebayes does not tolerate inconsistent order. Now re-run the 43 samples + sample-K88 using the same script for the 88 - sample-K88. 

# 7) SNP calling
# https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/Freebayes_Array_no_pop_late_silique.slurm
```

# analyze mapping result of batchE (including batch C + D)
```{r} 
# star_extract_stats.sh get star mapping stats 

source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R") 

# sample description 
sample_des_c <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_c.csv", header = T) 
sample_des_c$batch <- rep("C", nrow(sample_des_c))
sample_des_c_sub <- sample_des_c[,c("Sample.ID","TotalReads","batch")]
sample_des_d <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_d.csv", header = T)
sample_des_d$batch <- rep("D", nrow(sample_des_d))
sample_des_d_sub <- sample_des_d[,c("Sample.ID","TotalReads","batch")]
head(sample_des_d_sub)
sample_des_e <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_e.csv", header = T)
sample_des_e$batch <- rep("E", nrow(sample_des_e))
sample_des_e_sub <- sample_des_e[,c("Sample.ID","TotalReads","batch")]
head(sample_des_e_sub)

# mapping result
mapping_result <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/Star_Stats.tab", header = T)

mapping_result_e <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/Star_Stats_e.tab", header = T)

# mapping_result <- 
mapping_result <- rbind(mapping_result, mapping_result_e)
head(mapping_result)
mapping_result$Sample.ID <- gsub("(Sample_)([[:print:]])", "\\2", mapping_result$Sample) 

# merge sample description & mapping result for batch A and batch B 
batch.c.stats <- merge(sample_des_c_sub, mapping_result, by="Sample.ID")
batch.d.stats <- merge(sample_des_d_sub, mapping_result, by="Sample.ID")
batch.e.stats <- merge(sample_des_e_sub, mapping_result, by="Sample.ID")

mapping_result_505 <- rbind(batch.c.stats, batch.d.stats, batch.e.stats)

mapping_result_505$batch
mapping_result_505$Number_Input_Reads/(mapping_result_505$TotalReads/2)

### calculate average 
mean(mapping_result_505[mapping_result_505$batch=="C",]$TotalReads/2) # 29598998
mean(mapping_result_505[mapping_result_505$batch=="D",]$TotalReads/2) # 27072598 
mean(mapping_result_505[mapping_result_505$batch=="E",]$TotalReads/2) # 28474316 

# number of total reads 
library(ggplot2) 
p.total.reads <- ggplot(data=mapping_result_505)
p.total.reads <- p.total.reads + geom_bar(aes(x=reorder(Sample.ID, TotalReads/2), y=TotalReads/2, fill=batch), stat = "identity")
p.total.reads <- p.total.reads + labs(list(title = "", x = "", y = "Number of raw reads"))
p.total.reads <- p.total.reads + theme(axis.text.x = element_text(angle = 90, size = 8)) 
p.total.reads   
ggsave(p.total.reads, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/131_sample/total_reads_CDE.png", height = 8, width = 11) 

# look at HQ, mapped, multimapped in detail for each sample (memory too low, cannot view histogram of total reads...)
colnames(mapping_result_505)
mapping_result_505_sub <- data.frame(Sample.ID = mapping_result_505$Sample.ID, 
                                     batch=mapping_result_505$batch,
                                    trimmed.off = mapping_result_505$TotalReads/2 - mapping_result_505$Number_Input_Reads,
                                    unmapped = mapping_result_505$Number_Input_Reads * (as.numeric(mapping_result_505$Percent_Unmapped_Mismatches + mapping_result_505$Percent_Unmapped_Other + mapping_result_505$Percent_Unmapped_Too_Short)/100), 
                                    multi_too_many_mapped = mapping_result_505$Number_Multi_Mapped + mapping_result_505$Number_Too_Many_Multi_Mapped, 
                                    unqiuemapped = mapping_result_505$Number_Unique_Mapped
) 

mapping_result_505_sub 
dim(mapping_result_505_sub) # 131 6 

library(reshape2)
mapping_result_505_sub.long <- melt(mapping_result_505_sub, id.vars = c("Sample.ID", "batch"))
mapping_result_505_sub.long.c <- mapping_result_505_sub.long[mapping_result_505_sub.long$batch=="C",]
mapping_result_505_sub.long.d <- mapping_result_505_sub.long[mapping_result_505_sub.long$batch=="D",]
mapping_result_505_sub.long.e <- mapping_result_505_sub.long[mapping_result_505_sub.long$batch=="E",]


pl.mapping.indi.c <- ggplot(data = mapping_result_505_sub.long.c)
pl.mapping.indi.c <- pl.mapping.indi.c + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.c <- pl.mapping.indi.c + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_text(angle = 90, size = 4)) + ylim(0, 4.5e+7) 
# p.total.reads <- p.total.reads + theme(axis.text.x = element_text(angle = 90, size = 6)) 
pl.mapping.indi.c

pl.mapping.indi.d <- ggplot(data = mapping_result_505_sub.long.d)
pl.mapping.indi.d <- pl.mapping.indi.d + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.d <- pl.mapping.indi.d + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_text(angle = 90, size = 4)) + ylim(0, 4.5e+7)
pl.mapping.indi.d 

pl.mapping.indi.e <- ggplot(data = mapping_result_505_sub.long.e)
pl.mapping.indi.e <- pl.mapping.indi.e + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.e <- pl.mapping.indi.e + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_text(angle = 90, size = 4)) + ylim(0, 4.5e+7)
pl.mapping.indi.e 
 
library(cowplot)
pl.mapping.indi <- plot_grid(
  pl.mapping.indi.c +labs(title="mapping result of batch c")+theme(legend.position = "none"),
  pl.mapping.indi.d +labs(title="mapping result of batch d") +theme(legend.position = "none"),
  pl.mapping.indi.e +labs(title="mapping result of batch e") +theme(legend.position = "none"),
  align = 'vh', ncol=3, nrow = 1, labels=c("",""))

get_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}

legend <- get_legend(pl.mapping.indi.c)

pl.mapping.indi.final <- plot_grid(pl.mapping.indi, legend, rel_widths = c(3, 1))

pl.mapping.indi.final 
save_plot("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/131_sample/pl_mapping_indi_CDE.png", pl.mapping.indi.final, base_aspect_ratio = 1, base_width = 12)  


# look at histogram of HQ reads 
colnames(mapping_result_505)
p.HQ.perc <- ggplot(data=mapping_result_505)
p.HQ.perc <- p.HQ.perc + geom_histogram(aes(x=mapping_result_505$Number_Input_Reads/(mapping_result_505$TotalReads/2), fill=batch), binwidth = 0.01) 
p.HQ.perc <- p.HQ.perc + labs(list(title = "histogram of HQ reads percentage", x = "Percent of high qulity reads", y = "number of samples"))
p.HQ.perc 
# ggsave(p.HQ.perc, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/HQ_hist_CDE.png", height = 8, width = 11) 

# histogram of unqiue mapped reads 
p.uniq.mapping <- ggplot(data=mapping_result_505)
p.uniq.mapping <- p.uniq.mapping + geom_histogram(aes(x=Percent_Unique_Mapped, fill=batch), binwidth = 2.5) 
p.uniq.mapping <- p.uniq.mapping + labs(list(title = "histogram of uniquely mapped reads percentage", x = "Percent of uniquely mapped reads", y = "number of samples"))
p.uniq.mapping  
# ggsave(p.uniq.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/unique_mapped_CDE.png", height = 8, width = 11) 

# hisogram of multimapped & too many reads 
p.multi.too.many.mapping <- ggplot(data=mapping_result_505)
p.multi.too.many.mapping <- p.multi.too.many.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Multi_Mapped+mapping_result_505$Percent_Too_Many_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.multi.too.many.mapping <- p.multi.too.many.mapping + labs(list(title = "", x = "Percent of multi & too many mapped reads", y = "number of samples"))

# histogram of multi mapped 
p.multi.mapping <- ggplot(data=mapping_result_505)
p.multi.mapping <- p.multi.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.multi.mapping <- p.multi.mapping + labs(list(title = "histogram of multi mapped reads percentage", x = "Percent of multi mapped reads", y = "number of samples"))
p.multi.mapping
# ggsave(p.multi.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/multi_mapped_CDE.png", height = 8, width = 11) 

# histogram of too many mapped 
p.toomany.mapping <- ggplot(data=mapping_result_505)
p.toomany.mapping <- p.toomany.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Too_Many_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.toomany.mapping <- p.toomany.mapping + labs(list(title = "histogram of too many mapped reads percentage", x = "Percent of toomany mapped reads", y = "number of samples"))
p.toomany.mapping
# ggsave(p.toomany.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/toomany_mapped_CDE.png", height = 8, width = 11) 

# histogram of unmapped 
p.un.mapping <- ggplot(data=mapping_result_505)
p.un.mapping <- p.un.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Unmapped_Mismatches+mapping_result_505$Percent_Unmapped_Other+mapping_result_505$Percent_Unmapped_Too_Short, fill=batch), binwidth = 2.5) 
p.un.mapping <- p.un.mapping + labs(list(title = "histogram of unmapped reads percentage", x = "Percent of unmapped reads", y = "number of samples"))
p.un.mapping
# ggsave(p.un.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/un_mapped_CDE.png", height = 8, width = 11)

# mapping stats 
colnames(mapping_result_505)
# average percentage of HQ, unique, multi, toomany, unmapped 
average.mapping <- aggregate(mapping_result_505[,c("TotalReads","Number_Input_Reads", "Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short")], list(mapping_result_505$batch), mean)

rownames(average.mapping) <- average.mapping$Group.1
average.mapping$Percent_HQ <- average.mapping$Number_Input_Reads/(average.mapping$TotalReads/2)*100
average.mapping

average.mapping.long <- melt(average.mapping[,c("Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short","Percent_HQ","Group.1")], id.vars = "Group.1")

average.mapping.long

p.average.mapping <- ggplot(data = average.mapping.long)
p.average.mapping <- p.average.mapping + geom_bar(aes(x=variable, y=value, fill=variable), stat = "identity")
p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.mapping <- p.average.mapping + labs(list(title = "mapping stats of 505", x = "", y = "percentage"))
p.average.mapping <- p.average.mapping + theme(axis.text.x = element_text(angle = 90), legend.position = "none")
p.average.mapping <- p.average.mapping + geom_text(data = average.mapping.long, aes(x=variable, y=value, label=factor(round(value,0)))) 
p.average.mapping
# ggsave(p.average.mapping, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_mapping_CDE.png", height = 6, width = 9) 

### split this figure to HQ ratio & mapping ratio (two plots)
average.mapping.long.HQ <- 
  average.mapping.long[average.mapping.long$variable=="Percent_HQ",]

average.mapping.long.HQ

p.average.HQ <- ggplot(data = average.mapping.long.HQ)
p.average.HQ <- p.average.HQ + geom_bar(aes(x=Group.1, y=value, fill=variable), stat = "identity")
# p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.HQ <- p.average.HQ + labs(list(title = "", x = "", y = "percentage of total raw reads"))
p.average.HQ <- p.average.HQ + theme(legend.position = "none")
p.average.HQ <- p.average.HQ + geom_text(data = average.mapping.long.HQ, aes(x=Group.1, y=value, label=factor(round(value,0)))) 
p.average.HQ

# ggsave(p.average.HQ, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_HQ_late_silique_CDE.png", height = 4, width = 2) 

average.mapping.long.mapping <- 
  average.mapping.long[average.mapping.long$variable!="Percent_HQ",]

p.average.mapping <- ggplot(data = average.mapping.long.mapping)
p.average.mapping <- p.average.mapping + geom_bar(aes(x=variable, y=value, fill=variable), stat = "identity")
p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.mapping <- p.average.mapping + labs(list(title = "", x = "", y = "percentage of HQ reads"))
p.average.mapping <- p.average.mapping + theme(axis.text.x = element_text(angle = 90), legend.position = "none")
p.average.mapping <- p.average.mapping + geom_text(data = average.mapping.long.mapping, aes(x=variable, y=value, label=factor(round(value,0)))) 
p.average.mapping 

# ggsave(p.average.mapping, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_mapping_late_silique_CDE.png", height = 6, width = 5)  
```

# analyze data for GWAS 
```{r} 
# 1) scp from cluster to whitney 
# scp * ruijuanli@whitney.plb.ucdavis.edu:/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_no_pop

# compress & index 
# compress_index_vcf.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/compress_index_vcf.sh)

# combine into one big vcf file 
# concat.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/concat.sh)
# output: 505.vcf.gz  

# 6) filter 
# filter.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/filter.sh)
# output: 505.filtered.vcf.gz 
```

# stats, wait for late 
```{r}
#   
```

# chromosome distribution 
```{r}

```

# cluster analysis 
```{r} 
# reform_505_late_silique.R to reform vcf file for GWAS 
# output: late_silique_505.2 

# cluster analysis 
load("~/505/output/131_sample/late_silique_505.2.Rdata")
class(late_silique_505.2.t)
dim(late_silique_505.2.t) # 131 206222 
late_silique_505.2.t[1:10, 1:10]

geno.numeric <- apply(late_silique_505.2.t, 2, function(x) as.integer(x))
rnames <- rownames(late_silique_505.2.t)
rownames(geno.numeric) <- rnames

geno.numeric.df <- as.data.frame(geno.numeric)
geno.numeric.2 <- as.matrix(geno.numeric.df)
genDist <- as.matrix(dist(geno.numeric.2))

#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions
dim(geno.mds) # 131 2 
plot(geno.mds)  

############  
save(geno.numeric, file="~/505/output/131_sample/late_silique_505_geno_numeric.Rdata") 

## add country of origin, winter or spring type to the figure ... 
sample_des_c_revised <- read.csv("~/505/data/phenotype/batch_c_revised.csv", header=T, na.strings="")
sample_des_c_sub <- sample_des_c_revised[,c("Sample.ID", "Name","Sample.Description.Origin", "Sample.Description.Type")]
sample_des_d_revised <- read.csv("~/505/data/phenotype/batch_d_revised.csv", header=T, na.strings="")
sample_des_d_sub <- sample_des_d_revised[,c("Sample.ID", "Name", "Sample.Description.Origin", "Sample.Description.Type")]
sample_des_e <- read.csv("~/505/data/phenotype/batch_e.csv", header=T, na.strings="")
sample_des_e_sub <- sample_des_e[,c("Sample.ID", "Name", "Sample.Description.Origin", "Sample.Description.Type")]

dim(sample_des_c_sub)

sample_des_c_sub$Sample.ID
sample_des_d_sub$Sample.ID
sample_des_e_sub$Sample.ID

sample_des_revised <- rbind(sample_des_c_sub, sample_des_d_sub, sample_des_e_sub)
summary(sample_des_revised)

sample_des_revised <- sample_des_revised %>% 
  filter(!is.na(Sample.ID))

sample_des_revised$Sample.ID
save(sample_des_revised, file="~/505/output/131_sample/sample_des_revised.Rdata")

# MDS result 
# reformat MDS result sample ID (X remove & . to -)
geno.mds$Sample.ID <- gsub("\\.","\\-", rownames(geno.mds)) 
geno.mds$Sample.ID <- gsub("X", "", geno.mds$Sample.ID)
class(geno.mds$Sample.ID)

geno.mds$Sample.ID
sample_des_revised$Sample.ID

# merge 
geno.mds.all.des <- merge(geno.mds, sample_des_revised, by="Sample.ID") 
colnames(geno.mds.all.des)[1] <- "Taxa"
head(SN_Taxa_131_sample)
colnames(geno.mds.all.des)[5] <- "Location_of_origin"
geno.mds.all.des.2 <- merge(geno.mds.all.des, SN_Taxa_131_sample, by="Taxa")

library(ggrepel)
set.seed(111) 

# just dots, no text info 
p.mds.country.2 <- ggplot(data=geno.mds.all.des.2, aes(V1, V2, color=Location_of_origin)) + theme_bw()
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(SN))) + theme(legend.position=c(0.9, 0.2), legend.background=element_rect(fill="gray90", size=.5, linetype="dotted"))
p.mds.country.2
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
ggsave(p.mds.country.2, filename="~/505/output/figure/131_sample/p.mds.country.2.png", height=9, width=9)

# p.mds.country <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
# p.mds.country <- p.mds.country + geom_point(size=1) 
# p.mds.country <- p.mds.country + scale_color_brewer(type="qual",palette="Set1")
# p.mds.country <- p.mds.country + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
# p.mds.country 
# ggsave(p.mds.country, filename="~/505/output/figure/131_sample/p.mds.country.png", height=20, width=20)

p.mds.country.2 <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=Location_of_origin))
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country.2
ggsave(p.mds.country.2, filename="~/505/output/figure/131_sample/p.mds.country.2.png", height=20, width=20)

p.mds.country.2 <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country.2
# ggsave(p.mds.country.2, filename="~/505/output/figure/late_silique/p.mds.country.2.png", height=20, width=20) 

##### make hierarchical cluster tree for the cluster result presentation 
library(ggdendro)
dim(genDist)

hc.505 <- hclust(dist(geno.numeric))
ggdata.505 <- dendro_data(as.dendrogram(hc.505))

ggdata.505$labels
colnames(sample_des_revised)[1] <- "label"

ggdata.505$labels$label <- gsub("\\.","\\-", ggdata.505$labels$label) 
ggdata.505$labels$label <- gsub("X", "", ggdata.505$labels$label)
ggdata.505$labels <- merge(ggdata.505$labels, sample_des_revised, by = "label")
ggdata.505$labels
ggdata.505$labels

# start ggplot here 
p.dengro <- ggplot(data = segment(ggdata.505))
p.dendro <- p.dengro + geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) + theme_dendro()
p.dendro <- p.dendro + geom_text(data = label(ggdata.505), aes(x = x, y = y, label = Name, hjust=0, color=Sample.Description.Origin)) + coord_flip() + scale_y_reverse(expand=c(0.2, 0))
p.dendro 

ggsave("~/505/output/figure/clustering_common_leaf_late_silique_92.png", width = 15, height = 28)   

``` 

# GWAS ... 
```{r}
# reformat for GWAS 
# get hapmap file using tassel 

######## use tassel to get hapmap format ########### 
# run_pipeline.pl -SortGenotypeFilePlugin -inputFile 505_filtered.vcf.gz -outputFile 505_filtered_sorted -fileType VCF
# run_pipeline.pl -Xmx5g -fork1 -vcf 505_filtered_sorted.vcf -export -exportType Hapmap -runfork1 
# output: 505_filtered_sorted.hmp.txt  
# edit output file to remove "#" after rs & assembly 
#################################################### 
```

# GWAS run
```{r}
# GWAS_505_late_silique.R   
# run_1, all default parameter in GAPIT
# run_2, model selection activated 
# run_3, using PCA = 0 
```

# display GWAS result 
```{r}
# install.packages("qqman")
library("qqman")
Erucic_acid <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/filter_by_het/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/filter_by_het/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/filter_by_het/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/131_sample/Erucic_acid.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
    ) 
dev.off() 

# check the significant SNPs
sum(Erucic_acid$P < 0.00001)
unique(Erucic_acid[Erucic_acid$P < 0.00001,]$CHR) # no SNPs  
unique(Oil_content[Oil_content$P < 0.00001,]$CHR) # no SNPs  
```

# compare leaf & late silique 131 samples, how many lines are overlapped  
```{r}
load("~/505/output/pheno_data_505_leaf.R")
pheno_data_505_leaf <- pheno_data_505

load("~/505/output/131_sample/pheno_data_505_131_sample.R")
pheno_data_505_late_silique <- pheno_data_505

# use common taxa names to compare, check missing individuals 
load("~/505/output/leaf/sample_des_revised.Rdata")
sample_des_revised_leaf <- sample_des_revised 
load("~/505/output/131_sample/sample_des_revised.Rdata")
sample_des_revised_late_silique <- sample_des_revised

colnames(sample_des_revised_leaf)[1] <- "Taxa"
colnames(sample_des_revised_late_silique)[1] <- "Taxa"

lines_leaf_2 <- merge(pheno_data_505_leaf, sample_des_revised_leaf, by="Taxa")
lines_leaf_2

lines_late_silique_2 <- merge(pheno_data_505_late_silique, sample_des_revised_late_silique, by="Taxa")
lines_late_silique_2

# common lines
common_leaf_131_sample <- intersect(lines_leaf_2$Name, lines_late_silique_2$Name) 
save(common_leaf_131_sample, file="~/505/output/common_leaf_131_sample.Rdata") 
# unique lines in leaf 
unique_leaf <- lines_leaf_2$Name[!(lines_leaf_2$Name %in% common_leaf_131_sample)] 
# 

common.lines <- merge(lines_leaf_2, lines_late_silique_2, by="Name")[,c("Name", "Taxa.x", "Taxa.y")]
common.lines 
save(common.lines, file="~/505/output/common.lines.leaf.131.sample.Rdata") 
```

# GWAS on the same 92 lines between leaf & late silique 
```{r}
# 1) prep for GWAS 
# https://github.com/leejimmy93/KIAT/blob/master/505/prep_geno_92_sample.sh

# 2) GWAS leaf & late silique 
# https://github.com/leejimmy93/KIAT/blob/master/505/GWAS_505_leaf_92sample.R
# https://github.com/leejimmy93/KIAT/blob/master/505/GWAS_505_late_silique_92sample.R 
```

# plot the GWAS result for common lines 
```{r}
library("qqman")

##### leaf ###  
Erucic_acid <- read.csv("~/505/output/myGAPIT/leaf_92/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/leaf_92/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/leaf_92/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/leaf_92/Erucic_acid.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
    ) 
dev.off() 

# check the number of SNPs 
sum(Erucic_acid$P <= 0.00001) # 80 significant SNPs? 
sum(Oleic_acid$P <= 0.00001)

# overlaps between Erucic acid & Oleic acid? 
significant_SNP <- Erucic_acid[Erucic_acid$P <= 0.00001,]$SNP
Oleic_acid[Oleic_acid$P <= 0.00001,]$SNP %in% Erucic_acid[Erucic_acid$P <= 0.00001,]$SNP

####### late silique 
Erucic_acid <- read.csv("~/505/output/myGAPIT/late_silique_92//GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/late_silique_92/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/late_silique_92/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/late_silique_92/Erucic_acid.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
    ) 
dev.off() 

# check the number of SNPs 
sum(Erucic_acid$P <= 0.00001) 
sum(Oleic_acid$P <= 0.00001)
```

# cluster using common SNPs for the common 92 lines 
```{r} 
# 1) filter based on position  
# https://github.com/leejimmy93/KIAT/blob/master/505/get_common_POS_92.sh

# import into R, using vcfR for manipulation 
library("vcfR") 

late_silique_505_common <- read.vcfR("~/505/leaf_VS_late_silique/505_filtered_common_late_silique_92.vcf.recode.vcf.gz")
leaf_505_common <- read.vcfR("~/505/leaf_VS_late_silique/505_filtered_common_leaf_92.vcf.recode.vcf.gz")

late_silique_505_common.2 <- reform.vcf(late_silique_505_common)
leaf_505_common.2 <- reform.vcf(leaf_505_common)

# combine the two files 
dim(late_silique_505_common.2)

late_silique_505_common.2$feature <- paste(late_silique_505_common.2$CHROM, late_silique_505_common.2$POS, sep="-")
leaf_505_common.2$feature <- paste(leaf_505_common.2$CHROM, leaf_505_common.2$POS, sep="-")

combined_leaf_late_silique <- merge(leaf_505_common.2, late_silique_505_common.2, by="feature")
dim(combined_leaf_late_silique)

save(late_silique_505_common.2, file="~/505/output/late_silique_505_common.2.Rdata")
save(leaf_505_common.2, file="~/505/output/leaf_505_common.2.Rdata")
save(combined_leaf_late_silique, file="~/505/output/combined_leaf_late_silique.Rdata")

# prepare for cluster analysis 
colnames(combined_leaf_late_silique)

combined_leaf_late_silique.t <- data.frame(t(combined_leaf_late_silique))
combined_leaf_late_silique.t[1:10, 1:10]
rownames(combined_leaf_late_silique.t)
dim(combined_leaf_late_silique.t)

# combined_leaf_late_silique.t.2 <- combined_leaf_late_silique.t[c(6:63, 68:125),]

combined_leaf_late_silique.t.2 <- combined_leaf_late_silique.t[!(rownames(combined_leaf_late_silique.t) %in% grep("feature|CHROM|POS|REF|ALT", rownames(combined_leaf_late_silique.t), value=T)),]
dim(combined_leaf_late_silique.t.2)
combined_leaf_late_silique.t.2[1:10, 1:10]

geno.numeric <- apply(combined_leaf_late_silique.t.2, 2, function(x) as.integer(x))
rnames <- rownames(combined_leaf_late_silique.t.2)
rownames(geno.numeric) <- rnames

geno.numeric.df <- data.frame(geno.numeric)
geno.numeric <- as.matrix(geno.numeric.df)

genDist <- as.matrix(dist(geno.numeric))

#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions
dim(geno.mds)
plot(geno.mds)  

geno.mds

# add leaf & late silique 
commom_line_count <- 92
geno.mds$tissue <- c(rep("leaf", commom_line_count), rep("late_silique", commom_line_count))

# add name 
load("~/505/output/leaf/sample_des_revised.Rdata")
sample_des_revised.leaf <- sample_des_revised
dim(sample_des_revised.leaf) # 94 

load("~/505/output/131_sample/sample_des_revised.Rdata")
sample_des_revised.late_silique <- sample_des_revised 
dim(sample_des_revised.late_silique) # 88

colnames(sample_des_revised.leaf)
colnames(sample_des_revised.late_silique) 

sample_des_revised <- rbind(sample_des_revised.leaf, sample_des_revised.late_silique)
sample_des_revised
dim(sample_des_revised) # 182 

# MDS result 
# reformat MDS result sample ID (X remove & . to -)
geno.mds$Sample.ID <- gsub("\\.","\\-", rownames(geno.mds)) 
geno.mds$Sample.ID <- gsub("X", "", geno.mds$Sample.ID)

geno.mds.all.des <- merge(geno.mds, sample_des_revised, by="Sample.ID")
geno.mds.all.des
dim(geno.mds.all.des)

## use hclust to check sample seperation pattern 
library(ggdendro)
dim(genDist) 

hc.505.common <- hclust(dist(geno.numeric))
ggdata.505.common <- dendro_data(as.dendrogram(hc.505.common))

ggdata.505.common$labels$tissue[grep("X505", ggdata.505.common$labels$label)] <- "late_silique"
ggdata.505.common$labels$tissue[grep("X505", ggdata.505.common$labels$label, invert=T)] <- "leaf"

### add names to different samples 
save(sample_des_revised, file="~/505/output/sample_des_revised_common_leaf_silique.Rdata")
colnames(sample_des_revised)[1] <- "label"

ggdata.505.common$labels
ggdata.505.common$labels$label <- gsub("\\.","\\-", ggdata.505.common$labels$label) 
ggdata.505.common$labels$label <- gsub("X", "", ggdata.505.common$labels$label)
ggdata.505.common$labels <- merge(ggdata.505.common$labels, sample_des_revised, by = "label")
# ggdata.505.common$labels$

# start ggplot here 
p.dengro <- ggplot(data = segment(ggdata.505.common))
p.dendro <- p.dengro + geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) + theme_dendro()
p.dendro <- p.dendro + geom_text(data = label(ggdata.505.common), aes(x = x, y = y, label = Name, hjust=0, color=tissue)) + coord_flip() + scale_y_reverse(expand=c(0.2, 0))
p.dendro
ggsave("~/505/output/figure/clustering_common_leaf_late_silique_92.png", width = 15, height = 28)  
```

# GWAS using common SNPs and common lines  
```{r} 
# get hmp files for both leaf & late silique 
# https://github.com/leejimmy93/KIAT/blob/master/505/vcf_to_hmp.sh

# GWAS
# https://github.com/leejimmy93/KIAT/blob/master/505/GWAS_505_late_silique_92sample_102303.R
# https://github.com/leejimmy93/KIAT/blob/master/505/GWAS_505_leaf_92sample_102303.R 

# plot manhattan result 
library("qqman")

####### leaf 
Erucic_acid <- read.csv("~/505/output/myGAPIT/leaf_92_102303/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/leaf_92_102303/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/leaf_92_102303/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/leaf_92_102303/Erucic_acid.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
) 
dev.off()  

# check the number of SNPs 
sum(Erucic_acid$P <= 0.00001) # 39 
sum(Oleic_acid$P <= 0.00001) # 3

####### late_silique
Erucic_acid <- read.csv("~/505/output/myGAPIT/late_silique_92_102303/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/late_silique_92_102303/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/late_silique_92_102303/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/late_silique_92_102303/Erucic_acid.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
) 
dev.off()  
```

# are the significant SNPs for different leaf dataset very similar? (temp test)
```{r}
significant_SNP_leaf_94_all_SNP <- read.table("~/505/output/significant_SNP.txt")
significant_SNP_leaf_94_all_SNP <- significant_SNP_leaf_94_all_SNP$x

# significant_SNP_leaf_92_all_SNP <- Erucic_acid[Erucic_acid$P <= 0.00001,]$SNP # leaf 92 all SNP result 

# significant_SNP_leaf_92_common_SNP <- Erucic_acid[Erucic_acid$P <= 0.00001,]$SNP # leaf 92_common SNP result 

save(significant_SNP_leaf_92_common_SNP, significant_SNP_leaf_92_all_SNP, file="~/505/output/significant_SNP.Rdata")

length(significant_SNP_leaf_92_all_SNP)
length(significant_SNP_leaf_92_common_SNP)

tmp <- intersect(significant_SNP_leaf_92_common_SNP, significant_SNP_leaf_92_all_SNP)
significant_SNP_9 <- intersect(significant_SNP_leaf_94_all_SNP, tmp) # 9 SNPs are common across all leaf datasets 
significant_SNP_9
significant_SNP_leaf_94_all_SNP 
```

# check genotype for the 9 common significant SNPs (hmp file)
```{r}
library("tidyverse")
leaf_92_common <- read_table2("~/505/leaf_VS_late_silique/505_filtered_common_leaf_92.hmp.txt")
late_silique_92_common <- read_table2("~/505/leaf_VS_late_silique/505_filtered_common_late_silique_92.hmp.txt")

# extract significant SNPs 
late_silique_sigSNP_9 <- late_silique_92_common[late_silique_92_common$rs %in% significant_SNP_9,]
rownames(late_silique_sigSNP_9) <- late_silique_sigSNP_9$rs
View(late_silique_sigSNP_9)
leaf_sigSNP_9 <- leaf_92_common[leaf_92_common$rs %in% significant_SNP_9,]
rownames(leaf_sigSNP_9) <- leaf_sigSNP_9$rs
View(leaf_sigSNP_9)

# replace sampleID with SN ID
late_silique_sigSNP_9.t <- as.data.frame(t(late_silique_sigSNP_9))
late_silique_sigSNP_9.t$Taxa.y <- rownames(late_silique_sigSNP_9.t)

leaf_sigSNP_9.t <- as.data.frame(t(leaf_sigSNP_9))
leaf_sigSNP_9.t$Taxa.x <- rownames(leaf_sigSNP_9.t)

load("~/505/leaf_VS_late_silique/SN_Taxa_leaf_silique.Rdata")
SN_Taxa_leaf_silique

tmp1 <- merge(leaf_sigSNP_9.t, SN_Taxa_leaf_silique, by="Taxa.x")
tmp2 <- merge(late_silique_sigSNP_9.t, SN_Taxa_leaf_silique, by="Taxa.y")
sigSNP_9 <- merge(tmp1, tmp2, by="SN")
View(sigSNP_9)

write.csv(sigSNP_9, "~/505/output/sigSNP_9.csv")
```

### substitute leaf & late silique's 9 SNPs genotype & check association 
The idea is that there might be other factor influence the GWAS result... (Julin suggested) (a lot of input data in this chunk are legacy of the last chunk)
```{r}
# hmp file 
library("tidyverse")
leaf_92_common <- read_table2("~/505/leaf_VS_late_silique/505_filtered_common_leaf_92.hmp.txt")
late_silique_92_common <- read_table2("~/505/leaf_VS_late_silique/505_filtered_common_late_silique_92.hmp.txt")

dim(leaf_92_common) # 102303 103 
dim(late_silique_92_common) # 102303    103 
leaf_92_common[1:10, 1:20]

# replace Taxa ID with SN# 
head(SN_Taxa_leaf_silique)
temp1 <- data.frame(Taxa.x = colnames(leaf_92_common))
temp1$index <- rownames(temp1)
head(temp1)
dim(temp1)
temp1.2 <- merge(temp1, SN_Taxa_leaf_silique, by="Taxa.x", all=T)
temp1.2

for (i in 1:nrow(temp1.2)){ 
  temp1.2$SN_new[i] <- ifelse(!is.na(temp1.2$SN[i]), temp1.2$SN[i], as.character(temp1.2[i, "Taxa.x"]))  
}

colnames_new_leaf <- temp1.2[order(as.numeric(temp1.2$index)),]$SN_new
colnames(leaf_92_common) <- colnames_new_leaf
leaf_92_common[1:10, 1:20]
# write.table(leaf_92_common, sep="\t", "~/505/leaf_VS_late_silique/tmp1.hmp", row.names = F) 

### late silique 
temp2 <- data.frame(Taxa.y = colnames(late_silique_92_common))
temp2$index <- rownames(temp2)
head(temp2)
head(SN_Taxa_leaf_silique)
temp2.2 <- merge(temp2, SN_Taxa_leaf_silique, by="Taxa.y", all=T)

for (i in 1:nrow(temp2.2)){ 
  temp2.2$SN_new[i] <- ifelse(!is.na(temp2.2$SN[i]), temp2.2$SN[i], as.character(temp2.2[i, "Taxa.y"]))  
}

colnames_new_late_silique <- temp2.2[order(as.numeric(temp2.2$index)),]$SN_new
colnames(late_silique_92_common) <- colnames_new_late_silique
late_silique_92_common[1:10, 1:20] 
# write.table(late_silique_92_common, sep="\t", "~/505/leaf_VS_late_silique/tmp2.hmp", row.names = F) 

# add tissue to each file 
leaf_92_common$tissue <- rep("leaf", nrow(leaf_92_common))
leaf_92_common$index <- as.numeric(rownames(leaf_92_common))
late_silique_92_common$tissue <- rep("late_silique", nrow(late_silique_92_common))
late_silique_92_common$index <- as.numeric(rownames(late_silique_92_common))

### merge two data frames 
late_silique_92_common_sorted <- late_silique_92_common[,order(match(colnames(late_silique_92_common), colnames(leaf_92_common)))]
identical(colnames(late_silique_92_common_sorted), colnames(leaf_92_common)) # T 

# switch order back 
late_silique_92_common_back <- late_silique_92_common_sorted[,order(match(colnames(late_silique_92_common_sorted), colnames(late_silique_92_common)))]
identical(colnames(late_silique_92_common_back), colnames(late_silique_92_common))

late_silique_92_common_back <- 
  late_silique_92_common_back %>%
  select(-(tissue:index))
dim(late_silique_92_common_back) 

# export this file with order switched back for GWAS to double check. 
write.table(late_silique_92_common_back, sep="\t", "~/505/leaf_VS_late_silique/late_silique_92_common_back.hmp", row.names = F)
colnames(late_silique_92_common_back) <- NA
colnames(late_silique_92_common) <- NA
identical(late_silique_92_common_back, late_silique_92_common_back) 

common_92 <- rbind(leaf_92_common, late_silique_92_common_sorted)
write.table(common_92, sep="\t", "~/505/leaf_VS_late_silique/common_92.hmp", row.names = F) 

dim(common_92) # 204606    105 

# significant SNPs 
significant_SNP

# swap genotype for the 9 significant SNPs
# 9 sigSNP from leaf & other from late silique 
sigSNP <- common_92[common_92$rs %in% significant_SNP & common_92$tissue=="leaf",]
other <- common_92[!(common_92$rs %in% significant_SNP) & common_92$tissue=="late_silique",]

all <- rbind(sigSNP, other)
all.sorted <- all[order(all$index),]
all.sorted[1:10, 1:10]
sig_leaf_other_silique <- 
  all.sorted %>%
  select(-(tissue:index))
dim(sig_leaf_other_silique) # 102303 103 
write.table(sig_leaf_other_silique, sep="\t", "~/505/leaf_VS_late_silique/sig_leaf_other_silique.hmp", row.names = F) 
# cat sig_leaf_other_silique.hmp | sed 's/"//g' > sig_leaf_other_silique.hmp.txt 
   

# 9 sigSNP from silique & other from leaf 
sigSNP <- common_92[common_92$rs %in% significant_SNP & common_92$tissue=="late_silique",]
other <- common_92[!(common_92$rs %in% significant_SNP) & common_92$tissue=="leaf",]
all <- rbind(sigSNP, other) 
all.sorted <- all[order(all$index),]
all.sorted[1:10, 1:10]
sig_silique_other_leaf <- 
  all.sorted %>%
  select(-(tissue:index))
dim(sig_silique_other_leaf) # 102303 103 
write.table(sig_silique_other_leaf, sep="\t", "~/505/leaf_VS_late_silique/sig_silique_other_leaf.hmp", row.names = F) 
# cat sig_silique_other_leaf.hmp | sed 's/"//g' > sig_silique_other_leaf.hmp.txt 

# swap back 
# leaf 
dim(common_92) 
sigSNP <- common_92[common_92$rs %in% significant_SNP & common_92$tissue=="leaf",]
other <- common_92[!(common_92$rs %in% significant_SNP) & common_92$tissue=="leaf",]
sigSNP$tissue
unique(other$tissue)

all <- rbind(sigSNP, other) 
all.sorted <- all[order(all$index),]
all.sorted[1:10, 1:10]
sig_leaf_other_leaf <- 
  all.sorted %>%
  select(-(tissue:index))
dim(sig_leaf_other_leaf) # 102303 103 
sig_leaf_other_leaf[1:10, 1:20] 

write.table(sig_leaf_other_leaf, sep="\t", "~/505/leaf_VS_late_silique/sig_leaf_other_leaf.hmp", row.names = F) 
# cat sig_leaf_other_leaf.hmp | sed 's/"//g' > sig_leaf_other_leaf.hmp.txt 

# silique 
sigSNP <- common_92[common_92$rs %in% significant_SNP & common_92$tissue=="late_silique",]
other <- common_92[!(common_92$rs %in% significant_SNP) & common_92$tissue=="late_silique",]
sigSNP$tissue
unique(other$tissue)
all <- rbind(sigSNP, other) 
all.sorted <- all[order(all$index),]
all.sorted[1:10, 1:10]
sig_silique_other_silique <- 
  all.sorted %>% 
  select(-(tissue:index))
dim(sig_silique_other_silique) # 102303 103 
write.table(sig_silique_other_silique, sep="\t", "~/505/leaf_VS_late_silique/sig_silique_other_silique.hmp", row.names = F) 
# cat sig_silique_other_silique.hmp | sed 's/"//g' > sig_silique_other_silique.hmp.txt 
########### the above is correct ############ double checked !!!!

# GWAS 
# run in whitney 

# display result 
# sig leaf other silique 
Erucic_acid <- read.csv("~/505/output/myGAPIT/sig_leaf_other_silique/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/sig_leaf_other_silique/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/sig_leaf_other_silique/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/sig_leaf_other_silique.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
) 
dev.off() 

# sig silique other leaf 
Erucic_acid <- read.csv("~/505/output/myGAPIT/sig_silique_other_leaf/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/sig_silique_other_leaf/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/sig_silique_other_leaf/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/sig_silique_other_leaf.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
) 
dev.off()  

# swap genotypes & swap back found the no significant SNPs might be caused by not-matching column ID? redo GWAS for late silique ...

# sig leaf other leaf 
Erucic_acid <- read.csv("~/505/output/myGAPIT/sig_leaf_other_leaf/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/sig_leaf_other_leaf/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/sig_leaf_other_leaf/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/sig_leaf_other_leaf.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
) 
dev.off()  

# sig silique other silique
Erucic_acid <- read.csv("~/505/output/myGAPIT/sig_silique_other_silique/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/sig_silique_other_silique/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/sig_silique_other_silique/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/sig_silique_other_silique.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
) 
dev.off()  

``` 

# re-run GWAS for 505 131 samples again, using SN# as the sample ID 
```{r}
# firstly check whether the genotype & phenotype match 
geno_505_hmp <- read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/leaf_VS_late_silique/505_filtered_common_late_silique_92.hmp.txt", head=FALSE)
# get chrom name to numerics
geno_505_hmp$V3 <- as.numeric(geno_505_hmp$V3)
geno_505_hmp[1,3] <- "chrom"

# phenotype data
load("~/505/data/phenotype/gh.seed.bolt.table.wide.Rdata")
# reform phentype data to have match taxa ID with geno data
# get taxa ID from sample description file
sample_des_c <- read.csv("~/505/data/phenotype/batch_c.csv", header=T)
sample_des_d <- read.csv("~/505/data/phenotype/batch_d.csv", header=T)
sample_des_e <- read.csv("~/505/data/phenotype/batch_e.csv", header=T)

sample_des_c_sub2 <- sample_des_c[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_d_sub2 <- sample_des_d[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_e_sub2 <- sample_des_e[,c("Sample.ID", "No..of.Sowing", "Name")]

sample_des_d_sub2 <- sample_des_d_sub2[c(1:41),]
sample_des_sub2 <- rbind(sample_des_c_sub2, sample_des_d_sub2, sample_des_e_sub2)

# add name to to replace SN
sample_des_sub2$SN <- paste("K", sample_des_sub2$No..of.Sowing, sep="")
gh.seed.bolt.table.wide.merge <- merge(sample_des_sub2, gh.seed.bolt.table.wide, by = "SN")
pheno_data_505 <- gh.seed.bolt.table.wide.merge[,c("Sample.ID", "FATTY ACID_Erucic acid (C22:1n9)", "FATTY ACID_oil contents (%)", "FATTY ACID_Oleic acid (C18:1n9c)")]
colnames(pheno_data_505) <- c("Taxa", "Erucic_acid", "Oil_content", "Oleic_acid")
pheno_data_505
K250_K251 <- data.frame(Taxa = c("505_K_250", "505_K_251"),
  		Erucic_acid = rep(NA, 2),
			Oil_content = rep(NA, 2),
			Oleic_acid = rep(NA, 2))
pheno_data_505 <- rbind(pheno_data_505, K250_K251)

# get the 92 common lines between leaf & late silique
load("~/505/output/common.lines.leaf.131.sample.Rdata")
pheno_data_505_92 <- pheno_data_505[which(pheno_data_505$Taxa %in% common.lines$Taxa.y),]
rownames(pheno_data_505_92) <- 1:nrow(pheno_data_505_92)

myY <- pheno_data_505_92
myG <- geno_505_hmp

######## sorted back 
geno_505_hmp <- read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/leaf_VS_late_silique/late_silique_92_common_back.hmp.txt", head=FALSE)
# get chrom name to numerics
geno_505_hmp$V3 <- as.numeric(geno_505_hmp$V3)
geno_505_hmp[1,3] <- "chrom"

# phenotype data
load("~/505/leaf_VS_late_silique/common_92_pheno.Rdata")
pheno_data_505 <- common_92_pheno

myY2 <- pheno_data_505
myG2 <- geno_505_hmp

head(myY)
myY2
head(SN_Taxa_leaf_silique)

myY$Taxa.y <- myY$Taxa
tmp1 <- merge(myY, SN_Taxa_leaf_silique, by="Taxa.y") 
View(tmp1)
myY2$SN <- myY2$Taxa
View(myY2)
tmp2 <- merge(myY2, SN_Taxa_leaf_silique, by="SN") 
View(tmp2)
test <- merge(tmp1, tmp2, by="SN")

######## compare sorted back & the original one ########
myG[1:10, 1:20]
myG2[1:10, 1:20]

test1 <- as.vector(t(myG[1,]))
test2 <- as.vector(t(myG2[1,]))

myG[1,] <- NA
myG2[1,] <- NA
identical(myG, myG2)

for (i in 1:92){
  print(i)
  print(identical(as.character(myG[-1,11+i]), as.character(myG2[-1,11+i])))
}

dim(myG[])

identical(myG$V16, myG2$V16)
head(myG$V16)
head(myG2$V16)

identical(myG[-1,11+1], myG2[-1,11+1])
identical(as.character(myG[-1,11+1]), as.character(myG2[-1,11+1])) 
```

# check genotype for the 9 sigSNPs in the vcf file  
found that there might be problem when converting vcf to hmp format, get M,R,N etc in the genotype calls, which shoud not be there, so check the vcf genotype for the 9 sigSNPs in both leaf & silique data 
```{r}
leaf_505_common <- read.vcfR("~/505/leaf_VS_late_silique/505_filtered_common_leaf_92.vcf.recode.vcf.gz")
late_silique_505_common <- read.vcfR("~/505/leaf_VS_late_silique/505_filtered_common_late_silique_92.vcf.recode.vcf.gz")

significant_SNP_9

# reform vcf file to the right format 

leaf_505_common.2.2 <- reform.vcf.atcg(leaf_505_common)
leaf_505_common.2.2[1:10, 1:10]
late_silique_505_common.2.2 <- reform.vcf.atcg(late_silique_505_common)

# extract significant SNPs
# silique 
late_silique_505_common.2.new <- 
late_silique_505_common.2.2 %>% 
  unite(feature2, CHROM, POS)

late_silique_505_common.2.new$feature2 <- gsub("chr", "S", late_silique_505_common.2.new$feature2)
late_silique_505_common.2.new$feature2 <- toupper(late_silique_505_common.2.new$feature2)

vcf_sigSNP_9.silique <- late_silique_505_common.2.new[late_silique_505_common.2.new$feature2 %in% significant_SNP_9,]
rownames(vcf_sigSNP_9.silique) <- vcf_sigSNP_9.silique$feature2

vcf_sigSNP_9.silique.t <- as.data.frame(t(vcf_sigSNP_9.silique))
rownames(vcf_sigSNP_9.silique.t) <- gsub("\\.","\\-", rownames(vcf_sigSNP_9.silique.t))
rownames(vcf_sigSNP_9.silique.t) <- gsub("X", "", rownames(vcf_sigSNP_9.silique.t))  
vcf_sigSNP_9.silique.t 

# leaf 
leaf_505_common.2.new <-  
leaf_505_common.2.2 %>% 
  unite(feature2, CHROM, POS)

leaf_505_common.2.new$feature2 <- gsub("chr", "S", leaf_505_common.2.new$feature2)
leaf_505_common.2.new$feature2 <- toupper(leaf_505_common.2.new$feature2)

vcf_sigSNP_9.leaf <- leaf_505_common.2.new[leaf_505_common.2.new$feature2 %in% significant_SNP_9,]
rownames(vcf_sigSNP_9.leaf) <- vcf_sigSNP_9.leaf$feature2

vcf_sigSNP_9.leaf.t <- as.data.frame(t(vcf_sigSNP_9.leaf))
rownames(vcf_sigSNP_9.leaf.t) <- gsub("\\.","\\-", rownames(vcf_sigSNP_9.leaf.t))
rownames(vcf_sigSNP_9.leaf.t) <- gsub("X", "", rownames(vcf_sigSNP_9.leaf.t))  
# vcf_sigSNP_9.leaf.t$Taxa.x.x <- rownames(vcf_sigSNP_9.leaf.t)

# replace sample ID with SN ID  
load("~/505/leaf_VS_late_silique/SN_Taxa_leaf_silique.Rdata")
SN_Taxa_leaf_silique

vcf_sigSNP_9.leaf.t$Taxa.x <- rownames(vcf_sigSNP_9.leaf.t)
vcf_sigSNP_9.silique.t$Taxa.y <- rownames(vcf_sigSNP_9.silique.t)

tmp <- merge(vcf_sigSNP_9.leaf.t, SN_Taxa_leaf_silique, by="Taxa.x")
sigSNP_9_w_vcf <- merge(tmp, vcf_sigSNP_9.silique.t, by="Taxa.y")

View(sigSNP_9_w_vcf)

write.csv(sigSNP_9_w_vcf, "~/505/output/sigSNP_9_w_vcf.csv")

# compare hmp & vcf comparison for the 9 significant SNPs, does the disconcordance match? 

sigSNP_9_w_vcf.IUPAC <- IUPAC.code(sigSNP_9_w_vcf)
View(sigSNP_9_w_vcf.IUPAC)

sigSNP_9_w_vcf.IUPAC.2 <- 
sigSNP_9_w_vcf.IUPAC %>%
  select(-c(Taxa.y, Taxa.x, SN))
View(sigSNP_9_w_vcf.IUPAC.2)

View(sigSNP_9)
sig_SNP_hmp_vcf <- merge(sigSNP_9, sigSNP_9_w_vcf.IUPAC, by="SN")

write.csv(sig_SNP_hmp_vcf, "~/505/output/sig_SNP_hmp_vcf.csv")
identical(as.character(sig_SNP_hmp_vcf$SA08_9513662.x.x), as.character(sig_SNP_hmp_vcf$SA08_9513662.x.y))
### conclusion, not a problem of file conversion. 
# the different calls between leaf & silique are due to the difference in their depth.  
```

# check phenotype 
```{r}
# late silique 
sample_des_c <- read.csv("~/505/data/phenotype/batch_c.csv", header=T)
sample_des_d <- read.csv("~/505/data/phenotype/batch_d.csv", header=T)
sample_des_e <- read.csv("~/505/data/phenotype/batch_e.csv", header=T)

sample_des_c_sub2 <- sample_des_c[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_d_sub2 <- sample_des_d[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_e_sub2 <- sample_des_e[,c("Sample.ID", "No..of.Sowing", "Name")]

sample_des_d_sub2 <- sample_des_d_sub2[c(1:41),]
sample_des_sub2 <- rbind(sample_des_c_sub2, sample_des_d_sub2, sample_des_e_sub2)

# add name to to replace SN
sample_des_sub2$SN <- paste("K", sample_des_sub2$No..of.Sowing, sep="")
gh.seed.bolt.table.wide.merge <- merge(sample_des_sub2, gh.seed.bolt.table.wide, by = "SN")
pheno_data_505 <- gh.seed.bolt.table.wide.merge[,c("SN", "Sample.ID", "Name","FATTY ACID_Erucic acid (C22:1n9)", "FATTY ACID_oil contents (%)", "FATTY ACID_Oleic acid (C18:1n9c)")]
colnames(pheno_data_505) <- c("SN","Taxa", "Name","Erucic_acid", "Oil_content", "Oleic_acid")
pheno_data_505
K250_K251 <- data.frame(SN = c("K250", "K251"),
      Taxa = c("505_K_250", "505_K_251"),
      Name = c("K250", "K251"), 
  		Erucic_acid = rep(NA, 2),
			Oil_content = rep(NA, 2),
			Oleic_acid = rep(NA, 2))
pheno_data_505.silique <- rbind(pheno_data_505, K250_K251)
pheno_data_505.silique$Taxa <- as.character(pheno_data_505.silique$Taxa)
str(pheno_data_505.silique)
save(pheno_data_505.silique, file="~/505/output/131_sample/pheno_data_505.silique.Rdata")

# leaf 
sample_des_a_revised <- read.csv("~/505/data/phenotype/batch_a_revised.csv", header=T)
sample_des_b_revised <- read.csv("~/505/data/phenotype/batch_b_revised.csv", header=T)

sample_des_a_sub2 <- sample_des_a_revised[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_a_sub2 <- sample_des_a_sub2[!is.na(sample_des_a_sub2$Sample.ID),]
sample_des_b_sub2 <- sample_des_b_revised[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_sub2 <- rbind(sample_des_a_sub2, sample_des_b_sub2)

# add name to to replace SN
sample_des_sub2$SN <- paste("K", sample_des_sub2$No..of.Sowing, sep="")
gh.seed.bolt.table.wide.merge <- merge(sample_des_sub2, gh.seed.bolt.table.wide, by = "SN")
pheno_data_505.leaf <- gh.seed.bolt.table.wide.merge[,c("SN","Sample.ID","Name", "FATTY ACID_Erucic acid (C22:1n9)", "FATTY ACID_oil contents (%)", "FATTY ACID_Oleic acid (C18:1n9c)")]
colnames(pheno_data_505.leaf) <- c("SN","Taxa", "Name","Erucic_acid", "Oil_content", "Oleic_acid")
pheno_data_505.leaf 
save(pheno_data_505.leaf, file="~/505/output/leaf/pheno_data_505.leaf.Rdata")

dim(pheno_data_505.leaf)
dim(pheno_data_505.silique)
tmp <- merge(pheno_data_505.leaf, pheno_data_505.silique, by="SN")

common_92_pheno <- 
  tmp %>%
  select(SN, Erucic_acid.x, Oil_content.x, Oleic_acid.x)

colnames(common_92_pheno) <- c("Taxa", "Erucic_acid", "Oil_content", "Oleic_acid")
head(common_92_pheno)
save(common_92_pheno, file="~/505/leaf_VS_late_silique/common_92_pheno.Rdata")

SN_Taxa_leaf_silique <- tmp[,c("SN", "Taxa.x", "Taxa.y")]
save(SN_Taxa_leaf_silique, file="~/505/leaf_VS_late_silique/SN_Taxa_leaf_silique.Rdata")
View(gh.seed.bolt.table.wide) 
# conclusion: phenotype data has no problem  
```

# t-test on the 9 sigSNPs for leaf & silique 
The hypothesis is that leaf data has higher depth than silique data, so overall its P-value is more significant than silique data. So if we do t-test on a single SNP, compare it between leaf & silique data, the P-value should be smaller for leaf than in silique. (GWAS is just a collection of t-test using phenotype as the result, genotype as the predictor? I am actually not sure this is true, check notes need to... )  
```{r}
## load data 
# common phentype data 
load("~/505/output/pheno_data_505_late_silique.R")
load("~/505/output/pheno_data_505_leaf.R")

pheno_data_505_late_silique
pheno_data_505_leaf

load("~/505/output/common.lines.leaf.131.sample.Rdata")
pheno_data_505_92_leaf <- pheno_data_505_leaf[which(pheno_data_505_leaf$Taxa %in% common.lines$Taxa.x),]
pheno_data_505_92_leaf
rownames(pheno_data_505_92_leaf) <- 1:nrow(pheno_data_505_92_leaf)

pheno_data_505_92_late_silique <- pheno_data_505_late_silique[which(pheno_data_505_late_silique$Taxa %in% common.lines$Taxa.y),]
pheno_data_505_92_late_silique
rownames(pheno_data_505_92_late_silique) <- 1:nrow(pheno_data_505_92_late_silique)

## common genotype data (vcf format, genotype as numeric values) 
library("tidyverse")
leaf_505_common <- read.vcfR("~/505/leaf_VS_late_silique/505_filtered_common_leaf_92.vcf.recode.vcf.gz")
late_silique_505_common <- read.vcfR("~/505/leaf_VS_late_silique/505_filtered_common_late_silique_92.vcf.recode.vcf.gz")

significant_SNP_9

# vcf reform
leaf_505_common.2.2 <- reform.vcf(leaf_505_common)
late_silique_505_common.2.2 <- reform.vcf(late_silique_505_common) 

# extract significant SNPs 
leaf_505_common.t <- vcf_sigSNP_extract(vcf=leaf_505_common.2.2, significant_SNP_9)
leaf_505_common.t <- leaf_505_common.t[-c(1:3),]
dim(leaf_505_common.t)  # 92 9 
late_silique_505_common.t <- vcf_sigSNP_extract(vcf=late_silique_505_common.2.2, significant_SNP_9)
late_silique_505_common.t <- late_silique_505_common.t[-c(1:3),]
dim(late_silique_505_common.t) # 92 9 

# combine phenotype & genotype data 
late_silique_505_common.t$Taxa <- rownames(late_silique_505_common.t)
for_ttest_late_silique <- merge(late_silique_505_common.t, pheno_data_505_92_late_silique, by="Taxa")

leaf_505_common.t$Taxa <- rownames(leaf_505_common.t)
for_ttest_leaf <- merge(leaf_505_common.t, pheno_data_505_92_leaf, by="Taxa")

dim(for_ttest_leaf)
colnames(for_ttest_leaf)

## anova 
# cannot perform t-test bc. grouping factor is not 2 
significant_SNP
traits <- colnames(pheno_data_505_leaf)[2:4]

anova_result_leaf <- data.frame()

for (j in traits){
  for (i in significant_SNP){
      tmp <- aov(for_ttest_leaf[,j] ~ for_ttest_leaf[,i]) 
      anova_result_leaf[i,j] <- unlist(summary(tmp))["Pr(>F)1"]  
    } 
}
anova_result_leaf

anova_result_late_silique <- data.frame()

for (j in traits){
  for (i in significant_SNP){
      tmp <- aov(for_ttest_late_silique[,j] ~ for_ttest_late_silique[,i]) 
      anova_result_late_silique[i,j] <- unlist(summary(tmp))["Pr(>F)1"]  
    } 
}

anova_result_late_silique

### transform data for better result presentation 
anova_result_leaf.new <- 
anova_result_leaf %>%
  mutate(SNP = rownames(anova_result_leaf))

anova_result_leaf.new.melt <- melt(anova_result_leaf.new)
colnames(anova_result_leaf.new.melt)[3] <- "leaf"

anova_result_leaf.new.melt.new <- 
anova_result_leaf.new.melt %>%
  unite(ID, SNP, variable, sep="-")  

anova_result_late_silique.new <- 
anova_result_late_silique %>%
  mutate(SNP = rownames(anova_result_late_silique))

anova_result_late_silique.new.melt <- melt(anova_result_late_silique.new)
colnames(anova_result_late_silique.new.melt)[3] <- "late_silique"

anova_result_late_silique.new.melt.new <- 
anova_result_late_silique.new.melt %>%
  unite(ID, SNP, variable, sep="-")  

anova_result <- 
merge(anova_result_leaf.new.melt.new, anova_result_late_silique.new.melt.new, by="ID")

anova_result$note <- ifelse(anova_result$leaf < anova_result$late_silique, "T", "F")
anova_result 
write.csv(anova_result, "~/505/output/anova_result.csv")
``` 

* conclussion for GWAS trouble shooting 
The no signal for GWAS was caused by data processing. when I import the phenotype data into R, their TaxaID are factors, although by eye I only see 131 items, there are actually 133 factors there, that caused the data non-matching. So GAPIT didn't throw an error, but gave me no signal in the result. 

Lessons learned: should use stringAsFactors=F, as.is=T, OR tidyverse package to avoid problem of factors. 

# the right GWAS 
```{r}
###### replace genotype ID with SN# for the 131 samples and see whether get significant SNPs 
load("~/505/output/131_sample/pheno_data_505.silique.Rdata") 
pheno_data_505.silique
pheno_data_505.silique.SN <- pheno_data_505.silique[,-c(2:3)]
head(pheno_data_505.silique.SN)
colnames(pheno_data_505.silique.SN)[1] <- "Taxa"
save(pheno_data_505.silique.SN, file="~/505/data/pheno_data_505.silique.SN.Rdata")

SN_Taxa_131_sample <- pheno_data_505.silique[,c(1:2)]
SN_Taxa_131_sample

geno_505_131sample <- read_table2("~/505/vcf_late_silique_131_sample/combined/505_filtered_sorted.hmp.txt") 

temp1 <- data.frame(Taxa = colnames(geno_505_131sample))
temp1$index <- rownames(temp1)
temp1

temp1.2 <- merge(temp1, SN_Taxa_131_sample, by="Taxa", all=T)
temp1.2

for (i in 1:nrow(temp1.2)){ 
  temp1.2$SN_new[i] <- ifelse(!is.na(temp1.2$SN[i]), temp1.2$SN[i], as.character(temp1.2[i, "Taxa"]))  
}

colnames_131_sample <- temp1.2[order(as.numeric(temp1.2$index)),]$SN_new
colnames(geno_505_131sample) <- colnames_131_sample
geno_505_131sample[1:10, 1:20]

write.table(geno_505_131sample, sep="\t", "~/505/leaf_VS_late_silique/geno_505_131sample.hmp", row.names = F) 
# cat geno_505_131sample.hmp | sed 's/"//g' > geno_505_131sample.hmp.txt 

# GWAS using this new dataset 
# GWAS result 

Erucic_acid <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT/late_silique/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT/late_silique/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT/late_silique/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/131_sample_SN.png", width=12, height=10, units="in", res=300)
par(mfrow=c(2,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
) 

# manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
#           cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
# ) 
dev.off()  
```

# To avoid the duplicated sample problem, I also run GWAS using the sample ID, original data 
https://github.com/leejimmy93/KIAT/blob/master/505/GWAS_505_late_silique.R
display result 
```{r}
# GWAS result 

Erucic_acid <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/run_3/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/run_3/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/run_3/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/505/output/figure/131_sample.png", width=12, height=10, units="in", res=300)
par(mfrow=c(2,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
          cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
) 

# manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
#           cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
# ) 
dev.off() 
```


# examine GWAS result 
```{r}
sum(Erucic_acid$P <= 0.00001) # 258
sum(Oleic_acid$P <= 0.00001) # 125

Erucic_acid[Erucic_acid$P <= 0.00001,] 

# overlaps between Erucic acid & Oleic acid? 
significant_SNP <- unique(c(as.character(Erucic_acid[Erucic_acid$P <= 0.00001,]$SNP), as.character(Oleic_acid[Oleic_acid$P <= 0.00001,]$SNP))) 
significant_SNP[order(significant_SNP)]


Oleic_acid[Oleic_acid$P <= 0.00001,]$SNP %in% Erucic_acid[Erucic_acid$P <= 0.00001,]$SNP
write.table(significant_SNP, file = "~/505/output/131_sample/significant_SNP.txt")
```

# check the genes include the significant SNPs (my MAC)
```{r}
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures)
library("rtracklayer")

### get gff file with gene chrom & pos info, gff3 file must be sorted 
# Grange doesn't give me the corret order of genes, so I used unix command to get the gene info from sorted gff3 file 
# cat Brassica_napus.annotation_v5.sorted.gff3 | grep "mRNA" | awk '{print $1, $4, $5, $9}' | tr "=" "\t" | tr ";" "\t" | awk '{print $1, $2, $3, $5}' > gff.mRNA
# gff.mRNA <- read.table("~/Desktop/Brassica_project/IGV_file/gff.mRNA")
gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 

### get significant SNPs 
Erucic_acid <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/run_3/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/run_3/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

Erucic_Oleic <- rbind(Erucic_acid, Oleic_acid)

SNP.Erucic.Oleic <- data.frame(CHROM = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\2", Erucic_Oleic[Erucic_Oleic$P.value < 0.00001,]$SNP), 
                         POS = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\4", Erucic_Oleic[Erucic_Oleic$P.value < 0.00001,]$SNP))

SNP.Erucic.Oleic$CHROM <- paste("chr", SNP.Erucic.Oleic$CHROM, sep = "")
SNP.Erucic.Oleic$POS <- as.numeric(as.character(SNP.Erucic.Oleic$POS))

head(SNP.Erucic.Oleic) 
SNP.Erucic.Oleic.unique <- unique(SNP.Erucic.Oleic)
dim(SNP.Erucic.Oleic.unique) # 262 2 

genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), names = gff.mRNA$name)
genes

SNP <- GRanges(seqnames = Rle(SNP.Erucic.Oleic.unique$CHROM), ranges = IRanges(start = SNP.Erucic.Oleic.unique$POS, end = SNP.Erucic.Oleic.unique$POS), ID = paste(SNP.Erucic.Oleic.unique$CHROM, SNP.Erucic.Oleic.unique$POS, sep = "_"))
SNP

SNP_gene <- mergeByOverlaps(SNP, genes) # warning message here. look out!!!!!! 
SNP_gene

SNP_gene_df <- as.data.frame(SNP_gene)
SNP_gene_df

SNP_gene_final <- SNP_gene_df[,c("SNP.ID", "genes.seqnames", "SNP.start", "names")]
SNP_gene_final 

# napus_vs_ara.non_reciprocal <- read.table("~/Desktop/Brassica_project/reference/napus_vs_ara.non_reciprocal.table")
napus_vs_ara.non_reciprocal <- read.table("~/Reference/B.napus/napus_vs_ara.non_reciprocal.table")

head(napus_vs_ara.non_reciprocal)
dim(napus_vs_ara.non_reciprocal) # 64949 

colnames(napus_vs_ara.non_reciprocal)[1] <- "names"

SNP_gene_final_fxn <- merge(SNP_gene_final, napus_vs_ara.non_reciprocal, by = "names", all.x=T)
SNP_gene_final_fxn$SNP.ID <- gsub("chr", "S", SNP_gene_final_fxn$SNP.ID)

SNP_gene_final_fxn$Erucic_acid <- SNP_gene_final_fxn$SNP.ID %in% Erucic_acid[Erucic_acid$P.value < 0.00001,]$SNP

SNP_gene_final_fxn$Oleic_acid <- SNP_gene_final_fxn$SNP.ID %in% Oleic_acid[Oleic_acid$P.value < 0.00001,]$SNP 

colnames(SNP_gene_final_fxn) <- c("gene_ID", "snp_ID", "CHROM", "POS", "Arabidopsis_homolog", "function", "Erucic_acid", "Oleic_acid")
SNP_gene_final_fxn

write.csv(SNP_gene_final_fxn, file = "~/505/output/131_sample/GWAS_SNP_gene_fxn.csv")

# as a summary, 222 out of the 262 SNPs fall within genes, 201 of them have arabidopsis homologs.
``` 

# for the several significant SNPs, are they synonymous or not? what's the possible effect on the protein function? 
```{r}
# SNP annotation 
# in the snpEff folder 
# java -jar snpEff.jar napus /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_leaf_no_pop/combined/missing_0.5/505_filtered_sorted.vcf > /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_leaf_no_pop/combined/missing_0.5/505_filtered_sorted_ann.vcf 

# analyze snpEff annotation file 
# cat GWAS_SNP_gene_fxn.csv | sed 's/"//g' | sed 's/,/ /g' | awk '{print $3}' | tail -222 > significant_SNP
# grep -f significant_SNP /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_131_sample/combined/505_filtered_sorted_ann.vcf | awk '{print $3 "|" $4 "|" $5 "|" $8}' | awk 'BEGIN{FS="|"}{print $1, $2, $3, $5}' > significant_SNP_annotation 

sig_SNP_ann <- read_table2("~/505/output/131_sample/significant_SNP_annotation", col_names=F)
head(sig_SNP_ann)  
colnames(sig_SNP_ann) <- c("snp_ID", "ref", "alt", "mutation_type")

unique(sig_SNP_ann$mutation_type)                          

SNP_gene_final_fxn <- read.csv("~/505/output/131_sample/GWAS_SNP_gene_fxn.csv")
head(SNP_gene_final_fxn)
dim(SNP_gene_final_fxn)

SNP_gene_final_fxn.2 <- merge(SNP_gene_final_fxn, sig_SNP_ann, by="snp_ID")
SNP_gene_final_fxn.2 
```

# check the expression for genes include significant SNPs 
For the 24 genes that include the significant SNPs, we want to know whether they are differentially expressed among different lines, and what are their expression levels. 
```{r}
# 1) collect read count data for all lines 
# batch C(47), D(41), E(43)
library("edgeR")

read.count <- read.table("~/505/data/late_silique/read.count.505.tsv", header=T, row.names=1)
head(read.count)
dim(read.count) # 101040    131 

read.count.sample <- data.frame(group=factor(colnames(read.count)))
read.count.sample

# filter based on read count
read.count.small <- read.count[rowSums(read.count > 10) >= 3,]
dim(read.count.small) # 59934   131  
save(read.count.small, file="~/505/output/131_sample/read.count.small.Rdata")

# .... leave for later 
```

# correlation between gene expression & phenotypic trait value 
```{r}
# voom transformation of raw read count data 
library("DESeq2")  
load("~/505/output/131_sample/read.count.small.Rdata")

dds.505.silique <- DESeqDataSetFromMatrix(countData = round(read.count.small), colData = read.count.sample, design = ~0+group)

vsd.505.silique <- varianceStabilizingTransformation(dds.505.silique)
vstMat.505.silique <- assay(vsd.505.silique)
colnames(vstMat.505.silique) <- colnames(read.count.small)
# save(vstMat.505.silique, file = "~/505/output/131_sample/vstMat.505.silique.Rdata")
load("~/505/output/131_sample/vstMat.505.silique.Rdata")

colnames(vstMat.505.silique) <- gsub("Sample_", "", colnames(vstMat.505.silique))
colnames(vstMat.505.silique) <- gsub("\\.", "\\-", colnames(vstMat.505.silique))
colnames(vstMat.505.silique)

## pull out the vst transformed expression value for the sig genes (this way I am only getting genes that are expressed above certain value) 
SNP.gene.fxn <-read.csv("~/505/output/131_sample/GWAS_SNP_gene_fxn.csv", header=T)
vstMat.siggene <- as.data.frame(vstMat.505.silique[which(rownames(vstMat.505.silique) %in% SNP.gene.fxn$gene_ID),])
dim(vstMat.siggene) # 90 131  
head(vstMat.siggene)

# get the phenotypic trait values 
load("~/505/output/131_sample/pheno_data_505.silique.Rdata")
colnames(pheno_data_505.silique)[2] <- "Sample.ID"

vstMat.siggene.t <- data.frame(t(vstMat.siggene))
vstMat.siggene.t$Sample.ID <- rownames(vstMat.siggene.t) 
dim(vstMat.siggene.t) # 131 91 

trait.gene <- merge(vstMat.siggene.t, pheno_data_505.silique, by="Sample.ID")
dim(trait.gene) # 131 96  
summary(trait.gene)
trait.gene.sub <- trait.gene[!is.na(trait.gene$Erucic_acid),]
dim(trait.gene.sub) # 114 96 

# get spearman correlation test result 
phenotypes <- colnames(pheno_data_505.silique)[4:6]
genes <- rownames(vstMat.siggene)
spearman.result <- spearman.cor.gene.trait(trait.gene.sub=trait.gene.sub, phenotypes=phenotypes, genes=genes)
spearman.estimate <- spearman.result[[1]] 
spearman.pvalue <- spearman.result[[2]]

# reform for ggplot 
spearman.estimate$trait <- rownames(spearman.estimate)
spearman.estimate
spearman.result <- melt(spearman.estimate, id.vars="trait") 
spearman.result$pvalue <- as.data.frame(melt(spearman.pvalue))[,"value"]
spearman.result$sig <- ifelse(spearman.result$pvalue < 0.05, "*", " ")
spearman.result
dim(spearman.result) # 270 5 
save(spearman.result, file="~/505/output/131_sample/spearman.reult.Rdata")

# extract genes that have greater than 0.5 significant (p-value < 0.05) correlation with trait value 
genes.sig.cor <-
  spearman.result %>%
  filter(abs(value) >= 0.4 & pvalue <= 0.05)  

genes.sig.cor

# combine this info with annotation file 
genesID.sig.cor <- as.character(unique(genes.sig.cor$variable))
length(genesID.sig.cor)

View(SNP.gene.fxn)
View(SNP_gene_final_fxn.2)

gene_report <- SNP_gene_final_fxn.2[which(SNP_gene_final_fxn.2$gene_ID %in% genesID.sig.cor),]
colnames(gene_report)
gene_report.final <- 
  gene_report %>%
  select(snp_ID, gene_ID, CHROM, POS, Arabidopsis_homolog, function., Erucic_acid, Oleic_acid, mutation_type)

gene_report.final <- data.frame(sapply(gene_report.final, function(x) sub("FALSE","N",x))) 
gene_report.final <- data.frame(sapply(gene_report.final, function(x) sub("TRUE","Y",x))) 
gene_report.final
write.csv(gene_report.final, file="~/505/output/131_sample/gene_report.final.csv")
```

# filter based on max 0.2 missing data, also change the order of filter parameter, remove those with high het ratio 
```{r}
# https://github.com/leejimmy93/KIAT/blob/master/505/filter_late_silique.sh

# filter based on site heterozygosity 
# site that are highly heterozygous might be result of mis-mapping result of homeologous sequences 

# convert .vcf.gz file to hmp file --> import into tassel --> calculate site heterozygosity --> remove site with high heterozygosity (>0.2)
 
# convert to hmp format 
# run_pipeline.pl -SortGenotypeFilePlugin -inputFile 505_filtered_0.2_missing.vcf.gz -outputFile 505_filtered_0.2_missing_sorted -fileType VCF
# run_pipeline.pl -Xmx5g -fork1 -vcf 505_filtered_0.2_missing_sorted.vcf -export -exportType Hapmap -runfork1 

# mv 505_filtered_0.hmp.txt 505_filtered_0.2.hmp.txt

# use Tassel to calculate site heterzygosity ratio 
# export site summary file 

library("tidyverse")
site_summary <- read.table("~/505/genomic_prediction/data/site_summary_tassel.txt", sep="\t", header=T)
site_summary[1:10, 1:10]
dim(site_summary)  # 273363     37
hist(site_summary$Proportion.Heterozygous)

png("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/het.ratio.png", width=4, height=3, units="in", res=300) 
hist(site_summary$Proportion.Heterozygous, xlab = "heterozygosity ratio across samples", ylab="Number of SNPs", main="heterozygosity ratio", breaks=50)
abline(v=0.2, col="red", lwd=2)
dev.off()

colnames(site_summary) 

# SNP.ID.het.filterd <- 
# site_summary %>% 
#   filter(Proportion.Heterozygous < 0.1) %>%
#   select(Chromosome, Physical.Position) 

SNP.ID.het.filterd.0.2 <- 
site_summary %>% 
  filter(Proportion.Heterozygous < 0.2) %>%
  select(Chromosome, Physical.Position)

dim(SNP.ID.het.filterd.0.2) # 174397      2 
head(SNP.ID.het.filterd.0.2)
unique(SNP.ID.het.filterd.0.2$Chromosome)
SNP.ID.het.filterd.0.2$Chromosome <- paste("chr", SNP.ID.het.filterd.0.2$Chromosome, sep="")
SNP.ID.het.filterd.0.2$Chromosome <- gsub("NN", "nn", SNP.ID.het.filterd.0.2$Chromosome)
SNP.ID.het.filterd.0.2$Chromosome <- gsub("RANDOM", "random", SNP.ID.het.filterd.0.2$Chromosome)

write.table(SNP.ID.het.filterd.0.2, file= "~/505/genomic_prediction/output/SNP.ID.het.filterd.0.2.txt") 

# cat SNP.ID.het.filterd.0.2.txt | sed 's/"//g' | grep -v "Chromosome" | awk '{print $2 "\t" $3}' > SNP.ID.het.filterd.0.2

# vcftools --positions /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/genomic_prediction/output/SNP.ID.het.filterd.0.2 --gzvcf 505_filtered_0.2_missing.vcf.gz --recode --recode-INFO-all --out 505_filtered_het_0.2  

# gzip 505_filtered_het_0.2.recode.vcf
############################################


# dim(SNP.ID.het.filterd) # 98961      2
# head(SNP.ID.het.filterd) 
# SNP.ID.het.filterd$Chromosome
# 
# SNP.ID.het.filterd$Chromosome <- paste("chr", SNP.ID.het.filterd$Chromosome, sep="")
# 
# write.table(SNP.ID.het.filterd, file= "~/505/genomic_prediction/output/SNP.ID.het.filterd.txt") 

# filter based on het ratio 
# cat SNP.ID.het.filterd.txt | sed 's/"//g' | grep -v "Chromosome" | awk '{print "$2 "\t" $3}' > SNP.ID.het.filterd 
# filter to keep loci with high het ratio  
# vcftools --positions /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/genomic_prediction/output/SNP.ID.het.filterd --gzvcf 505_filtered_0.2_missing.vcf.gz --recode --recode-INFO-all --out 505_filtered_het_1 (wired that vcftool filtered based on position will lose positions...) 

# 505_filtered_het_1.recode.vcf OR 505_filtered_het.recode.vcf 
# import to genomic prediction script 
```

# re-run GWAS using new dataset, filtered using missing data 0.2 & het ratio of 0.2 
```{r}
# run_pipeline.pl -SortGenotypeFilePlugin -inputFile 505_filtered_het.recode.vcf -outputFile 505_filtered_het.recode.vcf_sorted -fileType VCF
# run_pipeline.pl -Xmx5g -fork1 -vcf 505_filtered_het_0.2.recode.sorted.vcf -export -exportType Hapmap -runfork1  

# edit output file to remove "#" after rs & assembly 
# run GWAS 
# https://github.com/leejimmy93/KIAT/blob/master/505/GWAS_505_late_silique.R 
# mahattan plot: https://github.com/leejimmy93/KIAT/blob/master/505/mahattan_plot.R  
``` 

# extract significant SNPs info & SNP annotation 
```{r}
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures)
library("rtracklayer")

### get gff file with gene chrom & pos info, gff3 file must be sorted 
gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 

### get significant SNPs 
Erucic_acid <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/filter_by_het/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/filter_by_het/GAPIT..Oleic_acid.GWAS.Results.csv", header = T) 

Erucic_Oleic <- rbind(Erucic_acid, Oleic_acid)

SNP.Erucic.Oleic <- data.frame(CHROM = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\2", Erucic_Oleic[Erucic_Oleic$P.value < 0.00001,]$SNP), 
                         POS = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\4", Erucic_Oleic[Erucic_Oleic$P.value < 0.00001,]$SNP)) 

SNP.Erucic.Oleic$CHROM <- paste("chr", SNP.Erucic.Oleic$CHROM, sep = "")
SNP.Erucic.Oleic$POS <- as.numeric(as.character(SNP.Erucic.Oleic$POS))

#### overlap between Erucic and Oleic acid 
length(Erucic_acid[Erucic_acid$P.value < 0.00001,]$SNP) # 231
length(Oleic_acid[Oleic_acid$P.value < 0.00001,]$SNP) # 104

head(SNP.Erucic.Oleic) 
SNP.Erucic.Oleic.unique <- unique(SNP.Erucic.Oleic)
dim(SNP.Erucic.Oleic.unique) # 235 2 
head(SNP.Erucic.Oleic.unique)

genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), names = gff.mRNA$name)
genes

SNP <- GRanges(seqnames = Rle(SNP.Erucic.Oleic.unique$CHROM), ranges = IRanges(start = SNP.Erucic.Oleic.unique$POS, end = SNP.Erucic.Oleic.unique$POS), ID = paste(SNP.Erucic.Oleic.unique$CHROM, SNP.Erucic.Oleic.unique$POS, sep = "_"))
SNP

SNP_gene <- mergeByOverlaps(SNP, genes) # warning message here. look out!!!!!! 
SNP_gene

SNP_gene_df <- as.data.frame(SNP_gene)
SNP_gene_df

SNP_gene_final <- SNP_gene_df[,c("SNP.ID", "genes.seqnames", "SNP.start", "names")]
SNP_gene_final 

# napus_vs_ara.non_reciprocal <- read.table("~/Desktop/Brassica_project/reference/napus_vs_ara.non_reciprocal.table")
napus_vs_ara.non_reciprocal <- read.table("~/Reference/B.napus/napus_vs_ara.non_reciprocal.table")

head(napus_vs_ara.non_reciprocal)
dim(napus_vs_ara.non_reciprocal) # 64949 

colnames(napus_vs_ara.non_reciprocal)[1] <- "names"

SNP_gene_final_fxn <- merge(SNP_gene_final, napus_vs_ara.non_reciprocal, by = "names", all.x=T)
SNP_gene_final_fxn$SNP.ID <- gsub("chr", "S", SNP_gene_final_fxn$SNP.ID)

SNP_gene_final_fxn$Erucic_acid <- SNP_gene_final_fxn$SNP.ID %in% Erucic_acid[Erucic_acid$P.value < 0.00001,]$SNP

SNP_gene_final_fxn$Oleic_acid <- SNP_gene_final_fxn$SNP.ID %in% Oleic_acid[Oleic_acid$P.value < 0.00001,]$SNP 

colnames(SNP_gene_final_fxn) <- c("gene_ID", "snp_ID", "CHROM", "POS", "Arabidopsis_homolog", "function", "Erucic_acid", "Oleic_acid")
View(SNP_gene_final_fxn) 
nrow(SNP_gene_final_fxn) # 204 

write.csv(SNP_gene_final_fxn, file = "~/505/output/131_sample/GWAS_SNP_gene_fxn.csv")

# as a summary, 204 out of the 235 SNPs fall within 84 genes, 69 of them have arabidopsis homologs.
########### come to this stage... 
``` 

# for the several significant SNPs, are they synonymous or not? what's the possible effect on the protein function? 
```{r} 
# SNP annotation 
# in the snpEff folder 
# java -jar snpEff.jar napus /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.sorted.vcf > /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.sorted.ann.vcf  

# analyze snpEff annotation file 

# number_of_SNP <- nrow(SNP_gene_final_fxn)  
# cat GWAS_SNP_gene_fxn.csv | sed 's/"//g' | sed 's/,/ /g' | awk '{print $3}' | tail -${number_of_SNP} > significant_SNP
# awk 'NR==FNR{a[$1]=$0}NR>FNR{if ($3 in a) print $0}' significant_SNP /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.sorted.ann.vcf | awk '{print $3 "|" $4 "|" $5 "|" $8}' | awk 'BEGIN{FS="|"}{print $1, $2, $3, $5}' > significant_SNP_annotation 

sig_SNP_ann <- read_table2("~/505/output/131_sample/significant_SNP_annotation", col_names=F)
head(sig_SNP_ann)  
dim(sig_SNP_ann) # 204 4 
colnames(sig_SNP_ann) <- c("snp_ID", "ref", "alt", "mutation_type")

unique(sig_SNP_ann$mutation_type)                          

SNP_gene_final_fxn <- read.csv("~/505/output/131_sample/GWAS_SNP_gene_fxn.csv")
head(SNP_gene_final_fxn)
dim(SNP_gene_final_fxn) # 204  

SNP_gene_final_fxn.2 <- merge(SNP_gene_final_fxn, sig_SNP_ann, by="snp_ID")
SNP_gene_final_fxn.2 
View(SNP_gene_final_fxn.2) 
save(SNP_gene_final_fxn.2, file="~/505/output/131_sample/SNP_gene_final_fxn.2.Rdata")
```

# overlaps between DEGs (genes differentially expressed Da-Ae & Da-Ol-1 at any stage) & GWAS loci
```{r}
# differentially expressed genes anlayzed collectively, significant for any... 
load("~/505/output/131_sample/DEgene.new.Rdata")
dim(DEgene.new.gt) # 15098     9 
dim(DEgene.new.interaction) # 1701    8 
dim(DEgene.new.tissue) # 22580     8 

# differentially expressed genes pairwise result, Da-Ol-1 as reference level 
load("~/505/output/131_sample/DEgene.pairwise.gt.Rdata")
dim(DEgene.young); dim(DEgene.bolting); dim(DEgene.flowering); dim(DEgene.early.silique); dim(DEgene.late.silique)
# [1] 8976    5
# [1] 4648    5
# [1] 6081    5
# [1] 6695    5
# [1] 7413    5

### make a list 
DE.list <- list(gt = DEgene.new.gt, tissue = DEgene.new.tissue, interaction = DEgene.new.interaction, young = DEgene.young, bolting = DEgene.bolting, flowering = DEgene.flowering, early_silique = DEgene.early.silique, late_silique = DEgene.late.silique)
length(DE.list) # 8 

load("~/505/output/131_sample/SNP_gene_final_fxn.2.Rdata")

# question to answer: 
# 1) Among genes that are significant for GWAS, how many are differentially expressed between Da-Ae and Da-Ol for gt effect, for each tissue type? for late & early silique? 
# 2) although we could not find fatty acid genes by using blast result from Arabidopsis, what is the GO & IPR term of these genes? The absence of fatty acid related genes might be due to the lack of such genes in Arabidopsis 


# find overlaps with DEGs 
gene_list <- list()

for (i in 1:length(DE.list)){
  DE.list[[i]]$gene_ID <- rownames(DE.list[[i]])
  tmp <- 
  SNP_gene_final_fxn.2 %>%
  left_join(DE.list[[i]], by="gene_ID")  
   
  gene_list[[i]] <- tmp[!is.na(tmp$FDR),][,"gene_ID"]
  gene_number <- nrow(tmp[!is.na(tmp$FDR),])
  cat(gene_number, "\n")
}

# 63 gt
# 45 tissue
# 2 interaction
# 34 young
# 8 bolting
# 28 flowering
# 26 early silique
# 25 late silique 

### use GO & IPR term to check genes 
SNP_gene_final_fxn.2 
B.napus.GO <- read.delim(file="~/Reference/B.napus/Brassica_napus_GO", header=F, sep=" ")
dim(B.napus.GO) # 150320      2 # GO annotation 
B.napus.IPR <- read.delim(file="~/Reference/B.napus/Brassica_napus_IPR", header=F, sep=" ")
dim(B.napus.IPR) # 188954      2  # InterProScan annotation 
colnames(B.napus.GO) <- c("gene_ID", "GO")
colnames(B.napus.IPR) <- c("gene_ID", "IPR")

tmp<-
SNP_gene_final_fxn.2 %>% 
  left_join(B.napus.GO, "gene_ID") 

unique(tmp$GO, na.rm=T)
colnames(SNP_gene_final_fxn.2) 
# GO:0015976 carbon utilization, might be important for oil synthesis 
# GO:0005975 

tmp[which(tmp$GO == "GO:0015976"),]
```

# overlaps between GWAS loci range with DEGs for gt effect 
```{r}
### identify the range of A08 & C03 QTLs 
### question to answer: the identified SNP loci range may provide us information, so check the genes within those region, and pull out their expression between Da-Ae & Da-Ol, check their Arabidopsis homologs  

str(SNP.Erucic.Oleic.unique) 
SNP.Erucic.Oleic.unique %>%
  filter(CHROM=="chrA08") %>%
  arrange(desc(POS)) # range 7053250 to 11476346 

SNP.Erucic.Oleic.unique %>%
  filter(CHROM=="chrC03") %>%
  arrange(desc(POS))  # range 52598214 to 56930228     

# get range as an range object 
SNP.range <- GRanges(seqnames = Rle(c("chrA08", "chrC03")), ranges = IRanges(start = c(7053250, 52598214), end = c(11476346, 56930228)))
SNP.range

### BnaC03g67820D: chrC03 57461343 57463972  2630      * BnaC03g67820D

SNP_range_gene <- mergeByOverlaps(SNP.range, genes)
SNP_range_gene

# enrichment of these different genes 

SNP_range_gene.df <- as.data.frame(SNP_range_gene)
head(SNP_range_gene.df)

save(SNP_range_gene.df, file="~/505/output/131_sample/SNP_range_gene.df.Rdata")
```

# GO enrichment of genes in A08 and C03 GWAS range 
```{r}
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/131_sample/SNP_range_gene.df.Rdata")
GOseq.Bn.ORA(SNP_range_gene.df$names) 
head(SNP_range_gene.df) 
```

# check expression of genes in A08 & C03 GWAS loci 
```{r}
# arabidopsis homologs 
head(napus_vs_ara.non_reciprocal)
dim(napus_vs_ara.non_reciprocal) # 64949 
colnames(napus_vs_ara.non_reciprocal)[1] <- "gene_ID"

colnames(SNP_range_gene.df)[12] <- "gene_ID"

## check the Arabidopsis homolog of these genes 
SNP_range_gene.df %>% 
  semi_join(napus_vs_ara.non_reciprocal, by="gene_ID") %>%
  View()


gene_list <- 
lapply(1:length(DE.list), function(i){
  DE.list[[i]]$gene_ID <- rownames(DE.list[[i]])  
  
  tmp <- 
    DE.list[[i]] %>%
    semi_join(SNP_range_gene.df, by="gene_ID") %>% # get genes in the GWAS loci 
    left_join(napus_vs_ara.non_reciprocal, by="gene_ID") # add Arabidopsis homolog information 
  
  gene_number <- nrow(tmp)
  cat(gene_number, "\n")
  
  return(tmp)
})

# 151 gt
# 211 tissue
# 13 interaction
# 80 young
# 36 bolting
# 48 flowering
# 49 early silique
# 64 late silique 

### check result 
View(gene_list[[1]])
gene_list[[8]][which(gene_list[[8]]$gene_ID=="BnaA08g12780D"),] # between Da-Ae & Da-Ol-1 late silique
<<<<<<< HEAD
gene_list[[2]][which(gene_list[[2]]$gene_ID=="BnaA08g12780D"),] # between Da-Ae & Da-Ol-1 late silique 
=======
gene_list[[2]][which(gene_list[[2]]$gene_ID=="BnaA08g12780D"),] # between Da-Ae & Da-Ol-1 late silique

### plot the expression of the two genes 
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/vstMat.parent.Rdata")
genes.GWAS <- c("BnaA08g12780D", "BnaC03g67820D")
genes.GWAS.df <- as.data.frame(genes.GWAS)
colnames(genes.GWAS.df) <- "V1"
genes.GWAS.df

expression.pattern.Bn.parent(genes.GWAS.df)
ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/131_sample/two_genes.png", width = 7, height = 5)  

expression.pattern.Bn.parent.bar(vstMat.parent = vstMat.parent, gene = genes.GWAS.df)
ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/131_sample/two_genes.png", width = 7, height = 5) 
>>>>>>> dcba47f81724688cb93464ba860a307637974ed2
```

# GWAS on more traits: 
bolting date, flowering date, fruition, and oil content 
```{r}
load("~/505/data/phenotype/mongolia_korea_data_means_combined_JM.Rdata")
pheno.505.four <- data.all[,grep("SN|flower_50|bolt_50|fruition|oil", colnames(data.all), value=T)]
colnames(pheno.505.four) <- c("SN", "bolting_time", "flowering_time", "fruition","oil_content")

str(pheno.505.four)
# get the right format for fruition 
pheno.505.four.2 <- 
pheno.505.four %>%
  separate(fruition, into=c("month", "day", "year")) %>%
  mutate(fruition=day) %>%
  dplyr::select(SN, bolting_time, flowering_time, fruition, oil_content) 
  
str(pheno.505.four.2)
# need to change fruition charater to numeric values 
pheno.505.four.2$fruition <- as.numeric(pheno.505.four.2$fruition)
str(pheno.505.four.2)
pheno.505.four.2$fruition # very few have data, don't think this can generate useful information 

# replace SN with sample ID in genotype data 
sample_des_c <- read.csv("~/505/data/phenotype/batch_c.csv", stringsAsFactors=F, header=T)
sample_des_d <- read.csv("~/505/data/phenotype/batch_d.csv", stringsAsFactors=F, header=T)
sample_des_e <- read.csv("~/505/data/phenotype/batch_e.csv", stringsAsFactors=F, header=T)

sample_des_c_sub2 <- sample_des_c[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_d_sub2 <- sample_des_d[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_e_sub2 <- sample_des_e[,c("Sample.ID", "No..of.Sowing", "Name")]

sample_des_d_sub2 <- sample_des_d_sub2[c(1:41),]
sample_des_sub2 <- rbind(sample_des_c_sub2, sample_des_d_sub2, sample_des_e_sub2)

sample_des_sub2$SN <- paste("K", sample_des_sub2$No..of.Sowing, sep="")
pheno.505.four.2.merge <- merge(sample_des_sub2, pheno.505.four.2, by = "SN")
colnames(pheno.505.four.2.merge)

pheno_data_505 <- pheno.505.four.2.merge[,c("Sample.ID", "bolting_time", "flowering_time", "fruition", "oil_content")]
colnames(pheno_data_505)[1] <- "Taxa"
pheno_data_505
K250_K251 <- data.frame(Taxa = c("505_K_250", "505_K_251"),
    	bolting_time = rep(NA, 2),
			flowering_time = rep(NA, 2),
			fruition = rep(NA, 2),
      oil_content = rep(NA, 2))
pheno_data_505 <- rbind(pheno_data_505, K250_K251)
str(pheno_data_505)
dim(pheno_data_505)
pheno_data_505

save(pheno_data_505, file="~/505/genomic_prediction/data/pheno_data_505_bolting_flowering_fruition_oil.Rdata") 

## combine oil content & flowering time for GWAS 
load("~/505/genomic_prediction/data/pheno.505.Rdata") # pheno data for fatty acid 
pheno.505$Taxa <- rownames(pheno.505)
pheno.505

identical(test$Oil_content, test$oil_content)
test$Oil_content
test$oil_content

test$Oil_content %in% test$oil_content # wierd that these two data are not the same, I will use the one that I am sure correct, "Oil_content" 

pheno.505.six <- 
pheno.505 %>%
  left_join(pheno_data_505, by=c("Taxa")) %>%
  dplyr::select(Taxa, Erucic_acid, Oleic_acid, Oil_content, bolting_time, flowering_time, fruition)

head(pheno.505.six)
save(pheno.505.six, file="~/505/genomic_prediction/data/pheno.505.six.Rdata")
```

# plot out phenotype data??? 
```{r}
hist(pheno_data_505$bolting_time)
hist(pheno_data_505$flowering_time)
hist(pheno_data_505$fruition)
hist(pheno_data_505$oil_content)
```

# run GWAS 
on bolting, flowering, fruition time and oil content 
```{r}
# GWAS 
# https://github.com/leejimmy93/KIAT/blob/master/505/GWAS_505_late_silique_bolting_flowering_fruition_oil_content.R

# plotting 
# https://github.com/leejimmy93/KIAT/blob/master/505/mahattan_plot_bolting_flower_oil.R
```

# extract genes for signficant SNPs 
```{r}
### get significant SNPs 
flowering_time <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/bolting_floweirng_fruition_oil/GAPIT..flowering_time.GWAS.Results.csv", header = T)

bolting_time <- read.csv("~/505/output/myGAPIT/late_silique_131_sample/bolting_floweirng_fruition_oil/GAPIT..bolting_time.GWAS.Results.csv", header = T) 

time <- rbind(flowering_time, bolting_time)

SNP.time <- data.frame(CHROM = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\2", time[time$P.value < 0.00001,]$SNP), 
                         POS = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\4", time[time$P.value < 0.00001,]$SNP)) 

SNP.time$CHROM <- paste("chr", SNP.time$CHROM, sep = "")
SNP.time$POS <- as.numeric(as.character(SNP.time$POS))

#### overlap between Erucic and Oleic acid 
length(flowering_time[flowering_time$P.value < 0.00001,]$SNP) # 1
length(bolting_time[bolting_time$P.value < 0.00001,]$SNP) # 5

######### work tonight to identify genes in those flowering and bolting time regions 
# maybe should still use the old code 



















SNP.time
SNP.time.unique <- unique(SNP.time)
dim(SNP.time.unique) # 6 2 
head(SNP.time.unique)

SNP_gene_final <- SNP.annotation(SNP_data=SNP.time.unique)
SNP_gene_final

# napus_vs_ara.non_reciprocal <- read.table("~/Desktop/Brassica_project/reference/napus_vs_ara.non_reciprocal.table")
napus_vs_ara.non_reciprocal <- read.table("~/Reference/B.napus/napus_vs_ara.non_reciprocal.table")

head(napus_vs_ara.non_reciprocal)
dim(napus_vs_ara.non_reciprocal) # 64949 

colnames(napus_vs_ara.non_reciprocal)[1] <- "gene_ID"

SNP_gene_final_fxn <- merge(SNP_gene_final, napus_vs_ara.non_reciprocal, by = "gene_ID", all.x=T)
SNP_gene_final_fxn$SNP.ID <- gsub("chr", "S", SNP_gene_final_fxn$SNP.ID)

SNP_gene_final_fxn$Erucic_acid <- SNP_gene_final_fxn$SNP.ID %in% Erucic_acid[Erucic_acid$P.value < 0.00001,]$SNP

SNP_gene_final_fxn$Oleic_acid <- SNP_gene_final_fxn$SNP.ID %in% Oleic_acid[Oleic_acid$P.value < 0.00001,]$SNP 

colnames(SNP_gene_final_fxn) <- c("gene_ID", "snp_ID", "CHROM", "POS", "Arabidopsis_homolog", "function", "Erucic_acid", "Oleic_acid")
View(SNP_gene_final_fxn) 
nrow(SNP_gene_final_fxn) # 204 

write.csv(SNP_gene_final_fxn, file = "~/505/output/131_sample/GWAS_SNP_gene_fxn.csv")

# as a summary, 204 out of the 235 SNPs fall within 84 genes, 69 of them have arabidopsis homologs.
########### come to this stage... 
``` 

# for the several significant SNPs, are they synonymous or not? what's the possible effect on the protein function? 
```{r} 
# SNP annotation 
# in the snpEff folder 
# java -jar snpEff.jar napus /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.sorted.vcf > /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.sorted.ann.vcf  

# analyze snpEff annotation file 

# number_of_SNP <- nrow(SNP_gene_final_fxn)  
# cat GWAS_SNP_gene_fxn.csv | sed 's/"//g' | sed 's/,/ /g' | awk '{print $3}' | tail -${number_of_SNP} > significant_SNP
# awk 'NR==FNR{a[$1]=$0}NR>FNR{if ($3 in a) print $0}' significant_SNP /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.sorted.ann.vcf | awk '{print $3 "|" $4 "|" $5 "|" $8}' | awk 'BEGIN{FS="|"}{print $1, $2, $3, $5}' > significant_SNP_annotation 

sig_SNP_ann <- read_table2("~/505/output/131_sample/significant_SNP_annotation", col_names=F)
head(sig_SNP_ann)  
dim(sig_SNP_ann) # 204 4 
colnames(sig_SNP_ann) <- c("snp_ID", "ref", "alt", "mutation_type")

unique(sig_SNP_ann$mutation_type)                          

SNP_gene_final_fxn <- read.csv("~/505/output/131_sample/GWAS_SNP_gene_fxn.csv")
head(SNP_gene_final_fxn)
dim(SNP_gene_final_fxn) # 204  

SNP_gene_final_fxn.2 <- merge(SNP_gene_final_fxn, sig_SNP_ann, by="snp_ID")
SNP_gene_final_fxn.2 
View(SNP_gene_final_fxn.2) 
save(SNP_gene_final_fxn.2, file="~/505/output/131_sample/SNP_gene_final_fxn.2.Rdata")
```










