---
title: "505_population_structure"
output: html_document
---

The goal of this script is to check the population structure of this 505 diversity panel, do this seperately for RNA & WGS datasets 

```{r}
library("tidyverse")
```

### check common SNPs between RNA & WGS datasets  
```{r} 
### filtered based on het ratio 
SNP_RNA<- read.table("~/505/vcf_late_silique_131_sample/combined/505_filtered_het_0.2.recode.hmp.txt", header = T, stringsAsFactors = F) 

SNP_WGS <- read.table("~/505/WGS/vcf_filtered/two_options/hmp/505_filtered_het_0.hmp.txt", header = T, stringsAsFactors = F) 

SNP_RNA %>% dim(); SNP_WGS %>% dim() 
# [1] 174397    142
# [1] 1529974     249  

common_WGS_RNA <- SNP_RNA %>% semi_join(SNP_WGS, by = c("alleles", "rs")) 
dim(common_WGS_RNA) # 11998   142     
```

### 2) population structure analysis (demo)
```{r}
# not filtered based on het ratio 
common_WGS_RNA <- common_WGS_RNA %>% dplyr::select(-(rs:QCcode)) 
common_WGS_RNA_2 <- common_WGS_RNA_2 %>% dplyr::select(-(rs:QCcode)) 

common_WGS_RNA <- common_WGS_RNA %>% t() 
common_WGS_RNA_2 <- common_WGS_RNA_2 %>% t() 

late_silique_505.2.t <- common_WGS_RNA 

geno.numeric <- apply(late_silique_505.2.t, 2, function(x) as.numeric(as.factor(x)))
rnames <- rownames(late_silique_505.2.t)
rownames(geno.numeric) <- rnames
geno.numeric.df <- as.data.frame(geno.numeric)
genDist <- as.matrix(dist(geno.numeric.df))
#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions 
dim(geno.mds)
plot(geno.mds)        

geno.mds$Sample.ID <- rownames(geno.mds)

library(ggrepel)
set.seed(111)  

sample_info <- geno.mds %>% rownames() %>% as.data.frame() 
sample_info$seq_info <- c(rep("RNA", 131), rep("WGS", 238))
colnames(sample_info)[1] <- "Sample.ID"

geno.mds.all.des <- merge(geno.mds, sample_info, by="Sample.ID")   
colnames(geno.mds.all.des)[1] <- "Sample.ID"

geno.mds.all.des$Sample.ID <- gsub("\\.","\\-", geno.mds.all.des$Sample.ID) 
geno.mds.all.des$Sample.ID <- gsub("X", "", geno.mds.all.des$Sample.ID)
class(geno.mds.all.des$Sample.ID)

load("~/505/output/late_silique/sample_des_revised.Rdata")
sample_des_revised$Sample.ID <- as.character(sample_des_revised$Sample.ID)

geno.mds.all.des$Sample.ID
sample_des_revised$Sample.ID

geno.mds.all.des.2 <- merge(geno.mds.all.des, sample_des_revised, by="Sample.ID")
head(geno.mds.all.des.2) 
dim(geno.mds.all.des.2) 

p.mds.country <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(seq_info)))
p.mds.country <- p.mds.country + geom_point(size=1) 
p.mds.country <- p.mds.country + scale_color_brewer(type="qual",palette="Set1")
p.mds.country 

p.mds.country <- ggplot(data=geno.mds.all.des.2, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country <- p.mds.country + geom_point(size=1) 
p.mds.country <- p.mds.country + scale_color_brewer(type="qual",palette="Set1")
# p.mds.country <- p.mds.country + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country     
```

* Do seperate population structure analysis using data from RNA and WGS data  

### to reduce time for STURCTURE
vcftools --vcf 505_filtered_het_0.2.vcf --max-missing 0.8 --recode --recode-INFO-all --out 505_WGS_pop (missing data < 0.2) 

148069 out of 1529974 Sites 

### MDS for population structure 
```{r}
SNP_WGS <- SNP_WGS %>% dplyr::select(-(rs:QCcode)) 

SNP_WGS <- SNP_WGS %>% t() 

geno.numeric <- apply(SNP_WGS, 2, function(x) as.numeric(as.factor(x)))
rnames <- rownames(SNP_WGS)
rownames(geno.numeric) <- rnames
geno.numeric.df <- as.data.frame(geno.numeric) 
genDist <- as.matrix(dist(geno.numeric.df))
#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions 
dim(geno.mds)
plot(geno.mds)        
geno.mds.WGS <- geno.mds

SNP_RNA <- SNP_RNA %>% dplyr::select(-(rs:QCcode)) 

SNP_RNA <- SNP_RNA %>% t() 

geno.numeric <- apply(SNP_RNA, 2, function(x) as.numeric(as.factor(x)))
rnames <- rownames(SNP_RNA)
rownames(geno.numeric) <- rnames
geno.numeric.df <- as.data.frame(geno.numeric)
genDist <- as.matrix(dist(geno.numeric.df))
#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions 
dim(geno.mds)
plot(geno.mds)        
geno.mds.RNA <- geno.mds 
```

### convert VCF to STRUCTURE 
http://www.cmpg.unibe.ch/software/PGDSpider/PGDSpider%20manual_vers%202-1-1-5.pdf
* make a spid file from GUI (open SGDSpider from local MAC)
java -Xmx50g -Xms50g -jar PGDSpider2-cli.jar -inputfile ~/505/WGS/vcf_filtered/two_options/505_filtered_het_0.2.vcf -inputformat VCF -outputfile ~/505/WGS/STRUCTURE/505_WGS_pop.txt -outputformat STRUCTURE -spid ~/505/WGS/STRUCTURE/505_WGS.spid

### install STRUCTURE 
https://web.stanford.edu/group/pritchardlab/structure_software/release_versions/v2.3.4/structure_kernel_source.tar.gz
tar xzvf structure_kernel_source.tar.gz 
cd structure_kernel_source 
make 

STURECTURE MANUAL 
http://burfordreiskind.com/wp-content/uploads/Structure_Manual_doc.pdf

K = 2,3,4,5,6 test for both WGS & RNA data 

./structure -K 3 -L 148069 -N 238 -i ~/505/WGS/STRUCTURE/505_WGS_pop.txt -o ~/505/WGS/STRUCTURE/505_WGS_pop_K3 

### RNA data 
vcftools --vcf 505_filtered_het_0.2.recode.sorted.vcf --max-missing 0.8 --recode --recode-INFO-all --out 505_RNA_pop  

53895 out of a possible 174397 Sites

java -Xmx50g -Xms50g -jar ~/bin/PGDSpider_2.1.1.5/PGDSpider2-cli.jar -inputfile 505_RNA_pop.recode.vcf -inputformat VCF -outputfile 505_RNA_pop.txt -outputformat STRUCTURE -spid 505_RNA.spid 

./structure -K 3 -L 53895 -N 131 -i ~/505/vcf_late_silique_131_sample/combined/STRUCTURE/505_RNA_pop.txt -o ~/505/vcf_late_silique_131_sample/combined/STRUCTURE/505_RNA_pop_K3 # takes two week for WGS and one week for RNA data 

### structure output analysis  

* Structure Harvester  
structureHarvester/structureHarvester.py --dir ~/505/WGS/STRUCTURE/ --out ~/505/WGS/STRUCTURE/ --evanno --clumpp    

structureHarvester/structureHarvester.py --dir ~/505/vcf_late_silique_131_sample/combined/STRUCTURE/ --out ~/505/vcf_late_silique_131_sample/combined/STRUCTURE/  --evanno --clumpp  

based on Delta K value from evanno.txt file, K =4 is the best K for WGS; K=4 is also the best K for RNAseq data  

### CLUMPP 
https://rosenberglab.stanford.edu/clumpp.html
./../../../bin/CLUMPP_Linux64.1.1.2/CLUMPP paramfile_WGS # modify paramfile for WGS data 
./../../../../bin/CLUMPP_Linux64.1.1.2/CLUMPP paramfile_RNA # modify paramfile for RNA data 

### distruct 
https://rosenberglab.stanford.edu/distruct.html 

















