---
title: "505_network_WGCNA_4"
output: html_document
---

purpose of this script is to combine GWAS analysis, WGCNA module-trait correlation, and module gene significance result to come up with biological interpretation, plot, plot, plot... 

### overlapped GWAS signal for trait-module pairs, which with interesting GO term enriched in that module 

goal: propose SNP/QTL/gene -- gene expression/biological pathway/GO term (module) -- phenotypic trait connection 

### GWAS plot for all traits & eigengene  
```{r}
library("qqman")

load("~/505/network_analysis/output/GWAS_eigengene.Rdata")
load("~/505/network_analysis/output/GWAS_trait.Rdata")

for(i in names(GWAS_eigengene)){ 

filename <- paste("~/505/network_analysis/output/figure/GWAS_eigengene_", i, ".png",  sep = "")
png(filename = filename, width=12, height=10, units="in", res=300)

colnames(GWAS_eigengene[[i]])[1:4] <- c("SNP", "CHR", "BP", "P")
manhattan(GWAS_eigengene[[i]], main = i, ylim = c(0, 8), cex = 0.6,
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F)

dev.off()
}

for(i in names(GWAS_trait)){ 

filename <- paste("~/505/network_analysis/output/figure/GWAS_trait_", i, ".png",  sep = "")
png(filename = filename, width=12, height=10, units="in", res=300)

colnames(GWAS_trait[[i]])[1:4] <- c("SNP", "CHR", "BP", "P")
manhattan(GWAS_trait[[i]], main = i, ylim = c(0, 8), cex = 0.6,
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F)

dev.off()
}
```

### load overlap data and take a look 
```{r} 
load("~/505/network_analysis/output/GWAS_eigengene_trait_overlap_gene.Rdata")

GWAS_eigengene_trait_overlap_gene$class %>% unique() %>% length() # 540 
GWAS_eigengene_trait_overlap_gene$class %>% unique() 

# important modules based on heatmap 
grep("MEdarkorange_", GWAS_eigengene_trait_overlap_gene$class %>% unique(), value = T) # flower eQTL, don't need to pay much attention to this 

load("~/505/network_analysis/output/blups.cor.5.Rdata")
load("~/505/network_analysis/output/geneTraitSignificance.final.Rdata")

blups.cor.5 %>% 
  as.data.frame() %>% 
  dplyr::select(MEyellow, MEsalmon, MEturquoise, MEdarkturquoise, MEdarkred, MEgreen, MEivory)

load("~/505/network_analysis/input/module.gene.GO.Rdata")
```

### propose network for trait regulated by salmon and yellow, Erucic acid and Oleic acid 
```{r}
grep("MEyellow_", GWAS_eigengene_trait_overlap_gene$class %>% unique(), value = T) # Erucic acid/ Oleic acid 
grep("MEsalmon_", GWAS_eigengene_trait_overlap_gene$class %>% unique(), value = T) # Erucic acid/ Oleic acid 
GO.module[["yellow"]]
GO.module[["salmon"]] 
```

salmon and yellow don't have overlapped QTL/SNP/gene with FA trait, so the hypothesis is that FA and defense can affect each other, correlation. 

### turquoise, doesn't seem very good module, GWAS result looks messy, correlation between FA, etc. trait with this module, only focus on WGCNA result    
```{r}
GO.module[["turquoise"]]
grep("MEturquoise_", GWAS_eigengene_trait_overlap_gene$class %>% unique(), value = T) ### 
# this module has high correlation with both FA and plant growth, also has overlapped QTL with many trait, 24 

tmp <- grep("MEturquoise_", GWAS_eigengene_trait_overlap_gene$class %>% unique(), value = T) 
tmp <- grep("acid", tmp, value = T) 
turquoise_FA <- 
GWAS_eigengene_trait_overlap_gene[GWAS_eigengene_trait_overlap_gene$class %in% tmp,] 

GWAS_eigengene.sig[["MEturquoise"]] 

# plot it out 
colnames(GWAS_eigengene[["MEturquoise"]])[1:4] <- c("SNP", "CHR", "BP", "P")
manhattan(GWAS_eigengene[["MEturquoise"]], main = "", ylim = c(0, 8), cex = 0.6,
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F)
```

### darkturquoise, useful, genes control defense and genes control FA, growth are linked or has pleotrophic effect
```{r}
GO.module[["darkturquoise"]]
grep("MEdarkturquoise_", GWAS_eigengene_trait_overlap_gene$class %>% unique(), value = T) ### 

colnames(GWAS_eigengene[["MEdarkturquoise"]] )[1:4] <- c("SNP", "CHR", "BP", "P")
manhattan(GWAS_eigengene[["MEdarkturquoise"]], main = "", ylim = c(0, 8), cex = 0.6,
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F) 

GWAS_eigengene_trait_overlap_gene %>% 
  filter(class == "MEdarkturquoise_Behenic_acid_year2" | 
           class == "MEdarkturquoise_Erucic_acid_year1" | 
           class == "MEdarkturquoise_Erucic_acid_year2" | 
           class == "MEdarkturquoise_vaccenic_acid_year2") %>% 
  dplyr::select(names) %>% 
  unique() %>% 
  left_join(BnapusAnnotation, by=c("names"="name")) %>% 
  View()

left_join()

```

### darkred, dosn't look quite good, messy GWAS result, only focus on WGCNA result 
```{r}
GO.module[["darkred"]]
grep("MEdarkred_", GWAS_eigengene_trait_overlap_gene$class %>% unique(), value = T) ### 

colnames(GWAS_eigengene[["MEdarkred"]] )[1:4] <- c("SNP", "CHR", "BP", "P")
manhattan(GWAS_eigengene[["MEdarkred"]], main = "", ylim = c(0, 8), cex = 0.6,
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F)  
```

### green, good result! MEgreen_plant_width_bottom_08_07_MG_year1 
```{r}
GO.module[["green"]]
grep("MEgreen_", GWAS_eigengene_trait_overlap_gene$class %>% unique(), value = T) ### 

colnames(GWAS_eigengene[["MEgreen"]] )[1:4] <- c("SNP", "CHR", "BP", "P")
manhattan(GWAS_eigengene[["MEgreen"]], main = "", ylim = c(0, 8), cex = 0.6,
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F)   

GWAS_eigengene_trait_overlap_gene %>% 
  filter(class == "MEgreen_plant_width_bottom_08_07_MG_year1") 
```

### ivory, not good to use GWAS result, focus on correlation 
```{r}
GO.module[["ivory"]]
grep("MEivory_", GWAS_eigengene_trait_overlap_gene$class %>% unique(), value = T) ### 

colnames(GWAS_eigengene[["MEivory"]] )[1:4] <- c("SNP", "CHR", "BP", "P")
manhattan(GWAS_eigengene[["MEivory"]], main = "", ylim = c(0, 8), cex = 0.6,
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F)   

GWAS_eigengene_trait_overlap_gene %>% 
  filter(class == "MEivory_bolting_height_12_15_KR_GH_year1") %>% 
  dplyr::select(genes.seqnames) %>% 
  unique()  
```

### look for overlap between GWAS candidate genes with genes from modules which are highly correlated with different traits, eQTL for different traits 

```{r}
load("~/505/network_analysis/output/GWAS_trait.sig.genes.Rdata")
load("~/505/network_analysis/output/geneTraitSignificance.final.Rdata")
  
geneTraitSignificance.final$Erucic_acid_year1$GS.Erucic_acid_year1
GWAS_trait.sig.genes$Behenic_acid_year2 

geneTraitSignificance.final$Erucic_acid_year1 %>% 
  semi_join(GWAS_trait.sig.genes$Erucic_acid_year1, by= c("geneID"="names")) %>% 
  View()

GWAS_trait_gene_GS <- 
lapply(names(geneTraitSignificance.final), function(i) {
  if((i %in% names(GWAS_trait.sig.genes)) && (nrow(GWAS_trait.sig.genes[[i]])>0)){
  geneTraitSignificance.final[[i]] %>% 
  semi_join(GWAS_trait.sig.genes[[i]], by= c("geneID"="names"))
  }
})

names(GWAS_trait_gene_GS) <- names(geneTraitSignificance.final)

test <- 
sapply(names(geneTraitSignificance.final), function(i) {
  if((i %in% names(GWAS_trait.sig.genes)) && (nrow(GWAS_trait.sig.genes[[i]])>0)){
  tmp <-   
  geneTraitSignificance.final[[i]] %>% 
  semi_join(GWAS_trait.sig.genes[[i]], by= c("geneID"="names")) %>% 
      nrow()
  }
})

test2 <- test %>% unlist() %>% as.data.frame() 

GWAS_trait_gene_GS$Palmitic_acid_year2 %>% View()
GWAS_trait_gene_GS$Heptadecanoic_acid_year2 %>% View()
GWAS_trait_gene_GS$Stearic_acid_year2 %>% View()
GWAS_trait_gene_GS$vaccenic_acid_year2 %>% View()
GWAS_trait_gene_GS$Linoleic_acid_year2 %>% View()
GWAS_trait_gene_GS$Linolenic_acid_year2 %>% View()
GWAS_trait_gene_GS$Arachidic_acid_year2 %>% View()
GWAS_trait_gene_GS$Behenic_acid_year2 %>% View()
GWAS_trait_gene_GS$Erucic_acid_year1 %>% View() # ***  
# BnaA08g30740D 
# BnaC03g66150D plays a role in membrane lipid modulation under Al stress.
GWAS_trait_gene_GS$Erucic_acid_year2 %>% View() # ***  
GWAS_trait_gene_GS$Oleic_acid_year1 %>% View() # *** 
GWAS_trait_gene_GS$Oleic_acid_year2 %>% View() # *** 
GWAS_trait_gene_GS$Erucic_acid_year1 %>% 
  filter(geneID == "BnaA08g30740D") %>% 
  dplyr::select(geneID, module, chrom, AGI, At_symbol, GS.Erucic_acid_year1) %>% 
  View()

GWAS_trait_gene_GS$bolt_50_days_MG_year1 %>% View()  
GWAS_trait_gene_GS$leafn_July_03_04_MG_year1 %>% View()  
GWAS_trait_gene_GS$flower_50_days_MG_year1 %>% View() 
GWAS_trait_gene_GS$bolting_length_08_07_MG_year1 %>% View() 
GWAS_trait_gene_GS$germ.rate_KR_GH_year1 %>% View()
GWAS_trait_gene_GS$germ.rate_KR_GH_year1[!is.na(GWAS_trait_gene_GS$germ.rate_KR_GH_year1$AGI),] %>% 
  filter(abs(GS.germ.rate_KR_GH_year1) > 0.2) %>% 
  dplyr::select(geneID, module, chrom, AGI, At_symbol, GS.germ.rate_KR_GH_year1) %>% 
  View()   
``` 



