---
title: "505_population_structure"
output: html_document
---

The goal of this script is to check the population structure of this 505 diversity panel. 

```{r}
library("tidyverse")
```

### 1) 
get the common SNPs between RNAseq and WGS for MDS analyiss, show the MDS plot 

SNP filtering for WGS data & RNAseq data 
QUAL > 40, lmiss < 0.5, mean read depth > 1 and < 100, MAF > 0.01  

* MAF 
SNPs due to sequencing errors will usually have major allele frequencies close to 1, because few genotypes will have an allele due to the error. So we could remove most SNPs due to sequencing errors by using this filter. If we do it, we will also filter out lots of real SNPs that are almost fixed in the population.

# stats for RNAseq data 
```{r} 
# QUAL score
QUAL <- read.delim("~/505/vcf_late_silique_131_sample/combined/505.lqual")
hist(log10(QUAL$QUAL))    
sum(QUAL$QUAL > 40)/nrow(QUAL) # 80% 
sum(QUAL$QUAL > 30)/nrow(QUAL) # 87% 

# MAF
MAF <- read.table("~/505/vcf_late_silique_131_sample/combined/MAF")
nrow(MAF)/nrow(QUAL) # 83% SNP out of all squence variations called from Freebayes

MAF <- 
MAF %>% 
  mutate(V7 = gsub("([[:print:]]+)(:)([[:print:]]+)", "\\3", V5), 
         V8 = gsub("([[:print:]]+)(:)([[:print:]]+)", "\\3", V6)) %>% 
  dplyr::select(V7, V8) 

MAF_minor <- as.numeric(apply(MAF, 1, min))  
sum(MAF_minor[!is.na(MAF_minor)] > 0.05) / nrow(QUAL) # 22% 
sum(MAF_minor[!is.na(MAF_minor)] > 0.01) / nrow(QUAL) # 54% # use this 
sum(MAF_minor[!is.na(MAF_minor)] > 0.001) / nrow(QUAL) # 80% # use this 

hist(as.numeric(apply(MAF, 1, min))) 

# mean read depth across individuals 
read_depth <- read.delim("~/505/vcf_late_silique_131_sample/combined/505.ldepth.mean")
mean(read_depth$MEAN_DEPTH) # 15  

hist(log10(read_depth$MEAN_DEPTH), breaks = 1000)
read_depth <- 
read_depth %>% 
  filter(MEAN_DEPTH > 1 & MEAN_DEPTH < 100) 

nrow(read_depth)/nrow(QUAL) # 97% 
hist(read_depth$MEAN_DEPTH, breaks = 500)

# missing rate 
# across taxa, on site level 
lmiss <- read.delim("~/505/vcf_late_silique_131_sample/combined/505.lmiss")
hist(lmiss$F_MISS) 
sum(lmiss$F_MISS < 0.5)/nrow(lmiss) # 79% left     
```

filter WGS data 
https://github.com/leejimmy93/KIAT/blob/master/505/WGS/filter.sh 

filter RNAseq data 
https://github.com/leejimmy93/KIAT/blob/master/505/filter_late_silique.sh 

convert to hmp --> calculate heterozygosity --> filter based on heterozygosity 
https://github.com/leejimmy93/KIAT/blob/master/505/WGS/vcf_to_hmp_2.sh 
https://github.com/leejimmy93/KIAT/blob/master/505/WGS/format_hmp_2.txt --> export site summary file 

# remove site with high heterozygosity (>0.2) 

```{r}
site_summary_RNA <- read.table("~/505/vcf_combine_WGS_RNAseq/RNA_filtered/hmp/RNA_het.txt", sep="\t", header=T)  
site_summary_WGS <- read.table("~/505/WGS/vcf_raw/filtere/hmp/WGS_het.txt", sep="\t", header=T)  
site_summary_WGS_2 <- read.table("~/505/WGS/vcf_raw/filtered/filter_1/hmp/WGS_2_het.txt", sep="\t", header=T)   

dim(site_summary_RNA) # 544432     37  
dim(site_summary_WGS)  # 303555     37  
dim(site_summary_WGS_2)  # 208755     37  

hist(site_summary_RNA$Proportion.Heterozygous)  
hist(site_summary_WGS$Proportion.Heterozygous)  
hist(site_summary_WGS_2$Proportion.Heterozygous)  

SNP.ID.het.filterd.0.2.RNA <-  
site_summary_RNA %>% 
  filter(Proportion.Heterozygous < 0.2) %>%
  select(Chromosome, Physical.Position)

SNP.ID.het.filterd.0.2.WGS <-  
site_summary_WGS %>% 
  filter(Proportion.Heterozygous < 0.2) %>%
  select(Chromosome, Physical.Position)

SNP.ID.het.filterd.0.2.WGS.2 <-  
site_summary_WGS_2 %>% 
  filter(Proportion.Heterozygous < 0.2) %>%
  dplyr::select(Chromosome, Physical.Position)

dim(SNP.ID.het.filterd.0.2.RNA) # 467669      2 
dim(SNP.ID.het.filterd.0.2.WGS) # 123126      2 
dim(SNP.ID.het.filterd.0.2.WGS.2) # 180429      2  

SNP.ID.het.filterd.0.2.RNA$Chromosome <- paste("chr", SNP.ID.het.filterd.0.2.RNA$Chromosome, sep="")
SNP.ID.het.filterd.0.2.RNA$Chromosome <- gsub("NN", "nn", SNP.ID.het.filterd.0.2.RNA$Chromosome)
SNP.ID.het.filterd.0.2.RNA$Chromosome <- gsub("RANDOM", "random", SNP.ID.het.filterd.0.2.RNA$Chromosome)

SNP.ID.het.filterd.0.2.WGS$Chromosome <- paste("chr", SNP.ID.het.filterd.0.2.WGS$Chromosome, sep="")
SNP.ID.het.filterd.0.2.WGS$Chromosome <- gsub("NN", "nn", SNP.ID.het.filterd.0.2.WGS$Chromosome)
SNP.ID.het.filterd.0.2.WGS$Chromosome <- gsub("RANDOM", "random", SNP.ID.het.filterd.0.2.WGS$Chromosome)

SNP.ID.het.filterd.0.2.WGS.2$Chromosome <- paste("chr", SNP.ID.het.filterd.0.2.WGS.2$Chromosome, sep="")
SNP.ID.het.filterd.0.2.WGS.2$Chromosome <- gsub("NN", "nn", SNP.ID.het.filterd.0.2.WGS.2$Chromosome)
SNP.ID.het.filterd.0.2.WGS.2$Chromosome <- gsub("RANDOM", "random", SNP.ID.het.filterd.0.2.WGS.2$Chromosome) 

write.table(SNP.ID.het.filterd.0.2.RNA, file= "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_combine_WGS_RNAseq/RNA_filtered/hmp/SNP.ID.het.filterd.0.2.txt")   

write.table(SNP.ID.het.filterd.0.2.WGS, file= "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/WGS/vcf_raw/filtered/hmp/SNP.ID.het.filterd.0.2.txt") 

# cat SNP.ID.het.filterd.0.2.txt | sed 's/"//g' | grep -v "Chromosome" | awk '{print $2 "\t" $3}' > SNP.ID.het.filterd.0.2  

# vcftools --positions /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/genomic_prediction/output/SNP.ID.het.filterd.0.2 --gzvcf 505_filtered_0.2_missing.vcf.gz --recode --recode-INFO-all --out 505_filtered_het_0.2    

# gzip 505_filtered_het_0.2.recode.vcf

# check read depth for high het sites 
read_depth <- read.delim("~/505/WGS/vcf_raw/combined/505.ldepth.mean", stringsAsFactors = F)
mean(read_depth$MEAN_DEPTH) # 3.56 

SNP.ID.het.filterd.0.2.WGS.2.depth <- 
SNP.ID.het.filterd.0.2.WGS.2 %>% 
  left_join(read_depth, c("Chromosome" = "CHROM", "Physical.Position" = "POS")) 

SNP.ID.het.filterd.0.2.WGS.depth <- 
SNP.ID.het.filterd.0.2.WGS %>% 
  left_join(read_depth, c("Chromosome" = "CHROM", "Physical.Position" = "POS")) 

min(SNP.ID.het.filterd.0.2.WGS.depth$MEAN_DEPTH) # het ratio < 0.2 
min(SNP.ID.het.filterd.0.2.WGS.2.depth$MEAN_DEPTH)

mean(SNP.ID.het.filterd.0.2.WGS.depth$MEAN_DEPTH)
mean(SNP.ID.het.filterd.0.2.WGS.2.depth$MEAN_DEPTH)

median(SNP.ID.het.filterd.0.2.WGS.depth$MEAN_DEPTH)
median(SNP.ID.het.filterd.0.2.WGS.2.depth$MEAN_DEPTH)

max(SNP.ID.het.filterd.0.2.WGS.depth$MEAN_DEPTH)
max(SNP.ID.het.filterd.0.2.WGS.2.depth$MEAN_DEPTH)  
SNP.ID.het.filterd.0.2.WGS.depth %>% 
  filter(MEAN_DEPTH > 100)

hist(log10(SNP.ID.het.filterd.0.2.WGS.depth$MEAN_DEPTH))
hist(log10(SNP.ID.het.filterd.0.2.WGS.2.depth$MEAN_DEPTH)) 
``` 

convert vcf to hmp 
https://github.com/leejimmy93/KIAT/blob/master/505/WGS/vcf_to_hmp_2.sh 
https://github.com/leejimmy93/KIAT/blob/master/505/WGS/format_hmp_2.txt 
sed 's/#//g' 505.hmp.txt > 505_final.hmp.txt 

### get common SNPs between RNA & WGS dataset --> check data seperation using MDS  
```{r} 
### filtered based on het ratio 
SNP_RNA_2 <- read.table("~/505/vcf_combine_WGS_RNAseq/RNA_filtered/hmp/505_final.hmp.txt", header = T, stringsAsFactors = F) 

SNP_WGS_2 <- read.table("~/505/WGS/vcf_raw/filtered/hmp/505_final.hmp.txt", header = T, stringsAsFactors = F)

SNP_RNA_2 %>% dim(); SNP_WGS_2 %>% dim() 
# [1] 467669    142
# [1] 123126    249  

common_WGS_RNA_2 <- SNP_RNA_2 %>% semi_join(SNP_WGS_2, by = c("alleles", "rs")) 
dim(common_WGS_RNA_2) # 391 142

common_WGS_RNA_2 <- common_WGS_RNA_2 %>% left_join(SNP_WGS_2) 

###### not filtered by het ratio 
SNP_RNA <- read.table("~/505/vcf_combine_WGS_RNAseq/RNA_filtered/505_final.hmp.txt", header = T, stringsAsFactors = F) 

SNP_WGS <- read.table("~/505/WGS/vcf_raw/filtered/505_final.hmp.txt", header = T, stringsAsFactors = F)

SNP_RNA %>% dim(); SNP_WGS %>% dim() 
# [1] 467669    142
# [1] 123126    249  

str(SNP_RNA)  
str(SNP_WGS) 

common_WGS_RNA <- SNP_RNA %>% semi_join(SNP_WGS, by = c("alleles", "rs")) 
dim(common_WGS_RNA) # 4569  142
common_WGS_RNA <- common_WGS_RNA %>% left_join(SNP_WGS) 
dim(common_WGS_RNA) # 4569  380  
```

decide how to conduct population structure analysis, use all data or just one of these two datasets 

### 2) population structure analysis  
```{r}
# not filtered based on het ratio 
common_WGS_RNA <- common_WGS_RNA %>% dplyr::select(-(rs:QCcode)) 
common_WGS_RNA_2 <- common_WGS_RNA_2 %>% dplyr::select(-(rs:QCcode)) 

common_WGS_RNA <- common_WGS_RNA %>% t() 
common_WGS_RNA_2 <- common_WGS_RNA_2 %>% t() 

late_silique_505.2.t <- common_WGS_RNA 

geno.numeric <- apply(late_silique_505.2.t, 2, function(x) as.numeric(as.factor(x)))
rnames <- rownames(late_silique_505.2.t)
rownames(geno.numeric) <- rnames
geno.numeric.df <- as.data.frame(geno.numeric)
genDist <- as.matrix(dist(geno.numeric.df))
#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions 
dim(geno.mds)
plot(geno.mds)        

geno.mds$Sample.ID <- rownames(geno.mds)

library(ggrepel)
set.seed(111)  

sample_info <- geno.mds %>% rownames() %>% as.data.frame() 
sample_info$seq_info <- c(rep("RNA", 131), rep("WGS", 238))
colnames(sample_info)[1] <- "Sample.ID"

geno.mds.all.des <- merge(geno.mds, sample_info, by="Sample.ID")   
colnames(geno.mds.all.des)[1] <- "Sample.ID"

geno.mds.all.des$Sample.ID <- gsub("\\.","\\-", geno.mds.all.des$Sample.ID) 
geno.mds.all.des$Sample.ID <- gsub("X", "", geno.mds.all.des$Sample.ID)
class(geno.mds.all.des$Sample.ID)

load("~/505/output/late_silique/sample_des_revised.Rdata")
sample_des_revised$Sample.ID <- as.character(sample_des_revised$Sample.ID)

geno.mds.all.des$Sample.ID
sample_des_revised$Sample.ID

geno.mds.all.des.2 <- merge(geno.mds.all.des, sample_des_revised, by="Sample.ID")
head(geno.mds.all.des.2) 
dim(geno.mds.all.des.2) 

p.mds.country <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(seq_info)))
p.mds.country <- p.mds.country + geom_point(size=1) 
p.mds.country <- p.mds.country + scale_color_brewer(type="qual",palette="Set1")
p.mds.country 


p.mds.country <- ggplot(data=geno.mds.all.des.2, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country <- p.mds.country + geom_point(size=1) 
p.mds.country <- p.mds.country + scale_color_brewer(type="qual",palette="Set1")
# p.mds.country <- p.mds.country + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country     
```

* I feel SNP calling from low coverage data is not credible: 1) low overlap with RNAseq data 2) very high het ratio sample wise and site wise, so determined to use RNAseq data only for downstream analysis 
### population structure 
```{r}

```


